# Comparing `tmp/cnmfsns-1.6.5.tar.gz` & `tmp/cnmfsns-1.7.0.tar.gz`

## filetype from file(1)

```diff
@@ -1 +1 @@
-gzip compressed data, was "cnmfsns-1.6.5.tar", last modified: Fri Jul 21 16:11:48 2023, max compression
+gzip compressed data, was "cnmfsns-1.7.0.tar", last modified: Mon Jul 31 21:04:29 2023, max compression
```

## Comparing `cnmfsns-1.6.5.tar` & `cnmfsns-1.7.0.tar`

### file list

```diff
@@ -1,25 +1,25 @@
-drwxrwxrwx   0        0        0        0 2023-07-21 16:11:48.063226 cnmfsns-1.6.5/
--rw-rw-rw-   0        0        0     1090 2023-04-24 20:31:27.000000 cnmfsns-1.6.5/LICENSE
--rw-rw-rw-   0        0        0     4320 2023-07-21 16:11:48.063226 cnmfsns-1.6.5/PKG-INFO
--rw-rw-rw-   0        0        0     3805 2023-07-21 16:02:46.000000 cnmfsns-1.6.5/README.md
--rw-rw-rw-   0        0        0       86 2023-04-24 20:31:27.000000 cnmfsns-1.6.5/pyproject.toml
--rw-rw-rw-   0        0        0      980 2023-07-21 16:11:48.063226 cnmfsns-1.6.5/setup.cfg
-drwxrwxrwx   0        0        0        0 2023-07-21 16:11:48.030243 cnmfsns-1.6.5/src/
-drwxrwxrwx   0        0        0        0 2023-07-21 16:11:48.046501 cnmfsns-1.6.5/src/cnmfsns/
--rw-rw-rw-   0        0        0      513 2023-05-12 18:06:16.000000 cnmfsns-1.6.5/src/cnmfsns/__init__.py
--rw-rw-rw-   0        0        0    52492 2023-07-21 16:08:05.000000 cnmfsns-1.6.5/src/cnmfsns/cli.py
--rw-rw-rw-   0        0        0    39682 2023-05-12 20:51:27.000000 cnmfsns-1.6.5/src/cnmfsns/cnmf.py
--rw-rw-rw-   0        0        0    16061 2023-04-24 20:31:27.000000 cnmfsns-1.6.5/src/cnmfsns/colors.py
--rw-rw-rw-   0        0        0     3257 2023-07-18 03:10:45.000000 cnmfsns-1.6.5/src/cnmfsns/config.py
--rw-rw-rw-   0        0        0    36407 2023-07-19 20:10:17.000000 cnmfsns-1.6.5/src/cnmfsns/dataset.py
--rw-rw-rw-   0        0        0    21163 2023-06-23 16:39:53.000000 cnmfsns-1.6.5/src/cnmfsns/integration.py
--rw-rw-rw-   0        0        0    59662 2023-07-10 17:04:09.000000 cnmfsns-1.6.5/src/cnmfsns/plots.py
--rw-rw-rw-   0        0        0    41881 2023-07-21 16:04:07.000000 cnmfsns-1.6.5/src/cnmfsns/sns.py
--rw-rw-rw-   0        0        0     2767 2023-07-10 02:54:46.000000 cnmfsns-1.6.5/src/cnmfsns/utils.py
-drwxrwxrwx   0        0        0        0 2023-07-21 16:11:48.063226 cnmfsns-1.6.5/src/cnmfsns.egg-info/
--rw-rw-rw-   0        0        0     4320 2023-07-21 16:11:48.000000 cnmfsns-1.6.5/src/cnmfsns.egg-info/PKG-INFO
--rw-rw-rw-   0        0        0      472 2023-07-21 16:11:48.000000 cnmfsns-1.6.5/src/cnmfsns.egg-info/SOURCES.txt
--rw-rw-rw-   0        0        0        1 2023-07-21 16:11:48.000000 cnmfsns-1.6.5/src/cnmfsns.egg-info/dependency_links.txt
--rw-rw-rw-   0        0        0       44 2023-07-21 16:11:48.000000 cnmfsns-1.6.5/src/cnmfsns.egg-info/entry_points.txt
--rw-rw-rw-   0        0        0      189 2023-07-21 16:11:48.000000 cnmfsns-1.6.5/src/cnmfsns.egg-info/requires.txt
--rw-rw-rw-   0        0        0        8 2023-07-21 16:11:48.000000 cnmfsns-1.6.5/src/cnmfsns.egg-info/top_level.txt
+drwxrwxrwx   0        0        0        0 2023-07-31 21:04:29.608713 cnmfsns-1.7.0/
+-rw-rw-rw-   0        0        0     1090 2023-04-24 20:31:27.000000 cnmfsns-1.7.0/LICENSE
+-rw-rw-rw-   0        0        0     4361 2023-07-31 21:04:29.608713 cnmfsns-1.7.0/PKG-INFO
+-rw-rw-rw-   0        0        0     3846 2023-07-31 21:01:55.000000 cnmfsns-1.7.0/README.md
+-rw-rw-rw-   0        0        0       86 2023-04-24 20:31:27.000000 cnmfsns-1.7.0/pyproject.toml
+-rw-rw-rw-   0        0        0      980 2023-07-31 21:04:29.609714 cnmfsns-1.7.0/setup.cfg
+drwxrwxrwx   0        0        0        0 2023-07-31 21:04:29.560163 cnmfsns-1.7.0/src/
+drwxrwxrwx   0        0        0        0 2023-07-31 21:04:29.579732 cnmfsns-1.7.0/src/cnmfsns/
+-rw-rw-rw-   0        0        0      513 2023-05-12 18:06:16.000000 cnmfsns-1.7.0/src/cnmfsns/__init__.py
+-rw-rw-rw-   0        0        0    52741 2023-07-31 21:02:15.000000 cnmfsns-1.7.0/src/cnmfsns/cli.py
+-rw-rw-rw-   0        0        0    39682 2023-07-24 22:05:31.000000 cnmfsns-1.7.0/src/cnmfsns/cnmf.py
+-rw-rw-rw-   0        0        0    16061 2023-04-24 20:31:27.000000 cnmfsns-1.7.0/src/cnmfsns/colors.py
+-rw-rw-rw-   0        0        0     3204 2023-07-31 19:07:44.000000 cnmfsns-1.7.0/src/cnmfsns/config.py
+-rw-rw-rw-   0        0        0    36443 2023-07-31 19:07:55.000000 cnmfsns-1.7.0/src/cnmfsns/dataset.py
+-rw-rw-rw-   0        0        0    22552 2023-07-31 20:58:13.000000 cnmfsns-1.7.0/src/cnmfsns/integration.py
+-rw-rw-rw-   0        0        0    62028 2023-07-31 20:59:13.000000 cnmfsns-1.7.0/src/cnmfsns/plots.py
+-rw-rw-rw-   0        0        0    41881 2023-07-28 20:18:37.000000 cnmfsns-1.7.0/src/cnmfsns/sns.py
+-rw-rw-rw-   0        0        0     2767 2023-07-10 02:54:46.000000 cnmfsns-1.7.0/src/cnmfsns/utils.py
+drwxrwxrwx   0        0        0        0 2023-07-31 21:04:29.607714 cnmfsns-1.7.0/src/cnmfsns.egg-info/
+-rw-rw-rw-   0        0        0     4361 2023-07-31 21:04:29.000000 cnmfsns-1.7.0/src/cnmfsns.egg-info/PKG-INFO
+-rw-rw-rw-   0        0        0      472 2023-07-31 21:04:29.000000 cnmfsns-1.7.0/src/cnmfsns.egg-info/SOURCES.txt
+-rw-rw-rw-   0        0        0        1 2023-07-31 21:04:29.000000 cnmfsns-1.7.0/src/cnmfsns.egg-info/dependency_links.txt
+-rw-rw-rw-   0        0        0       44 2023-07-31 21:04:29.000000 cnmfsns-1.7.0/src/cnmfsns.egg-info/entry_points.txt
+-rw-rw-rw-   0        0        0      189 2023-07-31 21:04:29.000000 cnmfsns-1.7.0/src/cnmfsns.egg-info/requires.txt
+-rw-rw-rw-   0        0        0        8 2023-07-31 21:04:29.000000 cnmfsns-1.7.0/src/cnmfsns.egg-info/top_level.txt
```

### Comparing `cnmfsns-1.6.5/LICENSE` & `cnmfsns-1.7.0/LICENSE`

 * *Files identical despite different names*

### Comparing `cnmfsns-1.6.5/PKG-INFO` & `cnmfsns-1.7.0/README.md`

 * *Files 18% similar despite different names*

```diff
@@ -1,270 +1,241 @@
-00000000: 4d65 7461 6461 7461 2d56 6572 7369 6f6e  Metadata-Version
-00000010: 3a20 322e 310d 0a4e 616d 653a 2063 6e6d  : 2.1..Name: cnm
-00000020: 6673 6e73 0d0a 5665 7273 696f 6e3a 2031  fsns..Version: 1
-00000030: 2e36 2e35 0d0a 5375 6d6d 6172 793a 2063  .6.5..Summary: c
-00000040: 4e4d 4620 536f 6c75 7469 6f6e 204e 6574  NMF Solution Net
-00000050: 776f 726b 2053 7061 6365 0d0a 486f 6d65  work Space..Home
-00000060: 2d70 6167 653a 2068 7474 7073 3a2f 2f67  -page: https://g
-00000070: 6974 6875 622e 636f 6d2f 4d6f 7272 6973  ithub.com/Morris
-00000080: 7379 4c61 622f 634e 4d46 2d53 4e53 0d0a  syLab/cNMF-SNS..
-00000090: 4175 7468 6f72 3a20 5465 6420 5665 7268  Author: Ted Verh
-000000a0: 6579 0d0a 4175 7468 6f72 2d65 6d61 696c  ey..Author-email
-000000b0: 3a20 7462 7665 7268 6579 4075 6361 6c67  : tbverhey@ucalg
-000000c0: 6172 792e 6361 0d0a 5072 6f6a 6563 742d  ary.ca..Project-
-000000d0: 5552 4c3a 2042 7567 2054 7261 636b 6572  URL: Bug Tracker
-000000e0: 2c20 6874 7470 733a 2f2f 6769 7468 7562  , https://github
-000000f0: 2e63 6f6d 2f4d 6f72 7269 7373 794c 6162  .com/MorrissyLab
-00000100: 2f63 4e4d 462d 534e 532f 6973 7375 6573  /cNMF-SNS/issues
-00000110: 0d0a 436c 6173 7369 6669 6572 3a20 5072  ..Classifier: Pr
-00000120: 6f67 7261 6d6d 696e 6720 4c61 6e67 7561  ogramming Langua
-00000130: 6765 203a 3a20 5079 7468 6f6e 203a 3a20  ge :: Python :: 
-00000140: 330d 0a43 6c61 7373 6966 6965 723a 204c  3..Classifier: L
-00000150: 6963 656e 7365 203a 3a20 4f53 4920 4170  icense :: OSI Ap
-00000160: 7072 6f76 6564 203a 3a20 4d49 5420 4c69  proved :: MIT Li
-00000170: 6365 6e73 650d 0a43 6c61 7373 6966 6965  cense..Classifie
-00000180: 723a 204f 7065 7261 7469 6e67 2053 7973  r: Operating Sys
-00000190: 7465 6d20 3a3a 204f 5320 496e 6465 7065  tem :: OS Indepe
-000001a0: 6e64 656e 740d 0a52 6571 7569 7265 732d  ndent..Requires-
-000001b0: 5079 7468 6f6e 3a20 3e3d 332e 390d 0a44  Python: >=3.9..D
-000001c0: 6573 6372 6970 7469 6f6e 2d43 6f6e 7465  escription-Conte
-000001d0: 6e74 2d54 7970 653a 2074 6578 742f 6d61  nt-Type: text/ma
-000001e0: 726b 646f 776e 0d0a 4c69 6365 6e73 652d  rkdown..License-
-000001f0: 4669 6c65 3a20 4c49 4345 4e53 450d 0a0d  File: LICENSE...
-00000200: 0a0d 0a21 5b63 4e4d 462d 534e 5320 6c6f  ...![cNMF-SNS lo
-00000210: 676f 5d28 6c6f 676f 2e70 6e67 290d 0a0d  go](logo.png)...
-00000220: 0a2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  .---------------
-00000230: 2d2d 0d0a 0d0a 2320 634e 4d46 2d53 4e53  --....# cNMF-SNS
-00000240: 3a20 706f 7765 7266 756c 2066 6163 746f  : powerful facto
-00000250: 7269 7a61 7469 6f6e 2d62 6173 6564 206d  rization-based m
-00000260: 756c 7469 2d6f 6d69 6373 2069 6e74 6567  ulti-omics integ
-00000270: 7261 7469 6f6e 2074 6f6f 6c6b 6974 0d0a  ration toolkit..
-00000280: 0d0a 215b 7665 7273 696f 6e20 6261 6467  ..![version badg
-00000290: 655d 2868 7474 7073 3a2f 2f69 6d67 2e73  e](https://img.s
-000002a0: 6869 656c 6473 2e69 6f2f 6261 6467 652f  hields.io/badge/
-000002b0: 7665 7273 696f 6e2d 312e 362e 352d 626c  version-1.6.5-bl
-000002c0: 7565 290d 0a5b 215b 5079 5049 204c 6174  ue)..[![PyPI Lat
-000002d0: 6573 7420 5265 6c65 6173 655d 2868 7474  est Release](htt
-000002e0: 7073 3a2f 2f69 6d67 2e73 6869 656c 6473  ps://img.shields
-000002f0: 2e69 6f2f 7079 7069 2f76 2f63 6e6d 6673  .io/pypi/v/cnmfs
-00000300: 6e73 2e73 7667 295d 2868 7474 7073 3a2f  ns.svg)](https:/
-00000310: 2f70 7970 692e 6f72 672f 7072 6f6a 6563  /pypi.org/projec
-00000320: 742f 636e 6d66 736e 732f 290d 0a5b 215b  t/cnmfsns/)..[![
-00000330: 436f 6e64 6120 4c61 7465 7374 2052 656c  Conda Latest Rel
-00000340: 6561 7365 5d28 6874 7470 733a 2f2f 696d  ease](https://im
-00000350: 672e 7368 6965 6c64 732e 696f 2f63 6f6e  g.shields.io/con
-00000360: 6461 2f76 6e2f 636f 6e64 612d 666f 7267  da/vn/conda-forg
-00000370: 652f 636e 6d66 736e 7329 5d28 6874 7470  e/cnmfsns)](http
-00000380: 733a 2f2f 616e 6163 6f6e 6461 2e6f 7267  s://anaconda.org
-00000390: 2f63 6f6e 6461 2d66 6f72 6765 2f63 6e6d  /conda-forge/cnm
-000003a0: 6673 6e73 2f29 0d0a 5b21 5b44 6f63 756d  fsns/)..[![Docum
-000003b0: 656e 7461 7469 6f6e 2073 7461 7475 735d  entation status]
-000003c0: 2868 7474 7073 3a2f 2f72 6561 6474 6865  (https://readthe
-000003d0: 646f 6373 2e6f 7267 2f70 726f 6a65 6374  docs.org/project
-000003e0: 732f 636e 6d66 2d73 6e73 2f62 6164 6765  s/cnmf-sns/badge
-000003f0: 2f3f 7665 7273 696f 6e3d 6c61 7465 7374  /?version=latest
-00000400: 2673 7479 6c65 3d66 6c61 7429 5d28 290d  &style=flat)]().
-00000410: 0a5b 215b 446f 776e 6c6f 6164 735d 2868  .[![Downloads](h
-00000420: 7474 7073 3a2f 2f73 7461 7469 632e 7065  ttps://static.pe
-00000430: 7079 2e74 6563 682f 6261 6467 652f 636e  py.tech/badge/cn
-00000440: 6d66 736e 7329 5d28 6874 7470 733a 2f2f  mfsns)](https://
-00000450: 7065 7079 2e74 6563 682f 7072 6f6a 6563  pepy.tech/projec
-00000460: 742f 636e 6d66 736e 7329 0d0a 5b21 5b4c  t/cnmfsns)..[![L
-00000470: 6963 656e 7365 5d28 6874 7470 733a 2f2f  icense](https://
-00000480: 696d 672e 7368 6965 6c64 732e 696f 2f70  img.shields.io/p
-00000490: 7970 692f 6c2f 636e 6d66 736e 732e 7376  ypi/l/cnmfsns.sv
-000004a0: 6729 5d28 6874 7470 733a 2f2f 6769 7468  g)](https://gith
-000004b0: 7562 2e63 6f6d 2f4d 6f72 7269 7373 794c  ub.com/MorrissyL
-000004c0: 6162 2f63 4e4d 462d 534e 532f 626c 6f62  ab/cNMF-SNS/blob
-000004d0: 2f6d 6169 6e2f 4c49 4345 4e53 4529 0d0a  /main/LICENSE)..
-000004e0: 0d0a 4175 7468 6f72 733a 205b 5465 6420  ..Authors: [Ted 
-000004f0: 5665 7268 6579 5d28 6874 7470 733a 2f2f  Verhey](https://
-00000500: 6769 7468 7562 2e63 6f6d 2f76 6572 6865  github.com/verhe
-00000510: 7974 6229 2c20 5b48 6565 776f 6e20 5365  ytb), [Heewon Se
-00000520: 6f5d 2868 7474 7073 3a2f 2f67 6974 6875  o](https://githu
-00000530: 622e 636f 6d2f 6c6f 6f74 7069 7a29 2c20  b.com/lootpiz), 
-00000540: 5b53 6f72 616e 6120 4d6f 7272 6973 7379  [Sorana Morrissy
-00000550: 5d28 6874 7470 733a 2f2f 6769 7468 7562  ](https://github
-00000560: 2e63 6f6d 2f61 6e63 6173 6f72 616e 6129  .com/ancasorana)
-00000570: 0d0a 0d0a 2a2a 634e 4d46 2d53 4e53 2a2a  ....**cNMF-SNS**
-00000580: 2028 636f 6e73 656e 7375 7320 4e6f 6e2d   (consensus Non-
-00000590: 6e65 6761 7469 7665 204d 6174 7269 7820  negative Matrix 
-000005a0: 4661 6374 6f72 697a 6174 696f 6e20 536f  Factorization So
-000005b0: 6c75 7469 6f6e 204e 6574 776f 726b 2053  lution Network S
-000005c0: 7061 6365 2920 6973 2061 2050 7974 686f  pace) is a Pytho
-000005d0: 6e20 7061 636b 6167 6520 656e 6162 6c69  n package enabli
-000005e0: 6e67 206d 6f73 6169 6320 696e 7465 6772  ng mosaic integr
-000005f0: 6174 696f 6e20 6f66 2062 756c 6b2c 2073  ation of bulk, s
-00000600: 696e 676c 652d 6365 6c6c 2c20 616e 640d  ingle-cell, and.
-00000610: 0a73 7061 7469 616c 2065 7870 7265 7373  .spatial express
-00000620: 696f 6e20 6461 7461 2062 6574 7765 656e  ion data between
-00000630: 2061 6e64 2077 6974 6869 6e20 6461 7461   and within data
-00000640: 7365 7473 2e20 4461 7461 7365 7473 2063  sets. Datasets c
-00000650: 616e 2068 6176 6520 7061 7274 6961 6c6c  an have partiall
-00000660: 7920 6f76 6572 6c61 7070 696e 6720 6665  y overlapping fe
-00000670: 6174 7572 6573 2028 6567 2e20 6765 6e65  atures (eg. gene
-00000680: 7329 2061 7320 7765 6c6c 2061 7320 6e6f  s) as well as no
-00000690: 6e2d 6f76 6572 6c61 7070 696e 6720 6665  n-overlapping fe
-000006a0: 6174 7572 6573 2e20 634e 4d46 2070 726f  atures. cNMF pro
-000006b0: 7669 6465 7320 6120 2a2a 726f 6275 7374  vides a **robust
-000006c0: 2c20 0d0a 756e 7375 7065 7276 6973 6564  , ..unsupervised
-000006d0: 2a2a 2064 6563 6f6e 766f 6c75 7469 6f6e  ** deconvolution
-000006e0: 206f 6620 6561 6368 2064 6174 6173 6574   of each dataset
-000006f0: 2069 6e74 6f20 6765 6e65 2065 7870 7265   into gene expre
-00000700: 7373 696f 6e20 7072 6f67 7261 6d73 2028  ssion programs (
-00000710: 4745 5073 292e 0d0a 2a2a 4e65 7477 6f72  GEPs)...**Networ
-00000720: 6b2d 6261 7365 6420 696e 7465 6772 6174  k-based integrat
-00000730: 696f 6e2a 2a20 6f66 2047 4550 7320 656e  ion** of GEPs en
-00000740: 6162 6c65 7320 666c 6578 6962 6c65 2069  ables flexible i
-00000750: 6e74 6567 7261 7469 6f6e 206f 6620 6d61  ntegration of ma
-00000760: 6e79 2064 6174 6173 6574 730d 0a61 6372  ny datasets..acr
-00000770: 6f73 7320 6173 7361 7973 2028 6567 2e20  oss assays (eg. 
-00000780: 5072 6f74 6569 6e2c 2052 4e41 2d53 6571  Protein, RNA-Seq
-00000790: 2c20 7363 524e 412d 5365 712c 2073 7061  , scRNA-Seq, spa
-000007a0: 7469 616c 2065 7870 7265 7373 696f 6e29  tial expression)
-000007b0: 2061 6e64 2070 6174 6965 6e74 2063 6f68   and patient coh
-000007c0: 6f72 7473 2e0d 0a0d 0a43 6f6d 6d75 6e69  orts.....Communi
-000007d0: 7469 6573 2077 6974 6820 4745 5073 2066  ties with GEPs f
-000007e0: 726f 6d20 6d75 6c74 6970 6c65 2064 6174  rom multiple dat
-000007f0: 6173 6574 7320 6361 6e20 6265 2061 6e6e  asets can be ann
-00000800: 6f74 6174 6564 2077 6974 6820 6461 7461  otated with data
-00000810: 7365 742d 7370 6563 6966 6963 0d0a 616e  set-specific..an
-00000820: 6e6f 7461 7469 6f6e 7320 746f 2066 6163  notations to fac
-00000830: 696c 6974 6174 6520 696e 7465 7270 7265  ilitate interpre
-00000840: 7461 7469 6f6e 2e0d 0a0d 0a23 2320 e29a  tation.....## ..
-00000850: a14d 6169 6e20 4665 6174 7572 6573 0d0a  .Main Features..
-00000860: 0d0a 4865 7265 2061 7265 206a 7573 7420  ..Here are just 
-00000870: 6120 6665 7720 6f66 2074 6865 2074 6869  a few of the thi
-00000880: 6e67 7320 7468 6174 2063 4e4d 462d 534e  ngs that cNMF-SN
-00000890: 5320 646f 6573 2077 656c 6c3a 0d0a 0d0a  S does well:....
-000008a0: 2d20 4964 656e 7469 6669 6573 2069 6e74  - Identifies int
-000008b0: 6572 7072 6574 6162 6c65 2c20 6164 6469  erpretable, addi
-000008c0: 7469 7665 206e 6f6e 2d6e 6567 6174 6976  tive non-negativ
-000008d0: 6520 6765 6e65 2065 7870 7265 7373 696f  e gene expressio
-000008e0: 6e20 7072 6f67 7261 6d73 0d0a 2d20 4d6f  n programs..- Mo
-000008f0: 7361 6963 2069 6e74 6567 7261 7469 6f6e  saic integration
-00000900: 2064 6f65 7320 6e6f 7420 7265 7175 6972   does not requir
-00000910: 6520 7375 6273 6574 7469 6e67 2066 6561  e subsetting fea
-00000920: 7475 7265 732f 6765 6e65 7320 746f 0d0a  tures/genes to..
-00000930: 2020 6120 7368 6172 6564 206f 7220 6f76    a shared or ov
-00000940: 6572 6469 7370 6572 7365 6420 7375 6273  erdispersed subs
-00000950: 6574 0d0a 2d20 4964 6561 6c20 666f 7220  et..- Ideal for 
-00000960: 696e 6372 656d 656e 7461 6c20 696e 7465  incremental inte
-00000970: 6772 6174 696f 6e20 2861 6464 696e 6720  gration (adding 
-00000980: 6461 7461 7365 7473 206f 6e65 2061 7420  datasets one at 
-00000990: 6120 7469 6d65 2920 7369 6e63 650d 0a20  a time) since.. 
-000009a0: 2064 6563 6f6e 766f 6c75 7469 6f6e 2069   deconvolution i
-000009b0: 7320 7065 7266 6f72 6d65 6420 696e 6465  s performed inde
-000009c0: 7065 6e64 656e 746c 7920 6f6e 2065 6163  pendently on eac
-000009d0: 6820 6461 7461 7365 7420 6765 6e65 7261  h dataset genera
-000009e0: 7469 6e67 2069 6e76 6172 6961 6e74 2047  ting invariant G
-000009f0: 4550 730d 0a2d 2049 6e74 6567 7261 7469  EPs..- Integrati
-00000a00: 6f6e 2064 6f65 7320 6e6f 7420 7265 7175  on does not requ
-00000a10: 6972 6520 636f 6d70 6172 6162 6c65 2064  ire comparable d
-00000a20: 6174 6120 7370 6172 7369 7479 206f 7220  ata sparsity or 
-00000a30: 7365 7175 656e 6369 6e67 2064 6570 7468  sequencing depth
-00000a40: 2028 7369 6e67 6c65 2d63 656c 6c2c 206d   (single-cell, m
-00000a50: 6963 726f 6172 7261 792c 2062 756c 6b29  icroarray, bulk)
-00000a60: 0d0a 2d20 5477 6f20 696e 7465 7266 6163  ..- Two interfac
-00000a70: 6573 3a20 636f 6d6d 616e 642d 6c69 6e65  es: command-line
-00000a80: 2069 6e74 6572 6661 6365 2066 6f72 2072   interface for r
-00000a90: 6170 6964 2064 6174 6120 6578 706c 6f72  apid data explor
-00000aa0: 6174 696f 6e20 616e 6420 7079 7468 6f6e  ation and python
-00000ab0: 0d0a 2020 696e 7465 7266 6163 6520 666f  ..  interface fo
-00000ac0: 7220 6578 7465 6e73 6962 696c 6974 7920  r extensibility 
-00000ad0: 616e 6420 666c 6578 6962 696c 6974 790d  and flexibility.
-00000ae0: 0a0d 0a23 2320 f09f 94a7 2049 6e73 7461  ...## .... Insta
-00000af0: 6c6c 0d0a 0d0a 2323 2320 e298 81ef b88f  ll....### ......
-00000b00: 2050 7562 6c69 6320 5265 6c65 6173 650d   Public Release.
-00000b10: 0a0d 0a49 6e73 7461 6c6c 2074 6865 2070  ...Install the p
-00000b20: 6163 6b61 6765 2077 6974 6820 636f 6e64  ackage with cond
-00000b30: 6120 2869 6e20 616e 2069 736f 6c61 7465  a (in an isolate
-00000b40: 6420 636f 6e64 6120 656e 7669 726f 6e6d  d conda environm
-00000b50: 656e 7429 0d0a 6060 6062 6173 680d 0a63  ent)..```bash..c
-00000b60: 6f6e 6461 2065 6e76 2063 7265 6174 6520  onda env create 
-00000b70: 636e 6d66 736e 7320 2d63 2063 6f6e 6461  cnmfsns -c conda
-00000b80: 2d66 6f72 6765 2063 6e6d 6673 6e73 0d0a  -forge cnmfsns..
-00000b90: 636f 6e64 6120 6163 7469 7661 7465 2063  conda activate c
-00000ba0: 6e6d 6673 6e73 0d0a 6060 600d 0a0d 0a23  nmfsns..```....#
-00000bb0: 2323 20e2 9ca8 204c 6174 6573 7420 7665  ## ... Latest ve
-00000bc0: 7273 696f 6e20 6672 6f6d 2047 6974 4875  rsion from GitHu
-00000bd0: 620d 0a0d 0a42 6566 6f72 6520 696e 7374  b....Before inst
-00000be0: 616c 6c69 6e67 2063 4e4d 462d 534e 5320  alling cNMF-SNS 
-00000bf0: 7573 696e 6720 7069 702c 2069 7420 6973  using pip, it is
-00000c00: 2072 6563 6f6d 6d65 6e64 6564 2074 6f20   recommended to 
-00000c10: 6669 7273 7420 7365 7420 7570 2061 2073  first set up a s
-00000c20: 6570 6172 6174 6520 636f 6e64 6120 656e  eparate conda en
-00000c30: 7669 726f 6e6d 656e 7420 616e 6420 6861  vironment and ha
-00000c40: 7665 2063 6f6e 6461 206d 616e 6167 6520  ve conda manage 
-00000c50: 6173 206d 616e 7920 6465 7065 6e64 656e  as many dependen
-00000c60: 6369 6573 2061 7320 706f 7373 6962 6c65  cies as possible
-00000c70: 2e0d 0a0d 0a60 6060 6261 7368 0d0a 636f  .....```bash..co
-00000c80: 6e64 6120 6372 6561 7465 202d 2d6e 616d  nda create --nam
-00000c90: 6520 636e 6d66 736e 7320 2d63 2063 6f6e  e cnmfsns -c con
-00000ca0: 6461 2d66 6f72 6765 2070 7974 686f 6e3d  da-forge python=
-00000cb0: 332e 3130 2061 6e6e 6461 7461 2070 616e  3.10 anndata pan
-00000cc0: 6461 7320 6e75 6d70 7920 7363 6970 7920  das numpy scipy 
-00000cd0: 6d61 7470 6c6f 746c 6962 2075 7073 6574  matplotlib upset
-00000ce0: 706c 6f74 2068 7474 706c 6962 3220 746f  plot httplib2 to
-00000cf0: 6d6c 6920 746f 6d6c 692d 7720 636c 6963  mli tomli-w clic
-00000d00: 6b20 7079 6772 6170 6876 697a 2070 7974  k pygraphviz pyt
-00000d10: 686f 6e2d 6967 7261 7068 2070 7979 616d  hon-igraph pyyam
-00000d20: 6c20 7363 696b 6974 2d6c 6561 726e 2066  l scikit-learn f
-00000d30: 6173 7463 6c75 7374 6572 2073 6361 6e70  astcluster scanp
-00000d40: 7920 7079 7961 6d6c 0d0a 636f 6e64 6120  y pyyaml..conda 
-00000d50: 6163 7469 7661 7465 2063 6e6d 6673 6e73  activate cnmfsns
-00000d60: 0d0a 7069 7020 696e 7374 616c 6c20 6769  ..pip install gi
-00000d70: 742b 6874 7470 733a 2f2f 6769 7468 7562  t+https://github
-00000d80: 2e63 6f6d 2f4d 6f72 7269 7373 794c 6162  .com/MorrissyLab
-00000d90: 2f63 4e4d 462d 534e 532e 6769 740d 0a60  /cNMF-SNS.git..`
-00000da0: 6060 0d0a 0d0a 2323 20f0 9f93 9620 446f  ``....## .... Do
-00000db0: 6375 6d65 6e74 6174 696f 6e0d 0a0d 0a23  cumentation....#
-00000dc0: 2323 20f0 9f97 9020 4461 7461 2067 7569  ## .... Data gui
-00000dd0: 6465 6c69 6e65 730d 0a0d 0a63 4e4d 462d  delines....cNMF-
-00000de0: 534e 5320 6361 6e20 6661 6374 6f72 697a  SNS can factoriz
-00000df0: 6520 6120 7769 6465 2076 6172 6965 7479  e a wide variety
-00000e00: 206f 6620 6461 7461 7365 7473 2c20 6275   of datasets, bu
-00000e10: 7420 7769 6c6c 2077 6f72 6b20 6f70 7469  t will work opti
-00000e20: 6d61 6c6c 7920 696e 2074 6865 7365 2063  mally in these c
-00000e30: 6f6e 6469 7469 6f6e 733a 0d0a 2020 2d20  onditions:..  - 
-00000e40: 5573 6520 756e 7472 616e 7366 6f72 6d65  Use untransforme
-00000e50: 6420 2872 6177 2920 6461 7461 2077 6865  d (raw) data whe
-00000e60: 7265 2070 6f73 7369 626c 652e 0d0a 2020  re possible...  
-00000e70: 2d20 466f 7220 7369 6e67 6c65 2d63 656c  - For single-cel
-00000e80: 6c20 6f72 2073 7061 7469 616c 2052 4e41  l or spatial RNA
-00000e90: 2d53 6571 2064 6174 612c 2075 7365 2066  -Seq data, use f
-00000ea0: 6561 7475 7265 2063 6f75 6e74 732c 206e  eature counts, n
-00000eb0: 6f74 2054 504d 206f 7220 6f74 6865 7220  ot TPM or other 
-00000ec0: 6c6f 672d 7472 616e 7366 6f72 6d65 6420  log-transformed 
-00000ed0: 7661 6c75 6573 2e0d 0a0d 0a0d 0a23 2323  values.......###
-00000ee0: 20f0 9f93 9320 5079 7468 6f6e 2069 6e74   .... Python int
-00000ef0: 6572 6661 6365 0d0a 0d0a 546f 2067 6574  erface....To get
-00000f00: 2073 7461 7274 6564 2c20 7361 6d70 6c65   started, sample
-00000f10: 2070 726f 7465 6f6d 6963 7320 6461 7461   proteomics data
-00000f20: 7365 7473 2061 6e64 2061 204a 7570 7974  sets and a Jupyt
-00000f30: 6572 206e 6f74 6562 6f6f 6b20 7475 746f  er notebook tuto
-00000f40: 7269 616c 2069 7320 6176 6169 6c61 626c  rial is availabl
-00000f50: 6520 5b68 6572 655d 282f 7475 746f 7269  e [here](/tutori
-00000f60: 616c 2f74 7574 6f72 6961 6c2e 6970 796e  al/tutorial.ipyn
-00000f70: 6229 2e0d 0a0d 0a44 6574 6169 6c65 6420  b).....Detailed 
-00000f80: 4150 4920 7265 6665 7265 6e63 6520 6361  API reference ca
-00000f90: 6e20 6265 2066 6f75 6e64 206f 6e20 5b52  n be found on [R
-00000fa0: 6561 6454 6865 446f 6373 5d28 6874 7470  eadTheDocs](http
-00000fb0: 733a 2f2f 636e 6d66 2d73 6e73 2e72 6561  s://cnmf-sns.rea
-00000fc0: 6474 6865 646f 6373 2e69 6f2f 292e 0d0a  dthedocs.io/)...
-00000fd0: 0d0a 0d0a 2323 2320 e28c a8ef b88f 2043  ....### ...... C
-00000fe0: 6f6d 6d61 6e64 206c 696e 6520 696e 7465  ommand line inte
-00000ff0: 7266 6163 650d 0a0d 0a53 6565 2074 6865  rface....See the
-00001000: 205b 636f 6d6d 616e 6420 6c69 6e65 2069   [command line i
-00001010: 6e74 6572 6661 6365 2064 6f63 756d 656e  nterface documen
-00001020: 7461 7469 6f6e 5d28 2f43 4c49 2e6d 6429  tation](/CLI.md)
-00001030: 2e0d 0a0d 0a23 2320 f09f 92ad 2047 6574  .....## .... Get
-00001040: 7469 6e67 2048 656c 700d 0a0d 0a46 6f72  ting Help....For
-00001050: 2065 7272 6f72 7320 6172 6973 696e 6720   errors arising 
-00001060: 6475 7269 6e67 2075 7365 206f 6620 634e  during use of cN
-00001070: 4d46 2d53 4e53 2c20 6372 6561 7465 2061  MF-SNS, create a
-00001080: 6e64 2062 726f 7773 6520 6973 7375 6573  nd browse issues
-00001090: 2069 6e20 7468 6520 5b47 6974 4875 6220   in the [GitHub 
-000010a0: 2269 7373 7565 7322 2074 6162 5d28 6874  "issues" tab](ht
-000010b0: 7470 733a 2f2f 6769 7468 7562 2e63 6f6d  tps://github.com
-000010c0: 2f4d 6f72 7269 7373 794c 6162 2f63 4e4d  /MorrissyLab/cNM
-000010d0: 462d 534e 532f 6973 7375 6573 292e 0d0a  F-SNS/issues)...
+00000000: 0d0a 215b 634e 4d46 2d53 4e53 206c 6f67  ..![cNMF-SNS log
+00000010: 6f5d 286c 6f67 6f2e 706e 6729 0d0a 0d0a  o](logo.png)....
+00000020: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
+00000030: 2d0d 0a0d 0a23 2063 4e4d 462d 534e 533a  -....# cNMF-SNS:
+00000040: 2070 6f77 6572 6675 6c20 6661 6374 6f72   powerful factor
+00000050: 697a 6174 696f 6e2d 6261 7365 6420 6d75  ization-based mu
+00000060: 6c74 692d 6f6d 6963 7320 696e 7465 6772  lti-omics integr
+00000070: 6174 696f 6e20 746f 6f6c 6b69 740d 0a0d  ation toolkit...
+00000080: 0a21 5b76 6572 7369 6f6e 2062 6164 6765  .![version badge
+00000090: 5d28 6874 7470 733a 2f2f 696d 672e 7368  ](https://img.sh
+000000a0: 6965 6c64 732e 696f 2f62 6164 6765 2f76  ields.io/badge/v
+000000b0: 6572 7369 6f6e 2d31 2e37 2e30 2d62 6c75  ersion-1.7.0-blu
+000000c0: 6529 0d0a 5b21 5b50 7950 4920 4c61 7465  e)..[![PyPI Late
+000000d0: 7374 2052 656c 6561 7365 5d28 6874 7470  st Release](http
+000000e0: 733a 2f2f 696d 672e 7368 6965 6c64 732e  s://img.shields.
+000000f0: 696f 2f70 7970 692f 762f 636e 6d66 736e  io/pypi/v/cnmfsn
+00000100: 732e 7376 6729 5d28 6874 7470 733a 2f2f  s.svg)](https://
+00000110: 7079 7069 2e6f 7267 2f70 726f 6a65 6374  pypi.org/project
+00000120: 2f63 6e6d 6673 6e73 2f29 0d0a 5b21 5b43  /cnmfsns/)..[![C
+00000130: 6f6e 6461 204c 6174 6573 7420 5265 6c65  onda Latest Rele
+00000140: 6173 655d 2868 7474 7073 3a2f 2f69 6d67  ase](https://img
+00000150: 2e73 6869 656c 6473 2e69 6f2f 636f 6e64  .shields.io/cond
+00000160: 612f 766e 2f63 6f6e 6461 2d66 6f72 6765  a/vn/conda-forge
+00000170: 2f63 6e6d 6673 6e73 295d 2868 7474 7073  /cnmfsns)](https
+00000180: 3a2f 2f61 6e61 636f 6e64 612e 6f72 672f  ://anaconda.org/
+00000190: 636f 6e64 612d 666f 7267 652f 636e 6d66  conda-forge/cnmf
+000001a0: 736e 732f 290d 0a5b 215b 446f 6375 6d65  sns/)..[![Docume
+000001b0: 6e74 6174 696f 6e20 7374 6174 7573 5d28  ntation status](
+000001c0: 6874 7470 733a 2f2f 7265 6164 7468 6564  https://readthed
+000001d0: 6f63 732e 6f72 672f 7072 6f6a 6563 7473  ocs.org/projects
+000001e0: 2f63 6e6d 662d 736e 732f 6261 6467 652f  /cnmf-sns/badge/
+000001f0: 3f76 6572 7369 6f6e 3d6c 6174 6573 7426  ?version=latest&
+00000200: 7374 796c 653d 666c 6174 295d 2829 0d0a  style=flat)]()..
+00000210: 5b21 5b44 6f77 6e6c 6f61 6473 5d28 6874  [![Downloads](ht
+00000220: 7470 733a 2f2f 7374 6174 6963 2e70 6570  tps://static.pep
+00000230: 792e 7465 6368 2f62 6164 6765 2f63 6e6d  y.tech/badge/cnm
+00000240: 6673 6e73 295d 2868 7474 7073 3a2f 2f70  fsns)](https://p
+00000250: 6570 792e 7465 6368 2f70 726f 6a65 6374  epy.tech/project
+00000260: 2f63 6e6d 6673 6e73 290d 0a5b 215b 4c69  /cnmfsns)..[![Li
+00000270: 6365 6e73 655d 2868 7474 7073 3a2f 2f69  cense](https://i
+00000280: 6d67 2e73 6869 656c 6473 2e69 6f2f 7079  mg.shields.io/py
+00000290: 7069 2f6c 2f63 6e6d 6673 6e73 2e73 7667  pi/l/cnmfsns.svg
+000002a0: 295d 2868 7474 7073 3a2f 2f67 6974 6875  )](https://githu
+000002b0: 622e 636f 6d2f 4d6f 7272 6973 7379 4c61  b.com/MorrissyLa
+000002c0: 622f 634e 4d46 2d53 4e53 2f62 6c6f 622f  b/cNMF-SNS/blob/
+000002d0: 6d61 696e 2f4c 4943 454e 5345 290d 0a0d  main/LICENSE)...
+000002e0: 0a41 7574 686f 7273 3a20 5b54 6564 2056  .Authors: [Ted V
+000002f0: 6572 6865 795d 2868 7474 7073 3a2f 2f67  erhey](https://g
+00000300: 6974 6875 622e 636f 6d2f 7665 7268 6579  ithub.com/verhey
+00000310: 7462 292c 205b 4865 6577 6f6e 2053 656f  tb), [Heewon Seo
+00000320: 5d28 6874 7470 733a 2f2f 6769 7468 7562  ](https://github
+00000330: 2e63 6f6d 2f6c 6f6f 7470 697a 292c 205b  .com/lootpiz), [
+00000340: 536f 7261 6e61 204d 6f72 7269 7373 795d  Sorana Morrissy]
+00000350: 2868 7474 7073 3a2f 2f67 6974 6875 622e  (https://github.
+00000360: 636f 6d2f 616e 6361 736f 7261 6e61 290d  com/ancasorana).
+00000370: 0a0d 0a2a 2a63 4e4d 462d 534e 532a 2a20  ...**cNMF-SNS** 
+00000380: 2863 6f6e 7365 6e73 7573 204e 6f6e 2d6e  (consensus Non-n
+00000390: 6567 6174 6976 6520 4d61 7472 6978 2046  egative Matrix F
+000003a0: 6163 746f 7269 7a61 7469 6f6e 2053 6f6c  actorization Sol
+000003b0: 7574 696f 6e20 4e65 7477 6f72 6b20 5370  ution Network Sp
+000003c0: 6163 6529 2069 7320 6120 5079 7468 6f6e  ace) is a Python
+000003d0: 2070 6163 6b61 6765 2065 6e61 626c 696e   package enablin
+000003e0: 6720 6d6f 7361 6963 2069 6e74 6567 7261  g mosaic integra
+000003f0: 7469 6f6e 206f 6620 6275 6c6b 2c20 7369  tion of bulk, si
+00000400: 6e67 6c65 2d63 656c 6c2c 2061 6e64 0d0a  ngle-cell, and..
+00000410: 7370 6174 6961 6c20 6578 7072 6573 7369  spatial expressi
+00000420: 6f6e 2064 6174 6120 6265 7477 6565 6e20  on data between 
+00000430: 616e 6420 7769 7468 696e 2064 6174 6173  and within datas
+00000440: 6574 732e 2044 6174 6173 6574 7320 6361  ets. Datasets ca
+00000450: 6e20 6861 7665 2070 6172 7469 616c 6c79  n have partially
+00000460: 206f 7665 726c 6170 7069 6e67 2066 6561   overlapping fea
+00000470: 7475 7265 7320 2865 672e 2067 656e 6573  tures (eg. genes
+00000480: 2920 6173 2077 656c 6c20 6173 206e 6f6e  ) as well as non
+00000490: 2d6f 7665 726c 6170 7069 6e67 2066 6561  -overlapping fea
+000004a0: 7475 7265 732e 2063 4e4d 4620 7072 6f76  tures. cNMF prov
+000004b0: 6964 6573 2061 202a 2a72 6f62 7573 742c  ides a **robust,
+000004c0: 200d 0a75 6e73 7570 6572 7669 7365 642a   ..unsupervised*
+000004d0: 2a20 6465 636f 6e76 6f6c 7574 696f 6e20  * deconvolution 
+000004e0: 6f66 2065 6163 6820 6461 7461 7365 7420  of each dataset 
+000004f0: 696e 746f 2067 656e 6520 6578 7072 6573  into gene expres
+00000500: 7369 6f6e 2070 726f 6772 616d 7320 2847  sion programs (G
+00000510: 4550 7329 2e0d 0a2a 2a4e 6574 776f 726b  EPs)...**Network
+00000520: 2d62 6173 6564 2069 6e74 6567 7261 7469  -based integrati
+00000530: 6f6e 2a2a 206f 6620 4745 5073 2065 6e61  on** of GEPs ena
+00000540: 626c 6573 2066 6c65 7869 626c 6520 696e  bles flexible in
+00000550: 7465 6772 6174 696f 6e20 6f66 206d 616e  tegration of man
+00000560: 7920 6461 7461 7365 7473 0d0a 6163 726f  y datasets..acro
+00000570: 7373 2061 7373 6179 7320 2865 672e 2050  ss assays (eg. P
+00000580: 726f 7465 696e 2c20 524e 412d 5365 712c  rotein, RNA-Seq,
+00000590: 2073 6352 4e41 2d53 6571 2c20 7370 6174   scRNA-Seq, spat
+000005a0: 6961 6c20 6578 7072 6573 7369 6f6e 2920  ial expression) 
+000005b0: 616e 6420 7061 7469 656e 7420 636f 686f  and patient coho
+000005c0: 7274 732e 0d0a 0d0a 436f 6d6d 756e 6974  rts.....Communit
+000005d0: 6965 7320 7769 7468 2047 4550 7320 6672  ies with GEPs fr
+000005e0: 6f6d 206d 756c 7469 706c 6520 6461 7461  om multiple data
+000005f0: 7365 7473 2063 616e 2062 6520 616e 6e6f  sets can be anno
+00000600: 7461 7465 6420 7769 7468 2064 6174 6173  tated with datas
+00000610: 6574 2d73 7065 6369 6669 630d 0a61 6e6e  et-specific..ann
+00000620: 6f74 6174 696f 6e73 2074 6f20 6661 6369  otations to faci
+00000630: 6c69 7461 7465 2069 6e74 6572 7072 6574  litate interpret
+00000640: 6174 696f 6e2e 0d0a 0d0a 2323 20e2 9aa1  ation.....## ...
+00000650: 4d61 696e 2046 6561 7475 7265 730d 0a0d  Main Features...
+00000660: 0a48 6572 6520 6172 6520 6a75 7374 2061  .Here are just a
+00000670: 2066 6577 206f 6620 7468 6520 7468 696e   few of the thin
+00000680: 6773 2074 6861 7420 634e 4d46 2d53 4e53  gs that cNMF-SNS
+00000690: 2064 6f65 7320 7765 6c6c 3a0d 0a0d 0a2d   does well:....-
+000006a0: 2049 6465 6e74 6966 6965 7320 696e 7465   Identifies inte
+000006b0: 7270 7265 7461 626c 652c 206e 6f6e 2d6e  rpretable, non-n
+000006c0: 6567 6174 6976 6520 7072 6f67 7261 6d73  egative programs
+000006d0: 2061 7420 6d75 6c74 6970 6c65 2072 6573   at multiple res
+000006e0: 6f6c 7574 696f 6e73 0d0a 2d20 4d6f 7361  olutions..- Mosa
+000006f0: 6963 2069 6e74 6567 7261 7469 6f6e 2064  ic integration d
+00000700: 6f65 7320 6e6f 7420 7265 7175 6972 6520  oes not require 
+00000710: 7375 6273 6574 7469 6e67 2066 6561 7475  subsetting featu
+00000720: 7265 732f 6765 6e65 7320 746f 0d0a 2020  res/genes to..  
+00000730: 6120 7368 6172 6564 206f 7220 6f76 6572  a shared or over
+00000740: 6469 7370 6572 7365 6420 7375 6273 6574  dispersed subset
+00000750: 0d0a 2d20 4964 6561 6c20 666f 7220 696e  ..- Ideal for in
+00000760: 6372 656d 656e 7461 6c20 696e 7465 6772  cremental integr
+00000770: 6174 696f 6e20 2861 6464 696e 6720 6461  ation (adding da
+00000780: 7461 7365 7473 206f 6e65 2061 7420 6120  tasets one at a 
+00000790: 7469 6d65 2920 7369 6e63 650d 0a20 2064  time) since..  d
+000007a0: 6563 6f6e 766f 6c75 7469 6f6e 2069 7320  econvolution is 
+000007b0: 7065 7266 6f72 6d65 6420 696e 6465 7065  performed indepe
+000007c0: 6e64 656e 746c 7920 6f6e 2065 6163 6820  ndently on each 
+000007d0: 6461 7461 7365 740d 0a2d 2049 6e74 6567  dataset..- Integ
+000007e0: 7261 7469 6f6e 2070 6572 666f 726d 7320  ration performs 
+000007f0: 7765 6c6c 2065 7665 6e20 7768 656e 2074  well even when t
+00000800: 6865 2064 6174 6173 6574 7320 6861 7665  he datasets have
+00000810: 206d 6973 6d61 7463 6865 6420 6665 6174   mismatched feat
+00000820: 7572 6573 0d0a 2020 2865 672e 204d 6963  ures..  (eg. Mic
+00000830: 726f 6172 7261 792c 2052 4e41 2d53 6571  roarray, RNA-Seq
+00000840: 2c20 5072 6f74 656f 6d69 6373 2920 6f72  , Proteomics) or
+00000850: 2073 7061 7273 6974 7920 2865 6720 7369   sparsity (eg si
+00000860: 6e67 6c65 2d63 656c 6c20 7673 2062 756c  ngle-cell vs bul
+00000870: 6b20 524e 412d 5365 7120 616e 6420 4154  k RNA-Seq and AT
+00000880: 4143 2d53 6571 290d 0a2d 2054 776f 2069  AC-Seq)..- Two i
+00000890: 6e74 6572 6661 6365 733a 2063 6f6d 6d61  nterfaces: comma
+000008a0: 6e64 2d6c 696e 6520 696e 7465 7266 6163  nd-line interfac
+000008b0: 6520 666f 7220 7261 7069 6420 6461 7461  e for rapid data
+000008c0: 2065 7870 6c6f 7261 7469 6f6e 2061 6e64   exploration and
+000008d0: 2070 7974 686f 6e0d 0a20 2069 6e74 6572   python..  inter
+000008e0: 6661 6365 2066 6f72 2065 7874 656e 7369  face for extensi
+000008f0: 6269 6c69 7479 2061 6e64 2066 6c65 7869  bility and flexi
+00000900: 6269 6c69 7479 0d0a 0d0a 2323 20f0 9f94  bility....## ...
+00000910: a720 496e 7374 616c 6c0d 0a0d 0a23 2323  . Install....###
+00000920: 20e2 9881 efb8 8f20 5075 626c 6963 2052   ...... Public R
+00000930: 656c 6561 7365 0d0a 0d0a 496e 7374 616c  elease....Instal
+00000940: 6c20 7468 6520 7061 636b 6167 6520 7769  l the package wi
+00000950: 7468 2063 6f6e 6461 2028 696e 2061 6e20  th conda (in an 
+00000960: 6973 6f6c 6174 6564 2063 6f6e 6461 2065  isolated conda e
+00000970: 6e76 6972 6f6e 6d65 6e74 290d 0a60 6060  nvironment)..```
+00000980: 6261 7368 0d0a 636f 6e64 6120 6372 6561  bash..conda crea
+00000990: 7465 202d 6e20 636e 6d66 736e 7320 2d63  te -n cnmfsns -c
+000009a0: 2063 6f6e 6461 2d66 6f72 6765 2063 6e6d   conda-forge cnm
+000009b0: 6673 6e73 0d0a 636f 6e64 6120 6163 7469  fsns..conda acti
+000009c0: 7661 7465 2063 6e6d 6673 6e73 0d0a 0d0a  vate cnmfsns....
+000009d0: 6060 600d 0a0d 0a23 2323 20e2 9ca8 204c  ```....### ... L
+000009e0: 6174 6573 7420 7665 7273 696f 6e20 6672  atest version fr
+000009f0: 6f6d 2047 6974 4875 620d 0a0d 0a42 6566  om GitHub....Bef
+00000a00: 6f72 6520 696e 7374 616c 6c69 6e67 2063  ore installing c
+00000a10: 4e4d 462d 534e 5320 7573 696e 6720 7069  NMF-SNS using pi
+00000a20: 702c 2069 7420 6973 2072 6563 6f6d 6d65  p, it is recomme
+00000a30: 6e64 6564 2074 6f20 6669 7273 7420 7365  nded to first se
+00000a40: 7420 7570 2061 2073 6570 6172 6174 6520  t up a separate 
+00000a50: 636f 6e64 6120 656e 7669 726f 6e6d 656e  conda environmen
+00000a60: 7420 616e 6420 6861 7665 2063 6f6e 6461  t and have conda
+00000a70: 206d 616e 6167 6520 6173 206d 616e 7920   manage as many 
+00000a80: 6465 7065 6e64 656e 6369 6573 2061 7320  dependencies as 
+00000a90: 706f 7373 6962 6c65 2e0d 0a0d 0a60 6060  possible.....```
+00000aa0: 6261 7368 0d0a 636f 6e64 6120 6372 6561  bash..conda crea
+00000ab0: 7465 202d 2d6e 616d 6520 636e 6d66 736e  te --name cnmfsn
+00000ac0: 7320 2d63 2063 6f6e 6461 2d66 6f72 6765  s -c conda-forge
+00000ad0: 2070 7974 686f 6e3d 332e 3130 2061 6e6e   python=3.10 ann
+00000ae0: 6461 7461 2070 616e 6461 7320 6e75 6d70  data pandas nump
+00000af0: 7920 7363 6970 7920 6d61 7470 6c6f 746c  y scipy matplotl
+00000b00: 6962 2075 7073 6574 706c 6f74 2068 7474  ib upsetplot htt
+00000b10: 706c 6962 3220 746f 6d6c 6920 746f 6d6c  plib2 tomli toml
+00000b20: 692d 7720 636c 6963 6b20 7079 6772 6170  i-w click pygrap
+00000b30: 6876 697a 2070 7974 686f 6e2d 6967 7261  hviz python-igra
+00000b40: 7068 2070 7979 616d 6c20 7363 696b 6974  ph pyyaml scikit
+00000b50: 2d6c 6561 726e 2066 6173 7463 6c75 7374  -learn fastclust
+00000b60: 6572 2073 6361 6e70 7920 7079 7961 6d6c  er scanpy pyyaml
+00000b70: 0d0a 636f 6e64 6120 6163 7469 7661 7465  ..conda activate
+00000b80: 2063 6e6d 6673 6e73 0d0a 7069 7020 696e   cnmfsns..pip in
+00000b90: 7374 616c 6c20 6769 742b 6874 7470 733a  stall git+https:
+00000ba0: 2f2f 6769 7468 7562 2e63 6f6d 2f4d 6f72  //github.com/Mor
+00000bb0: 7269 7373 794c 6162 2f63 4e4d 462d 534e  rissyLab/cNMF-SN
+00000bc0: 532e 6769 740d 0a60 6060 0d0a 0d0a 2323  S.git..```....##
+00000bd0: 20f0 9f93 9620 446f 6375 6d65 6e74 6174   .... Documentat
+00000be0: 696f 6e0d 0a0d 0a23 2323 20f0 9f97 9020  ion....### .... 
+00000bf0: 4461 7461 2067 7569 6465 6c69 6e65 730d  Data guidelines.
+00000c00: 0a0d 0a63 4e4d 462d 534e 5320 6361 6e20  ...cNMF-SNS can 
+00000c10: 6661 6374 6f72 697a 6520 6120 7769 6465  factorize a wide
+00000c20: 2076 6172 6965 7479 206f 6620 6461 7461   variety of data
+00000c30: 7365 7473 2c20 6275 7420 7769 6c6c 2077  sets, but will w
+00000c40: 6f72 6b20 6f70 7469 6d61 6c6c 7920 696e  ork optimally in
+00000c50: 2074 6865 7365 2063 6f6e 6469 7469 6f6e   these condition
+00000c60: 733a 0d0a 2020 2d20 5573 6520 756e 7472  s:..  - Use untr
+00000c70: 616e 7366 6f72 6d65 6420 2872 6177 2920  ansformed (raw) 
+00000c80: 6461 7461 2077 6865 7265 2070 6f73 7369  data where possi
+00000c90: 626c 652e 0d0a 2020 2d20 466f 7220 7369  ble...  - For si
+00000ca0: 6e67 6c65 2d63 656c 6c20 6f72 2073 7061  ngle-cell or spa
+00000cb0: 7469 616c 2052 4e41 2d53 6571 2064 6174  tial RNA-Seq dat
+00000cc0: 612c 2075 7365 2066 6561 7475 7265 2063  a, use feature c
+00000cd0: 6f75 6e74 732c 206e 6f74 2054 504d 206f  ounts, not TPM o
+00000ce0: 7220 6f74 6865 7220 6c6f 672d 7472 616e  r other log-tran
+00000cf0: 7366 6f72 6d65 6420 7661 6c75 6573 2e0d  sformed values..
+00000d00: 0a0d 0a0d 0a23 2323 20f0 9f93 9320 5079  .....### .... Py
+00000d10: 7468 6f6e 2069 6e74 6572 6661 6365 0d0a  thon interface..
+00000d20: 0d0a 546f 2067 6574 2073 7461 7274 6564  ..To get started
+00000d30: 2c20 7361 6d70 6c65 2070 726f 7465 6f6d  , sample proteom
+00000d40: 6963 7320 6461 7461 7365 7473 2061 6e64  ics datasets and
+00000d50: 2061 204a 7570 7974 6572 206e 6f74 6562   a Jupyter noteb
+00000d60: 6f6f 6b20 7475 746f 7269 616c 2069 7320  ook tutorial is 
+00000d70: 6176 6169 6c61 626c 6520 5b68 6572 655d  available [here]
+00000d80: 282f 7475 746f 7269 616c 2f74 7574 6f72  (/tutorial/tutor
+00000d90: 6961 6c2e 6970 796e 6229 2e0d 0a0d 0a44  ial.ipynb).....D
+00000da0: 6574 6169 6c65 6420 4150 4920 7265 6665  etailed API refe
+00000db0: 7265 6e63 6520 6361 6e20 6265 2066 6f75  rence can be fou
+00000dc0: 6e64 206f 6e20 5b52 6561 6454 6865 446f  nd on [ReadTheDo
+00000dd0: 6373 5d28 6874 7470 733a 2f2f 636e 6d66  cs](https://cnmf
+00000de0: 2d73 6e73 2e72 6561 6474 6865 646f 6373  -sns.readthedocs
+00000df0: 2e69 6f2f 292e 0d0a 0d0a 0d0a 2323 2320  .io/).......### 
+00000e00: e28c a8ef b88f 2043 6f6d 6d61 6e64 206c  ...... Command l
+00000e10: 696e 6520 696e 7465 7266 6163 650d 0a0d  ine interface...
+00000e20: 0a53 6565 2074 6865 205b 636f 6d6d 616e  .See the [comman
+00000e30: 6420 6c69 6e65 2069 6e74 6572 6661 6365  d line interface
+00000e40: 2064 6f63 756d 656e 7461 7469 6f6e 5d28   documentation](
+00000e50: 2f43 4c49 2e6d 6429 2e0d 0a0d 0a23 2320  /CLI.md).....## 
+00000e60: f09f 92ad 2047 6574 7469 6e67 2048 656c  .... Getting Hel
+00000e70: 700d 0a0d 0a46 6f72 2065 7272 6f72 7320  p....For errors 
+00000e80: 6172 6973 696e 6720 6475 7269 6e67 2075  arising during u
+00000e90: 7365 206f 6620 634e 4d46 2d53 4e53 2c20  se of cNMF-SNS, 
+00000ea0: 6372 6561 7465 2061 6e64 2062 726f 7773  create and brows
+00000eb0: 6520 6973 7375 6573 2069 6e20 7468 6520  e issues in the 
+00000ec0: 5b47 6974 4875 6220 2269 7373 7565 7322  [GitHub "issues"
+00000ed0: 2074 6162 5d28 6874 7470 733a 2f2f 6769   tab](https://gi
+00000ee0: 7468 7562 2e63 6f6d 2f4d 6f72 7269 7373  thub.com/Morriss
+00000ef0: 794c 6162 2f63 4e4d 462d 534e 532f 6973  yLab/cNMF-SNS/is
+00000f00: 7375 6573 292e                           sues).
```

### Comparing `cnmfsns-1.6.5/README.md` & `cnmfsns-1.7.0/PKG-INFO`

 * *Files 20% similar despite different names*

```diff
@@ -1,238 +1,273 @@
-00000000: 0d0a 215b 634e 4d46 2d53 4e53 206c 6f67  ..![cNMF-SNS log
-00000010: 6f5d 286c 6f67 6f2e 706e 6729 0d0a 0d0a  o](logo.png)....
-00000020: 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  ----------------
-00000030: 2d0d 0a0d 0a23 2063 4e4d 462d 534e 533a  -....# cNMF-SNS:
-00000040: 2070 6f77 6572 6675 6c20 6661 6374 6f72   powerful factor
-00000050: 697a 6174 696f 6e2d 6261 7365 6420 6d75  ization-based mu
-00000060: 6c74 692d 6f6d 6963 7320 696e 7465 6772  lti-omics integr
-00000070: 6174 696f 6e20 746f 6f6c 6b69 740d 0a0d  ation toolkit...
-00000080: 0a21 5b76 6572 7369 6f6e 2062 6164 6765  .![version badge
-00000090: 5d28 6874 7470 733a 2f2f 696d 672e 7368  ](https://img.sh
-000000a0: 6965 6c64 732e 696f 2f62 6164 6765 2f76  ields.io/badge/v
-000000b0: 6572 7369 6f6e 2d31 2e36 2e35 2d62 6c75  ersion-1.6.5-blu
-000000c0: 6529 0d0a 5b21 5b50 7950 4920 4c61 7465  e)..[![PyPI Late
-000000d0: 7374 2052 656c 6561 7365 5d28 6874 7470  st Release](http
-000000e0: 733a 2f2f 696d 672e 7368 6965 6c64 732e  s://img.shields.
-000000f0: 696f 2f70 7970 692f 762f 636e 6d66 736e  io/pypi/v/cnmfsn
-00000100: 732e 7376 6729 5d28 6874 7470 733a 2f2f  s.svg)](https://
-00000110: 7079 7069 2e6f 7267 2f70 726f 6a65 6374  pypi.org/project
-00000120: 2f63 6e6d 6673 6e73 2f29 0d0a 5b21 5b43  /cnmfsns/)..[![C
-00000130: 6f6e 6461 204c 6174 6573 7420 5265 6c65  onda Latest Rele
-00000140: 6173 655d 2868 7474 7073 3a2f 2f69 6d67  ase](https://img
-00000150: 2e73 6869 656c 6473 2e69 6f2f 636f 6e64  .shields.io/cond
-00000160: 612f 766e 2f63 6f6e 6461 2d66 6f72 6765  a/vn/conda-forge
-00000170: 2f63 6e6d 6673 6e73 295d 2868 7474 7073  /cnmfsns)](https
-00000180: 3a2f 2f61 6e61 636f 6e64 612e 6f72 672f  ://anaconda.org/
-00000190: 636f 6e64 612d 666f 7267 652f 636e 6d66  conda-forge/cnmf
-000001a0: 736e 732f 290d 0a5b 215b 446f 6375 6d65  sns/)..[![Docume
-000001b0: 6e74 6174 696f 6e20 7374 6174 7573 5d28  ntation status](
-000001c0: 6874 7470 733a 2f2f 7265 6164 7468 6564  https://readthed
-000001d0: 6f63 732e 6f72 672f 7072 6f6a 6563 7473  ocs.org/projects
-000001e0: 2f63 6e6d 662d 736e 732f 6261 6467 652f  /cnmf-sns/badge/
-000001f0: 3f76 6572 7369 6f6e 3d6c 6174 6573 7426  ?version=latest&
-00000200: 7374 796c 653d 666c 6174 295d 2829 0d0a  style=flat)]()..
-00000210: 5b21 5b44 6f77 6e6c 6f61 6473 5d28 6874  [![Downloads](ht
-00000220: 7470 733a 2f2f 7374 6174 6963 2e70 6570  tps://static.pep
-00000230: 792e 7465 6368 2f62 6164 6765 2f63 6e6d  y.tech/badge/cnm
-00000240: 6673 6e73 295d 2868 7474 7073 3a2f 2f70  fsns)](https://p
-00000250: 6570 792e 7465 6368 2f70 726f 6a65 6374  epy.tech/project
-00000260: 2f63 6e6d 6673 6e73 290d 0a5b 215b 4c69  /cnmfsns)..[![Li
-00000270: 6365 6e73 655d 2868 7474 7073 3a2f 2f69  cense](https://i
-00000280: 6d67 2e73 6869 656c 6473 2e69 6f2f 7079  mg.shields.io/py
-00000290: 7069 2f6c 2f63 6e6d 6673 6e73 2e73 7667  pi/l/cnmfsns.svg
-000002a0: 295d 2868 7474 7073 3a2f 2f67 6974 6875  )](https://githu
-000002b0: 622e 636f 6d2f 4d6f 7272 6973 7379 4c61  b.com/MorrissyLa
-000002c0: 622f 634e 4d46 2d53 4e53 2f62 6c6f 622f  b/cNMF-SNS/blob/
-000002d0: 6d61 696e 2f4c 4943 454e 5345 290d 0a0d  main/LICENSE)...
-000002e0: 0a41 7574 686f 7273 3a20 5b54 6564 2056  .Authors: [Ted V
-000002f0: 6572 6865 795d 2868 7474 7073 3a2f 2f67  erhey](https://g
-00000300: 6974 6875 622e 636f 6d2f 7665 7268 6579  ithub.com/verhey
-00000310: 7462 292c 205b 4865 6577 6f6e 2053 656f  tb), [Heewon Seo
-00000320: 5d28 6874 7470 733a 2f2f 6769 7468 7562  ](https://github
-00000330: 2e63 6f6d 2f6c 6f6f 7470 697a 292c 205b  .com/lootpiz), [
-00000340: 536f 7261 6e61 204d 6f72 7269 7373 795d  Sorana Morrissy]
-00000350: 2868 7474 7073 3a2f 2f67 6974 6875 622e  (https://github.
-00000360: 636f 6d2f 616e 6361 736f 7261 6e61 290d  com/ancasorana).
-00000370: 0a0d 0a2a 2a63 4e4d 462d 534e 532a 2a20  ...**cNMF-SNS** 
-00000380: 2863 6f6e 7365 6e73 7573 204e 6f6e 2d6e  (consensus Non-n
-00000390: 6567 6174 6976 6520 4d61 7472 6978 2046  egative Matrix F
-000003a0: 6163 746f 7269 7a61 7469 6f6e 2053 6f6c  actorization Sol
-000003b0: 7574 696f 6e20 4e65 7477 6f72 6b20 5370  ution Network Sp
-000003c0: 6163 6529 2069 7320 6120 5079 7468 6f6e  ace) is a Python
-000003d0: 2070 6163 6b61 6765 2065 6e61 626c 696e   package enablin
-000003e0: 6720 6d6f 7361 6963 2069 6e74 6567 7261  g mosaic integra
-000003f0: 7469 6f6e 206f 6620 6275 6c6b 2c20 7369  tion of bulk, si
-00000400: 6e67 6c65 2d63 656c 6c2c 2061 6e64 0d0a  ngle-cell, and..
-00000410: 7370 6174 6961 6c20 6578 7072 6573 7369  spatial expressi
-00000420: 6f6e 2064 6174 6120 6265 7477 6565 6e20  on data between 
-00000430: 616e 6420 7769 7468 696e 2064 6174 6173  and within datas
-00000440: 6574 732e 2044 6174 6173 6574 7320 6361  ets. Datasets ca
-00000450: 6e20 6861 7665 2070 6172 7469 616c 6c79  n have partially
-00000460: 206f 7665 726c 6170 7069 6e67 2066 6561   overlapping fea
-00000470: 7475 7265 7320 2865 672e 2067 656e 6573  tures (eg. genes
-00000480: 2920 6173 2077 656c 6c20 6173 206e 6f6e  ) as well as non
-00000490: 2d6f 7665 726c 6170 7069 6e67 2066 6561  -overlapping fea
-000004a0: 7475 7265 732e 2063 4e4d 4620 7072 6f76  tures. cNMF prov
-000004b0: 6964 6573 2061 202a 2a72 6f62 7573 742c  ides a **robust,
-000004c0: 200d 0a75 6e73 7570 6572 7669 7365 642a   ..unsupervised*
-000004d0: 2a20 6465 636f 6e76 6f6c 7574 696f 6e20  * deconvolution 
-000004e0: 6f66 2065 6163 6820 6461 7461 7365 7420  of each dataset 
-000004f0: 696e 746f 2067 656e 6520 6578 7072 6573  into gene expres
-00000500: 7369 6f6e 2070 726f 6772 616d 7320 2847  sion programs (G
-00000510: 4550 7329 2e0d 0a2a 2a4e 6574 776f 726b  EPs)...**Network
-00000520: 2d62 6173 6564 2069 6e74 6567 7261 7469  -based integrati
-00000530: 6f6e 2a2a 206f 6620 4745 5073 2065 6e61  on** of GEPs ena
-00000540: 626c 6573 2066 6c65 7869 626c 6520 696e  bles flexible in
-00000550: 7465 6772 6174 696f 6e20 6f66 206d 616e  tegration of man
-00000560: 7920 6461 7461 7365 7473 0d0a 6163 726f  y datasets..acro
-00000570: 7373 2061 7373 6179 7320 2865 672e 2050  ss assays (eg. P
-00000580: 726f 7465 696e 2c20 524e 412d 5365 712c  rotein, RNA-Seq,
-00000590: 2073 6352 4e41 2d53 6571 2c20 7370 6174   scRNA-Seq, spat
-000005a0: 6961 6c20 6578 7072 6573 7369 6f6e 2920  ial expression) 
-000005b0: 616e 6420 7061 7469 656e 7420 636f 686f  and patient coho
-000005c0: 7274 732e 0d0a 0d0a 436f 6d6d 756e 6974  rts.....Communit
-000005d0: 6965 7320 7769 7468 2047 4550 7320 6672  ies with GEPs fr
-000005e0: 6f6d 206d 756c 7469 706c 6520 6461 7461  om multiple data
-000005f0: 7365 7473 2063 616e 2062 6520 616e 6e6f  sets can be anno
-00000600: 7461 7465 6420 7769 7468 2064 6174 6173  tated with datas
-00000610: 6574 2d73 7065 6369 6669 630d 0a61 6e6e  et-specific..ann
-00000620: 6f74 6174 696f 6e73 2074 6f20 6661 6369  otations to faci
-00000630: 6c69 7461 7465 2069 6e74 6572 7072 6574  litate interpret
-00000640: 6174 696f 6e2e 0d0a 0d0a 2323 20e2 9aa1  ation.....## ...
-00000650: 4d61 696e 2046 6561 7475 7265 730d 0a0d  Main Features...
-00000660: 0a48 6572 6520 6172 6520 6a75 7374 2061  .Here are just a
-00000670: 2066 6577 206f 6620 7468 6520 7468 696e   few of the thin
-00000680: 6773 2074 6861 7420 634e 4d46 2d53 4e53  gs that cNMF-SNS
-00000690: 2064 6f65 7320 7765 6c6c 3a0d 0a0d 0a2d   does well:....-
-000006a0: 2049 6465 6e74 6966 6965 7320 696e 7465   Identifies inte
-000006b0: 7270 7265 7461 626c 652c 2061 6464 6974  rpretable, addit
-000006c0: 6976 6520 6e6f 6e2d 6e65 6761 7469 7665  ive non-negative
-000006d0: 2067 656e 6520 6578 7072 6573 7369 6f6e   gene expression
-000006e0: 2070 726f 6772 616d 730d 0a2d 204d 6f73   programs..- Mos
-000006f0: 6169 6320 696e 7465 6772 6174 696f 6e20  aic integration 
-00000700: 646f 6573 206e 6f74 2072 6571 7569 7265  does not require
-00000710: 2073 7562 7365 7474 696e 6720 6665 6174   subsetting feat
-00000720: 7572 6573 2f67 656e 6573 2074 6f0d 0a20  ures/genes to.. 
-00000730: 2061 2073 6861 7265 6420 6f72 206f 7665   a shared or ove
-00000740: 7264 6973 7065 7273 6564 2073 7562 7365  rdispersed subse
-00000750: 740d 0a2d 2049 6465 616c 2066 6f72 2069  t..- Ideal for i
-00000760: 6e63 7265 6d65 6e74 616c 2069 6e74 6567  ncremental integ
-00000770: 7261 7469 6f6e 2028 6164 6469 6e67 2064  ration (adding d
-00000780: 6174 6173 6574 7320 6f6e 6520 6174 2061  atasets one at a
-00000790: 2074 696d 6529 2073 696e 6365 0d0a 2020   time) since..  
-000007a0: 6465 636f 6e76 6f6c 7574 696f 6e20 6973  deconvolution is
-000007b0: 2070 6572 666f 726d 6564 2069 6e64 6570   performed indep
-000007c0: 656e 6465 6e74 6c79 206f 6e20 6561 6368  endently on each
-000007d0: 2064 6174 6173 6574 2067 656e 6572 6174   dataset generat
-000007e0: 696e 6720 696e 7661 7269 616e 7420 4745  ing invariant GE
-000007f0: 5073 0d0a 2d20 496e 7465 6772 6174 696f  Ps..- Integratio
-00000800: 6e20 646f 6573 206e 6f74 2072 6571 7569  n does not requi
-00000810: 7265 2063 6f6d 7061 7261 626c 6520 6461  re comparable da
-00000820: 7461 2073 7061 7273 6974 7920 6f72 2073  ta sparsity or s
-00000830: 6571 7565 6e63 696e 6720 6465 7074 6820  equencing depth 
-00000840: 2873 696e 676c 652d 6365 6c6c 2c20 6d69  (single-cell, mi
-00000850: 6372 6f61 7272 6179 2c20 6275 6c6b 290d  croarray, bulk).
-00000860: 0a2d 2054 776f 2069 6e74 6572 6661 6365  .- Two interface
-00000870: 733a 2063 6f6d 6d61 6e64 2d6c 696e 6520  s: command-line 
-00000880: 696e 7465 7266 6163 6520 666f 7220 7261  interface for ra
-00000890: 7069 6420 6461 7461 2065 7870 6c6f 7261  pid data explora
-000008a0: 7469 6f6e 2061 6e64 2070 7974 686f 6e0d  tion and python.
-000008b0: 0a20 2069 6e74 6572 6661 6365 2066 6f72  .  interface for
-000008c0: 2065 7874 656e 7369 6269 6c69 7479 2061   extensibility a
-000008d0: 6e64 2066 6c65 7869 6269 6c69 7479 0d0a  nd flexibility..
-000008e0: 0d0a 2323 20f0 9f94 a720 496e 7374 616c  ..## .... Instal
-000008f0: 6c0d 0a0d 0a23 2323 20e2 9881 efb8 8f20  l....### ...... 
-00000900: 5075 626c 6963 2052 656c 6561 7365 0d0a  Public Release..
-00000910: 0d0a 496e 7374 616c 6c20 7468 6520 7061  ..Install the pa
-00000920: 636b 6167 6520 7769 7468 2063 6f6e 6461  ckage with conda
-00000930: 2028 696e 2061 6e20 6973 6f6c 6174 6564   (in an isolated
-00000940: 2063 6f6e 6461 2065 6e76 6972 6f6e 6d65   conda environme
-00000950: 6e74 290d 0a60 6060 6261 7368 0d0a 636f  nt)..```bash..co
-00000960: 6e64 6120 656e 7620 6372 6561 7465 2063  nda env create c
-00000970: 6e6d 6673 6e73 202d 6320 636f 6e64 612d  nmfsns -c conda-
-00000980: 666f 7267 6520 636e 6d66 736e 730d 0a63  forge cnmfsns..c
-00000990: 6f6e 6461 2061 6374 6976 6174 6520 636e  onda activate cn
-000009a0: 6d66 736e 730d 0a60 6060 0d0a 0d0a 2323  mfsns..```....##
-000009b0: 2320 e29c a820 4c61 7465 7374 2076 6572  # ... Latest ver
-000009c0: 7369 6f6e 2066 726f 6d20 4769 7448 7562  sion from GitHub
-000009d0: 0d0a 0d0a 4265 666f 7265 2069 6e73 7461  ....Before insta
-000009e0: 6c6c 696e 6720 634e 4d46 2d53 4e53 2075  lling cNMF-SNS u
-000009f0: 7369 6e67 2070 6970 2c20 6974 2069 7320  sing pip, it is 
-00000a00: 7265 636f 6d6d 656e 6465 6420 746f 2066  recommended to f
-00000a10: 6972 7374 2073 6574 2075 7020 6120 7365  irst set up a se
-00000a20: 7061 7261 7465 2063 6f6e 6461 2065 6e76  parate conda env
-00000a30: 6972 6f6e 6d65 6e74 2061 6e64 2068 6176  ironment and hav
-00000a40: 6520 636f 6e64 6120 6d61 6e61 6765 2061  e conda manage a
-00000a50: 7320 6d61 6e79 2064 6570 656e 6465 6e63  s many dependenc
-00000a60: 6965 7320 6173 2070 6f73 7369 626c 652e  ies as possible.
-00000a70: 0d0a 0d0a 6060 6062 6173 680d 0a63 6f6e  ....```bash..con
-00000a80: 6461 2063 7265 6174 6520 2d2d 6e61 6d65  da create --name
-00000a90: 2063 6e6d 6673 6e73 202d 6320 636f 6e64   cnmfsns -c cond
-00000aa0: 612d 666f 7267 6520 7079 7468 6f6e 3d33  a-forge python=3
-00000ab0: 2e31 3020 616e 6e64 6174 6120 7061 6e64  .10 anndata pand
-00000ac0: 6173 206e 756d 7079 2073 6369 7079 206d  as numpy scipy m
-00000ad0: 6174 706c 6f74 6c69 6220 7570 7365 7470  atplotlib upsetp
-00000ae0: 6c6f 7420 6874 7470 6c69 6232 2074 6f6d  lot httplib2 tom
-00000af0: 6c69 2074 6f6d 6c69 2d77 2063 6c69 636b  li tomli-w click
-00000b00: 2070 7967 7261 7068 7669 7a20 7079 7468   pygraphviz pyth
-00000b10: 6f6e 2d69 6772 6170 6820 7079 7961 6d6c  on-igraph pyyaml
-00000b20: 2073 6369 6b69 742d 6c65 6172 6e20 6661   scikit-learn fa
-00000b30: 7374 636c 7573 7465 7220 7363 616e 7079  stcluster scanpy
-00000b40: 2070 7979 616d 6c0d 0a63 6f6e 6461 2061   pyyaml..conda a
-00000b50: 6374 6976 6174 6520 636e 6d66 736e 730d  ctivate cnmfsns.
-00000b60: 0a70 6970 2069 6e73 7461 6c6c 2067 6974  .pip install git
-00000b70: 2b68 7474 7073 3a2f 2f67 6974 6875 622e  +https://github.
-00000b80: 636f 6d2f 4d6f 7272 6973 7379 4c61 622f  com/MorrissyLab/
-00000b90: 634e 4d46 2d53 4e53 2e67 6974 0d0a 6060  cNMF-SNS.git..``
-00000ba0: 600d 0a0d 0a23 2320 f09f 9396 2044 6f63  `....## .... Doc
-00000bb0: 756d 656e 7461 7469 6f6e 0d0a 0d0a 2323  umentation....##
-00000bc0: 2320 f09f 9790 2044 6174 6120 6775 6964  # .... Data guid
-00000bd0: 656c 696e 6573 0d0a 0d0a 634e 4d46 2d53  elines....cNMF-S
-00000be0: 4e53 2063 616e 2066 6163 746f 7269 7a65  NS can factorize
-00000bf0: 2061 2077 6964 6520 7661 7269 6574 7920   a wide variety 
-00000c00: 6f66 2064 6174 6173 6574 732c 2062 7574  of datasets, but
-00000c10: 2077 696c 6c20 776f 726b 206f 7074 696d   will work optim
-00000c20: 616c 6c79 2069 6e20 7468 6573 6520 636f  ally in these co
-00000c30: 6e64 6974 696f 6e73 3a0d 0a20 202d 2055  nditions:..  - U
-00000c40: 7365 2075 6e74 7261 6e73 666f 726d 6564  se untransformed
-00000c50: 2028 7261 7729 2064 6174 6120 7768 6572   (raw) data wher
-00000c60: 6520 706f 7373 6962 6c65 2e0d 0a20 202d  e possible...  -
-00000c70: 2046 6f72 2073 696e 676c 652d 6365 6c6c   For single-cell
-00000c80: 206f 7220 7370 6174 6961 6c20 524e 412d   or spatial RNA-
-00000c90: 5365 7120 6461 7461 2c20 7573 6520 6665  Seq data, use fe
-00000ca0: 6174 7572 6520 636f 756e 7473 2c20 6e6f  ature counts, no
-00000cb0: 7420 5450 4d20 6f72 206f 7468 6572 206c  t TPM or other l
-00000cc0: 6f67 2d74 7261 6e73 666f 726d 6564 2076  og-transformed v
-00000cd0: 616c 7565 732e 0d0a 0d0a 0d0a 2323 2320  alues.......### 
-00000ce0: f09f 9393 2050 7974 686f 6e20 696e 7465  .... Python inte
-00000cf0: 7266 6163 650d 0a0d 0a54 6f20 6765 7420  rface....To get 
-00000d00: 7374 6172 7465 642c 2073 616d 706c 6520  started, sample 
-00000d10: 7072 6f74 656f 6d69 6373 2064 6174 6173  proteomics datas
-00000d20: 6574 7320 616e 6420 6120 4a75 7079 7465  ets and a Jupyte
-00000d30: 7220 6e6f 7465 626f 6f6b 2074 7574 6f72  r notebook tutor
-00000d40: 6961 6c20 6973 2061 7661 696c 6162 6c65  ial is available
-00000d50: 205b 6865 7265 5d28 2f74 7574 6f72 6961   [here](/tutoria
-00000d60: 6c2f 7475 746f 7269 616c 2e69 7079 6e62  l/tutorial.ipynb
-00000d70: 292e 0d0a 0d0a 4465 7461 696c 6564 2041  ).....Detailed A
-00000d80: 5049 2072 6566 6572 656e 6365 2063 616e  PI reference can
-00000d90: 2062 6520 666f 756e 6420 6f6e 205b 5265   be found on [Re
-00000da0: 6164 5468 6544 6f63 735d 2868 7474 7073  adTheDocs](https
-00000db0: 3a2f 2f63 6e6d 662d 736e 732e 7265 6164  ://cnmf-sns.read
-00000dc0: 7468 6564 6f63 732e 696f 2f29 2e0d 0a0d  thedocs.io/)....
-00000dd0: 0a0d 0a23 2323 20e2 8ca8 efb8 8f20 436f  ...### ...... Co
-00000de0: 6d6d 616e 6420 6c69 6e65 2069 6e74 6572  mmand line inter
-00000df0: 6661 6365 0d0a 0d0a 5365 6520 7468 6520  face....See the 
-00000e00: 5b63 6f6d 6d61 6e64 206c 696e 6520 696e  [command line in
-00000e10: 7465 7266 6163 6520 646f 6375 6d65 6e74  terface document
-00000e20: 6174 696f 6e5d 282f 434c 492e 6d64 292e  ation](/CLI.md).
-00000e30: 0d0a 0d0a 2323 20f0 9f92 ad20 4765 7474  ....## .... Gett
-00000e40: 696e 6720 4865 6c70 0d0a 0d0a 466f 7220  ing Help....For 
-00000e50: 6572 726f 7273 2061 7269 7369 6e67 2064  errors arising d
-00000e60: 7572 696e 6720 7573 6520 6f66 2063 4e4d  uring use of cNM
-00000e70: 462d 534e 532c 2063 7265 6174 6520 616e  F-SNS, create an
-00000e80: 6420 6272 6f77 7365 2069 7373 7565 7320  d browse issues 
-00000e90: 696e 2074 6865 205b 4769 7448 7562 2022  in the [GitHub "
-00000ea0: 6973 7375 6573 2220 7461 625d 2868 7474  issues" tab](htt
-00000eb0: 7073 3a2f 2f67 6974 6875 622e 636f 6d2f  ps://github.com/
-00000ec0: 4d6f 7272 6973 7379 4c61 622f 634e 4d46  MorrissyLab/cNMF
-00000ed0: 2d53 4e53 2f69 7373 7565 7329 2e         -SNS/issues).
+00000000: 4d65 7461 6461 7461 2d56 6572 7369 6f6e  Metadata-Version
+00000010: 3a20 322e 310d 0a4e 616d 653a 2063 6e6d  : 2.1..Name: cnm
+00000020: 6673 6e73 0d0a 5665 7273 696f 6e3a 2031  fsns..Version: 1
+00000030: 2e37 2e30 0d0a 5375 6d6d 6172 793a 2063  .7.0..Summary: c
+00000040: 4e4d 4620 536f 6c75 7469 6f6e 204e 6574  NMF Solution Net
+00000050: 776f 726b 2053 7061 6365 0d0a 486f 6d65  work Space..Home
+00000060: 2d70 6167 653a 2068 7474 7073 3a2f 2f67  -page: https://g
+00000070: 6974 6875 622e 636f 6d2f 4d6f 7272 6973  ithub.com/Morris
+00000080: 7379 4c61 622f 634e 4d46 2d53 4e53 0d0a  syLab/cNMF-SNS..
+00000090: 4175 7468 6f72 3a20 5465 6420 5665 7268  Author: Ted Verh
+000000a0: 6579 0d0a 4175 7468 6f72 2d65 6d61 696c  ey..Author-email
+000000b0: 3a20 7462 7665 7268 6579 4075 6361 6c67  : tbverhey@ucalg
+000000c0: 6172 792e 6361 0d0a 5072 6f6a 6563 742d  ary.ca..Project-
+000000d0: 5552 4c3a 2042 7567 2054 7261 636b 6572  URL: Bug Tracker
+000000e0: 2c20 6874 7470 733a 2f2f 6769 7468 7562  , https://github
+000000f0: 2e63 6f6d 2f4d 6f72 7269 7373 794c 6162  .com/MorrissyLab
+00000100: 2f63 4e4d 462d 534e 532f 6973 7375 6573  /cNMF-SNS/issues
+00000110: 0d0a 436c 6173 7369 6669 6572 3a20 5072  ..Classifier: Pr
+00000120: 6f67 7261 6d6d 696e 6720 4c61 6e67 7561  ogramming Langua
+00000130: 6765 203a 3a20 5079 7468 6f6e 203a 3a20  ge :: Python :: 
+00000140: 330d 0a43 6c61 7373 6966 6965 723a 204c  3..Classifier: L
+00000150: 6963 656e 7365 203a 3a20 4f53 4920 4170  icense :: OSI Ap
+00000160: 7072 6f76 6564 203a 3a20 4d49 5420 4c69  proved :: MIT Li
+00000170: 6365 6e73 650d 0a43 6c61 7373 6966 6965  cense..Classifie
+00000180: 723a 204f 7065 7261 7469 6e67 2053 7973  r: Operating Sys
+00000190: 7465 6d20 3a3a 204f 5320 496e 6465 7065  tem :: OS Indepe
+000001a0: 6e64 656e 740d 0a52 6571 7569 7265 732d  ndent..Requires-
+000001b0: 5079 7468 6f6e 3a20 3e3d 332e 390d 0a44  Python: >=3.9..D
+000001c0: 6573 6372 6970 7469 6f6e 2d43 6f6e 7465  escription-Conte
+000001d0: 6e74 2d54 7970 653a 2074 6578 742f 6d61  nt-Type: text/ma
+000001e0: 726b 646f 776e 0d0a 4c69 6365 6e73 652d  rkdown..License-
+000001f0: 4669 6c65 3a20 4c49 4345 4e53 450d 0a0d  File: LICENSE...
+00000200: 0a0d 0a21 5b63 4e4d 462d 534e 5320 6c6f  ...![cNMF-SNS lo
+00000210: 676f 5d28 6c6f 676f 2e70 6e67 290d 0a0d  go](logo.png)...
+00000220: 0a2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d 2d2d  .---------------
+00000230: 2d2d 0d0a 0d0a 2320 634e 4d46 2d53 4e53  --....# cNMF-SNS
+00000240: 3a20 706f 7765 7266 756c 2066 6163 746f  : powerful facto
+00000250: 7269 7a61 7469 6f6e 2d62 6173 6564 206d  rization-based m
+00000260: 756c 7469 2d6f 6d69 6373 2069 6e74 6567  ulti-omics integ
+00000270: 7261 7469 6f6e 2074 6f6f 6c6b 6974 0d0a  ration toolkit..
+00000280: 0d0a 215b 7665 7273 696f 6e20 6261 6467  ..![version badg
+00000290: 655d 2868 7474 7073 3a2f 2f69 6d67 2e73  e](https://img.s
+000002a0: 6869 656c 6473 2e69 6f2f 6261 6467 652f  hields.io/badge/
+000002b0: 7665 7273 696f 6e2d 312e 372e 302d 626c  version-1.7.0-bl
+000002c0: 7565 290d 0a5b 215b 5079 5049 204c 6174  ue)..[![PyPI Lat
+000002d0: 6573 7420 5265 6c65 6173 655d 2868 7474  est Release](htt
+000002e0: 7073 3a2f 2f69 6d67 2e73 6869 656c 6473  ps://img.shields
+000002f0: 2e69 6f2f 7079 7069 2f76 2f63 6e6d 6673  .io/pypi/v/cnmfs
+00000300: 6e73 2e73 7667 295d 2868 7474 7073 3a2f  ns.svg)](https:/
+00000310: 2f70 7970 692e 6f72 672f 7072 6f6a 6563  /pypi.org/projec
+00000320: 742f 636e 6d66 736e 732f 290d 0a5b 215b  t/cnmfsns/)..[![
+00000330: 436f 6e64 6120 4c61 7465 7374 2052 656c  Conda Latest Rel
+00000340: 6561 7365 5d28 6874 7470 733a 2f2f 696d  ease](https://im
+00000350: 672e 7368 6965 6c64 732e 696f 2f63 6f6e  g.shields.io/con
+00000360: 6461 2f76 6e2f 636f 6e64 612d 666f 7267  da/vn/conda-forg
+00000370: 652f 636e 6d66 736e 7329 5d28 6874 7470  e/cnmfsns)](http
+00000380: 733a 2f2f 616e 6163 6f6e 6461 2e6f 7267  s://anaconda.org
+00000390: 2f63 6f6e 6461 2d66 6f72 6765 2f63 6e6d  /conda-forge/cnm
+000003a0: 6673 6e73 2f29 0d0a 5b21 5b44 6f63 756d  fsns/)..[![Docum
+000003b0: 656e 7461 7469 6f6e 2073 7461 7475 735d  entation status]
+000003c0: 2868 7474 7073 3a2f 2f72 6561 6474 6865  (https://readthe
+000003d0: 646f 6373 2e6f 7267 2f70 726f 6a65 6374  docs.org/project
+000003e0: 732f 636e 6d66 2d73 6e73 2f62 6164 6765  s/cnmf-sns/badge
+000003f0: 2f3f 7665 7273 696f 6e3d 6c61 7465 7374  /?version=latest
+00000400: 2673 7479 6c65 3d66 6c61 7429 5d28 290d  &style=flat)]().
+00000410: 0a5b 215b 446f 776e 6c6f 6164 735d 2868  .[![Downloads](h
+00000420: 7474 7073 3a2f 2f73 7461 7469 632e 7065  ttps://static.pe
+00000430: 7079 2e74 6563 682f 6261 6467 652f 636e  py.tech/badge/cn
+00000440: 6d66 736e 7329 5d28 6874 7470 733a 2f2f  mfsns)](https://
+00000450: 7065 7079 2e74 6563 682f 7072 6f6a 6563  pepy.tech/projec
+00000460: 742f 636e 6d66 736e 7329 0d0a 5b21 5b4c  t/cnmfsns)..[![L
+00000470: 6963 656e 7365 5d28 6874 7470 733a 2f2f  icense](https://
+00000480: 696d 672e 7368 6965 6c64 732e 696f 2f70  img.shields.io/p
+00000490: 7970 692f 6c2f 636e 6d66 736e 732e 7376  ypi/l/cnmfsns.sv
+000004a0: 6729 5d28 6874 7470 733a 2f2f 6769 7468  g)](https://gith
+000004b0: 7562 2e63 6f6d 2f4d 6f72 7269 7373 794c  ub.com/MorrissyL
+000004c0: 6162 2f63 4e4d 462d 534e 532f 626c 6f62  ab/cNMF-SNS/blob
+000004d0: 2f6d 6169 6e2f 4c49 4345 4e53 4529 0d0a  /main/LICENSE)..
+000004e0: 0d0a 4175 7468 6f72 733a 205b 5465 6420  ..Authors: [Ted 
+000004f0: 5665 7268 6579 5d28 6874 7470 733a 2f2f  Verhey](https://
+00000500: 6769 7468 7562 2e63 6f6d 2f76 6572 6865  github.com/verhe
+00000510: 7974 6229 2c20 5b48 6565 776f 6e20 5365  ytb), [Heewon Se
+00000520: 6f5d 2868 7474 7073 3a2f 2f67 6974 6875  o](https://githu
+00000530: 622e 636f 6d2f 6c6f 6f74 7069 7a29 2c20  b.com/lootpiz), 
+00000540: 5b53 6f72 616e 6120 4d6f 7272 6973 7379  [Sorana Morrissy
+00000550: 5d28 6874 7470 733a 2f2f 6769 7468 7562  ](https://github
+00000560: 2e63 6f6d 2f61 6e63 6173 6f72 616e 6129  .com/ancasorana)
+00000570: 0d0a 0d0a 2a2a 634e 4d46 2d53 4e53 2a2a  ....**cNMF-SNS**
+00000580: 2028 636f 6e73 656e 7375 7320 4e6f 6e2d   (consensus Non-
+00000590: 6e65 6761 7469 7665 204d 6174 7269 7820  negative Matrix 
+000005a0: 4661 6374 6f72 697a 6174 696f 6e20 536f  Factorization So
+000005b0: 6c75 7469 6f6e 204e 6574 776f 726b 2053  lution Network S
+000005c0: 7061 6365 2920 6973 2061 2050 7974 686f  pace) is a Pytho
+000005d0: 6e20 7061 636b 6167 6520 656e 6162 6c69  n package enabli
+000005e0: 6e67 206d 6f73 6169 6320 696e 7465 6772  ng mosaic integr
+000005f0: 6174 696f 6e20 6f66 2062 756c 6b2c 2073  ation of bulk, s
+00000600: 696e 676c 652d 6365 6c6c 2c20 616e 640d  ingle-cell, and.
+00000610: 0a73 7061 7469 616c 2065 7870 7265 7373  .spatial express
+00000620: 696f 6e20 6461 7461 2062 6574 7765 656e  ion data between
+00000630: 2061 6e64 2077 6974 6869 6e20 6461 7461   and within data
+00000640: 7365 7473 2e20 4461 7461 7365 7473 2063  sets. Datasets c
+00000650: 616e 2068 6176 6520 7061 7274 6961 6c6c  an have partiall
+00000660: 7920 6f76 6572 6c61 7070 696e 6720 6665  y overlapping fe
+00000670: 6174 7572 6573 2028 6567 2e20 6765 6e65  atures (eg. gene
+00000680: 7329 2061 7320 7765 6c6c 2061 7320 6e6f  s) as well as no
+00000690: 6e2d 6f76 6572 6c61 7070 696e 6720 6665  n-overlapping fe
+000006a0: 6174 7572 6573 2e20 634e 4d46 2070 726f  atures. cNMF pro
+000006b0: 7669 6465 7320 6120 2a2a 726f 6275 7374  vides a **robust
+000006c0: 2c20 0d0a 756e 7375 7065 7276 6973 6564  , ..unsupervised
+000006d0: 2a2a 2064 6563 6f6e 766f 6c75 7469 6f6e  ** deconvolution
+000006e0: 206f 6620 6561 6368 2064 6174 6173 6574   of each dataset
+000006f0: 2069 6e74 6f20 6765 6e65 2065 7870 7265   into gene expre
+00000700: 7373 696f 6e20 7072 6f67 7261 6d73 2028  ssion programs (
+00000710: 4745 5073 292e 0d0a 2a2a 4e65 7477 6f72  GEPs)...**Networ
+00000720: 6b2d 6261 7365 6420 696e 7465 6772 6174  k-based integrat
+00000730: 696f 6e2a 2a20 6f66 2047 4550 7320 656e  ion** of GEPs en
+00000740: 6162 6c65 7320 666c 6578 6962 6c65 2069  ables flexible i
+00000750: 6e74 6567 7261 7469 6f6e 206f 6620 6d61  ntegration of ma
+00000760: 6e79 2064 6174 6173 6574 730d 0a61 6372  ny datasets..acr
+00000770: 6f73 7320 6173 7361 7973 2028 6567 2e20  oss assays (eg. 
+00000780: 5072 6f74 6569 6e2c 2052 4e41 2d53 6571  Protein, RNA-Seq
+00000790: 2c20 7363 524e 412d 5365 712c 2073 7061  , scRNA-Seq, spa
+000007a0: 7469 616c 2065 7870 7265 7373 696f 6e29  tial expression)
+000007b0: 2061 6e64 2070 6174 6965 6e74 2063 6f68   and patient coh
+000007c0: 6f72 7473 2e0d 0a0d 0a43 6f6d 6d75 6e69  orts.....Communi
+000007d0: 7469 6573 2077 6974 6820 4745 5073 2066  ties with GEPs f
+000007e0: 726f 6d20 6d75 6c74 6970 6c65 2064 6174  rom multiple dat
+000007f0: 6173 6574 7320 6361 6e20 6265 2061 6e6e  asets can be ann
+00000800: 6f74 6174 6564 2077 6974 6820 6461 7461  otated with data
+00000810: 7365 742d 7370 6563 6966 6963 0d0a 616e  set-specific..an
+00000820: 6e6f 7461 7469 6f6e 7320 746f 2066 6163  notations to fac
+00000830: 696c 6974 6174 6520 696e 7465 7270 7265  ilitate interpre
+00000840: 7461 7469 6f6e 2e0d 0a0d 0a23 2320 e29a  tation.....## ..
+00000850: a14d 6169 6e20 4665 6174 7572 6573 0d0a  .Main Features..
+00000860: 0d0a 4865 7265 2061 7265 206a 7573 7420  ..Here are just 
+00000870: 6120 6665 7720 6f66 2074 6865 2074 6869  a few of the thi
+00000880: 6e67 7320 7468 6174 2063 4e4d 462d 534e  ngs that cNMF-SN
+00000890: 5320 646f 6573 2077 656c 6c3a 0d0a 0d0a  S does well:....
+000008a0: 2d20 4964 656e 7469 6669 6573 2069 6e74  - Identifies int
+000008b0: 6572 7072 6574 6162 6c65 2c20 6e6f 6e2d  erpretable, non-
+000008c0: 6e65 6761 7469 7665 2070 726f 6772 616d  negative program
+000008d0: 7320 6174 206d 756c 7469 706c 6520 7265  s at multiple re
+000008e0: 736f 6c75 7469 6f6e 730d 0a2d 204d 6f73  solutions..- Mos
+000008f0: 6169 6320 696e 7465 6772 6174 696f 6e20  aic integration 
+00000900: 646f 6573 206e 6f74 2072 6571 7569 7265  does not require
+00000910: 2073 7562 7365 7474 696e 6720 6665 6174   subsetting feat
+00000920: 7572 6573 2f67 656e 6573 2074 6f0d 0a20  ures/genes to.. 
+00000930: 2061 2073 6861 7265 6420 6f72 206f 7665   a shared or ove
+00000940: 7264 6973 7065 7273 6564 2073 7562 7365  rdispersed subse
+00000950: 740d 0a2d 2049 6465 616c 2066 6f72 2069  t..- Ideal for i
+00000960: 6e63 7265 6d65 6e74 616c 2069 6e74 6567  ncremental integ
+00000970: 7261 7469 6f6e 2028 6164 6469 6e67 2064  ration (adding d
+00000980: 6174 6173 6574 7320 6f6e 6520 6174 2061  atasets one at a
+00000990: 2074 696d 6529 2073 696e 6365 0d0a 2020   time) since..  
+000009a0: 6465 636f 6e76 6f6c 7574 696f 6e20 6973  deconvolution is
+000009b0: 2070 6572 666f 726d 6564 2069 6e64 6570   performed indep
+000009c0: 656e 6465 6e74 6c79 206f 6e20 6561 6368  endently on each
+000009d0: 2064 6174 6173 6574 0d0a 2d20 496e 7465   dataset..- Inte
+000009e0: 6772 6174 696f 6e20 7065 7266 6f72 6d73  gration performs
+000009f0: 2077 656c 6c20 6576 656e 2077 6865 6e20   well even when 
+00000a00: 7468 6520 6461 7461 7365 7473 2068 6176  the datasets hav
+00000a10: 6520 6d69 736d 6174 6368 6564 2066 6561  e mismatched fea
+00000a20: 7475 7265 730d 0a20 2028 6567 2e20 4d69  tures..  (eg. Mi
+00000a30: 6372 6f61 7272 6179 2c20 524e 412d 5365  croarray, RNA-Se
+00000a40: 712c 2050 726f 7465 6f6d 6963 7329 206f  q, Proteomics) o
+00000a50: 7220 7370 6172 7369 7479 2028 6567 2073  r sparsity (eg s
+00000a60: 696e 676c 652d 6365 6c6c 2076 7320 6275  ingle-cell vs bu
+00000a70: 6c6b 2052 4e41 2d53 6571 2061 6e64 2041  lk RNA-Seq and A
+00000a80: 5441 432d 5365 7129 0d0a 2d20 5477 6f20  TAC-Seq)..- Two 
+00000a90: 696e 7465 7266 6163 6573 3a20 636f 6d6d  interfaces: comm
+00000aa0: 616e 642d 6c69 6e65 2069 6e74 6572 6661  and-line interfa
+00000ab0: 6365 2066 6f72 2072 6170 6964 2064 6174  ce for rapid dat
+00000ac0: 6120 6578 706c 6f72 6174 696f 6e20 616e  a exploration an
+00000ad0: 6420 7079 7468 6f6e 0d0a 2020 696e 7465  d python..  inte
+00000ae0: 7266 6163 6520 666f 7220 6578 7465 6e73  rface for extens
+00000af0: 6962 696c 6974 7920 616e 6420 666c 6578  ibility and flex
+00000b00: 6962 696c 6974 790d 0a0d 0a23 2320 f09f  ibility....## ..
+00000b10: 94a7 2049 6e73 7461 6c6c 0d0a 0d0a 2323  .. Install....##
+00000b20: 2320 e298 81ef b88f 2050 7562 6c69 6320  # ...... Public 
+00000b30: 5265 6c65 6173 650d 0a0d 0a49 6e73 7461  Release....Insta
+00000b40: 6c6c 2074 6865 2070 6163 6b61 6765 2077  ll the package w
+00000b50: 6974 6820 636f 6e64 6120 2869 6e20 616e  ith conda (in an
+00000b60: 2069 736f 6c61 7465 6420 636f 6e64 6120   isolated conda 
+00000b70: 656e 7669 726f 6e6d 656e 7429 0d0a 6060  environment)..``
+00000b80: 6062 6173 680d 0a63 6f6e 6461 2063 7265  `bash..conda cre
+00000b90: 6174 6520 2d6e 2063 6e6d 6673 6e73 202d  ate -n cnmfsns -
+00000ba0: 6320 636f 6e64 612d 666f 7267 6520 636e  c conda-forge cn
+00000bb0: 6d66 736e 730d 0a63 6f6e 6461 2061 6374  mfsns..conda act
+00000bc0: 6976 6174 6520 636e 6d66 736e 730d 0a0d  ivate cnmfsns...
+00000bd0: 0a60 6060 0d0a 0d0a 2323 2320 e29c a820  .```....### ... 
+00000be0: 4c61 7465 7374 2076 6572 7369 6f6e 2066  Latest version f
+00000bf0: 726f 6d20 4769 7448 7562 0d0a 0d0a 4265  rom GitHub....Be
+00000c00: 666f 7265 2069 6e73 7461 6c6c 696e 6720  fore installing 
+00000c10: 634e 4d46 2d53 4e53 2075 7369 6e67 2070  cNMF-SNS using p
+00000c20: 6970 2c20 6974 2069 7320 7265 636f 6d6d  ip, it is recomm
+00000c30: 656e 6465 6420 746f 2066 6972 7374 2073  ended to first s
+00000c40: 6574 2075 7020 6120 7365 7061 7261 7465  et up a separate
+00000c50: 2063 6f6e 6461 2065 6e76 6972 6f6e 6d65   conda environme
+00000c60: 6e74 2061 6e64 2068 6176 6520 636f 6e64  nt and have cond
+00000c70: 6120 6d61 6e61 6765 2061 7320 6d61 6e79  a manage as many
+00000c80: 2064 6570 656e 6465 6e63 6965 7320 6173   dependencies as
+00000c90: 2070 6f73 7369 626c 652e 0d0a 0d0a 6060   possible.....``
+00000ca0: 6062 6173 680d 0a63 6f6e 6461 2063 7265  `bash..conda cre
+00000cb0: 6174 6520 2d2d 6e61 6d65 2063 6e6d 6673  ate --name cnmfs
+00000cc0: 6e73 202d 6320 636f 6e64 612d 666f 7267  ns -c conda-forg
+00000cd0: 6520 7079 7468 6f6e 3d33 2e31 3020 616e  e python=3.10 an
+00000ce0: 6e64 6174 6120 7061 6e64 6173 206e 756d  ndata pandas num
+00000cf0: 7079 2073 6369 7079 206d 6174 706c 6f74  py scipy matplot
+00000d00: 6c69 6220 7570 7365 7470 6c6f 7420 6874  lib upsetplot ht
+00000d10: 7470 6c69 6232 2074 6f6d 6c69 2074 6f6d  tplib2 tomli tom
+00000d20: 6c69 2d77 2063 6c69 636b 2070 7967 7261  li-w click pygra
+00000d30: 7068 7669 7a20 7079 7468 6f6e 2d69 6772  phviz python-igr
+00000d40: 6170 6820 7079 7961 6d6c 2073 6369 6b69  aph pyyaml sciki
+00000d50: 742d 6c65 6172 6e20 6661 7374 636c 7573  t-learn fastclus
+00000d60: 7465 7220 7363 616e 7079 2070 7979 616d  ter scanpy pyyam
+00000d70: 6c0d 0a63 6f6e 6461 2061 6374 6976 6174  l..conda activat
+00000d80: 6520 636e 6d66 736e 730d 0a70 6970 2069  e cnmfsns..pip i
+00000d90: 6e73 7461 6c6c 2067 6974 2b68 7474 7073  nstall git+https
+00000da0: 3a2f 2f67 6974 6875 622e 636f 6d2f 4d6f  ://github.com/Mo
+00000db0: 7272 6973 7379 4c61 622f 634e 4d46 2d53  rrissyLab/cNMF-S
+00000dc0: 4e53 2e67 6974 0d0a 6060 600d 0a0d 0a23  NS.git..```....#
+00000dd0: 2320 f09f 9396 2044 6f63 756d 656e 7461  # .... Documenta
+00000de0: 7469 6f6e 0d0a 0d0a 2323 2320 f09f 9790  tion....### ....
+00000df0: 2044 6174 6120 6775 6964 656c 696e 6573   Data guidelines
+00000e00: 0d0a 0d0a 634e 4d46 2d53 4e53 2063 616e  ....cNMF-SNS can
+00000e10: 2066 6163 746f 7269 7a65 2061 2077 6964   factorize a wid
+00000e20: 6520 7661 7269 6574 7920 6f66 2064 6174  e variety of dat
+00000e30: 6173 6574 732c 2062 7574 2077 696c 6c20  asets, but will 
+00000e40: 776f 726b 206f 7074 696d 616c 6c79 2069  work optimally i
+00000e50: 6e20 7468 6573 6520 636f 6e64 6974 696f  n these conditio
+00000e60: 6e73 3a0d 0a20 202d 2055 7365 2075 6e74  ns:..  - Use unt
+00000e70: 7261 6e73 666f 726d 6564 2028 7261 7729  ransformed (raw)
+00000e80: 2064 6174 6120 7768 6572 6520 706f 7373   data where poss
+00000e90: 6962 6c65 2e0d 0a20 202d 2046 6f72 2073  ible...  - For s
+00000ea0: 696e 676c 652d 6365 6c6c 206f 7220 7370  ingle-cell or sp
+00000eb0: 6174 6961 6c20 524e 412d 5365 7120 6461  atial RNA-Seq da
+00000ec0: 7461 2c20 7573 6520 6665 6174 7572 6520  ta, use feature 
+00000ed0: 636f 756e 7473 2c20 6e6f 7420 5450 4d20  counts, not TPM 
+00000ee0: 6f72 206f 7468 6572 206c 6f67 2d74 7261  or other log-tra
+00000ef0: 6e73 666f 726d 6564 2076 616c 7565 732e  nsformed values.
+00000f00: 0d0a 0d0a 0d0a 2323 2320 f09f 9393 2050  ......### .... P
+00000f10: 7974 686f 6e20 696e 7465 7266 6163 650d  ython interface.
+00000f20: 0a0d 0a54 6f20 6765 7420 7374 6172 7465  ...To get starte
+00000f30: 642c 2073 616d 706c 6520 7072 6f74 656f  d, sample proteo
+00000f40: 6d69 6373 2064 6174 6173 6574 7320 616e  mics datasets an
+00000f50: 6420 6120 4a75 7079 7465 7220 6e6f 7465  d a Jupyter note
+00000f60: 626f 6f6b 2074 7574 6f72 6961 6c20 6973  book tutorial is
+00000f70: 2061 7661 696c 6162 6c65 205b 6865 7265   available [here
+00000f80: 5d28 2f74 7574 6f72 6961 6c2f 7475 746f  ](/tutorial/tuto
+00000f90: 7269 616c 2e69 7079 6e62 292e 0d0a 0d0a  rial.ipynb).....
+00000fa0: 4465 7461 696c 6564 2041 5049 2072 6566  Detailed API ref
+00000fb0: 6572 656e 6365 2063 616e 2062 6520 666f  erence can be fo
+00000fc0: 756e 6420 6f6e 205b 5265 6164 5468 6544  und on [ReadTheD
+00000fd0: 6f63 735d 2868 7474 7073 3a2f 2f63 6e6d  ocs](https://cnm
+00000fe0: 662d 736e 732e 7265 6164 7468 6564 6f63  f-sns.readthedoc
+00000ff0: 732e 696f 2f29 2e0d 0a0d 0a0d 0a23 2323  s.io/).......###
+00001000: 20e2 8ca8 efb8 8f20 436f 6d6d 616e 6420   ...... Command 
+00001010: 6c69 6e65 2069 6e74 6572 6661 6365 0d0a  line interface..
+00001020: 0d0a 5365 6520 7468 6520 5b63 6f6d 6d61  ..See the [comma
+00001030: 6e64 206c 696e 6520 696e 7465 7266 6163  nd line interfac
+00001040: 6520 646f 6375 6d65 6e74 6174 696f 6e5d  e documentation]
+00001050: 282f 434c 492e 6d64 292e 0d0a 0d0a 2323  (/CLI.md).....##
+00001060: 20f0 9f92 ad20 4765 7474 696e 6720 4865   .... Getting He
+00001070: 6c70 0d0a 0d0a 466f 7220 6572 726f 7273  lp....For errors
+00001080: 2061 7269 7369 6e67 2064 7572 696e 6720   arising during 
+00001090: 7573 6520 6f66 2063 4e4d 462d 534e 532c  use of cNMF-SNS,
+000010a0: 2063 7265 6174 6520 616e 6420 6272 6f77   create and brow
+000010b0: 7365 2069 7373 7565 7320 696e 2074 6865  se issues in the
+000010c0: 205b 4769 7448 7562 2022 6973 7375 6573   [GitHub "issues
+000010d0: 2220 7461 625d 2868 7474 7073 3a2f 2f67  " tab](https://g
+000010e0: 6974 6875 622e 636f 6d2f 4d6f 7272 6973  ithub.com/Morris
+000010f0: 7379 4c61 622f 634e 4d46 2d53 4e53 2f69  syLab/cNMF-SNS/i
+00001100: 7373 7565 7329 2e0d 0a                   ssues)...
```

### Comparing `cnmfsns-1.6.5/setup.cfg` & `cnmfsns-1.7.0/setup.cfg`

 * *Files 1% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 00000000: 5b6d 6574 6164 6174 615d 0d0a 6e61 6d65  [metadata]..name
 00000010: 203d 2063 6e6d 6673 6e73 0d0a 7665 7273   = cnmfsns..vers
-00000020: 696f 6e20 3d20 312e 362e 350d 0a61 7574  ion = 1.6.5..aut
+00000020: 696f 6e20 3d20 312e 372e 300d 0a61 7574  ion = 1.7.0..aut
 00000030: 686f 7220 3d20 5465 6420 5665 7268 6579  hor = Ted Verhey
 00000040: 0d0a 6175 7468 6f72 5f65 6d61 696c 203d  ..author_email =
 00000050: 2074 6276 6572 6865 7940 7563 616c 6761   tbverhey@ucalga
 00000060: 7279 2e63 610d 0a64 6573 6372 6970 7469  ry.ca..descripti
 00000070: 6f6e 203d 2063 4e4d 4620 536f 6c75 7469  on = cNMF Soluti
 00000080: 6f6e 204e 6574 776f 726b 2053 7061 6365  on Network Space
 00000090: 0d0a 6c6f 6e67 5f64 6573 6372 6970 7469  ..long_descripti
```

### Comparing `cnmfsns-1.6.5/src/cnmfsns/__init__.py` & `cnmfsns-1.7.0/src/cnmfsns/__init__.py`

 * *Files identical despite different names*

### Comparing `cnmfsns-1.6.5/src/cnmfsns/cli.py` & `cnmfsns-1.7.0/src/cnmfsns/cli.py`

 * *Files 2% similar despite different names*

```diff
@@ -463,20 +463,19 @@
     elif input_h5ad:
         config = Config.from_h5ad_files(input_h5ad)
 
     # Create dataset objects from config parameters
     datasets = {}
     for dsname, dsparams in config.datasets.items():
         
-        dataset = Dataset.from_h5ad(
-            dsparams["filename"],
-            backed=False,
-            patient_id_col = (dsparams["patient_id_col"] if "patient_id_col" in dsparams else None),
-            force_migrate = False
-        )
+        dataset = Dataset.from_h5ad(dsparams["filename"], force_migrate = False)
+            
+        if "patient_id_col" in dsparams:
+            dataset.patient_id_col = dsparams["patient_id_col"]
+
         datasets[dsname] = dataset
         
     # creates integration object from config parameters
     integration = Integration(
         datasets=datasets,
         corr_method = config.integrate["corr_method"],
         max_median_corr = config.integrate["max_median_corr"],
@@ -514,22 +513,30 @@
 
     # plot mirrored distributions with thresholds (which are computed on max-k filtered data only)
     fig = plot_pairwise_corr_overlaid(integration)
     fig.savefig(os.path.join(output_dir, "integrate", "pairwise_corr_overlaid.pdf"))
     fig.savefig(os.path.join(output_dir, "integrate", "pairwise_corr_overlaid.png"), dpi=600)
 
     # UpSet plot of odgenes and all genes in each dataset
+
+    df = integration.get_overdispersed_features_overlap_table()
+    df.to_csv(os.path.join(output_dir, "integrate", "overdispersed_features.txt"), sep="\t")
+    
+    df = integration.get_features_overlap_table()
+    df.to_csv(os.path.join(output_dir, "integrate", "features.txt"), sep="\t")
+    
     if integration.n_datasets > 1:
+
         fig = plot_overdispersed_features_upset(integration)
-        fig.savefig(os.path.join(output_dir, "integrate", "upsetplot_overdispersed_genes.pdf"))
-        fig.savefig(os.path.join(output_dir, "integrate", "upsetplot_overdispersed_genes.pdf" + ".png"), dpi=600)
+        fig.savefig(os.path.join(output_dir, "integrate", "overdispersed_features_upsetplot.pdf"))
+        fig.savefig(os.path.join(output_dir, "integrate", "overdispersed_features_upsetplot.png"), dpi=600)
         
         fig = plot_features_upset(integration)
-        fig.savefig(os.path.join(output_dir, "integrate", "upsetplot_all_genes.pdf"))
-        fig.savefig(os.path.join(output_dir, "integrate", "upsetplot_all_genes.pdf" + ".png"), dpi=600)
+        fig.savefig(os.path.join(output_dir, "integrate", "all_features_upsetplot.pdf"))
+        fig.savefig(os.path.join(output_dir, "integrate", "all_features_upsetplot.png"), dpi=600)
 
     nodetable = integration.get_node_table()
     nodetable.to_csv(os.path.join(output_dir, "integrate", "node_stats.txt"), sep="\t")
 
 @click.command(name="create-network")
 @click.option(
     '-o', '--output_dir', type=click.Path(file_okay=False, exists=True), required=True,
@@ -558,20 +565,17 @@
     # Create directory structures
     sns_output_dir = os.path.join(output_dir, "sns_networks", name)
     os.makedirs(sns_output_dir, exist_ok=True)
 
     # Create dataset objects from config parameters
     datasets = {}
     for dsname, dsparams in config.datasets.items():
-        ds_obj = Dataset.from_h5ad(
-            dsparams["filename"],
-            backed="r",
-            patient_id_col = (dsparams["patient_id_col"] if "patient_id_col" in dsparams else None),
-            force_migrate = False
-        )
+        ds_obj = Dataset.from_h5ad(dsparams["filename"], force_migrate = False)
+        if "patient_id_col" in dsparams:
+            dataset.patient_id_col = dsparams["patient_id_col"]
         datasets[dsname] = ds_obj
         
     # selected_k from config.toml file overrides defaults
     selected_k = {dsname: ("selected_k" in dsparams) for dsname, dsparams in config.datasets.items()}
     if all(list(selected_k.values())):
         selected_k = {dsname: dsparams["selected_k"] for dsname, dsparams in config.datasets.items()}
     elif any(list(selected_k.values())):
@@ -610,16 +614,18 @@
 
 
     snsmap.compute_layout(
         algorithm=config.sns["layout_algorithm"],
         shared_community_weight = config.sns["layouts"]["community_weighted_spring"]["within_community"],
         shared_dataset_weight = config.sns["layouts"]["community_weighted_spring"]["within_dataset"]
     )
-    snsmap.compute_community_network_layout()
     
+    # persist SNS object to file
+    snsmap.to_pkl(os.path.join(sns_output_dir, "sns.pkl"))
+
     # make figure legends for metadata
     colors = Colors.from_config(config)
     colors.add_missing_dataset_colors(datasets=integration)
     colors.add_missing_metadata_colors(datasets=integration)
     colors.add_missing_community_colors(snsmap=snsmap)
     colors.to_toml(os.path.join(sns_output_dir, "colors.toml"))
```

### Comparing `cnmfsns-1.6.5/src/cnmfsns/cnmf.py` & `cnmfsns-1.7.0/src/cnmfsns/cnmf.py`

 * *Files identical despite different names*

### Comparing `cnmfsns-1.6.5/src/cnmfsns/colors.py` & `cnmfsns-1.7.0/src/cnmfsns/colors.py`

 * *Files identical despite different names*

### Comparing `cnmfsns-1.6.5/src/cnmfsns/config.py` & `cnmfsns-1.7.0/src/cnmfsns/config.py`

 * *Files 10% similar despite different names*

```diff
@@ -50,17 +50,15 @@
         "pie_size_community": 50,
         "plot_size_community": [4, 4],
         },
     "datasets": {},
     "metadata_colors": {"missing_data": "#dddddd"},
     "metadata_colors_group": {}
 }
-dataset_defaults = {
-    "selected_k": [[1, 10, 1], [15, 500, 5]],
-    }
+dataset_defaults = {}
 
 def recursive_update(d, u):
     d = deepcopy(d)
     for k, v in u.items():
         if isinstance(v, Mapping):
             d[k] = recursive_update(d.get(k, {}), v)
         else:
```

### Comparing `cnmfsns-1.6.5/src/cnmfsns/dataset.py` & `cnmfsns-1.7.0/src/cnmfsns/dataset.py`

 * *Files 3% similar despite different names*

```diff
@@ -25,2252 +25,2254 @@
 00000180: 4761 6d2c 2042 5370 6c69 6e65 730d 0a0d  Gam, BSplines...
 00000190: 0a63 6c61 7373 2044 6174 6173 6574 2829  .class Dataset()
 000001a0: 3a0d 0a20 2020 200d 0a20 2020 2064 6566  :..    ..    def
 000001b0: 205f 5f69 6e69 745f 5f28 7365 6c66 2c0d   __init__(self,.
 000001c0: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
 000001d0: 2020 6164 6174 613a 2061 642e 416e 6e44    adata: ad.AnnD
 000001e0: 6174 612c 0d0a 2020 2020 2020 2020 2020  ata,..          
-000001f0: 2020 2020 2020 2070 6174 6965 6e74 5f69         patient_i
-00000200: 645f 636f 6c3a 204f 7074 696f 6e61 6c5b  d_col: Optional[
-00000210: 7374 725d 203d 204e 6f6e 652c 0d0a 2020  str] = None,..  
-00000220: 2020 2020 2020 2020 2020 2020 2020 2066                 f
-00000230: 6f72 6365 5f6d 6967 7261 7465 3a20 626f  orce_migrate: bo
-00000240: 6f6c 203d 2046 616c 7365 0d0a 2020 2020  ol = False..    
-00000250: 2020 2020 2020 2020 2020 2020 2029 3a0d               ):.
-00000260: 0a20 2020 2020 2020 2022 2222 4372 6561  .        """Crea
-00000270: 7465 7320 6120 3a63 6c61 7373 3a60 7e63  tes a :class:`~c
-00000280: 6e6d 6673 6e73 2e64 6174 6173 6574 2e44  nmfsns.dataset.D
-00000290: 6174 6173 6574 6020 6f62 6a65 6374 2066  ataset` object f
-000002a0: 726f 6d20 616e 2060 6164 2e41 6e6e 4461  rom an `ad.AnnDa
-000002b0: 7461 6020 6f62 6a65 6374 2e0d 0a0d 0a20  ta` object..... 
-000002c0: 2020 2020 2020 203a 7061 7261 6d20 6164         :param ad
-000002d0: 6174 613a 2041 6e6e 4461 7461 206f 626a  ata: AnnData obj
-000002e0: 6563 7420 7769 7468 2064 6174 610d 0a20  ect with data.. 
-000002f0: 2020 2020 2020 203a 7479 7065 2061 6461         :type ada
-00000300: 7461 3a20 6164 2e41 6e6e 4461 7461 0d0a  ta: ad.AnnData..
-00000310: 2020 2020 2020 2020 3a70 6172 616d 2070          :param p
-00000320: 6174 6965 6e74 5f69 645f 636f 6c3a 204e  atient_id_col: N
-00000330: 616d 6520 6f66 206d 6574 6164 6174 6120  ame of metadata 
-00000340: 6c61 7965 7220 7769 7468 2070 6174 6965  layer with patie
-00000350: 6e74 2049 4420 696e 666f 726d 6174 696f  nt ID informatio
-00000360: 6e2c 2064 6566 6175 6c74 7320 746f 204e  n, defaults to N
-00000370: 6f6e 650d 0a20 2020 2020 2020 203a 7479  one..        :ty
-00000380: 7065 2070 6174 6965 6e74 5f69 645f 636f  pe patient_id_co
-00000390: 6c3a 2073 7472 2c20 6f70 7469 6f6e 616c  l: str, optional
-000003a0: 0d0a 2020 2020 2020 2020 3a70 6172 616d  ..        :param
-000003b0: 2066 6f72 6365 5f6d 6967 7261 7465 3a20   force_migrate: 
-000003c0: 666f 7263 6573 2063 6f6e 7665 7273 696f  forces conversio
-000003d0: 6e20 6f66 2041 6e6e 4461 7461 206f 626a  n of AnnData obj
-000003e0: 6563 7473 2065 7665 6e20 7768 656e 2061  ects even when a
-000003f0: 6461 7461 2e58 2061 6e64 2061 6461 7461  data.X and adata
-00000400: 2e72 6177 2e58 2061 7265 206e 6f74 206c  .raw.X are not l
-00000410: 696e 6561 726c 7920 7363 616c 6564 2072  inearly scaled r
-00000420: 656c 6174 6976 6520 746f 2065 6163 6820  elative to each 
-00000430: 6f74 6865 722c 2064 6566 6175 6c74 7320  other, defaults 
-00000440: 746f 2046 616c 7365 0d0a 2020 2020 2020  to False..      
-00000450: 2020 3a74 7970 6520 666f 7263 655f 6d69    :type force_mi
-00000460: 6772 6174 653a 2062 6f6f 6c2c 206f 7074  grate: bool, opt
-00000470: 696f 6e61 6c0d 0a20 2020 2020 2020 203a  ional..        :
-00000480: 7261 6973 6573 2052 756e 7469 6d65 4572  raises RuntimeEr
-00000490: 726f 723a 2042 6163 6b65 642d 6d6f 6465  ror: Backed-mode
-000004a0: 2041 6e6e 6461 7461 206f 626a 6563 7473   Anndata objects
-000004b0: 2063 616e 6e6f 7420 6265 206d 6967 7261   cannot be migra
-000004c0: 7465 640d 0a20 2020 2020 2020 203a 7261  ted..        :ra
-000004d0: 6973 6573 2056 616c 7565 4572 726f 723a  ises ValueError:
-000004e0: 2045 7272 6f72 2069 7320 7261 6973 6564   Error is raised
-000004f0: 2077 6865 6e20 666f 7263 6520 6973 2046   when force is F
-00000500: 616c 7365 2061 6e64 2061 6461 7461 2069  alse and adata i
-00000510: 7320 6e6f 6e2d 6c69 6e65 6172 6c79 2073  s non-linearly s
-00000520: 6361 6c65 642e 0d0a 2020 2020 2020 2020  caled...        
-00000530: 3a72 6574 7572 6e3a 204f 626a 6563 7420  :return: Object 
-00000540: 7769 7468 2065 7870 7265 7373 696f 6e20  with expression 
-00000550: 616e 6420 6d65 7461 6461 7461 0d0a 2020  and metadata..  
-00000560: 2020 2020 2020 3a72 7479 7065 3a20 3a63        :rtype: :c
-00000570: 6c61 7373 3a60 7e63 6e6d 6673 6e73 2e64  lass:`~cnmfsns.d
-00000580: 6174 6173 6574 2e44 6174 6173 6574 600d  ataset.Dataset`.
-00000590: 0a20 2020 2020 2020 2022 2222 0d0a 0d0a  .        """....
-000005a0: 2020 2020 2020 2020 6966 2061 6461 7461          if adata
-000005b0: 2e58 2069 7320 4e6f 6e65 3a0d 0a20 2020  .X is None:..   
-000005c0: 2020 2020 2020 2020 206c 6f67 6769 6e67           logging
-000005d0: 2e65 7272 6f72 2866 2261 6461 7461 2063  .error(f"adata c
-000005e0: 6f6e 7461 696e 7320 6e6f 2065 7870 7265  ontains no expre
-000005f0: 7373 696f 6e20 6461 7461 2028 6164 6174  ssion data (adat
-00000600: 612e 5829 2229 0d0a 2020 2020 2020 2020  a.X)")..        
-00000610: 2020 2020 7261 6973 6520 5661 6c75 6545      raise ValueE
-00000620: 7272 6f72 2829 0d0a 2020 2020 2020 2020  rror()..        
-00000630: 0d0a 2020 2020 2020 2020 0d0a 2020 2020  ..        ..    
-00000640: 2020 2020 6966 2022 636e 6d66 736e 735f      if "cnmfsns_
-00000650: 7665 7273 696f 6e22 2069 6e20 6164 6174  version" in adat
-00000660: 612e 756e 7320 616e 6420 6164 6174 612e  a.uns and adata.
-00000670: 756e 735b 2263 6e6d 6673 6e73 5f76 6572  uns["cnmfsns_ver
-00000680: 7369 6f6e 225d 2069 7320 6e6f 7420 4e6f  sion"] is not No
-00000690: 6e65 3a0d 0a20 2020 2020 2020 2020 2020  ne:..           
-000006a0: 2073 656c 662e 6164 6174 6120 3d20 6164   self.adata = ad
-000006b0: 6174 610d 0a20 2020 2020 2020 2065 6c73  ata..        els
-000006c0: 653a 0d0a 2020 2020 2020 2020 2020 2020  e:..            
-000006d0: 2320 6368 6563 6b20 616e 6420 7570 6461  # check and upda
-000006e0: 7465 206f 6c64 206f 7220 6578 7465 726e  te old or extern
-000006f0: 616c 2068 3561 6420 6669 6c65 7320 666f  al h5ad files fo
-00000700: 7220 634e 4d46 2d53 4e53 2063 6f6d 706c  r cNMF-SNS compl
-00000710: 6961 6e63 650d 0a20 2020 200d 0a20 2020  iance..    ..   
-00000720: 2020 2020 2020 2020 2058 203d 2061 6461           X = ada
-00000730: 7461 2e74 6f5f 6466 2829 0d0a 2020 2020  ta.to_df()..    
-00000740: 2020 2020 2020 2020 0d0a 2020 2020 2020          ..      
-00000750: 2020 2020 2020 6966 2061 6461 7461 2e69        if adata.i
-00000760: 7362 6163 6b65 643a 0d0a 2020 2020 2020  sbacked:..      
-00000770: 2020 2020 2020 2020 2020 7261 6973 6520            raise 
-00000780: 5275 6e74 696d 6545 7272 6f72 2822 6164  RuntimeError("ad
-00000790: 6174 6120 6973 2061 2062 6163 6b65 6420  ata is a backed 
-000007a0: 416e 6e44 6174 6120 6f62 6a65 6374 2e20  AnnData object. 
-000007b0: 416e 6e44 6174 6120 6f62 6a65 6374 7320  AnnData objects 
-000007c0: 6f70 656e 6564 2069 6e20 6261 636b 6564  opened in backed
-000007d0: 206d 6f64 6520 6361 6e6e 6f74 2062 6520   mode cannot be 
-000007e0: 6d69 6772 6174 6564 2e22 290d 0a0d 0a20  migrated.").... 
-000007f0: 2020 2020 2020 2020 2020 2069 6620 6164             if ad
-00000800: 6174 612e 7261 7720 6973 204e 6f6e 653a  ata.raw is None:
-00000810: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
-00000820: 2020 6973 5f6e 6f72 6d61 6c69 7a65 643d    is_normalized=
-00000830: 4661 6c73 650d 0a20 2020 2020 2020 2020  False..         
-00000840: 2020 2020 2020 2072 6177 203d 2061 6461         raw = ada
-00000850: 7461 2e74 6f5f 6466 2829 0d0a 2020 2020  ta.to_df()..    
-00000860: 2020 2020 2020 2020 656c 7365 3a0d 0a20          else:.. 
-00000870: 2020 2020 2020 2020 2020 2020 2020 2072                 r
-00000880: 6177 203d 2070 642e 4461 7461 4672 616d  aw = pd.DataFram
-00000890: 6528 6164 6174 612e 7261 772e 582c 2069  e(adata.raw.X, i
-000008a0: 6e64 6578 3d58 2e69 6e64 6578 2c20 636f  ndex=X.index, co
-000008b0: 6c75 6d6e 733d 582e 636f 6c75 6d6e 7329  lumns=X.columns)
-000008c0: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
-000008d0: 2020 636f 7272 6469 7374 203d 2058 2e63    corrdist = X.c
-000008e0: 6f72 7277 6974 6828 7261 772c 2061 7869  orrwith(raw, axi
-000008f0: 733d 3129 0d0a 2020 2020 2020 2020 2020  s=1)..          
-00000900: 2020 2020 2020 2320 6368 6563 6b73 2074        # checks t
-00000910: 6861 7420 616c 6c20 7361 6d70 6c65 7320  hat all samples 
-00000920: 6172 6520 7065 7266 6563 746c 7920 636f  are perfectly co
-00000930: 7272 656c 6174 6564 2062 6574 7765 656e  rrelated between
-00000940: 2063 6f75 6e74 7320 616e 6420 6e6f 726d   counts and norm
-00000950: 616c 697a 6564 2064 6174 610d 0a20 2020  alized data..   
-00000960: 2020 2020 2020 2020 2020 2020 2069 6620               if 
-00000970: 2828 636f 7272 6469 7374 202d 2031 292e  ((corrdist - 1).
-00000980: 6162 7328 2920 3e20 3165 2d36 292e 616e  abs() > 1e-6).an
-00000990: 7928 293a 0d0a 2020 2020 2020 2020 2020  y():..          
-000009a0: 2020 2020 2020 2020 2020 6c6f 6767 696e            loggin
-000009b0: 672e 7761 726e 696e 6728 2243 6f75 6e74  g.warning("Count
-000009c0: 7320 616e 6420 6e6f 726d 616c 697a 6564  s and normalized
-000009d0: 2065 7870 7265 7373 696f 6e20 6d61 7472   expression matr
-000009e0: 6963 6573 2061 7265 206e 6f74 2070 6572  ices are not per
-000009f0: 6665 6374 6c79 2063 6f72 7265 6c61 7465  fectly correlate
-00000a00: 642e 2043 6f75 6e74 7320 6461 7461 2077  d. Counts data w
-00000a10: 696c 6c20 6265 2072 6574 6169 6e65 6420  ill be retained 
-00000a20: 696e 206d 6967 7261 7465 6420 6f62 6a65  in migrated obje
-00000a30: 6374 2e22 290d 0a20 2020 2020 2020 2020  ct.")..         
-00000a40: 2020 2020 2020 2020 2020 2069 6620 6e6f             if no
-00000a50: 7420 666f 7263 655f 6d69 6772 6174 653a  t force_migrate:
-00000a60: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
-00000a70: 2020 2020 2020 2020 2020 6572 726f 726d            errorm
-00000a80: 7367 203d 2022 436f 756c 6420 6e6f 7420  sg = "Could not 
-00000a90: 6d69 6772 6174 6520 416e 6e44 6174 6120  migrate AnnData 
-00000aa0: 6f62 6a65 6374 2e22 0d0a 2020 2020 2020  object."..      
-00000ab0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00000ac0: 2020 6c6f 6767 696e 672e 6572 726f 7228    logging.error(
-00000ad0: 6572 726f 726d 7367 290d 0a20 2020 2020  errormsg)..     
-00000ae0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00000af0: 2020 2072 6169 7365 2056 616c 7565 4572     raise ValueEr
-00000b00: 726f 7228 6572 726f 726d 7367 290d 0a20  ror(errormsg).. 
-00000b10: 2020 2020 2020 2020 2020 2020 2020 2023                 #
-00000b20: 2063 6865 636b 2066 6f72 2077 6865 7468   check for wheth
-00000b30: 6572 2075 7365 7220 6f72 6967 696e 616c  er user original
-00000b40: 6c79 2069 6e70 7574 206e 6f72 6d61 6c69  ly input normali
-00000b50: 7a65 6420 6f72 2063 6f75 6e74 7320 6461  zed or counts da
-00000b60: 7461 0d0a 2020 2020 2020 2020 2020 2020  ta..            
-00000b70: 2020 2020 6973 5f6e 6f72 6d61 6c69 7a65      is_normalize
-00000b80: 6420 3d20 2828 5820 2d20 7261 7729 2e61  d = ((X - raw).a
-00000b90: 6273 2829 203c 2031 652d 3629 2e61 6c6c  bs() < 1e-6).all
-00000ba0: 2829 2e61 6c6c 2829 0d0a 2020 2020 2020  ().all()..      
-00000bb0: 2020 2020 2020 2020 2020 585f 6973 5f74            X_is_t
-00000bc0: 706d 203d 2028 2858 2e73 756d 2861 7869  pm = ((X.sum(axi
-00000bd0: 733d 3129 202d 2031 6536 292e 6162 7328  s=1) - 1e6).abs(
-00000be0: 2920 3e20 3165 3229 2e61 6e79 2829 0d0a  ) > 1e2).any()..
-00000bf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00000c00: 6966 2069 735f 6e6f 726d 616c 697a 6564  if is_normalized
-00000c10: 2061 6e64 206e 6f74 2058 5f69 735f 7470   and not X_is_tp
-00000c20: 6d3a 0d0a 2020 2020 2020 2020 2020 2020  m:..            
-00000c30: 2020 2020 2020 2020 6c6f 6767 696e 672e          logging.
-00000c40: 7761 726e 696e 6728 2241 6e6e 4461 7461  warning("AnnData
-00000c50: 206f 626a 6563 7420 636f 6e74 6169 6e73   object contains
-00000c60: 206e 6f6e 2d54 504d 206e 6f72 6d61 6c69   non-TPM normali
-00000c70: 7a65 6420 6461 7461 2e20 4e65 7720 416e  zed data. New An
-00000c80: 6e44 6174 6120 6f62 6a65 6374 2077 696c  nData object wil
-00000c90: 6c20 7265 7461 696e 2074 6865 2063 6f75  l retain the cou
-00000ca0: 6e74 2028 756e 6e6f 726d 616c 697a 6564  nt (unnormalized
-00000cb0: 2920 6461 7461 206f 6e6c 792e 2229 0d0a  ) data only.")..
-00000cc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00000cd0: 0d0a 2020 2020 2020 2020 2020 2020 6966  ..            if
-00000ce0: 2022 6f64 6722 2069 6e20 6164 6174 612e   "odg" in adata.
-00000cf0: 756e 7320 616e 6420 2267 656e 655f 7374  uns and "gene_st
-00000d00: 6174 7322 2069 6e20 6164 6174 612e 756e  ats" in adata.un
-00000d10: 735b 226f 6467 225d 3a0d 0a20 2020 2020  s["odg"]:..     
-00000d20: 2020 2020 2020 2020 2020 2067 656e 655f             gene_
-00000d30: 7374 6174 5f63 6f6c 756d 6e73 203d 2061  stat_columns = a
-00000d40: 6461 7461 2e75 6e73 5b22 6f64 6722 5d5b  data.uns["odg"][
-00000d50: 2267 656e 655f 7374 6174 7322 5d2e 636f  "gene_stats"].co
-00000d60: 6c75 6d6e 730d 0a20 2020 2020 2020 2020  lumns..         
-00000d70: 2020 2020 2020 2069 6620 6164 6174 612e         if adata.
-00000d80: 7661 722e 636f 6c75 6d6e 732e 6973 696e  var.columns.isin
-00000d90: 2867 656e 655f 7374 6174 5f63 6f6c 756d  (gene_stat_colum
-00000da0: 6e73 292e 616e 7928 293a 0d0a 2020 2020  ns).any():..    
-00000db0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00000dc0: 6f76 6572 6c61 7070 696e 675f 636f 6c73  overlapping_cols
-00000dd0: 203d 2061 6461 7461 2e76 6172 2e63 6f6c   = adata.var.col
-00000de0: 756d 6e73 2e69 7369 6e28 6765 6e65 5f73  umns.isin(gene_s
-00000df0: 7461 745f 636f 6c75 6d6e 7329 0d0a 2020  tat_columns)..  
-00000e00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00000e10: 2020 6f76 6572 6c61 7070 696e 675f 636f    overlapping_co
-00000e20: 6c73 7472 203d 2022 2c20 222e 6a6f 696e  lstr = ", ".join
-00000e30: 286f 7665 726c 6170 7069 6e67 5f63 6f6c  (overlapping_col
-00000e40: 735b 6f76 6572 6c61 7070 696e 675f 636f  s[overlapping_co
-00000e50: 6c73 5d2e 696e 6465 782e 746f 5f6c 6973  ls].index.to_lis
-00000e60: 7428 2929 0d0a 2020 2020 2020 2020 2020  t())..          
-00000e70: 2020 2020 2020 2020 2020 6c6f 6767 696e            loggin
-00000e80: 672e 7761 726e 696e 6728 6622 416e 6e44  g.warning(f"AnnD
-00000e90: 6174 6120 6f62 6a65 6374 2063 6f6e 7461  ata object conta
-00000ea0: 696e 7320 634e 4d46 2067 656e 6520 7374  ins cNMF gene st
-00000eb0: 6174 7320 7768 6963 6820 7769 6c6c 206f  ats which will o
-00000ec0: 7665 7272 6964 6520 636f 6c75 6d6e 7320  verride columns 
-00000ed0: 696e 2061 6461 7461 2e76 6172 3a20 7b6f  in adata.var: {o
-00000ee0: 7665 726c 6170 7069 6e67 5f63 6f6c 7374  verlapping_colst
-00000ef0: 727d 2229 0d0a 2020 2020 2020 2020 2020  r}")..          
-00000f00: 2020 2020 2020 6164 6174 612e 7661 7220        adata.var 
-00000f10: 3d20 7064 2e6d 6572 6765 286c 6566 743d  = pd.merge(left=
-00000f20: 6164 6174 612e 7661 722c 2072 6967 6874  adata.var, right
-00000f30: 3d61 6461 7461 2e75 6e73 5b22 6f64 6722  =adata.uns["odg"
-00000f40: 5d5b 2267 656e 655f 7374 6174 7322 5d2c  ]["gene_stats"],
-00000f50: 2068 6f77 3d22 6c65 6674 222c 206c 6566   how="left", lef
-00000f60: 745f 696e 6465 783d 5472 7565 2c20 7269  t_index=True, ri
-00000f70: 6768 745f 696e 6465 783d 5472 7565 290d  ght_index=True).
-00000f80: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00000f90: 2064 656c 2061 6461 7461 2e75 6e73 5b22   del adata.uns["
-00000fa0: 6f64 6722 5d5b 2267 656e 655f 7374 6174  odg"]["gene_stat
-00000fb0: 7322 5d0d 0a20 2020 2020 2020 2020 2020  s"]..           
-00000fc0: 200d 0a20 2020 2020 2020 2020 2020 2023   ..            #
-00000fd0: 2063 7265 6174 6520 6e65 7720 416e 6e44   create new AnnD
-00000fe0: 6174 6120 6f62 6a65 6374 0d0a 2020 2020  ata object..    
-00000ff0: 2020 2020 2020 2020 6e65 775f 6164 6174          new_adat
-00001000: 6120 3d20 6164 2e41 6e6e 4461 7461 2858  a = ad.AnnData(X
-00001010: 3d72 6177 2c20 6f62 733d 6164 6174 612e  =raw, obs=adata.
-00001020: 6f62 732c 2076 6172 3d61 6461 7461 2e76  obs, var=adata.v
-00001030: 6172 2c20 7661 726d 3d61 6461 7461 2e76  ar, varm=adata.v
-00001040: 6172 6d2c 206f 6273 6d3d 6164 6174 612e  arm, obsm=adata.
-00001050: 6f62 736d 2c20 756e 733d 6164 6174 612e  obsm, uns=adata.
-00001060: 756e 7329 0d0a 2020 2020 2020 2020 2020  uns)..          
-00001070: 2020 6966 2022 6869 7374 6f72 7922 206e    if "history" n
-00001080: 6f74 2069 6e20 6e65 775f 6164 6174 612e  ot in new_adata.
-00001090: 756e 733a 0d0a 2020 2020 2020 2020 2020  uns:..          
-000010a0: 2020 2020 2020 6e65 775f 6164 6174 612e        new_adata.
-000010b0: 756e 735b 2268 6973 746f 7279 225d 203d  uns["history"] =
-000010c0: 207b 7d0d 0a0d 0a20 2020 2020 2020 2020   {}....         
-000010d0: 2020 2023 2075 7064 6174 6520 6461 7461     # update data
-000010e0: 7365 7420 6f62 6a65 6374 0d0a 2020 2020  set object..    
-000010f0: 2020 2020 2020 2020 7365 6c66 2e61 6461          self.ada
-00001100: 7461 203d 206e 6577 5f61 6461 7461 0d0a  ta = new_adata..
-00001110: 2020 2020 2020 2020 2020 2020 7365 6c66              self
-00001120: 2e69 735f 6e6f 726d 616c 697a 6564 203d  .is_normalized =
-00001130: 2069 735f 6e6f 726d 616c 697a 6564 0d0a   is_normalized..
-00001140: 2020 2020 2020 2020 2020 2020 7365 6c66              self
-00001150: 2e63 6e6d 6673 6e73 5f76 6572 7369 6f6e  .cnmfsns_version
-00001160: 203d 205f 5f76 6572 7369 6f6e 5f5f 0d0a   = __version__..
-00001170: 2020 2020 0d0a 2020 2020 2020 2020 6966      ..        if
-00001180: 2070 6174 6965 6e74 5f69 645f 636f 6c20   patient_id_col 
-00001190: 6973 206e 6f74 204e 6f6e 6520 616e 6420  is not None and 
-000011a0: 7061 7469 656e 745f 6964 5f63 6f6c 206e  patient_id_col n
-000011b0: 6f74 2069 6e20 7365 6c66 2e61 6461 7461  ot in self.adata
-000011c0: 2e6f 6273 2e63 6f6c 756d 6e73 3a0d 0a20  .obs.columns:.. 
-000011d0: 2020 2020 2020 2020 2020 2061 7661 696c             avail
-000011e0: 5f63 6f6c 756d 6e73 203d 2022 2c20 222e  _columns = ", ".
-000011f0: 6a6f 696e 2873 656c 662e 6164 6174 612e  join(self.adata.
-00001200: 6f62 732e 636f 6c75 6d6e 7329 0d0a 2020  obs.columns)..  
-00001210: 2020 2020 2020 2020 2020 7261 6973 6520            raise 
-00001220: 5661 6c75 6545 7272 6f72 2866 227b 7061  ValueError(f"{pa
-00001230: 7469 656e 745f 6964 5f63 6f6c 7d20 6973  tient_id_col} is
-00001240: 206e 6f74 2061 2076 616c 6964 2063 6f6c   not a valid col
-00001250: 756d 6e20 696e 2074 6865 206d 6574 6164  umn in the metad
-00001260: 6174 6120 6d61 7472 6978 2e20 4176 6169  ata matrix. Avai
-00001270: 6c61 626c 6520 636f 6c75 6d6e 7320 6172  lable columns ar
-00001280: 653a 207b 6176 6169 6c5f 636f 6c75 6d6e  e: {avail_column
-00001290: 737d 2229 0d0a 2020 2020 2020 2020 7365  s}")..        se
-000012a0: 6c66 2e70 6174 6965 6e74 5f69 645f 636f  lf.patient_id_co
-000012b0: 6c20 3d20 7061 7469 656e 745f 6964 5f63  l = patient_id_c
-000012c0: 6f6c 0d0a 2020 2020 0d0a 2020 2020 4063  ol..    ..    @c
-000012d0: 6c61 7373 6d65 7468 6f64 0d0a 2020 2020  lassmethod..    
-000012e0: 6465 6620 6672 6f6d 5f64 6628 636c 732c  def from_df(cls,
-000012f0: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
-00001300: 2020 2020 6461 7461 3a20 7064 2e44 6174      data: pd.Dat
-00001310: 6146 7261 6d65 2c0d 0a20 2020 2020 2020  aFrame,..       
-00001320: 2020 2020 2020 2020 2020 2069 735f 6e6f             is_no
-00001330: 726d 616c 697a 6564 3a20 626f 6f6c 2c0d  rmalized: bool,.
-00001340: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00001350: 2020 2073 7061 7273 6966 793a 2062 6f6f     sparsify: boo
-00001360: 6c20 3d20 4661 6c73 652c 0d0a 2020 2020  l = False,..    
-00001370: 2020 2020 2020 2020 2020 2020 2020 6f62                ob
-00001380: 733a 204f 7074 696f 6e61 6c5b 7064 2e44  s: Optional[pd.D
-00001390: 6174 6146 7261 6d65 5d20 3d20 4e6f 6e65  ataFrame] = None
-000013a0: 2c0d 0a20 2020 2020 2020 2020 2020 2020  ,..             
-000013b0: 2020 2020 2076 6172 3a20 4f70 7469 6f6e       var: Option
-000013c0: 616c 5b70 642e 4461 7461 4672 616d 655d  al[pd.DataFrame]
-000013d0: 203d 204e 6f6e 652c 0d0a 2020 2020 2020   = None,..      
-000013e0: 2020 2020 2020 2020 2020 2020 7061 7469              pati
-000013f0: 656e 745f 6964 5f63 6f6c 3a20 4f70 7469  ent_id_col: Opti
-00001400: 6f6e 616c 5b73 7472 5d20 3d20 4e6f 6e65  onal[str] = None
-00001410: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
-00001420: 2020 2020 293a 0d0a 2020 2020 2020 2020      ):..        
-00001430: 2222 2243 7265 6174 6573 2061 203a 636c  """Creates a :cl
-00001440: 6173 733a 607e 636e 6d66 736e 732e 6461  ass:`~cnmfsns.da
-00001450: 7461 7365 742e 4461 7461 7365 7460 206f  taset.Dataset` o
-00001460: 626a 6563 7420 6672 6f6d 2061 2070 616e  bject from a pan
-00001470: 6461 7320 4461 7461 4672 616d 652e 0d0a  das DataFrame...
-00001480: 0d0a 2020 2020 2020 2020 3a70 6172 616d  ..        :param
-00001490: 2064 6174 613a 2041 6e20 6f62 7365 7276   data: An observ
-000014a0: 6174 696f 6e73 20c3 9720 7661 7269 6162  ations .. variab
-000014b0: 6c65 7320 6461 7461 0d0a 2020 2020 2020  les data..      
-000014c0: 2020 3a74 7970 6520 6461 7461 3a20 7064    :type data: pd
-000014d0: 2e44 6174 6146 7261 6d65 0d0a 2020 2020  .DataFrame..    
-000014e0: 2020 2020 3a70 6172 616d 2069 735f 6e6f      :param is_no
-000014f0: 726d 616c 697a 6564 3a20 5370 6563 6966  rmalized: Specif
-00001500: 7920 6966 2064 6174 6120 6973 2061 6c72  y if data is alr
-00001510: 6561 6479 206e 6f72 6d61 6c69 7a65 6420  eady normalized 
-00001520: 6f72 2077 6865 7468 6572 206e 6f74 2e20  or whether not. 
-00001530: 5261 7720 6461 7461 2077 696c 6c20 6265  Raw data will be
-00001540: 2054 504d 206e 6f72 6d61 6c69 7a65 6420   TPM normalized 
-00001550: 7072 696f 7220 746f 206f 7665 7264 6973  prior to overdis
-00001560: 7065 7273 6564 2067 656e 6520 7365 6c65  persed gene sele
-00001570: 6374 696f 6e2c 2077 6865 7265 6173 2061  ction, whereas a
-00001580: 6c72 6561 6479 206e 6f72 6d61 6c69 7a65  lready normalize
-00001590: 6420 6461 7461 2077 696c 6c20 6e6f 742e  d data will not.
-000015a0: 0d0a 2020 2020 2020 2020 3a74 7970 6520  ..        :type 
-000015b0: 6973 5f6e 6f72 6d61 6c69 7a65 643a 2062  is_normalized: b
-000015c0: 6f6f 6c0d 0a20 2020 2020 2020 203a 7061  ool..        :pa
-000015d0: 7261 6d20 7370 6172 7369 6679 3a20 5374  ram sparsify: St
-000015e0: 6f72 6520 6461 7461 2061 7320 6120 7370  ore data as a sp
-000015f0: 6172 7365 206d 6174 7269 782e 205b 4e6f  arse matrix. [No
-00001600: 7465 2074 6861 7420 7468 6973 2066 6561  te that this fea
-00001610: 7475 7265 2069 7320 6578 7065 7269 6d65  ture is experime
-00001620: 6e74 616c 5d2c 2064 6566 6175 6c74 7320  ntal], defaults 
-00001630: 746f 2046 616c 7365 0d0a 2020 2020 2020  to False..      
-00001640: 2020 3a74 7970 6520 7370 6172 7369 6679    :type sparsify
-00001650: 3a20 626f 6f6c 2c20 6f70 7469 6f6e 616c  : bool, optional
-00001660: 0d0a 2020 2020 2020 2020 3a70 6172 616d  ..        :param
-00001670: 206f 6273 3a20 416e 206f 6273 6572 7661   obs: An observa
-00001680: 7469 6f6e 7320 c397 206d 6574 6164 6174  tions .. metadat
-00001690: 6120 6d61 7472 6978 2c20 6465 6661 756c  a matrix, defaul
-000016a0: 7473 2074 6f20 4e6f 6e65 0d0a 2020 2020  ts to None..    
-000016b0: 2020 2020 3a74 7970 6520 6f62 733a 2060      :type obs: `
-000016c0: 7064 2e44 6174 6146 7261 6d65 602c 206f  pd.DataFrame`, o
-000016d0: 7074 696f 6e61 6c0d 0a20 2020 2020 2020  ptional..       
-000016e0: 203a 7061 7261 6d20 7661 723a 2041 2076   :param var: A v
-000016f0: 6172 6961 626c 6573 20c3 9720 6d65 7461  ariables .. meta
-00001700: 6461 7461 206d 6174 7269 782c 2064 6566  data matrix, def
-00001710: 6175 6c74 7320 746f 204e 6f6e 650d 0a20  aults to None.. 
-00001720: 2020 2020 2020 203a 7479 7065 2076 6172         :type var
-00001730: 3a20 6070 642e 4461 7461 4672 616d 6560  : `pd.DataFrame`
-00001740: 2c20 6f70 7469 6f6e 616c 0d0a 2020 2020  , optional..    
-00001750: 2020 2020 3a70 6172 616d 2070 6174 6965      :param patie
-00001760: 6e74 5f69 645f 636f 6c3a 204e 616d 6520  nt_id_col: Name 
-00001770: 6f66 206d 6574 6164 6174 6120 6c61 7965  of metadata laye
-00001780: 7220 7769 7468 2070 6174 6965 6e74 2049  r with patient I
-00001790: 4420 696e 666f 726d 6174 696f 6e2c 2064  D information, d
-000017a0: 6566 6175 6c74 7320 746f 204e 6f6e 650d  efaults to None.
-000017b0: 0a20 2020 2020 2020 203a 7479 7065 2070  .        :type p
-000017c0: 6174 6965 6e74 5f69 645f 636f 6c3a 2073  atient_id_col: s
-000017d0: 7472 2c20 6f70 7469 6f6e 616c 0d0a 2020  tr, optional..  
-000017e0: 2020 2020 2020 3a72 6574 7572 6e3a 204f        :return: O
-000017f0: 626a 6563 7420 7769 7468 2065 7870 7265  bject with expre
-00001800: 7373 696f 6e20 616e 6420 6d65 7461 6461  ssion and metada
-00001810: 7461 0d0a 2020 2020 2020 2020 3a72 7479  ta..        :rty
-00001820: 7065 3a20 3a63 6c61 7373 3a60 7e63 6e6d  pe: :class:`~cnm
-00001830: 6673 6e73 2e64 6174 6173 6574 2e44 6174  fsns.dataset.Dat
-00001840: 6173 6574 600d 0a20 2020 2020 2020 2022  aset`..        "
-00001850: 2222 0d0a 2020 2020 2020 2020 6966 2076  ""..        if v
-00001860: 6172 2069 7320 6e6f 7420 4e6f 6e65 3a0d  ar is not None:.
-00001870: 0a20 2020 2020 2020 2020 2020 2076 6172  .            var
-00001880: 203d 2076 6172 2e72 6569 6e64 6578 2864   = var.reindex(d
-00001890: 6174 612e 636f 6c75 6d6e 7329 0d0a 2020  ata.columns)..  
-000018a0: 2020 2020 2020 6966 2073 7061 7273 6966        if sparsif
-000018b0: 793a 0d0a 2020 2020 2020 2020 2020 2020  y:..            
-000018c0: 6461 7461 203d 2073 702e 6373 725f 6d61  data = sp.csr_ma
-000018d0: 7472 6978 2864 6174 612e 7661 6c75 6573  trix(data.values
-000018e0: 290d 0a20 2020 2020 2020 2064 6174 6120  )..        data 
-000018f0: 3d20 6461 7461 2e61 7374 7970 6528 2266  = data.astype("f
-00001900: 6c6f 6174 3332 2229 0d0a 2020 2020 2020  loat32")..      
-00001910: 2020 756e 7320 3d20 7b22 6869 7374 6f72    uns = {"histor
-00001920: 7922 3a20 7b7d 2c20 226f 6467 223a 7b7d  y": {}, "odg":{}
-00001930: 7d0d 0a20 2020 2020 2020 2061 6461 7461  }..        adata
-00001940: 203d 2061 642e 416e 6e44 6174 6128 583d   = ad.AnnData(X=
-00001950: 6461 7461 2c20 7661 723d 7661 722c 2075  data, var=var, u
-00001960: 6e73 3d75 6e73 290d 0a20 2020 2020 2020  ns=uns)..       
-00001970: 2064 6174 6173 6574 203d 2063 6c73 2861   dataset = cls(a
-00001980: 6461 7461 3d61 6461 7461 2c20 7061 7469  data=adata, pati
-00001990: 656e 745f 6964 5f63 6f6c 3d70 6174 6965  ent_id_col=patie
-000019a0: 6e74 5f69 645f 636f 6c29 0d0a 2020 2020  nt_id_col)..    
-000019b0: 2020 2020 6966 206f 6273 2069 7320 6e6f      if obs is no
-000019c0: 7420 4e6f 6e65 3a0d 0a20 2020 2020 2020  t None:..       
-000019d0: 2020 2020 2064 6174 6173 6574 2e75 7064       dataset.upd
-000019e0: 6174 655f 6f62 7328 6f62 733d 6f62 7329  ate_obs(obs=obs)
-000019f0: 200d 0a20 2020 2020 2020 2072 6574 7572   ..        retur
-00001a00: 6e20 6461 7461 7365 740d 0a20 2020 200d  n dataset..    .
-00001a10: 0a20 2020 2040 636c 6173 736d 6574 686f  .    @classmetho
-00001a20: 640d 0a20 2020 2064 6566 2066 726f 6d5f  d..    def from_
-00001a30: 6835 6164 2863 6c73 2c0d 0a20 2020 2020  h5ad(cls,..     
-00001a40: 2020 2020 2020 2020 2020 2020 2068 3561               h5a
-00001a50: 645f 6669 6c65 3a20 7374 722c 0d0a 2020  d_file: str,..  
-00001a60: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00001a70: 7061 7469 656e 745f 6964 5f63 6f6c 3a20  patient_id_col: 
-00001a80: 4f70 7469 6f6e 616c 5b73 7472 5d20 3d20  Optional[str] = 
-00001a90: 4e6f 6e65 2c0d 0a20 2020 2020 2020 2020  None,..         
-00001aa0: 2020 2020 2020 2020 2066 6f72 6365 5f6d           force_m
-00001ab0: 6967 7261 7465 3d46 616c 7365 2c20 6261  igrate=False, ba
-00001ac0: 636b 6564 3d46 616c 7365 0d0a 2020 2020  cked=False..    
-00001ad0: 2020 2020 2020 2020 2020 2020 2020 293a                ):
-00001ae0: 0d0a 2020 2020 2020 2020 2222 2243 7265  ..        """Cre
-00001af0: 6174 6573 2061 203a 636c 6173 733a 607e  ates a :class:`~
-00001b00: 636e 6d66 736e 732e 6461 7461 7365 742e  cnmfsns.dataset.
-00001b10: 4461 7461 7365 7460 206f 626a 6563 7420  Dataset` object 
-00001b20: 6672 6f6d 2061 6e20 416e 6e44 6174 612d  from an AnnData-
-00001b30: 636f 6d70 6174 6962 6c65 202e 6835 6164  compatible .h5ad
-00001b40: 2066 696c 652e 0d0a 0d0a 2020 2020 2020   file.....      
-00001b50: 2020 3a70 6172 616d 2068 3561 645f 6669    :param h5ad_fi
-00001b60: 6c65 3a20 5061 7468 2074 6f20 2e68 3561  le: Path to .h5a
-00001b70: 6420 6669 6c65 2070 726f 6475 6365 6420  d file produced 
-00001b80: 6279 2073 6361 6e70 792c 2041 6e6e 4461  by scanpy, AnnDa
-00001b90: 7461 2c20 6f72 2063 4e4d 462d 534e 530d  ta, or cNMF-SNS.
-00001ba0: 0a20 2020 2020 2020 203a 7479 7065 2068  .        :type h
-00001bb0: 3561 645f 6669 6c65 3a20 7374 720d 0a20  5ad_file: str.. 
-00001bc0: 2020 2020 2020 203a 7061 7261 6d20 7061         :param pa
-00001bd0: 7469 656e 745f 6964 5f63 6f6c 3a20 4e61  tient_id_col: Na
-00001be0: 6d65 206f 6620 6d65 7461 6461 7461 206c  me of metadata l
-00001bf0: 6179 6572 2077 6974 6820 7061 7469 656e  ayer with patien
-00001c00: 7420 4944 2069 6e66 6f72 6d61 7469 6f6e  t ID information
-00001c10: 2c20 6465 6661 756c 7473 2074 6f20 4e6f  , defaults to No
-00001c20: 6e65 0d0a 2020 2020 2020 2020 3a74 7970  ne..        :typ
-00001c30: 6520 7061 7469 656e 745f 6964 5f63 6f6c  e patient_id_col
-00001c40: 3a20 7374 722c 206f 7074 696f 6e61 6c0d  : str, optional.
-00001c50: 0a20 2020 2020 2020 203a 7061 7261 6d20  .        :param 
-00001c60: 666f 7263 655f 6d69 6772 6174 653a 2066  force_migrate: f
-00001c70: 6f72 6365 7320 636f 6e76 6572 7369 6f6e  orces conversion
-00001c80: 206f 6620 416e 6e44 6174 6120 6f62 6a65   of AnnData obje
-00001c90: 6374 7320 6576 656e 2077 6865 6e20 6164  cts even when ad
-00001ca0: 6174 612e 5820 616e 6420 6164 6174 612e  ata.X and adata.
-00001cb0: 7261 772e 5820 6172 6520 6e6f 7420 6c69  raw.X are not li
-00001cc0: 6e65 6172 6c79 2073 6361 6c65 6420 7265  nearly scaled re
-00001cd0: 6c61 7469 7665 2074 6f20 6561 6368 206f  lative to each o
-00001ce0: 7468 6572 2c20 6465 6661 756c 7473 2074  ther, defaults t
-00001cf0: 6f20 4661 6c73 650d 0a20 2020 2020 2020  o False..       
-00001d00: 203a 7479 7065 2066 6f72 6365 5f6d 6967   :type force_mig
-00001d10: 7261 7465 3a20 626f 6f6c 2c20 6f70 7469  rate: bool, opti
-00001d20: 6f6e 616c 0d0a 2020 2020 2020 2020 3a70  onal..        :p
-00001d30: 6172 616d 2062 6163 6b65 643a 2055 7365  aram backed: Use
-00001d40: 2062 6163 6b65 6420 6d6f 6465 2074 6f20   backed mode to 
-00001d50: 6f70 656e 2068 3561 6420 6669 6c65 2e20  open h5ad file. 
-00001d60: 5468 6973 2063 616e 2073 6176 6520 6d65  This can save me
-00001d70: 6d6f 7279 2077 6865 6e20 7468 6520 6461  mory when the da
-00001d80: 7461 7365 7420 6973 2076 6572 7920 6c61  taset is very la
-00001d90: 7267 652c 2062 7574 2069 7320 6e6f 7420  rge, but is not 
-00001da0: 636f 6d70 6174 6962 6c65 2077 6974 6820  compatible with 
-00001db0: 6835 6164 2066 696c 6573 2070 726f 6475  h5ad files produ
-00001dc0: 6365 6420 6f75 7473 6964 6520 6f66 2063  ced outside of c
-00001dd0: 4e4d 462d 534e 532c 2064 6566 6175 6c74  NMF-SNS, default
-00001de0: 7320 746f 2046 616c 7365 0d0a 2020 2020  s to False..    
-00001df0: 2020 2020 3a74 7970 6520 6261 636b 6564      :type backed
-00001e00: 3a20 626f 6f6c 2c20 6f70 7469 6f6e 616c  : bool, optional
-00001e10: 0d0a 2020 2020 2020 2020 3a72 6574 7572  ..        :retur
-00001e20: 6e3a 204f 626a 6563 7420 7769 7468 2065  n: Object with e
-00001e30: 7870 7265 7373 696f 6e20 616e 6420 6d65  xpression and me
-00001e40: 7461 6461 7461 0d0a 2020 2020 2020 2020  tadata..        
-00001e50: 3a72 7479 7065 3a20 3a63 6c61 7373 3a60  :rtype: :class:`
-00001e60: 7e63 6e6d 6673 6e73 2e64 6174 6173 6574  ~cnmfsns.dataset
-00001e70: 2e44 6174 6173 6574 600d 0a20 2020 2020  .Dataset`..     
-00001e80: 2020 2022 2222 0d0a 2020 2020 2020 2020     """..        
-00001e90: 6164 6174 6120 3d20 6164 2e72 6561 645f  adata = ad.read_
-00001ea0: 6835 6164 2868 3561 645f 6669 6c65 2c20  h5ad(h5ad_file, 
-00001eb0: 6261 636b 6564 3d62 6163 6b65 6429 0d0a  backed=backed)..
-00001ec0: 2020 2020 2020 2020 6461 7461 7365 7420          dataset 
-00001ed0: 3d20 636c 7328 6164 6174 613d 6164 6174  = cls(adata=adat
-00001ee0: 612c 2070 6174 6965 6e74 5f69 645f 636f  a, patient_id_co
-00001ef0: 6c3d 7061 7469 656e 745f 6964 5f63 6f6c  l=patient_id_col
-00001f00: 2c20 666f 7263 655f 6d69 6772 6174 653d  , force_migrate=
-00001f10: 666f 7263 655f 6d69 6772 6174 6529 0d0a  force_migrate)..
-00001f20: 2020 2020 2020 2020 7265 7475 726e 2064          return d
-00001f30: 6174 6173 6574 0d0a 2020 2020 0d0a 2020  ataset..    ..  
-00001f40: 2020 4070 726f 7065 7274 790d 0a20 2020    @property..   
-00001f50: 2064 6566 2069 735f 6e6f 726d 616c 697a   def is_normaliz
-00001f60: 6564 2873 656c 6629 3a0d 0a20 2020 2020  ed(self):..     
-00001f70: 2020 2022 2222 4f75 7470 7574 7320 7468     """Outputs th
-00001f80: 6520 6e6f 726d 616c 697a 6174 696f 6e20  e normalization 
-00001f90: 7374 6174 7573 206f 6620 7468 6520 6461  status of the da
-00001fa0: 7461 7365 742e 0d0a 0d0a 2020 2020 2020  taset.....      
-00001fb0: 2020 3a72 6574 7572 6e3a 2054 7275 6520    :return: True 
-00001fc0: 6966 2064 6174 6173 6574 2063 6f6e 7461  if dataset conta
-00001fd0: 696e 7320 6e6f 726d 616c 697a 6564 2064  ins normalized d
-00001fe0: 6174 612c 2046 616c 7365 2069 6620 6974  ata, False if it
-00001ff0: 2069 7320 7261 7720 6461 7461 2e0d 0a20   is raw data... 
-00002000: 2020 2020 2020 203a 7274 7970 653a 2062         :rtype: b
-00002010: 6f6f 6c0d 0a20 2020 2020 2020 2022 2222  ool..        """
-00002020: 0d0a 2020 2020 2020 2020 7265 7475 726e  ..        return
-00002030: 2073 656c 662e 6164 6174 612e 756e 735b   self.adata.uns[
-00002040: 2269 735f 6e6f 726d 616c 697a 6564 225d  "is_normalized"]
-00002050: 0d0a 2020 2020 0d0a 2020 2020 4069 735f  ..    ..    @is_
-00002060: 6e6f 726d 616c 697a 6564 2e73 6574 7465  normalized.sette
-00002070: 720d 0a20 2020 2064 6566 2069 735f 6e6f  r..    def is_no
-00002080: 726d 616c 697a 6564 2873 656c 662c 2076  rmalized(self, v
-00002090: 616c 7565 3a20 626f 6f6c 293a 0d0a 2020  alue: bool):..  
-000020a0: 2020 2020 2020 7365 6c66 2e61 6461 7461        self.adata
-000020b0: 2e75 6e73 5b22 6973 5f6e 6f72 6d61 6c69  .uns["is_normali
-000020c0: 7a65 6422 5d20 3d20 7661 6c75 650d 0a20  zed"] = value.. 
-000020d0: 2020 200d 0a20 2020 2040 7072 6f70 6572     ..    @proper
-000020e0: 7479 0d0a 2020 2020 6465 6620 7061 7469  ty..    def pati
-000020f0: 656e 745f 6964 5f63 6f6c 2873 656c 6629  ent_id_col(self)
-00002100: 3a0d 0a20 2020 2020 2020 2022 2222 4f75  :..        """Ou
-00002110: 7470 7574 7320 7468 6520 6e6f 726d 616c  tputs the normal
-00002120: 697a 6174 696f 6e20 7374 6174 7573 206f  ization status o
-00002130: 6620 7468 6520 6461 7461 7365 742e 0d0a  f the dataset...
-00002140: 0d0a 2020 2020 2020 2020 3a72 6574 7572  ..        :retur
-00002150: 6e3a 2054 7275 6520 6966 2064 6174 6173  n: True if datas
-00002160: 6574 2063 6f6e 7461 696e 7320 6e6f 726d  et contains norm
-00002170: 616c 697a 6564 2064 6174 612c 2046 616c  alized data, Fal
-00002180: 7365 2069 6620 6974 2069 7320 7261 7720  se if it is raw 
-00002190: 6461 7461 2e0d 0a20 2020 2020 2020 203a  data...        :
-000021a0: 7274 7970 653a 2062 6f6f 6c0d 0a20 2020  rtype: bool..   
-000021b0: 2020 2020 2022 2222 0d0a 2020 2020 2020       """..      
-000021c0: 2020 7265 7475 726e 2073 656c 662e 6164    return self.ad
-000021d0: 6174 612e 756e 735b 2270 6174 6965 6e74  ata.uns["patient
-000021e0: 5f69 645f 636f 6c22 5d0d 0a20 2020 200d  _id_col"]..    .
-000021f0: 0a20 2020 2040 7061 7469 656e 745f 6964  .    @patient_id
-00002200: 5f63 6f6c 2e73 6574 7465 720d 0a20 2020  _col.setter..   
-00002210: 2064 6566 2070 6174 6965 6e74 5f69 645f   def patient_id_
-00002220: 636f 6c28 7365 6c66 2c20 7661 6c75 653a  col(self, value:
-00002230: 2073 7472 293a 0d0a 2020 2020 2020 2020   str):..        
-00002240: 7365 6c66 2e61 6461 7461 2e75 6e73 5b22  self.adata.uns["
-00002250: 7061 7469 656e 745f 6964 5f63 6f6c 225d  patient_id_col"]
-00002260: 203d 2076 616c 7565 0d0a 0d0a 2020 2020   = value....    
-00002270: 4070 726f 7065 7274 790d 0a20 2020 2064  @property..    d
-00002280: 6566 2063 6e6d 6673 6e73 5f76 6572 7369  ef cnmfsns_versi
-00002290: 6f6e 2873 656c 6629 3a0d 0a20 2020 2020  on(self):..     
-000022a0: 2020 2022 2222 634e 4d46 2d53 4e53 2076     """cNMF-SNS v
-000022b0: 6572 7369 6f6e 2075 7365 6420 746f 2063  ersion used to c
-000022c0: 7265 6174 6520 7468 6520 6461 7461 7365  reate the datase
-000022d0: 740d 0a0d 0a20 2020 2020 2020 203a 7265  t....        :re
-000022e0: 7475 726e 3a20 7665 7273 696f 6e0d 0a20  turn: version.. 
-000022f0: 2020 2020 2020 203a 7274 7970 653a 2073         :rtype: s
-00002300: 7472 0d0a 2020 2020 2020 2020 2222 220d  tr..        """.
-00002310: 0a20 2020 2020 2020 2069 6620 2263 6e6d  .        if "cnm
-00002320: 6673 6e73 5f76 6572 7369 6f6e 2220 696e  fsns_version" in
-00002330: 2073 656c 662e 6164 6174 612e 756e 733a   self.adata.uns:
-00002340: 0d0a 2020 2020 2020 2020 2020 2020 7665  ..            ve
-00002350: 7273 696f 6e20 3d20 7365 6c66 2e61 6461  rsion = self.ada
-00002360: 7461 2e75 6e73 5b22 636e 6d66 736e 735f  ta.uns["cnmfsns_
-00002370: 7665 7273 696f 6e22 5d0d 0a20 2020 2020  version"]..     
-00002380: 2020 2065 6c73 653a 0d0a 2020 2020 2020     else:..      
-00002390: 2020 2020 2020 7665 7273 696f 6e20 3d20        version = 
-000023a0: 4e6f 6e65 0d0a 2020 2020 2020 2020 7265  None..        re
-000023b0: 7475 726e 2076 6572 7369 6f6e 0d0a 2020  turn version..  
-000023c0: 2020 2020 2020 0d0a 2020 2020 4070 726f        ..    @pro
-000023d0: 7065 7274 790d 0a20 2020 2064 6566 2068  perty..    def h
-000023e0: 6173 5f63 6e6d 665f 7265 7375 6c74 7328  as_cnmf_results(
-000023f0: 7365 6c66 293a 0d0a 2020 2020 2020 2020  self):..        
-00002400: 2222 2254 6573 7420 666f 7220 7765 6874  """Test for weht
-00002410: 6865 7220 4461 7461 7365 7420 636f 6e74  her Dataset cont
-00002420: 6169 6e73 2063 4e4d 4620 7265 7375 6c74  ains cNMF result
-00002430: 7320 666f 7220 7468 6520 6461 7461 7365  s for the datase
-00002440: 740d 0a0d 0a20 2020 2020 2020 203a 7265  t....        :re
-00002450: 7475 726e 3a20 5768 6574 6865 7220 636f  turn: Whether co
-00002460: 6d70 6c65 7465 2063 4e4d 4620 7265 7375  mplete cNMF resu
-00002470: 6c74 7320 6172 6520 636f 6e74 6169 6e65  lts are containe
-00002480: 6420 666f 7220 6174 206c 6561 7374 2031  d for at least 1
-00002490: 2072 616e 6b20 286b 290d 0a20 2020 2020   rank (k)..     
-000024a0: 2020 203a 7274 7970 653a 2062 6f6f 6c0d     :rtype: bool.
-000024b0: 0a20 2020 2020 2020 2022 2222 0d0a 2020  .        """..  
-000024c0: 2020 2020 2020 6d61 7472 6978 5f63 6865        matrix_che
-000024d0: 636b 7320 3d20 5b0d 0a20 2020 2020 2020  cks = [..       
-000024e0: 2020 2020 2022 636e 6d66 5f75 7361 6765       "cnmf_usage
-000024f0: 2220 696e 2073 656c 662e 6164 6174 612e  " in self.adata.
-00002500: 6f62 736d 2c0d 0a20 2020 2020 2020 2020  obsm,..         
-00002510: 2020 2022 636e 6d66 5f67 6570 5f73 636f     "cnmf_gep_sco
-00002520: 7265 2220 696e 2073 656c 662e 6164 6174  re" in self.adat
-00002530: 612e 7661 726d 2c0d 0a20 2020 2020 2020  a.varm,..       
-00002540: 2020 2020 2022 636e 6d66 5f67 6570 5f74       "cnmf_gep_t
-00002550: 706d 2220 696e 2073 656c 662e 6164 6174  pm" in self.adat
-00002560: 612e 7661 726d 2c0d 0a20 2020 2020 2020  a.varm,..       
-00002570: 2020 2020 2022 636e 6d66 5f67 6570 5f72       "cnmf_gep_r
-00002580: 6177 2220 696e 2073 656c 662e 6164 6174  aw" in self.adat
-00002590: 612e 7661 726d 2c0d 0a20 2020 2020 2020  a.varm,..       
-000025a0: 2020 2020 2022 6b76 616c 7322 2069 6e20       "kvals" in 
-000025b0: 7365 6c66 2e61 6461 7461 2e75 6e73 0d0a  self.adata.uns..
-000025c0: 2020 2020 2020 2020 5d0d 0a20 2020 2020          ]..     
-000025d0: 2020 2072 6574 7572 6e20 616c 6c28 6d61     return all(ma
-000025e0: 7472 6978 5f63 6865 636b 7329 0d0a 2020  trix_checks)..  
-000025f0: 2020 2020 2020 0d0a 2020 2020 4070 726f        ..    @pro
-00002600: 7065 7274 790d 0a20 2020 2064 6566 206f  perty..    def o
-00002610: 7665 7264 6973 7065 7273 6564 5f67 656e  verdispersed_gen
-00002620: 6573 2873 656c 6629 3a0d 0a20 2020 2020  es(self):..     
-00002630: 2020 2022 2222 4f76 6572 6469 7370 6572     """Overdisper
-00002640: 7365 6420 6765 6e65 206c 6973 7420 7573  sed gene list us
-00002650: 6564 2066 6f72 2063 4e4d 460d 0a0d 0a20  ed for cNMF.... 
-00002660: 2020 2020 2020 203a 7265 7475 726e 3a20         :return: 
-00002670: 6765 6e65 206c 6973 740d 0a20 2020 2020  gene list..     
-00002680: 2020 203a 7274 7970 653a 206c 6973 740d     :rtype: list.
-00002690: 0a20 2020 2020 2020 2022 2222 0d0a 2020  .        """..  
-000026a0: 2020 2020 2020 7265 7475 726e 2073 656c        return sel
-000026b0: 662e 6164 6174 612e 756e 735b 2267 656e  f.adata.uns["gen
-000026c0: 655f 6c69 7374 225d 0d0a 0d0a 0d0a 2020  e_list"]......  
-000026d0: 2020 4063 6e6d 6673 6e73 5f76 6572 7369    @cnmfsns_versi
-000026e0: 6f6e 2e73 6574 7465 720d 0a20 2020 2064  on.setter..    d
-000026f0: 6566 2063 6e6d 6673 6e73 5f76 6572 7369  ef cnmfsns_versi
-00002700: 6f6e 2873 656c 662c 2076 616c 7565 3a20  on(self, value: 
-00002710: 626f 6f6c 293a 0d0a 2020 2020 2020 2020  bool):..        
-00002720: 7365 6c66 2e61 6461 7461 2e75 6e73 5b22  self.adata.uns["
-00002730: 636e 6d66 736e 735f 7665 7273 696f 6e22  cnmfsns_version"
-00002740: 5d20 3d20 7661 6c75 650d 0a0d 0a20 2020  ] = value....   
-00002750: 2020 2020 200d 0a20 2020 2064 6566 2075       ..    def u
-00002760: 7064 6174 655f 6f62 7328 7365 6c66 2c20  pdate_obs(self, 
-00002770: 6f62 7329 3a0d 0a20 2020 2020 2020 2022  obs):..        "
-00002780: 2222 5570 6461 7465 2074 6865 206f 6273  ""Update the obs
-00002790: 6572 7661 7469 6f6e 206d 6574 6164 6174  ervation metadat
-000027a0: 6120 7769 7468 2061 206e 6577 206d 6574  a with a new met
-000027b0: 6164 6174 6120 6d61 7472 6978 0d0a 0d0a  adata matrix....
-000027c0: 2020 2020 2020 2020 3a70 6172 616d 206f          :param o
-000027d0: 6273 3a20 416e 206f 6273 6572 7661 7469  bs: An observati
-000027e0: 6f6e 7320 c397 206d 6574 6164 6174 6120  ons .. metadata 
-000027f0: 6d61 7472 6978 2c20 6465 6661 756c 7473  matrix, defaults
-00002800: 2074 6f20 4e6f 6e65 0d0a 2020 2020 2020   to None..      
-00002810: 2020 3a74 7970 6520 6f62 733a 2060 7064    :type obs: `pd
-00002820: 2e44 6174 6146 7261 6d65 602c 206f 7074  .DataFrame`, opt
-00002830: 696f 6e61 6c0d 0a20 2020 2020 2020 2022  ional..        "
-00002840: 2222 0d0a 2020 2020 2020 2020 2320 636f  ""..        # co
-00002850: 6e76 6572 7420 276f 626a 6563 7427 2064  nvert 'object' d
-00002860: 7479 7065 2074 6f20 6361 7465 676f 7269  type to categori
-00002870: 6361 6c2c 2063 6f6e 7665 7274 696e 6720  cal, converting 
-00002880: 626f 6f6c 2076 616c 7565 7320 746f 2073  bool values to s
-00002890: 7472 696e 6773 2061 7320 7468 6573 6520  trings as these 
-000028a0: 6172 6520 6e6f 7420 7375 7070 6f72 7465  are not supporte
-000028b0: 6420 6279 2041 6e6e 4461 7461 206f 6e2d  d by AnnData on-
-000028c0: 6469 736b 2066 6f72 6d61 740d 0a20 2020  disk format..   
-000028d0: 2020 2020 2066 6f72 2063 6f6c 2069 6e20       for col in 
-000028e0: 6f62 732e 7365 6c65 6374 5f64 7479 7065  obs.select_dtype
-000028f0: 7328 696e 636c 7564 653d 2822 626f 6f6c  s(include=("bool
-00002900: 222c 2022 6f62 6a65 6374 2229 292e 636f  ", "object")).co
-00002910: 6c75 6d6e 733a 0d0a 2020 2020 2020 2020  lumns:..        
-00002920: 2020 2020 6f62 735b 636f 6c5d 203d 206f      obs[col] = o
-00002930: 6273 5b63 6f6c 5d2e 6173 7479 7065 2822  bs[col].astype("
-00002940: 7374 7222 292e 6173 7479 7065 2822 6361  str").astype("ca
-00002950: 7465 676f 7279 2229 0d0a 2020 2020 2020  tegory")..      
-00002960: 2020 6d69 7373 696e 675f 7361 6d70 6c65    missing_sample
-00002970: 735f 696e 5f58 203d 206f 6273 2e69 6e64  s_in_X = obs.ind
-00002980: 6578 2e64 6966 6665 7265 6e63 6528 7365  ex.difference(se
-00002990: 6c66 2e61 6461 7461 2e6f 6273 2e69 6e64  lf.adata.obs.ind
-000029a0: 6578 292e 6173 7479 7065 2873 7472 292e  ex).astype(str).
-000029b0: 746f 5f6c 6973 7428 290d 0a20 2020 2020  to_list()..     
-000029c0: 2020 2069 6620 6d69 7373 696e 675f 7361     if missing_sa
-000029d0: 6d70 6c65 735f 696e 5f58 3a0d 0a20 2020  mples_in_X:..   
-000029e0: 2020 2020 2020 2020 206c 6f67 6769 6e67           logging
-000029f0: 2e77 6172 6e69 6e67 2822 5468 6520 666f  .warning("The fo
-00002a00: 6c6c 6f77 696e 6720 7361 6d70 6c65 7320  llowing samples 
-00002a10: 696e 2074 6865 206d 6574 6164 6174 6120  in the metadata 
-00002a20: 7765 7265 206e 6f74 2070 7265 7365 6e74  were not present
-00002a30: 2069 6e20 7468 6520 6461 7461 2028 6061   in the data (`a
-00002a40: 6461 7461 2e58 6029 3a5c 6e20 202d 2022  data.X`):\n  - "
-00002a50: 202b 2022 5c6e 2020 2d20 222e 6a6f 696e   + "\n  - ".join
-00002a60: 286d 6973 7369 6e67 5f73 616d 706c 6573  (missing_samples
-00002a70: 5f69 6e5f 5829 290d 0a20 2020 2020 2020  _in_X))..       
-00002a80: 206d 6973 7369 6e67 5f73 616d 706c 6573   missing_samples
-00002a90: 5f69 6e5f 6d64 203d 2073 656c 662e 6164  _in_md = self.ad
-00002aa0: 6174 612e 6f62 732e 696e 6465 782e 6469  ata.obs.index.di
-00002ab0: 6666 6572 656e 6365 286f 6273 2e69 6e64  fference(obs.ind
-00002ac0: 6578 292e 6173 7479 7065 2873 7472 292e  ex).astype(str).
-00002ad0: 746f 5f6c 6973 7428 290d 0a20 2020 2020  to_list()..     
-00002ae0: 2020 2069 6620 6d69 7373 696e 675f 7361     if missing_sa
-00002af0: 6d70 6c65 735f 696e 5f6d 643a 0d0a 2020  mples_in_md:..  
-00002b00: 2020 2020 2020 2020 2020 6c6f 6767 696e            loggin
-00002b10: 672e 7761 726e 696e 6728 2254 6865 2066  g.warning("The f
-00002b20: 6f6c 6c6f 7769 6e67 2073 616d 706c 6573  ollowing samples
-00002b30: 2069 6e20 7468 6520 6461 7461 2028 6061   in the data (`a
-00002b40: 6461 7461 2e58 6029 2077 6572 6520 6162  data.X`) were ab
-00002b50: 7365 6e74 2069 6e20 7468 6520 6d65 7461  sent in the meta
-00002b60: 6461 7461 3a5c 6e20 202d 2022 202b 2022  data:\n  - " + "
-00002b70: 5c6e 2020 2d20 222e 6a6f 696e 286d 6973  \n  - ".join(mis
-00002b80: 7369 6e67 5f73 616d 706c 6573 5f69 6e5f  sing_samples_in_
-00002b90: 6d64 2929 0d0a 2020 2020 2020 2020 7365  md))..        se
-00002ba0: 6c66 2e61 6461 7461 2e6f 6273 203d 206f  lf.adata.obs = o
-00002bb0: 6273 2e72 6569 6e64 6578 2873 656c 662e  bs.reindex(self.
-00002bc0: 6164 6174 612e 6f62 732e 696e 6465 7829  adata.obs.index)
-00002bd0: 0d0a 2020 2020 0d0a 2020 2020 6465 6620  ..    ..    def 
-00002be0: 6765 745f 6d65 7461 6461 7461 5f74 7970  get_metadata_typ
-00002bf0: 655f 7375 6d6d 6172 7928 7365 6c66 293a  e_summary(self):
-00002c00: 0d0a 2020 2020 2020 2020 2222 2252 6574  ..        """Ret
-00002c10: 7572 6e20 6120 7072 696e 7461 626c 6520  urn a printable 
-00002c20: 7375 6d6d 6172 7920 6f66 206d 6574 6164  summary of metad
-00002c30: 6174 6120 616e 6420 7661 6c75 6520 7479  ata and value ty
-00002c40: 7065 7320 666f 7220 6561 6368 206d 6574  pes for each met
-00002c50: 6164 6174 6120 6c61 7965 722e 0d0a 0d0a  adata layer.....
-00002c60: 2020 2020 2020 2020 3a72 6574 7572 6e3a          :return:
-00002c70: 2053 756d 6d61 7279 206f 6620 6d65 7461   Summary of meta
-00002c80: 6461 7461 0d0a 2020 2020 2020 2020 3a72  data..        :r
-00002c90: 7479 7065 3a20 7374 720d 0a20 2020 2020  type: str..     
-00002ca0: 2020 2022 2222 0d0a 2020 2020 2020 2020     """..        
-00002cb0: 6d73 6720 3d20 2222 0d0a 2020 2020 2020  msg = ""..      
-00002cc0: 2020 666f 7220 636f 6c20 696e 2073 656c    for col in sel
-00002cd0: 662e 6164 6174 612e 6f62 732e 636f 6c75  f.adata.obs.colu
-00002ce0: 6d6e 733a 0d0a 2020 2020 2020 2020 2020  mns:..          
-00002cf0: 2020 6d73 6720 2b3d 2022 2020 2020 436f    msg += "    Co
-00002d00: 6c75 6d6e 3a20 2220 2b20 636f 6c20 2b20  lumn: " + col + 
-00002d10: 225c 6e22 0d0a 2020 2020 2020 2020 2020  "\n"..          
-00002d20: 2020 666f 7220 7661 6c75 655f 7479 7065    for value_type
-00002d30: 2c20 636f 756e 7420 696e 2073 656c 662e  , count in self.
-00002d40: 6164 6174 612e 6f62 735b 636f 6c5d 2e64  adata.obs[col].d
-00002d50: 726f 706e 6128 292e 6d61 7028 7479 7065  ropna().map(type
-00002d60: 292e 7661 6c75 655f 636f 756e 7473 2829  ).value_counts()
-00002d70: 2e69 7465 6d73 2829 3a0d 0a20 2020 2020  .items():..     
-00002d80: 2020 2020 2020 2020 2020 206d 7367 202b             msg +
-00002d90: 3d20 6622 2020 2020 2020 2020 7b76 616c  = f"        {val
-00002da0: 7565 5f74 7970 657d 3a20 7b63 6f75 6e74  ue_type}: {count
-00002db0: 7d5c 6e22 0d0a 2020 2020 2020 2020 7265  }\n"..        re
-00002dc0: 7475 726e 206d 7367 0d0a 2020 2020 0d0a  turn msg..    ..
-00002dd0: 2020 2020 6465 6620 7772 6974 655f 6835      def write_h5
-00002de0: 6164 2873 656c 662c 0d0a 2020 2020 2020  ad(self,..      
-00002df0: 2020 2020 2020 2020 2020 2020 2066 696c               fil
-00002e00: 656e 616d 653a 2073 7472 293a 0d0a 2020  ename: str):..  
-00002e10: 2020 2020 2020 2222 2257 7269 7465 2064        """Write d
-00002e20: 6174 6173 6574 2074 6f20 2e68 3561 6420  ataset to .h5ad 
-00002e30: 6669 6c65 2e0d 0a0d 0a20 2020 2020 2020  file.....       
-00002e40: 203a 7061 7261 6d20 6669 6c65 6e61 6d65   :param filename
-00002e50: 3a20 6669 6c65 7061 7468 0d0a 2020 2020  : filepath..    
-00002e60: 2020 2020 3a74 7970 6520 6669 6c65 6e61      :type filena
-00002e70: 6d65 3a20 7374 720d 0a20 2020 2020 2020  me: str..       
-00002e80: 2022 2222 0d0a 2020 2020 2020 2020 6669   """..        fi
-00002e90: 6c65 6e61 6d65 203d 206f 732e 7061 7468  lename = os.path
-00002ea0: 2e61 6273 7061 7468 2866 696c 656e 616d  .abspath(filenam
-00002eb0: 6529 0d0a 2020 2020 2020 2020 6c6f 6767  e)..        logg
-00002ec0: 696e 672e 696e 666f 2866 2257 7269 7469  ing.info(f"Writi
-00002ed0: 6e67 2074 6f20 7b66 696c 656e 616d 657d  ng to {filename}
-00002ee0: 2229 0d0a 2020 2020 2020 2020 7365 6c66  ")..        self
-00002ef0: 2e61 6461 7461 2e77 7269 7465 5f68 3561  .adata.write_h5a
-00002f00: 6428 6669 6c65 6e61 6d65 290d 0a20 2020  d(filename)..   
-00002f10: 2020 2020 206c 6f67 6769 6e67 2e69 6e66       logging.inf
-00002f20: 6f28 6622 446f 6e65 2229 0d0a 2020 2020  o(f"Done")..    
-00002f30: 0d0a 2020 2020 6465 6620 746f 5f64 6628  ..    def to_df(
-00002f40: 7365 6c66 2c0d 0a20 2020 2020 2020 2020  self,..         
-00002f50: 2020 2020 206e 6f72 6d61 6c69 7a65 643a       normalized:
-00002f60: 2062 6f6f 6c20 3d20 4661 6c73 6529 3a0d   bool = False):.
-00002f70: 0a20 2020 2020 2020 2022 2222 4765 7420  .        """Get 
-00002f80: 6461 7461 206d 6174 7269 7820 6173 2061  data matrix as a
-00002f90: 2060 7064 2e44 6174 6146 7261 6d65 600d   `pd.DataFrame`.
-00002fa0: 0a0d 0a20 2020 2020 2020 203a 7061 7261  ...        :para
-00002fb0: 6d20 6e6f 726d 616c 697a 6564 3a20 5365  m normalized: Se
-00002fc0: 7420 7472 7565 2066 6f72 2054 504d 206e  t true for TPM n
-00002fd0: 6f72 6d61 6c69 7a65 6420 6f75 7470 7574  ormalized output
-00002fe0: 2c20 6465 6661 756c 7473 2074 6f20 4661  , defaults to Fa
-00002ff0: 6c73 650d 0a20 2020 2020 2020 203a 7479  lse..        :ty
-00003000: 7065 206e 6f72 6d61 6c69 7a65 643a 2062  pe normalized: b
-00003010: 6f6f 6c2c 206f 7074 696f 6e61 6c0d 0a20  ool, optional.. 
-00003020: 2020 2020 2020 203a 7265 7475 726e 3a20         :return: 
-00003030: 6f62 7365 7276 6174 696f 6e73 20c3 9720  observations .. 
-00003040: 7661 7269 6162 6c65 7320 6461 7461 206d  variables data m
-00003050: 6174 7269 780d 0a20 2020 2020 2020 203a  atrix..        :
-00003060: 7274 7970 653a 2070 642e 4461 7461 4672  rtype: pd.DataFr
-00003070: 616d 650d 0a20 2020 2020 2020 2022 2222  ame..        """
-00003080: 0d0a 2020 2020 2020 2020 6466 203d 2073  ..        df = s
-00003090: 656c 662e 6164 6174 612e 746f 5f64 6628  elf.adata.to_df(
-000030a0: 290d 0a20 2020 2020 2020 2069 6620 6e6f  )..        if no
-000030b0: 726d 616c 697a 6564 2061 6e64 206e 6f74  rmalized and not
-000030c0: 2073 656c 662e 6973 5f6e 6f72 6d61 6c69   self.is_normali
-000030d0: 7a65 643a 0d0a 2020 2020 2020 2020 2020  zed:..          
-000030e0: 2020 6466 203d 2064 662e 6469 7628 6466    df = df.div(df
-000030f0: 2e73 756d 2861 7869 733d 3129 2c20 6178  .sum(axis=1), ax
-00003100: 6973 3d30 2920 2a20 3165 3620 2023 2054  is=0) * 1e6  # T
-00003110: 504d 206e 6f72 6d61 6c69 7a61 7469 6f6e  PM normalization
-00003120: 0d0a 2020 2020 2020 2020 7265 7475 726e  ..        return
-00003130: 2064 660d 0a20 2020 2020 2020 200d 0a20   df..        .. 
-00003140: 2020 2064 6566 2072 656d 6f76 655f 756e     def remove_un
-00003150: 6661 6374 6f72 697a 6162 6c65 5f67 656e  factorizable_gen
-00003160: 6573 2873 656c 6629 3a0d 0a20 2020 2020  es(self):..     
-00003170: 2020 2022 2222 5265 6d6f 7665 7320 6765     """Removes ge
-00003180: 6e65 7320 7769 7468 206d 6973 7369 6e67  nes with missing
-00003190: 2076 616c 7565 7320 6f72 207a 6572 6f20   values or zero 
-000031a0: 7661 7269 616e 6365 2066 726f 6d20 7468  variance from th
-000031b0: 6520 6461 7461 206d 6174 7269 782e 0d0a  e data matrix...
-000031c0: 2020 2020 2020 2020 2222 220d 0a20 2020          """..   
-000031d0: 2020 2020 2064 6620 3d20 7365 6c66 2e74       df = self.t
-000031e0: 6f5f 6466 286e 6f72 6d61 6c69 7a65 643d  o_df(normalized=
-000031f0: 4661 6c73 6529 0d0a 2020 2020 2020 2020  False)..        
-00003200: 2320 4368 6563 6b20 666f 7220 7661 7269  # Check for vari
-00003210: 6162 6c65 7320 7769 7468 206d 6973 7369  ables with missi
-00003220: 6e67 2076 616c 7565 730d 0a20 2020 2020  ng values..     
-00003230: 2020 2067 656e 6573 5f77 6974 685f 6d69     genes_with_mi
-00003240: 7373 696e 6776 616c 7565 7320 3d20 6466  ssingvalues = df
-00003250: 2e69 736e 756c 6c28 292e 616e 7928 290d  .isnull().any().
-00003260: 0a20 2020 2020 2020 200d 0a20 2020 2020  .        ..     
-00003270: 2020 2069 6620 6765 6e65 735f 7769 7468     if genes_with
-00003280: 5f6d 6973 7369 6e67 7661 6c75 6573 2e61  _missingvalues.a
-00003290: 6e79 2829 3a0d 0a20 2020 2020 2020 2020  ny():..         
-000032a0: 2020 206e 5f6d 6973 7369 6e67 203d 2067     n_missing = g
-000032b0: 656e 6573 5f77 6974 685f 6d69 7373 696e  enes_with_missin
-000032c0: 6776 616c 7565 732e 7375 6d28 290d 0a20  gvalues.sum().. 
-000032d0: 2020 2020 2020 2020 2020 206c 6f67 6769             loggi
-000032e0: 6e67 2e77 6172 6e69 6e67 2866 227b 6e5f  ng.warning(f"{n_
-000032f0: 6d69 7373 696e 677d 206f 6620 7b73 656c  missing} of {sel
-00003300: 662e 6164 6174 612e 6e5f 7661 7273 7d20  f.adata.n_vars} 
-00003310: 7661 7269 6162 6c65 7320 6172 6520 6d69  variables are mi
-00003320: 7373 696e 6720 7661 6c75 6573 2028 6061  ssing values (`a
-00003330: 6461 7461 2e58 6029 2e22 290d 0a20 2020  data.X`).")..   
-00003340: 2020 2020 2020 2020 206c 6f67 6769 6e67           logging
-00003350: 2e77 6172 6e69 6e67 2866 2253 7562 7365  .warning(f"Subse
-00003360: 7474 696e 6720 7661 7269 6162 6c65 7320  tting variables 
-00003370: 746f 2074 686f 7365 2077 6974 6820 6e6f  to those with no
-00003380: 206d 6973 7369 6e67 2076 616c 7565 732e   missing values.
-00003390: 2229 0d0a 2020 2020 2020 2020 2020 2020  ")..            
-000033a0: 2020 2020 0d0a 2020 2020 2020 2020 2320      ..        # 
-000033b0: 4368 6563 6b20 666f 7220 6765 6e65 7320  Check for genes 
-000033c0: 7769 7468 207a 6572 6f20 7661 7269 616e  with zero varian
-000033d0: 6365 0d0a 2020 2020 2020 2020 7a65 726f  ce..        zero
-000033e0: 7661 7267 656e 6573 203d 2028 6466 2e76  vargenes = (df.v
-000033f0: 6172 2829 203d 3d20 3029 0d0a 2020 2020  ar() == 0)..    
-00003400: 2020 2020 6966 207a 6572 6f76 6172 6765      if zerovarge
-00003410: 6e65 732e 616e 7928 293a 0d0a 2020 2020  nes.any():..    
-00003420: 2020 2020 2020 2020 6e5f 7a65 726f 7661          n_zerova
-00003430: 7220 3d20 7a65 726f 7661 7267 656e 6573  r = zerovargenes
-00003440: 2e73 756d 2829 0d0a 2020 2020 2020 2020  .sum()..        
-00003450: 2020 2020 6c6f 6767 696e 672e 7761 726e      logging.warn
-00003460: 696e 6728 6622 7b6e 5f7a 6572 6f76 6172  ing(f"{n_zerovar
-00003470: 7d20 6f66 207b 7365 6c66 2e61 6461 7461  } of {self.adata
-00003480: 2e6e 5f76 6172 737d 2076 6172 6961 626c  .n_vars} variabl
-00003490: 6573 2068 6176 6520 6120 7661 7269 616e  es have a varian
-000034a0: 6365 206f 6620 7a65 726f 2069 6e20 636f  ce of zero in co
-000034b0: 756e 7473 2064 6174 6120 2860 6164 6174  unts data (`adat
-000034c0: 612e 7261 772e 5860 292e 2229 0d0a 2020  a.raw.X`).")..  
-000034d0: 2020 2020 2020 2020 2020 6c6f 6767 696e            loggin
-000034e0: 672e 7761 726e 696e 6728 6622 5375 6273  g.warning(f"Subs
-000034f0: 6574 7469 6e67 2076 6172 6961 626c 6573  etting variables
-00003500: 2074 6f20 7468 6f73 6520 7769 7468 206e   to those with n
-00003510: 6f6e 7a65 726f 2076 6172 6961 6e63 652e  onzero variance.
-00003520: 2229 0d0a 2020 2020 2020 2020 0d0a 2020  ")..        ..  
-00003530: 2020 2020 2020 6765 6e65 735f 746f 5f6b        genes_to_k
-00003540: 6565 7020 3d20 287e 6765 6e65 735f 7769  eep = (~genes_wi
-00003550: 7468 5f6d 6973 7369 6e67 7661 6c75 6573  th_missingvalues
-00003560: 2920 2620 287e 7a65 726f 7661 7267 656e  ) & (~zerovargen
-00003570: 6573 290d 0a20 2020 2020 2020 200d 0a20  es)..        .. 
-00003580: 2020 2020 2020 2073 656c 662e 6164 6174         self.adat
-00003590: 6120 3d20 7365 6c66 2e61 6461 7461 5b3a  a = self.adata[:
-000035a0: 2c67 656e 6573 5f74 6f5f 6b65 6570 5d0d  ,genes_to_keep].
-000035b0: 0a0d 0a20 2020 2064 6566 2063 6f6d 7075  ...    def compu
-000035c0: 7465 5f67 656e 655f 7374 6174 7328 7365  te_gene_stats(se
-000035d0: 6c66 2c20 6f64 675f 6465 6661 756c 745f  lf, odg_default_
-000035e0: 7370 6c69 6e65 5f64 6567 7265 653a 2069  spline_degree: i
-000035f0: 6e74 203d 2033 2c20 6f64 675f 6465 6661  nt = 3, odg_defa
-00003600: 756c 745f 646f 663a 2069 6e74 203d 2038  ult_dof: int = 8
-00003610: 293a 0d0a 2020 2020 2020 2020 2222 220d  ):..        """.
-00003620: 0a20 2020 2020 2020 2043 6f6d 7075 7465  .        Compute
-00003630: 7320 6765 6e65 2073 7461 7469 7374 6963  s gene statistic
-00003640: 7320 616e 6420 6669 7473 2074 776f 206d  s and fits two m
-00003650: 6f64 656c 7320 6f66 206d 6561 6e20 616e  odels of mean an
-00003660: 6420 7661 7269 616e 6365 206f 6620 6765  d variance of ge
-00003670: 6e65 7320 696e 2074 6865 2064 6174 6173  nes in the datas
-00003680: 6574 2e20 5468 6520 6669 7273 7420 6d65  et. The first me
-00003690: 7468 6f64 2069 7320 7468 650d 0a20 2020  thod is the..   
-000036a0: 2020 2020 2067 656e 6572 616c 697a 6564       generalized
-000036b0: 2061 6464 6974 6976 6520 6d6f 6465 6c20   additive model 
-000036c0: 7769 7468 2073 6d6f 6f74 6820 636f 6d70  with smooth comp
-000036d0: 6f6e 656e 7473 2028 422d 7370 6c69 6e65  onents (B-spline
-000036e0: 7329 2074 6f20 6d6f 6465 6c20 7468 6520  s) to model the 
-000036f0: 7265 6c61 7469 6f6e 7368 6970 206f 6620  relationship of 
-00003700: 6d65 616e 2061 6e64 2076 6172 6961 6e63  mean and varianc
-00003710: 650d 0a20 2020 2020 2020 2062 6574 7765  e..        betwe
-00003720: 656e 2067 656e 6573 2069 6e20 7468 6520  en genes in the 
-00003730: 6461 7461 7365 742e 2049 7420 7072 6f64  dataset. It prod
-00003740: 7563 6573 2061 6e20 6f64 7363 6f72 6520  uces an odscore 
-00003750: 6d65 7472 6963 2066 6f72 206f 7665 7264  metric for overd
-00003760: 6973 7065 7273 696f 6e2e 2054 6865 2073  ispersion. The s
-00003770: 6563 6f6e 6420 6973 2074 6865 2063 6f75  econd is the cou
-00003780: 6e74 2d73 7461 7469 7374 6963 730d 0a20  nt-statistics.. 
-00003790: 2020 2020 2020 206d 6574 686f 6420 666f         method fo
-000037a0: 756e 6420 696e 2074 6865 2063 4e4d 4620  und in the cNMF 
-000037b0: 7061 636b 6167 652c 2077 6869 6368 2070  package, which p
-000037c0: 726f 6475 6365 7320 6120 6d6f 6469 6669  roduces a modifi
-000037d0: 6564 2076 2d73 636f 7265 206d 6574 7269  ed v-score metri
-000037e0: 632e 2041 6c6c 2067 656e 6520 7374 6174  c. All gene stat
-000037f0: 6973 7469 6373 2061 7265 2073 746f 7265  istics are store
-00003800: 6420 7769 7468 696e 2074 6865 0d0a 2020  d within the..  
-00003810: 2020 2020 2020 6461 7461 7365 7420 6f62        dataset ob
-00003820: 6a65 6374 2061 6e64 2061 7265 2061 6363  ject and are acc
-00003830: 6573 7369 626c 6520 7573 696e 6720 6064  essible using `d
-00003840: 6174 6173 6574 2e61 6e6e 6461 7461 2e76  ataset.anndata.v
-00003850: 6172 602e 0d0a 0d0a 2020 2020 2020 2020  ar`.....        
-00003860: 3a70 6172 616d 206f 6467 5f64 6566 6175  :param odg_defau
-00003870: 6c74 5f73 706c 696e 655f 6465 6772 6565  lt_spline_degree
-00003880: 3a20 422d 5370 6c69 6e65 2064 6567 7265  : B-Spline degre
-00003890: 6520 666f 7220 474c 4d2d 4741 4d20 6d6f  e for GLM-GAM mo
-000038a0: 6465 6c6c 696e 6720 6f66 206d 6561 6e2d  delling of mean-
-000038b0: 7661 7269 616e 6365 2072 656c 6174 696f  variance relatio
-000038c0: 6e73 6869 702c 2064 6566 6175 6c74 7320  nship, defaults 
-000038d0: 746f 2033 0d0a 2020 2020 2020 2020 3a74  to 3..        :t
-000038e0: 7970 6520 6f64 675f 6465 6661 756c 745f  ype odg_default_
-000038f0: 7370 6c69 6e65 5f64 6567 7265 653a 2069  spline_degree: i
-00003900: 6e74 2c20 6f70 7469 6f6e 616c 0d0a 2020  nt, optional..  
-00003910: 2020 2020 2020 3a70 6172 616d 206f 6467        :param odg
-00003920: 5f64 6566 6175 6c74 5f64 6f66 3a20 4465  _default_dof: De
-00003930: 6772 6565 7320 6f66 2066 7265 6564 6f6d  grees of freedom
-00003940: 2066 6f72 2047 4c4d 2d47 414d 206d 6f64   for GLM-GAM mod
-00003950: 656c 6c69 6e67 206f 6620 6d65 616e 2d76  elling of mean-v
-00003960: 6172 616e 6365 2c20 6465 6661 756c 7473  arance, defaults
-00003970: 2074 6f20 380d 0a20 2020 2020 2020 203a   to 8..        :
-00003980: 7479 7065 206f 6467 5f64 6566 6175 6c74  type odg_default
-00003990: 5f64 6f66 3a20 696e 742c 206f 7074 696f  _dof: int, optio
-000039a0: 6e61 6c0d 0a20 2020 2020 2020 2022 2222  nal..        """
-000039b0: 0d0a 2020 2020 2020 2020 6461 7461 5f72  ..        data_r
-000039c0: 6177 203d 2073 656c 662e 746f 5f64 6628  aw = self.to_df(
-000039d0: 290d 0a20 2020 2020 2020 2064 6174 615f  )..        data_
-000039e0: 6e6f 726d 616c 697a 6564 203d 2073 656c  normalized = sel
-000039f0: 662e 746f 5f64 6628 6e6f 726d 616c 697a  f.to_df(normaliz
-00003a00: 6564 3d54 7275 6529 0d0a 2020 2020 2020  ed=True)..      
-00003a10: 2020 0d0a 2020 2020 2020 2020 2320 6372    ..        # cr
-00003a20: 6561 7465 2064 6174 6166 7261 6d65 206f  eate dataframe o
-00003a30: 6620 7065 722d 6765 6e65 2073 7461 7469  f per-gene stati
-00003a40: 7374 6963 730d 0a20 2020 2020 2020 2073  stics..        s
-00003a50: 656c 662e 6164 6174 612e 7661 725b 226d  elf.adata.var["m
-00003a60: 6561 6e22 5d20 3d20 6461 7461 5f6e 6f72  ean"] = data_nor
-00003a70: 6d61 6c69 7a65 642e 6d65 616e 2829 0d0a  malized.mean()..
-00003a80: 2020 2020 2020 2020 7365 6c66 2e61 6461          self.ada
-00003a90: 7461 2e76 6172 5b22 7261 6e6b 5f6d 6561  ta.var["rank_mea
-00003aa0: 6e22 5d20 3d20 7365 6c66 2e61 6461 7461  n"] = self.adata
-00003ab0: 2e76 6172 5b22 6d65 616e 225d 2e72 616e  .var["mean"].ran
-00003ac0: 6b28 290d 0a20 2020 2020 2020 2073 656c  k()..        sel
-00003ad0: 662e 6164 6174 612e 7661 725b 2276 6172  f.adata.var["var
-00003ae0: 6961 6e63 6522 5d20 3d20 6461 7461 5f6e  iance"] = data_n
-00003af0: 6f72 6d61 6c69 7a65 642e 7661 7228 290d  ormalized.var().
-00003b00: 0a20 2020 2020 2020 2073 656c 662e 6164  .        self.ad
-00003b10: 6174 612e 7661 725b 2273 6422 5d20 3d20  ata.var["sd"] = 
-00003b20: 6461 7461 5f6e 6f72 6d61 6c69 7a65 642e  data_normalized.
-00003b30: 7374 6428 290d 0a20 2020 2020 2020 2073  std()..        s
-00003b40: 656c 662e 6164 6174 612e 7661 725b 226d  elf.adata.var["m
-00003b50: 6973 7369 6e67 6e65 7373 225d 203d 2064  issingness"] = d
-00003b60: 6174 615f 6e6f 726d 616c 697a 6564 2e69  ata_normalized.i
-00003b70: 736e 756c 6c28 292e 7375 6d28 2920 2f20  snull().sum() / 
-00003b80: 6461 7461 5f6e 6f72 6d61 6c69 7a65 642e  data_normalized.
-00003b90: 7368 6170 655b 305d 0d0a 2020 2020 2020  shape[0]..      
-00003ba0: 2020 7365 6c66 2e61 6461 7461 2e76 6172    self.adata.var
-00003bb0: 5b5b 226c 6f67 5f6d 6561 6e22 2c20 226c  [["log_mean", "l
-00003bc0: 6f67 5f76 6172 6961 6e63 6522 5d5d 203d  og_variance"]] =
-00003bd0: 206e 702e 6c6f 6731 3028 7365 6c66 2e61   np.log10(self.a
-00003be0: 6461 7461 2e76 6172 5b5b 226d 6561 6e22  data.var[["mean"
-00003bf0: 2c20 2276 6172 6961 6e63 6522 5d5d 290d  , "variance"]]).
-00003c00: 0a20 2020 2020 2020 2073 656c 662e 6164  .        self.ad
-00003c10: 6174 612e 7661 725b 226d 6561 6e5f 636f  ata.var["mean_co
-00003c20: 756e 7473 225d 203d 2064 6174 615f 7261  unts"] = data_ra
-00003c30: 772e 6d65 616e 2829 0d0a 2020 2020 2020  w.mean()..      
-00003c40: 2020 7365 6c66 2e61 6461 7461 2e76 6172    self.adata.var
-00003c50: 5b22 6f64 7363 6f72 655f 6578 636c 7564  ["odscore_exclud
-00003c60: 6564 225d 203d 2028 2873 656c 662e 6164  ed"] = ((self.ad
-00003c70: 6174 612e 7661 725b 226d 6973 7369 6e67  ata.var["missing
-00003c80: 6e65 7373 225d 203e 2030 2920 7c0d 0a20  ness"] > 0) |.. 
-00003c90: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003ca0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003cb0: 2020 2020 2020 2020 2020 2020 2073 656c               sel
-00003cc0: 662e 6164 6174 612e 7661 725b 226c 6f67  f.adata.var["log
-00003cd0: 5f6d 6561 6e22 5d2e 6973 6e75 6c6c 2829  _mean"].isnull()
-00003ce0: 207c 0d0a 2020 2020 2020 2020 2020 2020   |..            
-00003cf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003d00: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003d10: 2020 2873 656c 662e 6164 6174 612e 7661    (self.adata.va
-00003d20: 725b 226d 6561 6e22 5d20 3d3d 2030 2920  r["mean"] == 0) 
-00003d30: 7c0d 0a20 2020 2020 2020 2020 2020 2020  |..             
-00003d40: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003d50: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00003d60: 2073 656c 662e 6164 6174 612e 7661 725b   self.adata.var[
-00003d70: 226c 6f67 5f76 6172 6961 6e63 6522 5d2e  "log_variance"].
-00003d80: 6973 6e75 6c6c 2829 290d 0a0d 0a20 2020  isnull())....   
-00003d90: 2020 2020 2023 206d 6f64 656c 206d 6561       # model mea
-00003da0: 6e2d 7661 7269 616e 6365 2072 656c 6174  n-variance relat
-00003db0: 696f 6e73 6869 7020 7573 696e 6720 6765  ionship using ge
-00003dc0: 6e65 7261 6c69 7a65 6420 6164 6469 7469  neralized additi
-00003dd0: 7665 206d 6f64 656c 2077 6974 6820 736d  ve model with sm
-00003de0: 6f6f 7468 2063 6f6d 706f 6e65 6e74 730d  ooth components.
-00003df0: 0a20 2020 2020 2020 2064 665f 6d6f 6465  .        df_mode
-00003e00: 6c20 3d20 7365 6c66 2e61 6461 7461 2e76  l = self.adata.v
-00003e10: 6172 5b7e 7365 6c66 2e61 6461 7461 2e76  ar[~self.adata.v
-00003e20: 6172 5b22 6f64 7363 6f72 655f 6578 636c  ar["odscore_excl
-00003e30: 7564 6564 225d 5d0d 0a20 2020 2020 2020  uded"]]..       
-00003e40: 2062 7320 3d20 4253 706c 696e 6573 2864   bs = BSplines(d
-00003e50: 665f 6d6f 6465 6c5b 226d 6561 6e22 5d2c  f_model["mean"],
-00003e60: 2064 663d 6f64 675f 6465 6661 756c 745f   df=odg_default_
-00003e70: 646f 662c 2064 6567 7265 653d 6f64 675f  dof, degree=odg_
-00003e80: 6465 6661 756c 745f 7370 6c69 6e65 5f64  default_spline_d
-00003e90: 6567 7265 6529 0d0a 2020 2020 2020 2020  egree)..        
-00003ea0: 6761 6d20 3d20 474c 4d47 616d 2e66 726f  gam = GLMGam.fro
-00003eb0: 6d5f 666f 726d 756c 6128 226c 6f67 5f76  m_formula("log_v
-00003ec0: 6172 6961 6e63 6520 7e20 6c6f 675f 6d65  ariance ~ log_me
-00003ed0: 616e 222c 2064 6174 613d 6466 5f6d 6f64  an", data=df_mod
-00003ee0: 656c 2c20 736d 6f6f 7468 6572 3d62 7329  el, smoother=bs)
-00003ef0: 2e66 6974 2829 0d0a 2020 2020 2020 2020  .fit()..        
-00003f00: 7365 6c66 2e61 6461 7461 2e76 6172 5b22  self.adata.var["
-00003f10: 7265 7369 645f 6c6f 675f 7661 7269 616e  resid_log_varian
-00003f20: 6365 225d 203d 2067 616d 2e72 6573 6964  ce"] = gam.resid
-00003f30: 5f72 6573 706f 6e73 650d 0a20 2020 2020  _response..     
-00003f40: 2020 2073 656c 662e 6164 6174 612e 7661     self.adata.va
-00003f50: 725b 226f 6473 636f 7265 225d 203d 206e  r["odscore"] = n
-00003f60: 702e 7371 7274 2831 3020 2a2a 2073 656c  p.sqrt(10 ** sel
-00003f70: 662e 6164 6174 612e 7661 725b 2272 6573  f.adata.var["res
-00003f80: 6964 5f6c 6f67 5f76 6172 6961 6e63 6522  id_log_variance"
-00003f90: 5d29 0d0a 2020 2020 2020 2020 7365 6c66  ])..        self
-00003fa0: 2e61 6461 7461 2e76 6172 5b22 6761 6d5f  .adata.var["gam_
-00003fb0: 6669 7474 6564 7661 6c75 6573 225d 203d  fittedvalues"] =
-00003fc0: 2067 616d 2e66 6974 7465 6476 616c 7565   gam.fittedvalue
-00003fd0: 730d 0a0d 0a0d 0a20 2020 2020 2020 2023  s......        #
-00003fe0: 206d 6f64 656c 206d 6561 6e2d 7661 7269   model mean-vari
-00003ff0: 616e 6365 2072 656c 6174 696f 6e73 6869  ance relationshi
-00004000: 7020 7573 696e 6720 634e 4d46 2773 206d  p using cNMF's m
-00004010: 6574 686f 6420 6261 7365 6420 6f6e 2076  ethod based on v
-00004020: 2d73 636f 7265 2061 6e64 206d 696e 696d  -score and minim
-00004030: 756d 2065 7870 7265 7373 696f 6e20 7468  um expression th
-00004040: 7265 7368 6f6c 640d 0a20 2020 2020 2020  reshold..       
-00004050: 2076 7363 6f72 655f 7374 6174 7320 3d20   vscore_stats = 
-00004060: 7064 2e44 6174 6146 7261 6d65 2863 6e6d  pd.DataFrame(cnm
-00004070: 662e 6765 745f 6869 6768 7661 725f 6765  f.get_highvar_ge
-00004080: 6e65 7328 696e 7075 745f 636f 756e 7473  nes(input_counts
-00004090: 3d64 6174 615f 6e6f 726d 616c 697a 6564  =data_normalized
-000040a0: 2e76 616c 7565 732c 206d 696e 696d 616c  .values, minimal
-000040b0: 5f6d 6561 6e3d 3029 5b30 5d29 0d0a 2020  _mean=0)[0])..  
-000040c0: 2020 2020 2020 7673 636f 7265 5f73 7461        vscore_sta
-000040d0: 7473 2e69 6e64 6578 203d 2064 6174 615f  ts.index = data_
-000040e0: 6e6f 726d 616c 697a 6564 2e63 6f6c 756d  normalized.colum
-000040f0: 6e73 0d0a 2020 2020 2020 2020 7365 6c66  ns..        self
-00004100: 2e61 6461 7461 2e76 6172 5b22 7673 636f  .adata.var["vsco
-00004110: 7265 225d 203d 2076 7363 6f72 655f 7374  re"] = vscore_st
-00004120: 6174 735b 2266 616e 6f5f 7261 7469 6f22  ats["fano_ratio"
-00004130: 5d0d 0a20 2020 2020 2020 2073 656c 662e  ]..        self.
-00004140: 6164 6174 612e 756e 735b 226f 6467 225d  adata.uns["odg"]
-00004150: 5b22 6f64 675f 6465 6661 756c 745f 7370  ["odg_default_sp
-00004160: 6c69 6e65 5f64 6567 7265 6522 5d20 3d20  line_degree"] = 
-00004170: 6f64 675f 6465 6661 756c 745f 7370 6c69  odg_default_spli
-00004180: 6e65 5f64 6567 7265 650d 0a20 2020 2020  ne_degree..     
-00004190: 2020 2073 656c 662e 6164 6174 612e 756e     self.adata.un
-000041a0: 735b 226f 6467 225d 5b22 6f64 675f 6465  s["odg"]["odg_de
-000041b0: 6661 756c 745f 646f 6622 5d20 3d20 6f64  fault_dof"] = od
-000041c0: 675f 6465 6661 756c 745f 646f 660d 0a20  g_default_dof.. 
-000041d0: 2020 2020 2020 2073 656c 662e 6170 7065         self.appe
-000041e0: 6e64 5f74 6f5f 6869 7374 6f72 7928 2247  nd_to_history("G
-000041f0: 656e 652d 6c65 7665 6c20 7374 6174 6973  ene-level statis
-00004200: 7469 6373 2061 6e64 206f 7665 7264 6973  tics and overdis
-00004210: 7065 7273 696f 6e20 6d6f 6465 6c6c 696e  persion modellin
-00004220: 6720 636f 6d70 6c65 7465 642e 2229 0d0a  g completed.")..
-00004230: 2020 2020 2020 2020 0d0a 2020 2020 6465          ..    de
-00004240: 6620 7365 6c65 6374 5f6f 7665 7264 6973  f select_overdis
-00004250: 7065 7273 6564 5f67 656e 6573 5f66 726f  persed_genes_fro
-00004260: 6d5f 6765 6e65 6c69 7374 2873 656c 662c  m_genelist(self,
-00004270: 2067 656e 6573 3a20 436f 6c6c 6563 7469   genes: Collecti
-00004280: 6f6e 2c20 6d69 6e5f 6d65 616e 3d30 293a  on, min_mean=0):
-00004290: 0d0a 2020 2020 2020 2020 2222 2253 656c  ..        """Sel
-000042a0: 6563 7420 6f76 6572 6469 7370 6572 7365  ect overdisperse
-000042b0: 6420 6765 6e65 732f 6665 6174 7572 6573  d genes/features
-000042c0: 2075 7369 6e67 2061 2063 7573 746f 6d20   using a custom 
-000042d0: 6c69 7374 2e20 4765 6e65 732f 6665 6174  list. Genes/feat
-000042e0: 7572 6573 206e 6f74 2070 7265 7365 6e74  ures not present
-000042f0: 2069 6e20 7468 6520 6461 7461 7365 7420   in the dataset 
-00004300: 6172 6520 6175 746f 6d61 7469 6361 6c6c  are automaticall
-00004310: 7920 6669 6c74 6572 6564 206f 7574 2e0d  y filtered out..
-00004320: 0a0d 0a20 2020 2020 2020 203a 7061 7261  ...        :para
-00004330: 6d20 6765 6e65 733a 2067 656e 6520 6c69  m genes: gene li
-00004340: 7374 0d0a 2020 2020 2020 2020 3a74 7970  st..        :typ
-00004350: 6520 6765 6e65 733a 2043 6f6c 6c65 6374  e genes: Collect
-00004360: 696f 6e0d 0a20 2020 2020 2020 203a 7061  ion..        :pa
-00004370: 7261 6d20 6d69 6e5f 6d65 616e 3a20 6d69  ram min_mean: mi
-00004380: 6e69 6d75 6d20 6765 6e65 2065 7870 7265  nimum gene expre
-00004390: 7373 696f 6e20 666f 7220 6765 6e65 7320  ssion for genes 
-000043a0: 746f 2062 6520 636f 756e 7465 6420 6173  to be counted as
-000043b0: 206f 7665 7264 6973 7065 7273 6564 2c20   overdispersed, 
-000043c0: 6465 6661 756c 7473 2074 6f20 300d 0a20  defaults to 0.. 
-000043d0: 2020 2020 2020 203a 7479 7065 206d 696e         :type min
-000043e0: 5f6d 6561 6e3a 2069 6e74 2c20 6f70 7469  _mean: int, opti
-000043f0: 6f6e 616c 0d0a 2020 2020 2020 2020 2222  onal..        ""
-00004400: 220d 0a20 2020 2020 2020 2073 656c 662e  "..        self.
-00004410: 6164 6174 612e 7661 725b 2273 656c 6563  adata.var["selec
-00004420: 7465 6422 5d20 3d20 7365 6c66 2e61 6461  ted"] = self.ada
-00004430: 7461 2e76 6172 2e69 6e64 6578 2e69 7369  ta.var.index.isi
-00004440: 6e28 6765 6e65 7329 2026 2073 656c 662e  n(genes) & self.
-00004450: 6164 6174 612e 7661 725b 226d 6561 6e5f  adata.var["mean_
-00004460: 636f 756e 7473 225d 203e 3d20 6d69 6e5f  counts"] >= min_
-00004470: 6d65 616e 0d0a 2020 2020 2020 2020 7365  mean..        se
-00004480: 6c66 2e61 6461 7461 2e75 6e73 5b22 6f64  lf.adata.uns["od
-00004490: 6722 5d5b 226f 7665 7264 6973 7065 7273  g"]["overdispers
-000044a0: 696f 6e5f 6d65 7472 6963 225d 203d 2022  ion_metric"] = "
-000044b0: 220d 0a20 2020 2020 2020 2073 656c 662e  "..        self.
-000044c0: 6164 6174 612e 756e 735b 226f 6467 225d  adata.uns["odg"]
-000044d0: 5b22 6d69 6e5f 6d65 616e 225d 203d 206d  ["min_mean"] = m
-000044e0: 696e 5f6d 6561 6e0d 0a20 2020 2020 2020  in_mean..       
-000044f0: 2073 656c 662e 6164 6174 612e 756e 735b   self.adata.uns[
-00004500: 226f 6467 225d 5b22 6d69 6e5f 7363 6f72  "odg"]["min_scor
-00004510: 6522 5d20 3d20 2222 0d0a 2020 2020 2020  e"] = ""..      
-00004520: 2020 7365 6c66 2e61 6461 7461 2e75 6e73    self.adata.uns
-00004530: 5b22 6f64 6722 5d5b 2274 6f70 5f6e 225d  ["odg"]["top_n"]
-00004540: 203d 2022 220d 0a20 2020 2020 2020 2073   = ""..        s
-00004550: 656c 662e 6164 6174 612e 756e 735b 226f  elf.adata.uns["o
-00004560: 6467 225d 5b22 7175 616e 7469 6c65 225d  dg"]["quantile"]
-00004570: 203d 2022 220d 0a20 2020 2020 2020 2073   = ""..        s
-00004580: 656c 662e 6170 7065 6e64 5f74 6f5f 6869  elf.append_to_hi
-00004590: 7374 6f72 7928 224f 7665 7264 6973 7065  story("Overdispe
-000045a0: 7273 6564 2067 656e 6573 2073 656c 6563  rsed genes selec
-000045b0: 7465 6420 6672 6f6d 2063 7573 746f 6d20  ted from custom 
-000045c0: 6765 6e65 206c 6973 7422 290d 0a20 2020  gene list")..   
-000045d0: 2020 2020 200d 0a20 2020 2064 6566 2073       ..    def s
-000045e0: 656c 6563 745f 6f76 6572 6469 7370 6572  elect_overdisper
-000045f0: 7365 645f 6765 6e65 7328 7365 6c66 2c0d  sed_genes(self,.
-00004600: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00004610: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00004620: 2020 2020 6f76 6572 6469 7370 6572 7369      overdispersi
-00004630: 6f6e 5f6d 6574 7269 633a 2073 7472 203d  on_metric: str =
-00004640: 2022 6f64 7363 6f72 6522 2c0d 0a20 2020   "odscore",..   
-00004650: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00004660: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00004670: 6d69 6e5f 6d65 616e 3a20 666c 6f61 7420  min_mean: float 
-00004680: 3d20 302c 0d0a 2020 2020 2020 2020 2020  = 0,..          
-00004690: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000046a0: 2020 2020 2020 2020 206d 696e 5f73 636f           min_sco
-000046b0: 7265 3a20 666c 6f61 7420 3d20 312e 302c  re: float = 1.0,
-000046c0: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
-000046d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000046e0: 2020 2020 2074 6f70 5f6e 3a20 696e 7420       top_n: int 
-000046f0: 3d20 4e6f 6e65 2c0d 0a20 2020 2020 2020  = None,..       
-00004700: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00004710: 2020 2020 2020 2020 2020 2020 7175 616e              quan
-00004720: 7469 6c65 3a20 666c 6f61 7420 3d20 4e6f  tile: float = No
-00004730: 6e65 293a 0d0a 2020 2020 2020 2020 2222  ne):..        ""
-00004740: 2253 656c 6563 7420 6f76 6572 6469 7370  "Select overdisp
-00004750: 6572 7365 6420 6765 6e65 732f 6665 6174  ersed genes/feat
-00004760: 7572 6573 2075 7369 6e67 2061 6e20 6f76  ures using an ov
-00004770: 6572 6469 7370 6572 7369 6f6e 206d 6574  erdispersion met
-00004780: 7269 632e 204f 7074 696f 6e61 6c6c 7920  ric. Optionally 
-00004790: 7365 7420 6120 6d69 6e69 6d75 6d20 6765  set a minimum ge
-000047a0: 6e65 2065 7870 7265 7373 696f 6e20 6c65  ne expression le
-000047b0: 7665 6c2e 0d0a 2020 2020 2020 2020 5365  vel...        Se
-000047c0: 7420 6120 7468 7265 7368 6f6c 6420 7573  t a threshold us
-000047d0: 696e 6720 7468 6520 746f 7020 4e20 2827  ing the top N ('
-000047e0: 746f 705f 6e27 292c 206d 696e 696d 756d  top_n'), minimum
-000047f0: 2073 636f 7265 2028 276d 696e 5f73 636f   score ('min_sco
-00004800: 7265 2729 2c20 6f72 2070 726f 706f 7274  re'), or proport
-00004810: 696f 6e20 6f66 2066 6561 7475 7265 7320  ion of features 
-00004820: 2827 7175 616e 7469 6c65 2729 206d 6574  ('quantile') met
-00004830: 686f 6473 2e0d 0a20 2020 2020 2020 204f  hods...        O
-00004840: 7665 7264 6973 7065 7273 6564 2067 656e  verdispersed gen
-00004850: 6520 6c69 7374 2069 7320 7361 7665 6420  e list is saved 
-00004860: 696e 2074 6865 2044 6174 6173 6574 206f  in the Dataset o
-00004870: 626a 6563 742e 0d0a 0d0a 2020 2020 2020  bject.....      
-00004880: 2020 3a70 6172 616d 206f 7665 7264 6973    :param overdis
-00004890: 7065 7273 696f 6e5f 6d65 7472 6963 3a20  persion_metric: 
-000048a0: 226f 6473 636f 7265 2220 6f72 2022 7673  "odscore" or "vs
-000048b0: 636f 7265 222c 2064 6566 6175 6c74 7320  core", defaults 
-000048c0: 746f 2022 6f64 7363 6f72 6522 0d0a 2020  to "odscore"..  
-000048d0: 2020 2020 2020 3a74 7970 6520 6f76 6572        :type over
-000048e0: 6469 7370 6572 7369 6f6e 5f6d 6574 7269  dispersion_metri
-000048f0: 633a 2073 7472 2c20 6f70 7469 6f6e 616c  c: str, optional
-00004900: 0d0a 2020 2020 2020 2020 3a70 6172 616d  ..        :param
-00004910: 206d 696e 5f6d 6561 6e3a 206d 696e 696d   min_mean: minim
-00004920: 756d 2067 656e 6520 6578 7072 6573 7369  um gene expressi
-00004930: 6f6e 2066 6f72 2067 656e 6573 2074 6f20  on for genes to 
-00004940: 6265 2063 6f75 6e74 6564 2061 7320 6f76  be counted as ov
-00004950: 6572 6469 7370 6572 7365 642c 2064 6566  erdispersed, def
-00004960: 6175 6c74 7320 746f 2030 0d0a 2020 2020  aults to 0..    
-00004970: 2020 2020 3a74 7970 6520 6d69 6e5f 6d65      :type min_me
-00004980: 616e 3a20 696e 742c 206f 7074 696f 6e61  an: int, optiona
-00004990: 6c0d 0a20 2020 2020 2020 203a 7061 7261  l..        :para
-000049a0: 6d20 6d69 6e5f 7363 6f72 653a 206d 696e  m min_score: min
-000049b0: 696d 756d 2073 636f 7265 2066 6f72 206f  imum score for o
-000049c0: 7665 7264 6973 7065 7273 696f 6e2c 2064  verdispersion, d
-000049d0: 6566 6175 6c74 7320 746f 2031 2e30 0d0a  efaults to 1.0..
-000049e0: 2020 2020 2020 2020 3a74 7970 6520 6d69          :type mi
-000049f0: 6e5f 7363 6f72 653a 2066 6c6f 6174 2c20  n_score: float, 
-00004a00: 6f70 7469 6f6e 616c 0d0a 2020 2020 2020  optional..      
-00004a10: 2020 3a70 6172 616d 2074 6f70 5f6e 3a20    :param top_n: 
-00004a20: 4368 6f6f 7365 2074 6865 2074 6f70 204e  Choose the top N
-00004a30: 206d 6f73 7420 6f76 6572 6469 7370 6572   most overdisper
-00004a40: 7365 6420 6765 6e65 732c 2064 6566 6175  sed genes, defau
-00004a50: 6c74 7320 746f 204e 6f6e 650d 0a20 2020  lts to None..   
-00004a60: 2020 2020 203a 7479 7065 2074 6f70 5f6e       :type top_n
-00004a70: 3a20 696e 742c 206f 7074 696f 6e61 6c0d  : int, optional.
-00004a80: 0a20 2020 2020 2020 203a 7061 7261 6d20  .        :param 
-00004a90: 7175 616e 7469 6c65 3a20 4368 6f6f 7365  quantile: Choose
-00004aa0: 2061 2071 7561 6e74 696c 6520 6f66 206f   a quantile of o
-00004ab0: 7665 7264 6973 7065 7273 696f 6e2e 2046  verdispersion. F
-00004ac0: 6f72 2065 7861 6d70 6c65 2c20 7468 6520  or example, the 
-00004ad0: 746f 7020 3130 2520 6f66 206f 7665 7264  top 10% of overd
-00004ae0: 6973 7065 7273 6564 2067 656e 6573 2077  ispersed genes w
-00004af0: 6f75 6c64 2062 6520 302e 3130 2e20 4465  ould be 0.10. De
-00004b00: 6661 756c 7473 2074 6f20 4e6f 6e65 0d0a  faults to None..
-00004b10: 2020 2020 2020 2020 3a74 7970 6520 7175          :type qu
-00004b20: 616e 7469 6c65 3a20 666c 6f61 742c 206f  antile: float, o
-00004b30: 7074 696f 6e61 6c0d 0a20 2020 2020 2020  ptional..       
-00004b40: 203a 7261 6973 6573 2056 616c 7565 4572   :raises ValueEr
-00004b50: 726f 723a 2045 7272 6f72 2069 6620 696e  ror: Error if in
-00004b60: 7661 6c69 6420 6f76 6572 6469 7370 6572  valid overdisper
-00004b70: 7369 6f6e 206d 6574 7269 6320 6973 2063  sion metric is c
-00004b80: 686f 7365 6e2e 0d0a 2020 2020 2020 2020  hosen...        
-00004b90: 2222 220d 0a20 2020 2020 2020 200d 0a20  """..        .. 
-00004ba0: 2020 2020 2020 2069 6620 6f76 6572 6469         if overdi
-00004bb0: 7370 6572 7369 6f6e 5f6d 6574 7269 6320  spersion_metric 
-00004bc0: 6e6f 7420 696e 2073 656c 662e 6164 6174  not in self.adat
-00004bd0: 612e 7661 722e 636f 6c75 6d6e 733a 0d0a  a.var.columns:..
-00004be0: 2020 2020 2020 2020 2020 2020 6966 206f              if o
-00004bf0: 7665 7264 6973 7065 7273 696f 6e5f 6d65  verdispersion_me
-00004c00: 7472 6963 2069 6e20 2822 6f64 7363 6f72  tric in ("odscor
-00004c10: 6522 2c20 2276 7363 6f72 6522 293a 0d0a  e", "vscore"):..
-00004c20: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00004c30: 7261 6973 6520 5661 6c75 6545 7272 6f72  raise ValueError
-00004c40: 280d 0a20 2020 2020 2020 2020 2020 2020  (..             
-00004c50: 2020 2020 2020 2066 227b 6f76 6572 6469         f"{overdi
-00004c60: 7370 6572 7369 6f6e 5f6d 6574 7269 637d  spersion_metric}
-00004c70: 2068 6173 206e 6f74 2062 6565 6e20 6361   has not been ca
-00004c80: 6c63 756c 6174 6564 2066 6f72 2074 6869  lculated for thi
-00004c90: 7320 6461 7461 7365 742e 2022 0d0a 2020  s dataset. "..  
-00004ca0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00004cb0: 2020 2245 6e73 7572 6520 7468 6174 2079    "Ensure that y
-00004cc0: 6f75 2063 616c 6c20 7468 6520 6044 6174  ou call the `Dat
-00004cd0: 6173 6574 2e63 6f6d 7075 7465 5f67 656e  aset.compute_gen
-00004ce0: 655f 7374 6174 7328 2960 2066 6972 7374  e_stats()` first
-00004cf0: 2e22 0d0a 2020 2020 2020 2020 2020 2020  ."..            
-00004d00: 2020 2020 2020 2020 290d 0a20 2020 2020          )..     
-00004d10: 2020 2020 2020 2065 6c73 653a 0d0a 2020         else:..  
-00004d20: 2020 2020 2020 2020 2020 2020 2020 7261                ra
-00004d30: 6973 6520 5661 6c75 6545 7272 6f72 280d  ise ValueError(.
-00004d40: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00004d50: 2020 2020 2066 227b 6f76 6572 6469 7370       f"{overdisp
-00004d60: 6572 7369 6f6e 5f6d 6574 7269 637d 2069  ersion_metric} i
-00004d70: 7320 6e6f 7420 6120 7661 6c69 6420 6f76  s not a valid ov
-00004d80: 6572 6469 7370 6572 7369 6f6e 206d 6574  erdispersion met
-00004d90: 7269 632e 220d 0a20 2020 2020 2020 2020  ric."..         
-00004da0: 2020 2020 2020 2020 2020 2029 0d0a 2020             )..  
-00004db0: 2020 2020 2020 0d0a 2020 2020 2020 2020        ..        
-00004dc0: 2320 7761 726e 2069 6620 6d75 6c74 6970  # warn if multip
-00004dd0: 6c65 206d 6574 686f 6473 2061 7265 2073  le methods are s
-00004de0: 656c 6563 7465 640d 0a20 2020 2020 2020  elected..       
-00004df0: 2073 656c 6563 7465 645f 6d65 7468 6f64   selected_method
-00004e00: 7320 3d20 5b5d 0d0a 2020 2020 2020 2020  s = []..        
-00004e10: 6966 206d 696e 5f73 636f 7265 2069 7320  if min_score is 
-00004e20: 6e6f 7420 4e6f 6e65 3a0d 0a20 2020 2020  not None:..     
-00004e30: 2020 2020 2020 2073 656c 6563 7465 645f         selected_
-00004e40: 6d65 7468 6f64 732e 6170 7065 6e64 2822  methods.append("
-00004e50: 6d69 6e5f 7363 6f72 6522 290d 0a20 2020  min_score")..   
-00004e60: 2020 2020 2069 6620 746f 705f 6e20 6973       if top_n is
-00004e70: 206e 6f74 204e 6f6e 653a 0d0a 2020 2020   not None:..    
-00004e80: 2020 2020 2020 2020 7365 6c65 6374 6564          selected
-00004e90: 5f6d 6574 686f 6473 2e61 7070 656e 6428  _methods.append(
-00004ea0: 2274 6f70 5f6e 2229 0d0a 2020 2020 2020  "top_n")..      
-00004eb0: 2020 6966 2071 7561 6e74 696c 6520 6973    if quantile is
-00004ec0: 206e 6f74 204e 6f6e 653a 0d0a 2020 2020   not None:..    
-00004ed0: 2020 2020 2020 2020 7365 6c65 6374 6564          selected
-00004ee0: 5f6d 6574 686f 6473 2e61 7070 656e 6428  _methods.append(
-00004ef0: 2271 7561 6e74 696c 6522 290d 0a20 2020  "quantile")..   
-00004f00: 2020 2020 2069 6620 6c65 6e28 7365 6c65       if len(sele
-00004f10: 6374 6564 5f6d 6574 686f 6473 2920 3e20  cted_methods) > 
-00004f20: 313a 0d0a 2020 2020 2020 2020 2020 2020  1:..            
-00004f30: 6d65 7468 6f64 7761 726e 7374 7220 3d20  methodwarnstr = 
-00004f40: 222c 2022 2e6a 6f69 6e28 7365 6c65 6374  ", ".join(select
-00004f50: 6564 5f6d 6574 686f 6473 290d 0a20 2020  ed_methods)..   
-00004f60: 2020 2020 2020 2020 206c 6f67 6769 6e67           logging
-00004f70: 2e77 6172 6e69 6e67 2866 224d 756c 7469  .warning(f"Multi
-00004f80: 706c 6520 636f 6e66 6c69 6374 696e 6720  ple conflicting 
-00004f90: 6f76 6572 6469 7370 6572 7365 6420 6765  overdispersed ge
-00004fa0: 6e65 2073 656c 6563 7469 6f6e 2063 7269  ne selection cri
-00004fb0: 7465 7269 6120 6861 7665 2062 6565 6e20  teria have been 
-00004fc0: 7365 6c65 6374 6564 3a20 7b6d 6574 686f  selected: {metho
-00004fd0: 6477 6172 6e73 7472 7d2e 2022 0d0a 2020  dwarnstr}. "..  
-00004fe0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00004ff0: 2020 2020 2020 2020 2020 224f 6e6c 7920            "Only 
-00005000: 7468 6520 696e 7465 7273 6563 7469 6f6e  the intersection
-00005010: 206f 6620 7468 6573 6520 6d65 7468 6f64   of these method
-00005020: 7320 7769 6c6c 2062 6520 7365 6c65 6374  s will be select
-00005030: 6564 2e22 290d 0a20 2020 2020 2020 2020  ed.")..         
-00005040: 2020 2020 2020 200d 0a20 2020 2020 2020         ..       
-00005050: 2023 206d 696e 5f6d 6561 6e20 6669 6c74   # min_mean filt
-00005060: 6572 0d0a 2020 2020 2020 2020 7365 6c65  er..        sele
-00005070: 6374 6564 5f67 656e 6573 203d 2073 656c  cted_genes = sel
-00005080: 662e 6164 6174 612e 7661 725b 226d 6561  f.adata.var["mea
-00005090: 6e5f 636f 756e 7473 225d 203e 3d20 6d69  n_counts"] >= mi
-000050a0: 6e5f 6d65 616e 0d0a 2020 2020 2020 2020  n_mean..        
-000050b0: 2320 6d69 6e5f 7363 6f72 6520 6669 6c74  # min_score filt
-000050c0: 6572 0d0a 2020 2020 2020 2020 6966 206d  er..        if m
-000050d0: 696e 5f73 636f 7265 2069 7320 6e6f 7420  in_score is not 
-000050e0: 4e6f 6e65 3a0d 0a20 2020 2020 2020 2020  None:..         
-000050f0: 2020 2073 656c 6563 7465 645f 6765 6e65     selected_gene
-00005100: 7320 3d20 7365 6c65 6374 6564 5f67 656e  s = selected_gen
-00005110: 6573 2026 2028 7365 6c66 2e61 6461 7461  es & (self.adata
-00005120: 2e76 6172 5b6f 7665 7264 6973 7065 7273  .var[overdispers
-00005130: 696f 6e5f 6d65 7472 6963 5d20 3e3d 206d  ion_metric] >= m
-00005140: 696e 5f73 636f 7265 290d 0a20 2020 2020  in_score)..     
-00005150: 2020 2023 2074 6f70 5f6e 2066 696c 7465     # top_n filte
-00005160: 720d 0a20 2020 2020 2020 2069 6620 746f  r..        if to
-00005170: 705f 6e20 6973 206e 6f74 204e 6f6e 653a  p_n is not None:
-00005180: 0d0a 2020 2020 2020 2020 2020 2020 6765  ..            ge
-00005190: 6e65 7320 3d20 7365 6c66 2e61 6461 7461  nes = self.adata
-000051a0: 2e76 6172 5b6f 7665 7264 6973 7065 7273  .var[overdispers
-000051b0: 696f 6e5f 6d65 7472 6963 5d2e 736f 7274  ion_metric].sort
-000051c0: 5f76 616c 7565 7328 6173 6365 6e64 696e  _values(ascendin
-000051d0: 673d 4661 6c73 6529 2e68 6561 6428 696e  g=False).head(in
-000051e0: 7428 746f 705f 6e29 292e 696e 6465 780d  t(top_n)).index.
-000051f0: 0a20 2020 2020 2020 2020 2020 2073 656c  .            sel
-00005200: 6563 7465 645f 6765 6e65 7320 3d20 7365  ected_genes = se
-00005210: 6c65 6374 6564 5f67 656e 6573 2026 2073  lected_genes & s
-00005220: 656c 662e 6164 6174 612e 7661 722e 696e  elf.adata.var.in
-00005230: 6465 782e 6973 696e 2867 656e 6573 290d  dex.isin(genes).
-00005240: 0a20 2020 2020 2020 2023 2071 7561 6e74  .        # quant
-00005250: 696c 6520 6669 6c74 6572 0d0a 2020 2020  ile filter..    
-00005260: 2020 2020 6966 2071 7561 6e74 696c 6520      if quantile 
-00005270: 6973 206e 6f74 204e 6f6e 653a 0d0a 2020  is not None:..  
-00005280: 2020 2020 2020 2020 2020 6e5f 746f 7461            n_tota
-00005290: 6c5f 6765 6e65 7320 3d20 7365 6c66 2e61  l_genes = self.a
-000052a0: 6461 7461 2e76 6172 5b6f 7665 7264 6973  data.var[overdis
-000052b0: 7065 7273 696f 6e5f 6d65 7472 6963 5d2e  persion_metric].
-000052c0: 6e6f 746e 756c 6c28 292e 7375 6d28 290d  notnull().sum().
-000052d0: 0a20 2020 2020 2020 2020 2020 2067 656e  .            gen
-000052e0: 6573 203d 2073 656c 662e 6164 6174 612e  es = self.adata.
-000052f0: 7661 725b 6f76 6572 6469 7370 6572 7369  var[overdispersi
-00005300: 6f6e 5f6d 6574 7269 635d 2e73 6f72 745f  on_metric].sort_
-00005310: 7661 6c75 6573 2861 7363 656e 6469 6e67  values(ascending
-00005320: 3d46 616c 7365 292e 6865 6164 2869 6e74  =False).head(int
-00005330: 2871 7561 6e74 696c 6520 2a20 6e5f 746f  (quantile * n_to
-00005340: 7461 6c5f 6765 6e65 7329 292e 696e 6465  tal_genes)).inde
-00005350: 780d 0a20 2020 2020 2020 2020 2020 2073  x..            s
-00005360: 656c 6563 7465 645f 6765 6e65 7320 3d20  elected_genes = 
-00005370: 7365 6c65 6374 6564 5f67 656e 6573 2026  selected_genes &
-00005380: 2073 656c 662e 6164 6174 612e 7661 722e   self.adata.var.
-00005390: 696e 6465 782e 6973 696e 2867 656e 6573  index.isin(genes
-000053a0: 290d 0a0d 0a20 2020 2020 2020 206e 5f73  )....        n_s
-000053b0: 656c 6563 7465 645f 6765 6e65 7320 3d20  elected_genes = 
-000053c0: 7365 6c65 6374 6564 5f67 656e 6573 2e73  selected_genes.s
-000053d0: 756d 2829 0d0a 2020 2020 2020 2020 6c6f  um()..        lo
-000053e0: 6767 696e 672e 696e 666f 2866 227b 6e5f  gging.info(f"{n_
-000053f0: 7365 6c65 6374 6564 5f67 656e 6573 7d20  selected_genes} 
-00005400: 6765 6e65 7320 7365 6c65 6374 6564 2066  genes selected f
-00005410: 6f72 2066 6163 746f 7269 7a61 7469 6f6e  or factorization
-00005420: 2229 0d0a 2020 2020 2020 2020 0d0a 2020  ")..        ..  
-00005430: 2020 2020 2020 2320 6d61 6b65 2063 6861        # make cha
-00005440: 6e67 6573 2074 6f20 4461 7461 7365 7420  nges to Dataset 
-00005450: 6f62 6a65 6374 0d0a 2020 2020 2020 2020  object..        
-00005460: 7365 6c66 2e61 6461 7461 2e76 6172 5b22  self.adata.var["
-00005470: 7365 6c65 6374 6564 225d 203d 2073 656c  selected"] = sel
-00005480: 6563 7465 645f 6765 6e65 730d 0a20 2020  ected_genes..   
-00005490: 2020 2020 2073 656c 662e 6164 6174 612e       self.adata.
-000054a0: 756e 735b 226f 6467 225d 5b22 6f76 6572  uns["odg"]["over
-000054b0: 6469 7370 6572 7369 6f6e 5f6d 6574 7269  dispersion_metri
-000054c0: 6322 5d20 3d20 6f76 6572 6469 7370 6572  c"] = overdisper
-000054d0: 7369 6f6e 5f6d 6574 7269 630d 0a20 2020  sion_metric..   
-000054e0: 2020 2020 2073 656c 662e 6164 6174 612e       self.adata.
-000054f0: 756e 735b 226f 6467 225d 5b22 6d69 6e5f  uns["odg"]["min_
-00005500: 6d65 616e 225d 203d 206d 696e 5f6d 6561  mean"] = min_mea
-00005510: 6e0d 0a20 2020 2020 2020 2073 656c 662e  n..        self.
-00005520: 6164 6174 612e 756e 735b 226f 6467 225d  adata.uns["odg"]
-00005530: 5b22 6d69 6e5f 7363 6f72 6522 5d20 3d20  ["min_score"] = 
-00005540: 6d69 6e5f 7363 6f72 6520 6966 206d 696e  min_score if min
-00005550: 5f73 636f 7265 2069 7320 6e6f 7420 4e6f  _score is not No
-00005560: 6e65 2065 6c73 6520 2222 0d0a 2020 2020  ne else ""..    
-00005570: 2020 2020 7365 6c66 2e61 6461 7461 2e75      self.adata.u
-00005580: 6e73 5b22 6f64 6722 5d5b 2274 6f70 5f6e  ns["odg"]["top_n
-00005590: 225d 203d 2074 6f70 5f6e 2069 6620 746f  "] = top_n if to
-000055a0: 705f 6e20 6973 206e 6f74 204e 6f6e 6520  p_n is not None 
-000055b0: 656c 7365 2022 220d 0a20 2020 2020 2020  else ""..       
-000055c0: 2073 656c 662e 6164 6174 612e 756e 735b   self.adata.uns[
-000055d0: 226f 6467 225d 5b22 7175 616e 7469 6c65  "odg"]["quantile
-000055e0: 225d 203d 2071 7561 6e74 696c 6520 6966  "] = quantile if
-000055f0: 2071 7561 6e74 696c 6520 6973 206e 6f74   quantile is not
-00005600: 204e 6f6e 6520 656c 7365 2022 220d 0a20   None else "".. 
-00005610: 2020 2020 2020 2073 656c 662e 6170 7065         self.appe
-00005620: 6e64 5f74 6f5f 6869 7374 6f72 7928 224f  nd_to_history("O
-00005630: 7665 7264 6973 7065 7273 6564 2067 656e  verdispersed gen
-00005640: 6573 2073 656c 6563 7465 6422 290d 0a20  es selected").. 
-00005650: 2020 200d 0a20 2020 2064 6566 2069 6e69     ..    def ini
-00005660: 7469 616c 697a 655f 636e 6d66 2873 656c  tialize_cnmf(sel
-00005670: 662c 2063 6e6d 665f 6f75 7470 7574 5f64  f, cnmf_output_d
-00005680: 6972 3a20 7374 722c 0d0a 2020 2020 2020  ir: str,..      
-00005690: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000056a0: 2020 636e 6d66 5f6e 616d 653a 2073 7472    cnmf_name: str
-000056b0: 2c0d 0a20 2020 2020 2020 2020 2020 2020  ,..             
-000056c0: 2020 2020 2020 2020 2020 206b 7661 6c73             kvals
-000056d0: 3a20 436f 6c6c 6563 7469 6f6e 203d 2072  : Collection = r
-000056e0: 616e 6765 2832 2c20 3631 292c 0d0a 2020  ange(2, 61),..  
-000056f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00005700: 2020 2020 2020 6e5f 6974 6572 3a20 696e        n_iter: in
-00005710: 7420 3d20 3230 302c 0d0a 2020 2020 2020  t = 200,..      
-00005720: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00005730: 2020 6265 7461 5f6c 6f73 733a 2073 7472    beta_loss: str
-00005740: 203d 2022 6b75 6c6c 6261 636b 2d6c 6569   = "kullback-lei
-00005750: 626c 6572 222c 0d0a 2020 2020 2020 2020  bler",..        
-00005760: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00005770: 7365 6564 3a20 4f70 7469 6f6e 616c 5b69  seed: Optional[i
-00005780: 6e74 5d20 3d20 4e6f 6e65 2920 2d3e 2063  nt] = None) -> c
-00005790: 6e6d 662e 634e 4d46 3a0d 0a20 2020 2020  nmf.cNMF:..     
-000057a0: 2020 2022 2222 496e 6974 6961 6c69 7a65     """Initialize
-000057b0: 2061 2063 4e4d 4620 7275 6e20 666f 7220   a cNMF run for 
-000057c0: 7375 6273 6571 7565 6e74 2066 6163 746f  subsequent facto
-000057d0: 7269 7a61 7469 6f6e 2e0d 0a0d 0a20 2020  rization.....   
-000057e0: 2020 2020 203a 7061 7261 6d20 636e 6d66       :param cnmf
-000057f0: 5f6f 7574 7075 745f 6469 723a 204f 7574  _output_dir: Out
-00005800: 7075 7420 6469 7265 6374 6f72 7920 666f  put directory fo
-00005810: 7220 634e 4d46 2072 6573 756c 7473 0d0a  r cNMF results..
-00005820: 2020 2020 2020 2020 3a74 7970 6520 636e          :type cn
-00005830: 6d66 5f6f 7574 7075 745f 6469 723a 2073  mf_output_dir: s
-00005840: 7472 0d0a 2020 2020 2020 2020 3a70 6172  tr..        :par
-00005850: 616d 2063 6e6d 665f 6e61 6d65 3a20 4e61  am cnmf_name: Na
-00005860: 6d65 206f 6620 7468 6520 634e 4d46 2072  me of the cNMF r
-00005870: 6573 756c 7473 2e20 4669 6c65 7320 7769  esults. Files wi
-00005880: 6c6c 2062 6520 6f75 7470 7574 2074 6f20  ll be output to 
-00005890: 5b63 6e6d 665f 6f75 7470 7574 5f64 6972  [cnmf_output_dir
-000058a0: 5d2f 5b63 6e6d 665f 6e61 6d65 5d2f 0d0a  ]/[cnmf_name]/..
-000058b0: 2020 2020 2020 2020 3a74 7970 6520 636e          :type cn
-000058c0: 6d66 5f6e 616d 653a 2073 7472 0d0a 2020  mf_name: str..  
-000058d0: 2020 2020 2020 3a70 6172 616d 206b 7661        :param kva
-000058e0: 6c73 3a20 5261 6e6b 7320 666f 7220 634e  ls: Ranks for cN
-000058f0: 4d46 2066 6163 746f 7269 7a61 7469 6f6e  MF factorization
-00005900: 2c20 6465 6661 756c 7473 2074 6f20 7261  , defaults to ra
-00005910: 6e67 6528 322c 2036 3129 0d0a 2020 2020  nge(2, 61)..    
-00005920: 2020 2020 3a74 7970 6520 6b76 616c 733a      :type kvals:
-00005930: 2043 6f6c 6c65 6374 696f 6e2c 206f 7074   Collection, opt
-00005940: 696f 6e61 6c0d 0a20 2020 2020 2020 203a  ional..        :
-00005950: 7061 7261 6d20 6e5f 6974 6572 3a20 4e75  param n_iter: Nu
-00005960: 6d62 6572 206f 6620 6974 6572 6174 696f  mber of iteratio
-00005970: 6e73 2066 726f 6d20 7768 6963 6820 746f  ns from which to
-00005980: 2062 7569 6c64 2061 2063 6f6e 7365 6e73   build a consens
-00005990: 7573 2073 6f6c 7574 696f 6e2c 2064 6566  us solution, def
-000059a0: 6175 6c74 7320 746f 2032 3030 0d0a 2020  aults to 200..  
-000059b0: 2020 2020 2020 3a74 7970 6520 6e5f 6974        :type n_it
-000059c0: 6572 3a20 696e 742c 206f 7074 696f 6e61  er: int, optiona
-000059d0: 6c0d 0a20 2020 2020 2020 203a 7061 7261  l..        :para
-000059e0: 6d20 6265 7461 5f6c 6f73 733a 2062 6574  m beta_loss: bet
-000059f0: 612d 6c6f 7373 2066 756e 6374 696f 6e2c  a-loss function,
-00005a00: 2065 6974 6865 7220 226b 756c 6c62 6163   either "kullbac
-00005a10: 6b2d 6c65 6962 6c65 7222 206f 7220 2266  k-leibler" or "f
-00005a20: 726f 6265 6e69 7573 222e 2044 6566 6175  robenius". Defau
-00005a30: 6c74 7320 746f 2022 6b75 6c6c 6261 636b  lts to "kullback
-00005a40: 2d6c 6569 626c 6572 220d 0a20 2020 2020  -leibler"..     
-00005a50: 2020 203a 7479 7065 2062 6574 615f 6c6f     :type beta_lo
-00005a60: 7373 3a20 7374 722c 206f 7074 696f 6e61  ss: str, optiona
-00005a70: 6c0d 0a20 2020 2020 2020 203a 7061 7261  l..        :para
-00005a80: 6d20 7365 6564 3a20 5261 6e64 6f6d 2073  m seed: Random s
-00005a90: 6565 6420 666f 7220 7265 7072 6f64 7563  eed for reproduc
-00005aa0: 6962 696c 6974 792c 2064 6566 6175 6c74  ibility, default
-00005ab0: 7320 746f 204e 6f6e 650d 0a20 2020 2020  s to None..     
-00005ac0: 2020 203a 7479 7065 2073 6565 643a 204f     :type seed: O
-00005ad0: 7074 696f 6e61 6c5b 696e 745d 2c20 6f70  ptional[int], op
-00005ae0: 7469 6f6e 616c 0d0a 2020 2020 2020 2020  tional..        
-00005af0: 3a72 6574 7572 6e3a 2063 4e4d 4620 6f62  :return: cNMF ob
-00005b00: 6a65 6374 0d0a 2020 2020 2020 2020 3a72  ject..        :r
-00005b10: 7479 7065 3a20 3a63 6c61 7373 3a60 636e  type: :class:`cn
-00005b20: 6d66 736e 732e 636e 6d66 2e63 4e4d 4660  mfsns.cnmf.cNMF`
-00005b30: 0d0a 2020 2020 2020 2020 2222 220d 0a20  ..        """.. 
-00005b40: 2020 2020 2020 2063 6e6d 665f 6f62 6a20         cnmf_obj 
-00005b50: 3d20 636e 6d66 2e63 4e4d 4628 6f75 7470  = cnmf.cNMF(outp
-00005b60: 7574 5f64 6972 3d63 6e6d 665f 6f75 7470  ut_dir=cnmf_outp
-00005b70: 7574 5f64 6972 2c20 6e61 6d65 3d63 6e6d  ut_dir, name=cnm
-00005b80: 665f 6e61 6d65 290d 0a20 2020 2020 2020  f_name)..       
-00005b90: 200d 0a20 2020 2020 2020 2023 2077 7269   ..        # wri
-00005ba0: 7465 2054 504d 2028 6e6f 726d 616c 697a  te TPM (normaliz
-00005bb0: 6564 2920 6461 7461 0d0a 2020 2020 2020  ed) data..      
-00005bc0: 2020 7470 6d20 3d20 6164 2e41 6e6e 4461    tpm = ad.AnnDa
-00005bd0: 7461 2873 656c 662e 746f 5f64 6628 6e6f  ta(self.to_df(no
-00005be0: 726d 616c 697a 6564 3d54 7275 6529 290d  rmalized=True)).
-00005bf0: 0a20 2020 2020 2020 2074 706d 2e77 7269  .        tpm.wri
-00005c00: 7465 5f68 3561 6428 636e 6d66 5f6f 626a  te_h5ad(cnmf_obj
-00005c10: 2e70 6174 6873 5b22 7470 6d22 5d29 0d0a  .paths["tpm"])..
-00005c20: 0d0a 2020 2020 2020 2020 6765 6e65 5f74  ..        gene_t
-00005c30: 706d 5f6d 6561 6e20 3d20 6e70 2e61 7272  pm_mean = np.arr
-00005c40: 6179 2874 706d 2e58 2e6d 6561 6e28 6178  ay(tpm.X.mean(ax
-00005c50: 6973 3d30 2929 2e72 6573 6861 7065 282d  is=0)).reshape(-
-00005c60: 3129 0d0a 2020 2020 2020 2020 6765 6e65  1)..        gene
-00005c70: 5f74 706d 5f73 7464 6465 7620 3d20 6e70  _tpm_stddev = np
-00005c80: 2e61 7272 6179 2874 706d 2e58 2e73 7464  .array(tpm.X.std
-00005c90: 2861 7869 733d 302c 2064 646f 663d 3029  (axis=0, ddof=0)
-00005ca0: 292e 7265 7368 6170 6528 2d31 290d 0a20  ).reshape(-1).. 
-00005cb0: 2020 2020 2020 2069 6e70 7574 5f74 706d         input_tpm
-00005cc0: 5f73 7461 7473 203d 2070 642e 4461 7461  _stats = pd.Data
-00005cd0: 4672 616d 6528 5b67 656e 655f 7470 6d5f  Frame([gene_tpm_
-00005ce0: 6d65 616e 2c20 6765 6e65 5f74 706d 5f73  mean, gene_tpm_s
-00005cf0: 7464 6465 765d 2c20 696e 6465 7820 3d20  tddev], index = 
-00005d00: 5b27 5f5f 6d65 616e 272c 2027 5f5f 7374  ['__mean', '__st
-00005d10: 6427 5d29 2e54 0d0a 2020 2020 2020 2020  d']).T..        
-00005d20: 7574 696c 732e 7361 7665 5f64 665f 746f  utils.save_df_to
-00005d30: 5f6e 707a 2869 6e70 7574 5f74 706d 5f73  _npz(input_tpm_s
-00005d40: 7461 7473 2c20 636e 6d66 5f6f 626a 2e70  tats, cnmf_obj.p
-00005d50: 6174 6873 5b27 7470 6d5f 7374 6174 7327  aths['tpm_stats'
-00005d60: 5d29 0d0a 2020 2020 2020 2020 6f76 6572  ])..        over
-00005d70: 6469 7370 6572 7365 645f 6765 6e65 7320  dispersed_genes 
-00005d80: 3d20 7365 6c66 2e61 6461 7461 2e76 6172  = self.adata.var
-00005d90: 5b22 7365 6c65 6374 6564 225d 5b73 656c  ["selected"][sel
-00005da0: 662e 6164 6174 612e 7661 725b 2273 656c  f.adata.var["sel
-00005db0: 6563 7465 6422 5d5d 2e69 6e64 6578 0d0a  ected"]].index..
-00005dc0: 2020 2020 2020 2020 6e6f 726d 5f63 6f75          norm_cou
-00005dd0: 6e74 7320 3d20 636e 6d66 5f6f 626a 2e67  nts = cnmf_obj.g
-00005de0: 6574 5f6e 6f72 6d5f 636f 756e 7473 2873  et_norm_counts(s
-00005df0: 656c 662e 6164 6174 612c 2074 706d 2c20  elf.adata, tpm, 
-00005e00: 6869 6768 5f76 6172 6961 6e63 655f 6765  high_variance_ge
-00005e10: 6e65 735f 6669 6c74 6572 3d6f 7665 7264  nes_filter=overd
-00005e20: 6973 7065 7273 6564 5f67 656e 6573 290d  ispersed_genes).
-00005e30: 0a20 2020 2020 2020 2069 6620 6e6f 726d  .        if norm
-00005e40: 5f63 6f75 6e74 732e 582e 6474 7970 6520  _counts.X.dtype 
-00005e50: 213d 206e 702e 666c 6f61 7436 343a 0d0a  != np.float64:..
-00005e60: 2020 2020 2020 2020 2020 2020 6e6f 726d              norm
-00005e70: 5f63 6f75 6e74 732e 5820 3d20 6e6f 726d  _counts.X = norm
-00005e80: 5f63 6f75 6e74 732e 582e 6173 7479 7065  _counts.X.astype
-00005e90: 286e 702e 666c 6f61 7436 3429 0d0a 2020  (np.float64)..  
-00005ea0: 2020 2020 2020 636e 6d66 5f6f 626a 2e73        cnmf_obj.s
-00005eb0: 6176 655f 6e6f 726d 5f63 6f75 6e74 7328  ave_norm_counts(
-00005ec0: 6e6f 726d 5f63 6f75 6e74 7329 0d0a 0d0a  norm_counts)....
-00005ed0: 2020 2020 2020 2020 2320 7361 7665 2070          # save p
-00005ee0: 6172 616d 6574 6572 7320 666f 7220 6661  arameters for fa
-00005ef0: 6374 6f72 697a 6174 696f 6e20 7374 6570  ctorization step
-00005f00: 0d0a 2020 2020 2020 2020 636e 6d66 5f6f  ..        cnmf_o
-00005f10: 626a 2e73 6176 655f 6e6d 665f 6974 6572  bj.save_nmf_iter
-00005f20: 5f70 6172 616d 7328 2a63 6e6d 665f 6f62  _params(*cnmf_ob
-00005f30: 6a2e 6765 745f 6e6d 665f 6974 6572 5f70  j.get_nmf_iter_p
-00005f40: 6172 616d 7328 6b73 3d6b 7661 6c73 2c20  arams(ks=kvals, 
-00005f50: 6e5f 6974 6572 3d6e 5f69 7465 722c 2072  n_iter=n_iter, r
-00005f60: 616e 646f 6d5f 7374 6174 655f 7365 6564  andom_state_seed
-00005f70: 3d73 6565 642c 2062 6574 615f 6c6f 7373  =seed, beta_loss
-00005f80: 3d62 6574 615f 6c6f 7373 2929 0d0a 0d0a  =beta_loss))....
-00005f90: 2020 2020 2020 2020 2320 7361 7665 2070          # save p
-00005fa0: 6172 616d 6574 6572 7320 696e 2041 6e6e  arameters in Ann
-00005fb0: 4461 7461 206f 626a 6563 740d 0a20 2020  Data object..   
-00005fc0: 2020 2020 2073 656c 662e 6164 6174 612e       self.adata.
-00005fd0: 756e 735b 2263 6e6d 6622 5d20 3d20 636e  uns["cnmf"] = cn
-00005fe0: 6d66 5f6f 626a 2e67 6574 5f6e 6d66 5f69  mf_obj.get_nmf_i
-00005ff0: 7465 725f 7061 7261 6d73 286b 733d 6b76  ter_params(ks=kv
-00006000: 616c 732c 206e 5f69 7465 723d 6e5f 6974  als, n_iter=n_it
-00006010: 6572 2c20 7261 6e64 6f6d 5f73 7461 7465  er, random_state
-00006020: 5f73 6565 643d 7365 6564 2c20 6265 7461  _seed=seed, beta
-00006030: 5f6c 6f73 733d 6265 7461 5f6c 6f73 7329  _loss=beta_loss)
-00006040: 5b31 5d20 2023 2064 6963 7420 6f66 2063  [1]  # dict of c
-00006050: 6e6d 6620 7061 7261 6d65 7465 7273 0d0a  nmf parameters..
-00006060: 2020 2020 2020 2020 0d0a 2020 2020 2020          ..      
-00006070: 2020 7365 6c66 2e61 7070 656e 645f 746f    self.append_to
-00006080: 5f68 6973 746f 7279 2866 2263 4e4d 4620  _history(f"cNMF 
-00006090: 7061 7261 6d65 7465 7273 2061 6464 6564  parameters added
-000060a0: 2e20 634e 4d46 2069 6e70 7574 7320 696e  . cNMF inputs in
-000060b0: 6974 6961 6c69 7a65 6420 696e 207b 636e  itialized in {cn
-000060c0: 6d66 5f6f 7574 7075 745f 6469 727d 2f7b  mf_output_dir}/{
-000060d0: 636e 6d66 5f6e 616d 657d 2229 0d0a 2020  cnmf_name}")..  
-000060e0: 2020 2020 2020 7265 7475 726e 2063 6e6d        return cnm
-000060f0: 665f 6f62 6a0d 0a20 2020 200d 0a20 2020  f_obj..    ..   
-00006100: 2064 6566 2061 6464 5f63 6e6d 665f 7265   def add_cnmf_re
-00006110: 7375 6c74 7328 7365 6c66 2c20 636e 6d66  sults(self, cnmf
-00006120: 5f6f 7574 7075 745f 6469 722c 2063 6e6d  _output_dir, cnm
-00006130: 665f 6e61 6d65 2c20 6c6f 6361 6c5f 6465  f_name, local_de
-00006140: 6e73 6974 795f 7468 7265 7368 6f6c 643a  nsity_threshold:
-00006150: 2066 6c6f 6174 203d 204e 6f6e 652c 206c   float = None, l
-00006160: 6f63 616c 5f6e 6569 6768 626f 7268 6f6f  ocal_neighborhoo
-00006170: 645f 7369 7a65 3a20 666c 6f61 7420 3d20  d_size: float = 
-00006180: 4e6f 6e65 293a 0d0a 2020 2020 2020 2020  None):..        
-00006190: 2222 220d 0a20 2020 2020 2020 2041 6674  """..        Aft
-000061a0: 6572 2066 6163 746f 7269 7a61 7469 6f6e  er factorization
-000061b0: 2c20 6164 6420 636f 6d70 6c65 7465 6420  , add completed 
-000061c0: 634e 4d46 2072 6573 756c 7473 2069 6e20  cNMF results in 
-000061d0: 5b63 6e6d 665f 6f75 7470 7574 5f64 6972  [cnmf_output_dir
-000061e0: 5d2f 5b63 6e6d 665f 6e61 6d65 5d20 746f  ]/[cnmf_name] to
-000061f0: 2074 6865 2064 6174 6173 6574 206f 626a   the dataset obj
-00006200: 6563 742e 0d0a 0d0a 2020 2020 2020 2020  ect.....        
-00006210: 3a70 6172 616d 2063 6e6d 665f 6f75 7470  :param cnmf_outp
-00006220: 7574 5f64 6972 3a20 4f75 7470 7574 2064  ut_dir: Output d
-00006230: 6972 6563 746f 7279 2066 6f72 2063 4e4d  irectory for cNM
-00006240: 4620 7265 7375 6c74 730d 0a20 2020 2020  F results..     
-00006250: 2020 203a 7479 7065 2063 6e6d 665f 6f75     :type cnmf_ou
-00006260: 7470 7574 5f64 6972 3a20 7374 720d 0a20  tput_dir: str.. 
-00006270: 2020 2020 2020 203a 7061 7261 6d20 636e         :param cn
-00006280: 6d66 5f6e 616d 653a 204e 616d 6520 6f66  mf_name: Name of
-00006290: 2074 6865 2063 4e4d 4620 7265 7375 6c74   the cNMF result
-000062a0: 732e 2046 696c 6573 2077 696c 6c20 6265  s. Files will be
-000062b0: 206f 7574 7075 7420 746f 205b 636e 6d66   output to [cnmf
-000062c0: 5f6f 7574 7075 745f 6469 725d 2f5b 636e  _output_dir]/[cn
-000062d0: 6d66 5f6e 616d 655d 2f0d 0a20 2020 2020  mf_name]/..     
-000062e0: 2020 203a 7479 7065 2063 6e6d 665f 6e61     :type cnmf_na
-000062f0: 6d65 3a20 7374 720d 0a20 2020 2020 2020  me: str..       
-00006300: 203a 7061 7261 6d20 6c6f 6361 6c5f 6465   :param local_de
-00006310: 6e73 6974 795f 7468 7265 7368 6f6c 643a  nsity_threshold:
-00006320: 2054 6872 6573 686f 6c64 2066 6f72 2074   Threshold for t
-00006330: 6865 206c 6f63 616c 2064 656e 7369 7479  he local density
-00006340: 2066 696c 7465 7269 6e67 2070 7269 6f72   filtering prior
-00006350: 2074 6f20 4745 5020 636f 6e73 656e 7375   to GEP consensu
-00006360: 732e 2041 6363 6570 7461 626c 6520 7468  s. Acceptable th
-00006370: 7265 7368 6f6c 6473 2061 7265 203e 2030  resholds are > 0
-00006380: 2061 6e64 203c 3d20 3220 2832 2e30 2069   and <= 2 (2.0 i
-00006390: 7320 6e6f 2066 696c 7465 7269 6e67 292e  s no filtering).
-000063a0: 2044 6566 6175 6c74 7320 746f 204e 6f6e   Defaults to Non
-000063b0: 652e 0d0a 2020 2020 2020 2020 3a74 7970  e...        :typ
-000063c0: 6520 6c6f 6361 6c5f 6465 6e73 6974 795f  e local_density_
-000063d0: 7468 7265 7368 6f6c 643a 2066 6c6f 6174  threshold: float
-000063e0: 2c20 6f70 7469 6f6e 616c 0d0a 2020 2020  , optional..    
-000063f0: 2020 2020 3a70 6172 616d 206c 6f63 616c      :param local
-00006400: 5f6e 6569 6768 626f 7268 6f6f 645f 7369  _neighborhood_si
-00006410: 7a65 3a20 4672 6163 7469 6f6e 206f 6620  ze: Fraction of 
-00006420: 7468 6520 6e75 6d62 6572 206f 6620 7265  the number of re
-00006430: 706c 6963 6174 6573 2074 6f20 7573 6520  plicates to use 
-00006440: 6173 206e 6561 7265 7374 206e 6569 6768  as nearest neigh
-00006450: 626f 7273 2066 6f72 206c 6f63 616c 2064  bors for local d
-00006460: 656e 7369 7479 2066 696c 7465 7269 6e67  ensity filtering
-00006470: 2e20 4465 6661 756c 7473 2074 6f20 4e6f  . Defaults to No
-00006480: 6e65 0d0a 2020 2020 2020 2020 3a74 7970  ne..        :typ
-00006490: 6520 6c6f 6361 6c5f 6e65 6967 6862 6f72  e local_neighbor
-000064a0: 686f 6f64 5f73 697a 653a 2066 6c6f 6174  hood_size: float
-000064b0: 2c20 6f70 7469 6f6e 616c 0d0a 2020 2020  , optional..    
-000064c0: 2020 2020 2222 220d 0a20 2020 2020 2020      """..       
-000064d0: 2073 656c 662e 6164 6174 612e 756e 735b   self.adata.uns[
-000064e0: 2263 6e6d 665f 6e61 6d65 225d 203d 2063  "cnmf_name"] = c
-000064f0: 6e6d 665f 6e61 6d65 0d0a 0d0a 2020 2020  nmf_name....    
-00006500: 2020 2020 2320 696e 6665 7220 6672 6f6d      # infer from
-00006510: 2066 696c 656e 616d 6573 2077 6869 6368   filenames which
-00006520: 206c 6f63 616c 2064 656e 7369 7479 2074   local density t
-00006530: 6872 6573 686f 6c64 2077 6173 2075 7365  hreshold was use
-00006540: 640d 0a20 2020 2020 2020 2073 656e 7365  d..        sense
-00006550: 645f 6c64 7473 203d 2073 6574 2829 0d0a  d_ldts = set()..
-00006560: 2020 2020 2020 2020 666f 7220 666e 2069          for fn i
-00006570: 6e20 676c 6f62 286f 732e 7061 7468 2e6a  n glob(os.path.j
-00006580: 6f69 6e28 636e 6d66 5f6f 7574 7075 745f  oin(cnmf_output_
-00006590: 6469 722c 2063 6e6d 665f 6e61 6d65 2c20  dir, cnmf_name, 
-000065a0: 6622 7b63 6e6d 665f 6e61 6d65 7d2a 2e2a  f"{cnmf_name}*.*
-000065b0: 7370 6563 7472 612a 2e6b 5f2a 2229 293a  spectra*.k_*")):
-000065c0: 0d0a 2020 2020 2020 2020 2020 2020 6c64  ..            ld
-000065d0: 745f 7374 7220 3d20 6f73 2e70 6174 682e  t_str = os.path.
-000065e0: 6261 7365 6e61 6d65 2866 6e29 2e73 706c  basename(fn).spl
-000065f0: 6974 2822 2e22 295b 2d33 5d0d 0a20 2020  it(".")[-3]..   
-00006600: 2020 2020 2020 2020 2074 7279 3a0d 0a20           try:.. 
-00006610: 2020 2020 2020 2020 2020 2020 2020 206c                 l
-00006620: 6474 203d 2066 6c6f 6174 286c 6474 5f73  dt = float(ldt_s
-00006630: 7472 2e72 6570 6c61 6365 2822 6474 5f22  tr.replace("dt_"
-00006640: 2c20 2222 292e 7265 706c 6163 6528 225f  , "").replace("_
-00006650: 222c 2022 2e22 2929 0d0a 2020 2020 2020  ", "."))..      
-00006660: 2020 2020 2020 6578 6365 7074 2056 616c        except Val
-00006670: 7565 4572 726f 723a 0d0a 2020 2020 2020  ueError:..      
-00006680: 2020 2020 2020 2020 2020 7061 7373 0d0a            pass..
-00006690: 2020 2020 2020 2020 2020 2020 656c 7365              else
-000066a0: 3a0d 0a20 2020 2020 2020 2020 2020 2020  :..             
-000066b0: 2020 2073 656e 7365 645f 6c64 7473 2e61     sensed_ldts.a
-000066c0: 6464 2828 6c64 745f 7374 722c 206c 6474  dd((ldt_str, ldt
-000066d0: 2929 0d0a 2020 2020 2020 2020 6966 206c  ))..        if l
-000066e0: 6f63 616c 5f64 656e 7369 7479 5f74 6872  ocal_density_thr
-000066f0: 6573 686f 6c64 2069 7320 4e6f 6e65 2061  eshold is None a
-00006700: 6e64 206c 656e 2873 656e 7365 645f 6c64  nd len(sensed_ld
-00006710: 7473 2920 3d3d 2031 3a0d 0a20 2020 2020  ts) == 1:..     
-00006720: 2020 2020 2020 206c 6474 5f73 7472 2c20         ldt_str, 
-00006730: 6c64 7420 3d20 7365 6e73 6564 5f6c 6474  ldt = sensed_ldt
-00006740: 732e 706f 7028 290d 0a20 2020 2020 2020  s.pop()..       
-00006750: 2065 6c69 6620 6c6f 6361 6c5f 6465 6e73   elif local_dens
-00006760: 6974 795f 7468 7265 7368 6f6c 6420 696e  ity_threshold in
-00006770: 2028 6c64 745b 315d 2066 6f72 206c 6474   (ldt[1] for ldt
-00006780: 2069 6e20 7365 6e73 6564 5f6c 6474 7329   in sensed_ldts)
-00006790: 3a0d 0a20 2020 2020 2020 2020 2020 206c  :..            l
-000067a0: 6474 5f73 7472 2c20 6c64 7420 3d20 5b28  dt_str, ldt = [(
-000067b0: 6c64 745f 7374 722c 206c 6474 2920 666f  ldt_str, ldt) fo
-000067c0: 7220 6c64 745f 7374 722c 206c 6474 2069  r ldt_str, ldt i
-000067d0: 6e20 7365 6e73 6564 5f6c 6474 7320 6966  n sensed_ldts if
-000067e0: 206c 6474 203d 3d20 6c6f 6361 6c5f 6465   ldt == local_de
-000067f0: 6e73 6974 795f 7468 7265 7368 6f6c 645d  nsity_threshold]
-00006800: 2e70 6f70 2829 0d0a 2020 2020 2020 2020  .pop()..        
-00006810: 656c 7365 3a0d 0a20 2020 2020 2020 2020  else:..         
-00006820: 2020 206c 6f67 6769 6e67 2e65 7272 6f72     logging.error
-00006830: 2866 226c 6f63 616c 5f64 656e 7369 7479  (f"local_density
-00006840: 5f74 6872 6573 686f 6c64 206f 6620 7b6c  _threshold of {l
-00006850: 6f63 616c 5f64 656e 7369 7479 5f74 6872  ocal_density_thr
-00006860: 6573 686f 6c64 7d20 646f 6573 206e 6f74  eshold} does not
-00006870: 206d 6174 6368 2077 6861 7420 6973 2069   match what is i
-00006880: 6e20 7468 6520 634e 4d46 2072 6573 756c  n the cNMF resul
-00006890: 7420 6469 7265 6374 6f72 793a 207b 7365  t directory: {se
-000068a0: 6e73 6564 5f6c 6474 737d 2229 0d0a 2020  nsed_ldts}")..  
-000068b0: 2020 2020 2020 2020 2020 7379 732e 6578            sys.ex
-000068c0: 6974 2831 290d 0a20 2020 2020 2020 2073  it(1)..        s
-000068d0: 656c 662e 6164 6174 612e 756e 735b 226c  elf.adata.uns["l
-000068e0: 6474 225d 203d 206c 6474 0d0a 2020 2020  dt"] = ldt..    
-000068f0: 2020 2020 7365 6c66 2e61 6461 7461 2e75      self.adata.u
-00006900: 6e73 5b22 6c6e 7322 5d20 3d20 6c6f 6361  ns["lns"] = loca
-00006910: 6c5f 6e65 6967 6862 6f72 686f 6f64 5f73  l_neighborhood_s
-00006920: 697a 650d 0a20 2020 2020 2020 2020 2020  ize..           
-00006930: 200d 0a20 2020 2020 2020 2023 2049 6d70   ..        # Imp
-00006940: 6f72 7420 4745 5073 0d0a 2020 2020 2020  ort GEPs..      
-00006950: 2020 7265 7375 6c74 5f74 7970 6573 203d    result_types =
-00006960: 207b 0d0a 2020 2020 2020 2020 2020 2020   {..            
-00006970: 2267 656e 655f 7370 6563 7472 615f 7363  "gene_spectra_sc
-00006980: 6f72 6522 3a20 2263 6e6d 665f 6765 705f  ore": "cnmf_gep_
-00006990: 7363 6f72 6522 2c0d 0a20 2020 2020 2020  score",..       
-000069a0: 2020 2020 2022 6765 6e65 5f73 7065 6374       "gene_spect
-000069b0: 7261 5f74 706d 223a 2022 636e 6d66 5f67  ra_tpm": "cnmf_g
-000069c0: 6570 5f74 706d 222c 0d0a 2020 2020 2020  ep_tpm",..      
-000069d0: 2020 2020 2020 2273 7065 6374 7261 223a        "spectra":
-000069e0: 2022 636e 6d66 5f67 6570 5f72 6177 220d   "cnmf_gep_raw".
-000069f0: 0a20 2020 2020 2020 2020 2020 207d 0d0a  .            }..
-00006a00: 2020 2020 2020 2020 666f 7220 6d61 7463          for matc
-00006a10: 6873 7472 2c20 7265 7375 6c74 5f74 7970  hstr, result_typ
-00006a20: 6520 696e 2072 6573 756c 745f 7479 7065  e in result_type
-00006a30: 732e 6974 656d 7328 293a 0d0a 2020 2020  s.items():..    
-00006a40: 2020 2020 2020 2020 6c6f 6767 696e 672e          logging.
-00006a50: 696e 666f 2866 2249 6d70 6f72 7469 6e67  info(f"Importing
-00006a60: 2047 4550 733a 207b 6d61 7463 6873 7472   GEPs: {matchstr
-00006a70: 7d22 2920 200d 0a20 2020 2020 2020 2020  }")  ..         
-00006a80: 2020 206d 6574 615f 7720 3d20 5b5d 0d0a     meta_w = []..
-00006a90: 2020 2020 2020 2020 2020 2020 666f 7220              for 
-00006aa0: 666e 2069 6e20 676c 6f62 286f 732e 7061  fn in glob(os.pa
-00006ab0: 7468 2e6a 6f69 6e28 636e 6d66 5f6f 7574  th.join(cnmf_out
-00006ac0: 7075 745f 6469 722c 2063 6e6d 665f 6e61  put_dir, cnmf_na
-00006ad0: 6d65 2c20 6622 7b63 6e6d 665f 6e61 6d65  me, f"{cnmf_name
-00006ae0: 7d2a 2e7b 6d61 7463 6873 7472 7d2e 6b5f  }*.{matchstr}.k_
-00006af0: 2a2e 7b6c 6474 5f73 7472 7d2e 2a74 7874  *.{ldt_str}.*txt
-00006b00: 2229 293a 0d0a 2020 2020 2020 2020 2020  ")):..          
-00006b10: 2020 2020 2020 6b20 3d20 696e 7428 6f73        k = int(os
-00006b20: 2e70 6174 682e 6261 7365 6e61 6d65 2866  .path.basename(f
-00006b30: 6e29 2e72 656d 6f76 6570 7265 6669 7828  n).removeprefix(
-00006b40: 6622 7b63 6e6d 665f 6e61 6d65 7d2e 7b6d  f"{cnmf_name}.{m
-00006b50: 6174 6368 7374 727d 2e22 292e 7370 6c69  atchstr}.").spli
-00006b60: 7428 222e 2229 5b30 5d2e 7265 706c 6163  t(".")[0].replac
-00006b70: 6528 226b 5f22 2c20 2222 2929 0d0a 2020  e("k_", ""))..  
-00006b80: 2020 2020 2020 2020 2020 2020 2020 7720                w 
-00006b90: 3d20 7064 2e72 6561 645f 7461 626c 6528  = pd.read_table(
-00006ba0: 666e 2c20 696e 6465 785f 636f 6c3d 3029  fn, index_col=0)
-00006bb0: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
-00006bc0: 2020 772e 696e 6465 7820 3d20 7374 7228    w.index = str(
-00006bd0: 6b29 202b 2022 2e22 202b 2077 2e69 6e64  k) + "." + w.ind
-00006be0: 6578 2e61 7374 7970 6528 7374 7229 0d0a  ex.astype(str)..
-00006bf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00006c00: 6d65 7461 5f77 2e61 7070 656e 6428 7729  meta_w.append(w)
-00006c10: 0d0a 2020 2020 2020 2020 2020 2020 6d65  ..            me
-00006c20: 7461 5f77 203d 2070 642e 636f 6e63 6174  ta_w = pd.concat
-00006c30: 286d 6574 615f 772c 2061 7869 733d 3029  (meta_w, axis=0)
-00006c40: 2e54 2e72 6569 6e64 6578 2873 656c 662e  .T.reindex(self.
-00006c50: 6164 6174 612e 7661 722e 696e 6465 7829  adata.var.index)
-00006c60: 2e72 656e 616d 655f 6178 6973 285b 226b  .rename_axis(["k
-00006c70: 2e67 6570 225d 2c20 6178 6973 3d31 290d  .gep"], axis=1).
-00006c80: 0a20 2020 2020 2020 2020 2020 2073 656c  .            sel
-00006c90: 662e 6164 6174 612e 7661 726d 5b72 6573  f.adata.varm[res
-00006ca0: 756c 745f 7479 7065 5d20 3d20 6d65 7461  ult_type] = meta
-00006cb0: 5f77 0d0a 0d0a 2020 2020 2020 2020 2320  _w....        # 
-00006cc0: 496d 706f 7274 2055 7361 6765 730d 0a20  Import Usages.. 
-00006cd0: 2020 2020 2020 206c 6f67 6769 6e67 2e69         logging.i
-00006ce0: 6e66 6f28 6622 496d 706f 7274 696e 6720  nfo(f"Importing 
-00006cf0: 5573 6167 6573 2229 2020 0d0a 2020 2020  Usages")  ..    
-00006d00: 2020 2020 7573 6167 6520 3d20 5b5d 0d0a      usage = []..
-00006d10: 2020 2020 2020 2020 666f 7220 666e 2069          for fn i
-00006d20: 6e20 676c 6f62 286f 732e 7061 7468 2e6a  n glob(os.path.j
-00006d30: 6f69 6e28 636e 6d66 5f6f 7574 7075 745f  oin(cnmf_output_
-00006d40: 6469 722c 2063 6e6d 665f 6e61 6d65 2c20  dir, cnmf_name, 
-00006d50: 6622 7b63 6e6d 665f 6e61 6d65 7d2a 2e75  f"{cnmf_name}*.u
-00006d60: 7361 6765 732e 6b5f 2a2e 7b6c 6474 5f73  sages.k_*.{ldt_s
-00006d70: 7472 7d2e 2a74 7874 2229 293a 0d0a 2020  tr}.*txt")):..  
-00006d80: 2020 2020 2020 2020 2020 6b20 3d20 696e            k = in
-00006d90: 7428 6f73 2e70 6174 682e 6261 7365 6e61  t(os.path.basena
-00006da0: 6d65 2866 6e29 2e72 656d 6f76 6570 7265  me(fn).removepre
-00006db0: 6669 7828 6622 7b63 6e6d 665f 6e61 6d65  fix(f"{cnmf_name
-00006dc0: 7d2e 7573 6167 6573 2e22 292e 7370 6c69  }.usages.").spli
-00006dd0: 7428 222e 2229 5b30 5d2e 7265 706c 6163  t(".")[0].replac
-00006de0: 6528 226b 5f22 2c20 2222 2929 0d0a 2020  e("k_", ""))..  
-00006df0: 2020 2020 2020 2020 2020 6820 3d20 7064            h = pd
-00006e00: 2e72 6561 645f 7461 626c 6528 666e 2c20  .read_table(fn, 
-00006e10: 696e 6465 785f 636f 6c3d 3029 0d0a 2020  index_col=0)..  
-00006e20: 2020 2020 2020 2020 2020 682e 636f 6c75            h.colu
-00006e30: 6d6e 7320 3d20 7374 7228 6b29 202b 2022  mns = str(k) + "
-00006e40: 2e22 202b 2068 2e63 6f6c 756d 6e73 2e61  ." + h.columns.a
-00006e50: 7374 7970 6528 7374 7229 0d0a 2020 2020  stype(str)..    
-00006e60: 2020 2020 2020 2020 7573 6167 652e 6170          usage.ap
-00006e70: 7065 6e64 2868 290d 0a20 2020 2020 2020  pend(h)..       
-00006e80: 2073 656c 662e 6164 6174 612e 6f62 736d   self.adata.obsm
-00006e90: 5b22 636e 6d66 5f75 7361 6765 225d 203d  ["cnmf_usage"] =
-00006ea0: 2070 642e 636f 6e63 6174 2875 7361 6765   pd.concat(usage
-00006eb0: 2c20 6178 6973 3d31 292e 736f 7274 5f69  , axis=1).sort_i
-00006ec0: 6e64 6578 2861 7869 733d 3129 2e72 656e  ndex(axis=1).ren
-00006ed0: 616d 655f 6178 6973 285b 226b 2e67 6570  ame_axis(["k.gep
-00006ee0: 225d 2c20 6178 6973 3d31 290d 0a20 2020  "], axis=1)..   
-00006ef0: 2020 2020 200d 0a20 2020 2020 2020 2023       ..        #
-00006f00: 2049 6d70 6f72 7420 6765 6e65 206c 6973   Import gene lis
-00006f10: 7420 7573 6564 2066 6f72 2066 6163 746f  t used for facto
-00006f20: 7269 7a61 7469 6f6e 0d0a 2020 2020 2020  rization..      
-00006f30: 2020 7769 7468 206f 7065 6e28 6f73 2e70    with open(os.p
-00006f40: 6174 682e 6a6f 696e 2863 6e6d 665f 6f75  ath.join(cnmf_ou
-00006f50: 7470 7574 5f64 6972 2c20 636e 6d66 5f6e  tput_dir, cnmf_n
-00006f60: 616d 652c 2066 227b 636e 6d66 5f6e 616d  ame, f"{cnmf_nam
-00006f70: 657d 2e6f 7665 7264 6973 7065 7273 6564  e}.overdispersed
-00006f80: 5f67 656e 6573 2e74 7874 2229 2920 6173  _genes.txt")) as
-00006f90: 2066 3a0d 0a20 2020 2020 2020 2020 2020   f:..           
-00006fa0: 2073 656c 662e 6164 6174 612e 756e 735b   self.adata.uns[
-00006fb0: 2267 656e 655f 6c69 7374 225d 203d 205b  "gene_list"] = [
-00006fc0: 6c69 6e65 2e73 7472 6970 2829 2066 6f72  line.strip() for
-00006fd0: 206c 696e 6520 696e 2066 2e72 6561 646c   line in f.readl
-00006fe0: 696e 6573 2829 5d0d 0a0d 0a20 2020 2020  ines()]....     
-00006ff0: 2020 2023 2049 6d70 6f72 7420 4b2d 7365     # Import K-se
-00007000: 6c65 6374 696f 6e20 7374 6174 730d 0a20  lection stats.. 
-00007010: 2020 2020 2020 206b 7661 6c73 203d 2070         kvals = p
-00007020: 642e 4461 7461 4672 616d 6528 2a2a 6e70  d.DataFrame(**np
-00007030: 2e6c 6f61 6428 6f73 2e70 6174 682e 6a6f  .load(os.path.jo
-00007040: 696e 2863 6e6d 665f 6f75 7470 7574 5f64  in(cnmf_output_d
-00007050: 6972 2c20 636e 6d66 5f6e 616d 652c 2066  ir, cnmf_name, f
-00007060: 227b 636e 6d66 5f6e 616d 657d 2e6b 5f73  "{cnmf_name}.k_s
-00007070: 656c 6563 7469 6f6e 5f73 7461 7473 2e64  election_stats.d
-00007080: 662e 6e70 7a22 292c 2061 6c6c 6f77 5f70  f.npz"), allow_p
-00007090: 6963 6b6c 653d 5472 7565 2929 2e73 6574  ickle=True)).set
-000070a0: 5f69 6e64 6578 2822 6b22 295b 5b22 7374  _index("k")[["st
-000070b0: 6162 696c 6974 7922 2c20 2270 7265 6469  ability", "predi
-000070c0: 6374 696f 6e5f 6572 726f 7222 5d5d 0d0a  ction_error"]]..
-000070d0: 2020 2020 2020 2020 6b76 616c 732e 696e          kvals.in
-000070e0: 6465 7820 3d20 6b76 616c 732e 696e 6465  dex = kvals.inde
-000070f0: 782e 6173 7479 7065 2869 6e74 290d 0a20  x.astype(int).. 
-00007100: 2020 2020 2020 2073 656c 662e 6164 6174         self.adat
-00007110: 612e 756e 735b 226b 7661 6c73 225d 203d  a.uns["kvals"] =
-00007120: 206b 7661 6c73 0d0a 2020 2020 2020 2020   kvals..        
-00007130: 7365 6c66 2e61 7070 656e 645f 746f 5f68  self.append_to_h
-00007140: 6973 746f 7279 2822 634e 4d46 2072 6573  istory("cNMF res
-00007150: 756c 7473 2061 6464 6564 2066 726f 6d20  ults added from 
-00007160: 6f75 7470 7574 2064 6972 6563 746f 7279  output directory
-00007170: 207b 636e 6d66 5f6f 7574 7075 745f 6469   {cnmf_output_di
-00007180: 727d 2f7b 636e 6d66 5f6e 616d 657d 2229  r}/{cnmf_name}")
-00007190: 0d0a 0d0a 2020 2020 6465 6620 6765 745f  ....    def get_
-000071a0: 7573 6167 6573 2873 656c 662c 0d0a 2020  usages(self,..  
-000071b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000071c0: 206b 3a20 556e 696f 6e5b 696e 742c 2049   k: Union[int, I
-000071d0: 7465 7261 626c 655d 203d 204e 6f6e 652c  terable] = None,
-000071e0: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
-000071f0: 2020 2020 2064 6973 6372 6574 697a 653a       discretize:
-00007200: 2062 6f6f 6c20 3d20 4661 6c73 652c 0d0a   bool = False,..
-00007210: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007220: 2020 206e 6f72 6d61 6c69 7a65 3a20 626f     normalize: bo
-00007230: 6f6c 203d 2046 616c 7365 0d0a 2020 2020  ol = False..    
-00007240: 2020 2020 2020 2020 2020 2020 2020 2029                 )
-00007250: 202d 3e20 7064 2e44 6174 6146 7261 6d65   -> pd.DataFrame
-00007260: 3a0d 0a20 2020 2020 2020 2022 2222 0d0a  :..        """..
-00007270: 2020 2020 2020 2020 4361 6c63 756c 6174          Calculat
-00007280: 6520 7573 6167 6520 6f66 2065 6163 6820  e usage of each 
-00007290: 4745 5020 696e 2065 6163 6820 7361 6d70  GEP in each samp
-000072a0: 6c65 2f6f 6273 6572 7661 7469 6f6e 2e0d  le/observation..
-000072b0: 0a0d 0a20 2020 2020 2020 203a 7061 7261  ...        :para
-000072c0: 6d20 6b3a 2049 6620 616e 2069 6e74 6567  m k: If an integ
-000072d0: 6572 206f 7220 6c69 7374 206f 6620 696e  er or list of in
-000072e0: 7465 6765 7273 2c20 7265 7475 726e 7320  tegers, returns 
-000072f0: 7573 6167 6573 206f 6e6c 7920 666f 7220  usages only for 
-00007300: 7370 6563 6966 6965 6420 7261 6e6b 732e  specified ranks.
-00007310: 204f 7468 6572 7769 7365 2c20 7265 7475   Otherwise, retu
-00007320: 726e 7320 7573 6167 6520 6f66 2061 6c6c  rns usage of all
-00007330: 2047 4550 7320 6163 726f 7373 2072 616e   GEPs across ran
-00007340: 6b73 2e20 4465 6661 756c 7473 2074 6f20  ks. Defaults to 
-00007350: 4e6f 6e65 0d0a 2020 2020 2020 2020 3a74  None..        :t
-00007360: 7970 6520 6b3a 2069 6e74 2c20 6f70 7469  ype k: int, opti
-00007370: 6f6e 616c 0d0a 2020 2020 2020 2020 3a70  onal..        :p
-00007380: 6172 616d 2064 6973 6372 6574 697a 653a  aram discretize:
-00007390: 2044 6973 6372 6574 697a 6573 2074 6865   Discretizes the
-000073a0: 2075 7361 6765 206d 6174 7269 7820 7375   usage matrix su
-000073b0: 6368 2074 6861 7420 666f 7220 6561 6368  ch that for each
-000073c0: 2076 616c 7565 206f 6620 6b2c 2065 6163   value of k, eac
-000073d0: 6820 7361 6d70 6c65 2068 6173 2075 7361  h sample has usa
-000073e0: 6765 206f 6620 6f6e 6c79 2031 2047 4550  ge of only 1 GEP
-000073f0: 2028 7468 6520 6f6e 6520 7769 7468 2074   (the one with t
-00007400: 6865 206d 6178 696d 756d 2075 7361 6765  he maximum usage
-00007410: 292e 2044 6566 6175 6c74 7320 746f 2046  ). Defaults to F
-00007420: 616c 7365 0d0a 2020 2020 2020 2020 3a74  alse..        :t
-00007430: 7970 6520 6469 7363 7265 7469 7a65 3a20  ype discretize: 
-00007440: 626f 6f6c 2c20 6f70 7469 6f6e 616c 0d0a  bool, optional..
-00007450: 2020 2020 2020 2020 3a70 6172 616d 206e          :param n
-00007460: 6f72 6d61 6c69 7a65 3a20 4e6f 726d 616c  ormalize: Normal
-00007470: 697a 6520 7468 6520 4745 5020 7573 6167  ize the GEP usag
-00007480: 6520 6d61 7472 6978 2073 7563 6820 7468  e matrix such th
-00007490: 6174 2066 6f72 2065 6163 6820 7661 6c75  at for each valu
-000074a0: 6520 6f66 206b 2c20 7573 6167 6520 6f66  e of k, usage of
-000074b0: 2061 6c6c 2047 4550 7320 7375 6d73 2074   all GEPs sums t
-000074c0: 6f20 312e 2044 6566 6175 6c74 7320 746f  o 1. Defaults to
-000074d0: 2046 616c 7365 0d0a 2020 2020 2020 2020   False..        
-000074e0: 3a74 7970 6520 6e6f 726d 616c 697a 653a  :type normalize:
-000074f0: 2062 6f6f 6c2c 206f 7074 696f 6e61 6c0d   bool, optional.
-00007500: 0a20 2020 2020 2020 203a 7265 7475 726e  .        :return
-00007510: 3a20 6f62 7365 7276 6174 696f 6e20 c397  : observation ..
-00007520: 2047 4550 206d 6174 7269 780d 0a20 2020   GEP matrix..   
-00007530: 2020 2020 203a 7274 7970 653a 2070 642e       :rtype: pd.
-00007540: 4461 7461 4672 616d 650d 0a20 2020 2020  DataFrame..     
-00007550: 2020 2022 2222 0d0a 2020 2020 2020 2020     """..        
-00007560: 6466 203d 2073 656c 662e 6164 6174 612e  df = self.adata.
-00007570: 6f62 736d 5b22 636e 6d66 5f75 7361 6765  obsm["cnmf_usage
-00007580: 225d 2e63 6f70 7928 290d 0a20 2020 2020  "].copy()..     
-00007590: 2020 2064 662e 636f 6c75 6d6e 7320 3d20     df.columns = 
-000075a0: 7064 2e4d 756c 7469 496e 6465 782e 6672  pd.MultiIndex.fr
-000075b0: 6f6d 5f74 7570 6c65 7328 6466 2e63 6f6c  om_tuples(df.col
-000075c0: 756d 6e73 2e73 7472 2e73 706c 6974 2822  umns.str.split("
-000075d0: 2e22 292e 746f 5f6c 6973 7428 2929 0d0a  .").to_list())..
-000075e0: 2020 2020 2020 2020 6466 2e63 6f6c 756d          df.colum
-000075f0: 6e73 203d 2064 662e 636f 6c75 6d6e 732e  ns = df.columns.
-00007600: 7365 745f 6c65 7665 6c73 285b 6c2e 6173  set_levels([l.as
-00007610: 7479 7065 2822 696e 7422 2920 666f 7220  type("int") for 
-00007620: 6c20 696e 2064 662e 636f 6c75 6d6e 732e  l in df.columns.
-00007630: 6c65 7665 6c73 5d29 0d0a 2020 2020 2020  levels])..      
-00007640: 2020 6966 206e 6f72 6d61 6c69 7a65 3a0d    if normalize:.
-00007650: 0a20 2020 2020 2020 2020 2020 206e 6f72  .            nor
-00007660: 6d61 6c69 7a65 6420 3d20 5b5d 0d0a 2020  malized = []..  
-00007670: 2020 2020 2020 2020 2020 666f 7220 5f2c            for _,
-00007680: 2073 7562 6466 2069 6e20 6466 2e67 726f   subdf in df.gro
-00007690: 7570 6279 2861 7869 733d 312c 206c 6576  upby(axis=1, lev
-000076a0: 656c 3d30 293a 0d0a 2020 2020 2020 2020  el=0):..        
-000076b0: 2020 2020 2020 2020 6e6f 726d 616c 697a          normaliz
-000076c0: 6564 2e61 7070 656e 6428 7375 6264 662e  ed.append(subdf.
-000076d0: 6469 7628 7375 6264 662e 7375 6d28 6178  div(subdf.sum(ax
-000076e0: 6973 3d31 292c 2061 7869 733d 3029 290d  is=1), axis=0)).
-000076f0: 0a20 2020 2020 2020 2020 2020 2064 6620  .            df 
-00007700: 3d20 7064 2e63 6f6e 6361 7428 6e6f 726d  = pd.concat(norm
-00007710: 616c 697a 6564 2c20 6178 6973 3d31 290d  alized, axis=1).
-00007720: 0a20 2020 2020 2020 2069 6620 6469 7363  .        if disc
-00007730: 7265 7469 7a65 3a0d 0a20 2020 2020 2020  retize:..       
-00007740: 2020 2020 2064 6973 6372 6574 697a 6564       discretized
-00007750: 203d 205b 5d0d 0a20 2020 2020 2020 2020   = []..         
-00007760: 2020 2066 6f72 205f 2c20 7375 6264 6620     for _, subdf 
-00007770: 696e 2064 662e 6772 6f75 7062 7928 6178  in df.groupby(ax
-00007780: 6973 3d31 2c20 6c65 7665 6c3d 3029 3a0d  is=1, level=0):.
-00007790: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-000077a0: 2064 6973 6372 6574 697a 6564 2e61 7070   discretized.app
-000077b0: 656e 6428 7375 6264 662e 6571 2873 7562  end(subdf.eq(sub
-000077c0: 6466 2e6d 6178 2861 7869 733d 3129 2c20  df.max(axis=1), 
-000077d0: 6178 6973 3d30 292e 6173 7479 7065 2869  axis=0).astype(i
-000077e0: 6e74 2929 0d0a 2020 2020 2020 2020 2020  nt))..          
-000077f0: 2020 6466 203d 2070 642e 636f 6e63 6174    df = pd.concat
-00007800: 2864 6973 6372 6574 697a 6564 2c20 6178  (discretized, ax
-00007810: 6973 3d31 2920 2020 2020 2020 200d 0a20  is=1)        .. 
-00007820: 2020 2020 2020 2069 6620 6b20 6973 206e         if k is n
-00007830: 6f74 204e 6f6e 653a 0d0a 2020 2020 2020  ot None:..      
-00007840: 2020 2020 2020 6466 203d 2064 662e 6c6f        df = df.lo
-00007850: 635b 3a2c 206b 5d0d 0a20 2020 2020 2020  c[:, k]..       
-00007860: 2064 6620 3d20 6466 2e73 6f72 745f 696e   df = df.sort_in
-00007870: 6465 7828 6178 6973 3d30 292e 736f 7274  dex(axis=0).sort
-00007880: 5f69 6e64 6578 2861 7869 733d 3129 2020  _index(axis=1)  
-00007890: 200d 0a20 2020 2020 2020 2072 6574 7572   ..        retur
-000078a0: 6e20 6466 0d0a 2020 2020 0d0a 2020 2020  n df..    ..    
-000078b0: 6465 6620 6765 745f 6765 7073 2873 656c  def get_geps(sel
-000078c0: 662c 0d0a 2020 2020 2020 2020 2020 2020  f,..            
-000078d0: 2020 2020 206b 3a20 556e 696f 6e5b 696e       k: Union[in
-000078e0: 742c 2049 7465 7261 626c 655d 203d 204e  t, Iterable] = N
-000078f0: 6f6e 652c 0d0a 2020 2020 2020 2020 2020  one,..          
-00007900: 2020 2020 2020 2074 7970 653d 2263 6e6d         type="cnm
-00007910: 665f 6765 705f 7363 6f72 6522 0d0a 2020  f_gep_score"..  
-00007920: 2020 2020 2020 2020 2020 2020 2020 2029                 )
-00007930: 202d 3e20 7064 2e44 6174 6146 7261 6d65   -> pd.DataFrame
-00007940: 3a0d 0a20 2020 2020 2020 2022 2222 0d0a  :..        """..
-00007950: 2020 2020 2020 2020 4765 7420 4745 5073          Get GEPs
-00007960: 2e0d 0a0d 0a20 2020 2020 2020 203a 7061  .....        :pa
-00007970: 7261 6d20 6b3a 2049 6620 616e 2069 6e74  ram k: If an int
-00007980: 6567 6572 206f 7220 6c69 7374 206f 6620  eger or list of 
-00007990: 696e 7465 6765 7273 2c20 7265 7475 726e  integers, return
-000079a0: 7320 4745 5073 206f 6e6c 7920 666f 7220  s GEPs only for 
-000079b0: 7370 6563 6966 6965 6420 7261 6e6b 732e  specified ranks.
-000079c0: 204f 7468 6572 7769 7365 2c20 7265 7475   Otherwise, retu
-000079d0: 726e 7320 4745 5073 2066 726f 6d20 616c  rns GEPs from al
-000079e0: 6c20 7261 6e6b 732e 2044 6566 6175 6c74  l ranks. Default
-000079f0: 7320 746f 204e 6f6e 650d 0a20 2020 2020  s to None..     
-00007a00: 2020 203a 7479 7065 206b 3a20 556e 696f     :type k: Unio
-00007a10: 6e5b 696e 742c 2049 7465 7261 626c 655d  n[int, Iterable]
-00007a20: 2c20 6f70 7469 6f6e 616c 0d0a 2020 2020  , optional..    
-00007a30: 2020 2020 3a70 6172 616d 2074 7970 653a      :param type:
-00007a40: 2022 636e 6d66 5f67 6570 5f73 636f 7265   "cnmf_gep_score
-00007a50: 2220 6f72 2022 636e 6d66 5f67 6570 5f74  " or "cnmf_gep_t
-00007a60: 706d 222c 2064 6566 6175 6c74 7320 746f  pm", defaults to
-00007a70: 2022 636e 6d66 5f67 6570 5f73 636f 7265   "cnmf_gep_score
-00007a80: 220d 0a20 2020 2020 2020 203a 7479 7065  "..        :type
-00007a90: 2074 7970 653a 2073 7472 2c20 6f70 7469   type: str, opti
-00007aa0: 6f6e 616c 0d0a 2020 2020 2020 2020 3a72  onal..        :r
-00007ab0: 6574 7572 6e3a 2066 6561 7475 7265 7320  eturn: features 
-00007ac0: c397 2047 4550 206d 6174 7269 780d 0a20  .. GEP matrix.. 
-00007ad0: 2020 2020 2020 203a 7274 7970 653a 2070         :rtype: p
-00007ae0: 642e 4461 7461 4672 616d 650d 0a20 2020  d.DataFrame..   
-00007af0: 2020 2020 2022 2222 0d0a 2020 2020 2020       """..      
-00007b00: 2020 6466 203d 2073 656c 662e 6164 6174    df = self.adat
-00007b10: 612e 7661 726d 5b74 7970 655d 2e63 6f70  a.varm[type].cop
-00007b20: 7928 290d 0a20 2020 2020 2020 2064 662e  y()..        df.
-00007b30: 636f 6c75 6d6e 7320 3d20 7064 2e4d 756c  columns = pd.Mul
-00007b40: 7469 496e 6465 782e 6672 6f6d 5f74 7570  tiIndex.from_tup
-00007b50: 6c65 7328 6466 2e63 6f6c 756d 6e73 2e73  les(df.columns.s
-00007b60: 7472 2e73 706c 6974 2822 2e22 292e 746f  tr.split(".").to
-00007b70: 5f6c 6973 7428 2929 0d0a 2020 2020 2020  _list())..      
-00007b80: 2020 6466 2e63 6f6c 756d 6e73 203d 2064    df.columns = d
-00007b90: 662e 636f 6c75 6d6e 732e 7365 745f 6c65  f.columns.set_le
-00007ba0: 7665 6c73 285b 6c2e 6173 7479 7065 2822  vels([l.astype("
-00007bb0: 696e 7422 2920 666f 7220 6c20 696e 2064  int") for l in d
-00007bc0: 662e 636f 6c75 6d6e 732e 6c65 7665 6c73  f.columns.levels
-00007bd0: 5d29 0d0a 2020 2020 2020 2020 6966 2069  ])..        if i
-00007be0: 7369 6e73 7461 6e63 6528 6b2c 2028 696e  sinstance(k, (in
-00007bf0: 742c 2049 7465 7261 626c 6529 293a 0d0a  t, Iterable)):..
-00007c00: 2020 2020 2020 2020 2020 2020 6466 203d              df =
-00007c10: 2064 662e 6c6f 635b 3a2c 206b 5d0d 0a20   df.loc[:, k].. 
-00007c20: 2020 2020 2020 2064 6620 3d20 6466 2e73         df = df.s
-00007c30: 6f72 745f 696e 6465 7828 6178 6973 3d31  ort_index(axis=1
-00007c40: 290d 0a20 2020 2020 2020 2072 6574 7572  )..        retur
-00007c50: 6e20 6466 0d0a 2020 2020 0d0a 2020 2020  n df..    ..    
-00007c60: 6465 6620 6765 745f 6d65 7461 6461 7461  def get_metadata
-00007c70: 5f64 6628 7365 6c66 2c0d 0a20 2020 2020  _df(self,..     
-00007c80: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007c90: 2020 2069 6e63 6c75 6465 5f63 6174 6567     include_categ
-00007ca0: 6f72 6963 616c 3a20 626f 6f6c 203d 2054  orical: bool = T
-00007cb0: 7275 652c 0d0a 2020 2020 2020 2020 2020  rue,..          
-00007cc0: 2020 2020 2020 2020 2020 2020 2020 696e                in
-00007cd0: 636c 7564 655f 6e75 6d65 7269 6361 6c3a  clude_numerical:
-00007ce0: 2062 6f6f 6c20 3d20 5472 7565 0d0a 2020   bool = True..  
-00007cf0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00007d00: 2020 2020 2020 2920 2d3e 2070 642e 4461        ) -> pd.Da
-00007d10: 7461 4672 616d 653a 0d0a 2020 2020 2020  taFrame:..      
-00007d20: 2020 2222 2247 6574 2073 616d 706c 652f    """Get sample/
-00007d30: 6f62 7365 7276 6174 696f 6e20 6d65 7461  observation meta
-00007d40: 6461 7461 2e0d 0a0d 0a20 2020 2020 2020  data.....       
-00007d50: 203a 7061 7261 6d20 696e 636c 7564 655f   :param include_
-00007d60: 6361 7465 676f 7269 6361 6c3a 2049 6e63  categorical: Inc
-00007d70: 6c75 6465 2063 6174 6567 6f72 6963 616c  lude categorical
-00007d80: 206d 6574 6164 6174 6120 6c61 7965 7273   metadata layers
-00007d90: 2c20 6465 6661 756c 7473 2074 6f20 5472  , defaults to Tr
-00007da0: 7565 0d0a 2020 2020 2020 2020 3a74 7970  ue..        :typ
-00007db0: 6520 696e 636c 7564 655f 6361 7465 676f  e include_catego
-00007dc0: 7269 6361 6c3a 2062 6f6f 6c2c 206f 7074  rical: bool, opt
-00007dd0: 696f 6e61 6c0d 0a20 2020 2020 2020 203a  ional..        :
-00007de0: 7061 7261 6d20 696e 636c 7564 655f 6e75  param include_nu
-00007df0: 6d65 7269 6361 6c3a 2049 6e63 6c75 6465  merical: Include
-00007e00: 206e 756d 6572 6963 616c 206d 6574 6164   numerical metad
-00007e10: 6174 6120 6c61 7965 7273 2c20 6465 6661  ata layers, defa
-00007e20: 756c 7473 2074 6f20 5472 7565 0d0a 2020  ults to True..  
-00007e30: 2020 2020 2020 3a74 7970 6520 696e 636c        :type incl
-00007e40: 7564 655f 6e75 6d65 7269 6361 6c3a 2062  ude_numerical: b
-00007e50: 6f6f 6c2c 206f 7074 696f 6e61 6c0d 0a20  ool, optional.. 
-00007e60: 2020 2020 2020 203a 7261 6973 6573 2056         :raises V
-00007e70: 616c 7565 4572 726f 723a 2045 7272 6f72  alueError: Error
-00007e80: 2069 6620 6d65 7461 6461 7461 2074 7970   if metadata typ
-00007e90: 6573 2061 7265 206e 6f74 2072 6563 6f67  es are not recog
-00007ea0: 6e69 7a65 640d 0a20 2020 2020 2020 203a  nized..        :
-00007eb0: 7265 7475 726e 3a20 6f62 7365 7276 6174  return: observat
-00007ec0: 696f 6e73 20c3 9720 6d65 7461 6461 7461  ions .. metadata
-00007ed0: 206d 6174 7269 780d 0a20 2020 2020 2020   matrix..       
-00007ee0: 203a 7274 7970 653a 2070 642e 4461 7461   :rtype: pd.Data
-00007ef0: 4672 616d 650d 0a20 2020 2020 2020 2022  Frame..        "
-00007f00: 2222 0d0a 2020 2020 2020 2020 6474 7970  ""..        dtyp
-00007f10: 6573 203d 205b 5d0d 0a20 2020 2020 2020  es = []..       
-00007f20: 2069 6620 696e 636c 7564 655f 6361 7465   if include_cate
-00007f30: 676f 7269 6361 6c3a 0d0a 2020 2020 2020  gorical:..      
-00007f40: 2020 2020 2020 6474 7970 6573 2e61 7070        dtypes.app
-00007f50: 656e 6428 2263 6174 6567 6f72 7922 290d  end("category").
-00007f60: 0a20 2020 2020 2020 2069 6620 696e 636c  .        if incl
-00007f70: 7564 655f 6e75 6d65 7269 6361 6c3a 0d0a  ude_numerical:..
-00007f80: 2020 2020 2020 2020 2020 2020 6474 7970              dtyp
-00007f90: 6573 202b 3d20 5b22 666c 6f61 7422 2c20  es += ["float", 
-00007fa0: 2269 6e74 225d 0d0a 2020 2020 2020 2020  "int"]..        
-00007fb0: 756e 6578 706c 6169 6e65 645f 636f 6c73  unexplained_cols
-00007fc0: 203d 2073 656c 662e 6164 6174 612e 6f62   = self.adata.ob
-00007fd0: 732e 7365 6c65 6374 5f64 7479 7065 7328  s.select_dtypes(
-00007fe0: 6578 636c 7564 653d 2822 6361 7465 676f  exclude=("catego
-00007ff0: 7279 222c 2022 666c 6f61 7422 2c20 2269  ry", "float", "i
-00008000: 6e74 2229 292e 636f 6c75 6d6e 730d 0a20  nt")).columns.. 
-00008010: 2020 2020 2020 2069 6620 6c65 6e28 756e         if len(un
-00008020: 6578 706c 6169 6e65 645f 636f 6c73 2920  explained_cols) 
-00008030: 3e20 303a 0d0a 2020 2020 2020 2020 2020  > 0:..          
-00008040: 2020 756e 6578 706c 6169 6e65 645f 636f    unexplained_co
-00008050: 6c5f 7374 7220 3d20 222c 2022 2e6a 6f69  l_str = ", ".joi
-00008060: 6e28 756e 6578 706c 6169 6e65 645f 636f  n(unexplained_co
-00008070: 6c73 290d 0a20 2020 2020 2020 2020 2020  ls)..           
-00008080: 2072 6169 7365 2056 616c 7565 4572 726f   raise ValueErro
-00008090: 7228 6622 7b75 6e65 7870 6c61 696e 6564  r(f"{unexplained
-000080a0: 5f63 6f6c 5f73 7472 7d20 6d65 7461 6461  _col_str} metada
-000080b0: 7461 2063 6f6c 756d 6e73 2068 6176 6520  ta columns have 
-000080c0: 756e 7265 636f 676e 697a 6564 2064 7479  unrecognized dty
-000080d0: 7065 732e 2229 0d0a 2020 2020 2020 2020  pes.")..        
-000080e0: 6466 203d 2073 656c 662e 6164 6174 612e  df = self.adata.
-000080f0: 6f62 732e 7365 6c65 6374 5f64 7479 7065  obs.select_dtype
-00008100: 7328 696e 636c 7564 653d 6474 7970 6573  s(include=dtypes
-00008110: 290d 0a20 2020 2020 2020 2064 6620 3d20  )..        df = 
-00008120: 6466 2e64 726f 706e 6128 6178 6973 3d31  df.dropna(axis=1
-00008130: 2c20 686f 773d 2261 6c6c 2229 0d0a 2020  , how="all")..  
-00008140: 2020 2020 2020 7265 7475 726e 2064 660d        return df.
-00008150: 0a20 2020 200d 0a20 2020 2064 6566 2067  .    ..    def g
-00008160: 6574 5f63 6174 6567 6f72 795f 6f76 6572  et_category_over
-00008170: 7265 7072 6573 656e 7461 7469 6f6e 2873  representation(s
-00008180: 656c 662c 0d0a 2020 2020 2020 2020 2020  elf,..          
-00008190: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000081a0: 2020 2020 2020 2020 2020 2020 2020 6c61                la
-000081b0: 7965 723a 2073 7472 2c0d 0a20 2020 2020  yer: str,..     
+000001f0: 2020 2020 2020 2066 6f72 6365 5f6d 6967         force_mig
+00000200: 7261 7465 3a20 626f 6f6c 203d 2046 616c  rate: bool = Fal
+00000210: 7365 0d0a 2020 2020 2020 2020 2020 2020  se..            
+00000220: 2020 2020 2029 3a0d 0a20 2020 2020 2020       ):..       
+00000230: 2022 2222 4372 6561 7465 7320 6120 3a63   """Creates a :c
+00000240: 6c61 7373 3a60 7e63 6e6d 6673 6e73 2e64  lass:`~cnmfsns.d
+00000250: 6174 6173 6574 2e44 6174 6173 6574 6020  ataset.Dataset` 
+00000260: 6f62 6a65 6374 2066 726f 6d20 616e 2060  object from an `
+00000270: 6164 2e41 6e6e 4461 7461 6020 6f62 6a65  ad.AnnData` obje
+00000280: 6374 2e0d 0a0d 0a20 2020 2020 2020 203a  ct.....        :
+00000290: 7061 7261 6d20 6164 6174 613a 2041 6e6e  param adata: Ann
+000002a0: 4461 7461 206f 626a 6563 7420 7769 7468  Data object with
+000002b0: 2064 6174 610d 0a20 2020 2020 2020 203a   data..        :
+000002c0: 7479 7065 2061 6461 7461 3a20 6164 2e41  type adata: ad.A
+000002d0: 6e6e 4461 7461 0d0a 2020 2020 2020 2020  nnData..        
+000002e0: 3a70 6172 616d 2066 6f72 6365 5f6d 6967  :param force_mig
+000002f0: 7261 7465 3a20 666f 7263 6573 2063 6f6e  rate: forces con
+00000300: 7665 7273 696f 6e20 6f66 2041 6e6e 4461  version of AnnDa
+00000310: 7461 206f 626a 6563 7473 2065 7665 6e20  ta objects even 
+00000320: 7768 656e 2061 6461 7461 2e58 2061 6e64  when adata.X and
+00000330: 2061 6461 7461 2e72 6177 2e58 2061 7265   adata.raw.X are
+00000340: 206e 6f74 206c 696e 6561 726c 7920 7363   not linearly sc
+00000350: 616c 6564 2072 656c 6174 6976 6520 746f  aled relative to
+00000360: 2065 6163 6820 6f74 6865 722c 2064 6566   each other, def
+00000370: 6175 6c74 7320 746f 2046 616c 7365 0d0a  aults to False..
+00000380: 2020 2020 2020 2020 3a74 7970 6520 666f          :type fo
+00000390: 7263 655f 6d69 6772 6174 653a 2062 6f6f  rce_migrate: boo
+000003a0: 6c2c 206f 7074 696f 6e61 6c0d 0a20 2020  l, optional..   
+000003b0: 2020 2020 203a 7261 6973 6573 2052 756e       :raises Run
+000003c0: 7469 6d65 4572 726f 723a 2042 6163 6b65  timeError: Backe
+000003d0: 642d 6d6f 6465 2041 6e6e 6461 7461 206f  d-mode Anndata o
+000003e0: 626a 6563 7473 2063 616e 6e6f 7420 6265  bjects cannot be
+000003f0: 206d 6967 7261 7465 640d 0a20 2020 2020   migrated..     
+00000400: 2020 203a 7261 6973 6573 2056 616c 7565     :raises Value
+00000410: 4572 726f 723a 2045 7272 6f72 2069 7320  Error: Error is 
+00000420: 7261 6973 6564 2077 6865 6e20 666f 7263  raised when forc
+00000430: 6520 6973 2046 616c 7365 2061 6e64 2061  e is False and a
+00000440: 6461 7461 2069 7320 6e6f 6e2d 6c69 6e65  data is non-line
+00000450: 6172 6c79 2073 6361 6c65 642e 0d0a 2020  arly scaled...  
+00000460: 2020 2020 2020 3a72 6574 7572 6e3a 204f        :return: O
+00000470: 626a 6563 7420 7769 7468 2065 7870 7265  bject with expre
+00000480: 7373 696f 6e20 616e 6420 6d65 7461 6461  ssion and metada
+00000490: 7461 0d0a 2020 2020 2020 2020 3a72 7479  ta..        :rty
+000004a0: 7065 3a20 3a63 6c61 7373 3a60 7e63 6e6d  pe: :class:`~cnm
+000004b0: 6673 6e73 2e64 6174 6173 6574 2e44 6174  fsns.dataset.Dat
+000004c0: 6173 6574 600d 0a20 2020 2020 2020 2022  aset`..        "
+000004d0: 2222 0d0a 0d0a 2020 2020 2020 2020 6966  ""....        if
+000004e0: 2061 6461 7461 2e58 2069 7320 4e6f 6e65   adata.X is None
+000004f0: 3a0d 0a20 2020 2020 2020 2020 2020 206c  :..            l
+00000500: 6f67 6769 6e67 2e65 7272 6f72 2866 2261  ogging.error(f"a
+00000510: 6461 7461 2063 6f6e 7461 696e 7320 6e6f  data contains no
+00000520: 2065 7870 7265 7373 696f 6e20 6461 7461   expression data
+00000530: 2028 6164 6174 612e 5829 2229 0d0a 2020   (adata.X)")..  
+00000540: 2020 2020 2020 2020 2020 7261 6973 6520            raise 
+00000550: 5661 6c75 6545 7272 6f72 2829 0d0a 2020  ValueError()..  
+00000560: 2020 2020 2020 0d0a 2020 2020 2020 2020        ..        
+00000570: 0d0a 2020 2020 2020 2020 6966 2022 636e  ..        if "cn
+00000580: 6d66 736e 735f 7665 7273 696f 6e22 2069  mfsns_version" i
+00000590: 6e20 6164 6174 612e 756e 7320 616e 6420  n adata.uns and 
+000005a0: 6164 6174 612e 756e 735b 2263 6e6d 6673  adata.uns["cnmfs
+000005b0: 6e73 5f76 6572 7369 6f6e 225d 2069 7320  ns_version"] is 
+000005c0: 6e6f 7420 4e6f 6e65 3a0d 0a20 2020 2020  not None:..     
+000005d0: 2020 2020 2020 2073 656c 662e 6164 6174         self.adat
+000005e0: 6120 3d20 6164 6174 610d 0a20 2020 2020  a = adata..     
+000005f0: 2020 2065 6c73 653a 0d0a 2020 2020 2020     else:..      
+00000600: 2020 2020 2020 2320 6368 6563 6b20 616e        # check an
+00000610: 6420 7570 6461 7465 206f 6c64 206f 7220  d update old or 
+00000620: 6578 7465 726e 616c 2068 3561 6420 6669  external h5ad fi
+00000630: 6c65 7320 666f 7220 634e 4d46 2d53 4e53  les for cNMF-SNS
+00000640: 2063 6f6d 706c 6961 6e63 650d 0a20 2020   compliance..   
+00000650: 200d 0a20 2020 2020 2020 2020 2020 2058   ..            X
+00000660: 203d 2061 6461 7461 2e74 6f5f 6466 2829   = adata.to_df()
+00000670: 0d0a 2020 2020 2020 2020 2020 2020 0d0a  ..            ..
+00000680: 2020 2020 2020 2020 2020 2020 6966 2061              if a
+00000690: 6461 7461 2e69 7362 6163 6b65 643a 0d0a  data.isbacked:..
+000006a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000006b0: 7261 6973 6520 5275 6e74 696d 6545 7272  raise RuntimeErr
+000006c0: 6f72 2822 6164 6174 6120 6973 2061 2062  or("adata is a b
+000006d0: 6163 6b65 6420 416e 6e44 6174 6120 6f62  acked AnnData ob
+000006e0: 6a65 6374 2e20 416e 6e44 6174 6120 6f62  ject. AnnData ob
+000006f0: 6a65 6374 7320 6f70 656e 6564 2069 6e20  jects opened in 
+00000700: 6261 636b 6564 206d 6f64 6520 6361 6e6e  backed mode cann
+00000710: 6f74 2062 6520 6d69 6772 6174 6564 2e22  ot be migrated."
+00000720: 290d 0a0d 0a20 2020 2020 2020 2020 2020  )....           
+00000730: 2069 6620 6164 6174 612e 7261 7720 6973   if adata.raw is
+00000740: 204e 6f6e 653a 0d0a 2020 2020 2020 2020   None:..        
+00000750: 2020 2020 2020 2020 6973 5f6e 6f72 6d61          is_norma
+00000760: 6c69 7a65 643d 4661 6c73 650d 0a20 2020  lized=False..   
+00000770: 2020 2020 2020 2020 2020 2020 2072 6177               raw
+00000780: 203d 2061 6461 7461 2e74 6f5f 6466 2829   = adata.to_df()
+00000790: 0d0a 2020 2020 2020 2020 2020 2020 656c  ..            el
+000007a0: 7365 3a20 2023 2072 6563 6f6e 6369 6c65  se:  # reconcile
+000007b0: 7320 7477 6f20 6d61 7472 6963 6573 3a20  s two matrices: 
+000007c0: 5820 616e 6420 7261 772e 580d 0a20 2020  X and raw.X..   
+000007d0: 2020 2020 2020 2020 2020 2020 2072 6177               raw
+000007e0: 203d 2070 642e 4461 7461 4672 616d 6528   = pd.DataFrame(
+000007f0: 6164 6174 612e 7261 772e 582c 2069 6e64  adata.raw.X, ind
+00000800: 6578 3d58 2e69 6e64 6578 2c20 636f 6c75  ex=X.index, colu
+00000810: 6d6e 733d 582e 636f 6c75 6d6e 7329 0d0a  mns=X.columns)..
+00000820: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00000830: 636f 7272 6469 7374 203d 2058 2e63 6f72  corrdist = X.cor
+00000840: 7277 6974 6828 7261 772c 2061 7869 733d  rwith(raw, axis=
+00000850: 3129 0d0a 2020 2020 2020 2020 2020 2020  1)..            
+00000860: 2020 2020 2320 6368 6563 6b73 2074 6861      # checks tha
+00000870: 7420 616c 6c20 7361 6d70 6c65 7320 6172  t all samples ar
+00000880: 6520 7065 7266 6563 746c 7920 636f 7272  e perfectly corr
+00000890: 656c 6174 6564 2062 6574 7765 656e 2063  elated between c
+000008a0: 6f75 6e74 7320 616e 6420 6e6f 726d 616c  ounts and normal
+000008b0: 697a 6564 2064 6174 610d 0a20 2020 2020  ized data..     
+000008c0: 2020 2020 2020 2020 2020 2069 6620 2828             if ((
+000008d0: 636f 7272 6469 7374 202d 2031 292e 6162  corrdist - 1).ab
+000008e0: 7328 2920 3e20 3165 2d36 292e 616e 7928  s() > 1e-6).any(
+000008f0: 293a 0d0a 2020 2020 2020 2020 2020 2020  ):..            
+00000900: 2020 2020 2020 2020 6c6f 6767 696e 672e          logging.
+00000910: 7761 726e 696e 6728 2243 6f75 6e74 7320  warning("Counts 
+00000920: 616e 6420 6e6f 726d 616c 697a 6564 2065  and normalized e
+00000930: 7870 7265 7373 696f 6e20 6d61 7472 6963  xpression matric
+00000940: 6573 2061 7265 206e 6f74 2070 6572 6665  es are not perfe
+00000950: 6374 6c79 2063 6f72 7265 6c61 7465 642e  ctly correlated.
+00000960: 2043 6f75 6e74 7320 6461 7461 2077 696c   Counts data wil
+00000970: 6c20 6265 2072 6574 6169 6e65 6420 696e  l be retained in
+00000980: 206d 6967 7261 7465 6420 6f62 6a65 6374   migrated object
+00000990: 2e22 290d 0a20 2020 2020 2020 2020 2020  .")..           
+000009a0: 2020 2020 2020 2020 2069 6620 6e6f 7420           if not 
+000009b0: 666f 7263 655f 6d69 6772 6174 653a 0d0a  force_migrate:..
+000009c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000009d0: 2020 2020 2020 2020 6572 726f 726d 7367          errormsg
+000009e0: 203d 2022 436f 756c 6420 6e6f 7420 6d69   = "Could not mi
+000009f0: 6772 6174 6520 416e 6e44 6174 6120 6f62  grate AnnData ob
+00000a00: 6a65 6374 2e22 0d0a 2020 2020 2020 2020  ject."..        
+00000a10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00000a20: 6c6f 6767 696e 672e 6572 726f 7228 6572  logging.error(er
+00000a30: 726f 726d 7367 290d 0a20 2020 2020 2020  rormsg)..       
+00000a40: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00000a50: 2072 6169 7365 2056 616c 7565 4572 726f   raise ValueErro
+00000a60: 7228 6572 726f 726d 7367 290d 0a20 2020  r(errormsg)..   
+00000a70: 2020 2020 2020 2020 2020 2020 2023 2063               # c
+00000a80: 6865 636b 2066 6f72 2077 6865 7468 6572  heck for whether
+00000a90: 2075 7365 7220 6f72 6967 696e 616c 6c79   user originally
+00000aa0: 2069 6e70 7574 206e 6f72 6d61 6c69 7a65   input normalize
+00000ab0: 6420 6f72 2063 6f75 6e74 7320 6461 7461  d or counts data
+00000ac0: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
+00000ad0: 2020 6973 5f6e 6f72 6d61 6c69 7a65 6420    is_normalized 
+00000ae0: 3d20 2828 5820 2d20 7261 7729 2e61 6273  = ((X - raw).abs
+00000af0: 2829 203c 2031 652d 3629 2e61 6c6c 2829  () < 1e-6).all()
+00000b00: 2e61 6c6c 2829 0d0a 2020 2020 2020 2020  .all()..        
+00000b10: 2020 2020 2020 2020 585f 6973 5f74 706d          X_is_tpm
+00000b20: 203d 2028 2858 2e73 756d 2861 7869 733d   = ((X.sum(axis=
+00000b30: 3129 202d 2031 6536 292e 6162 7328 2920  1) - 1e6).abs() 
+00000b40: 3e20 3165 3229 2e61 6e79 2829 0d0a 2020  > 1e2).any()..  
+00000b50: 2020 2020 2020 2020 2020 2020 2020 6966                if
+00000b60: 2069 735f 6e6f 726d 616c 697a 6564 2061   is_normalized a
+00000b70: 6e64 206e 6f74 2058 5f69 735f 7470 6d3a  nd not X_is_tpm:
+00000b80: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
+00000b90: 2020 2020 2020 6c6f 6767 696e 672e 7761        logging.wa
+00000ba0: 726e 696e 6728 2241 6e6e 4461 7461 206f  rning("AnnData o
+00000bb0: 626a 6563 7420 636f 6e74 6169 6e73 206e  bject contains n
+00000bc0: 6f6e 2d54 504d 206e 6f72 6d61 6c69 7a65  on-TPM normalize
+00000bd0: 6420 6461 7461 2e20 4e65 7720 416e 6e44  d data. New AnnD
+00000be0: 6174 6120 6f62 6a65 6374 2077 696c 6c20  ata object will 
+00000bf0: 7265 7461 696e 2074 6865 2063 6f75 6e74  retain the count
+00000c00: 2028 756e 6e6f 726d 616c 697a 6564 2920   (unnormalized) 
+00000c10: 6461 7461 206f 6e6c 792e 2229 0d0a 2020  data only.")..  
+00000c20: 2020 2020 2020 2020 2020 2020 2020 0d0a                ..
+00000c30: 2020 2020 2020 2020 2020 2020 6966 2022              if "
+00000c40: 6f64 6722 2069 6e20 6164 6174 612e 756e  odg" in adata.un
+00000c50: 7320 616e 6420 2267 656e 655f 7374 6174  s and "gene_stat
+00000c60: 7322 2069 6e20 6164 6174 612e 756e 735b  s" in adata.uns[
+00000c70: 226f 6467 225d 3a0d 0a20 2020 2020 2020  "odg"]:..       
+00000c80: 2020 2020 2020 2020 2067 656e 655f 7374           gene_st
+00000c90: 6174 5f63 6f6c 756d 6e73 203d 2061 6461  at_columns = ada
+00000ca0: 7461 2e75 6e73 5b22 6f64 6722 5d5b 2267  ta.uns["odg"]["g
+00000cb0: 656e 655f 7374 6174 7322 5d2e 636f 6c75  ene_stats"].colu
+00000cc0: 6d6e 730d 0a20 2020 2020 2020 2020 2020  mns..           
+00000cd0: 2020 2020 2069 6620 6164 6174 612e 7661       if adata.va
+00000ce0: 722e 636f 6c75 6d6e 732e 6973 696e 2867  r.columns.isin(g
+00000cf0: 656e 655f 7374 6174 5f63 6f6c 756d 6e73  ene_stat_columns
+00000d00: 292e 616e 7928 293a 0d0a 2020 2020 2020  ).any():..      
+00000d10: 2020 2020 2020 2020 2020 2020 2020 6f76                ov
+00000d20: 6572 6c61 7070 696e 675f 636f 6c73 203d  erlapping_cols =
+00000d30: 2061 6461 7461 2e76 6172 2e63 6f6c 756d   adata.var.colum
+00000d40: 6e73 2e69 7369 6e28 6765 6e65 5f73 7461  ns.isin(gene_sta
+00000d50: 745f 636f 6c75 6d6e 7329 0d0a 2020 2020  t_columns)..    
+00000d60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00000d70: 6f76 6572 6c61 7070 696e 675f 636f 6c73  overlapping_cols
+00000d80: 7472 203d 2022 2c20 222e 6a6f 696e 286f  tr = ", ".join(o
+00000d90: 7665 726c 6170 7069 6e67 5f63 6f6c 735b  verlapping_cols[
+00000da0: 6f76 6572 6c61 7070 696e 675f 636f 6c73  overlapping_cols
+00000db0: 5d2e 696e 6465 782e 746f 5f6c 6973 7428  ].index.to_list(
+00000dc0: 2929 0d0a 2020 2020 2020 2020 2020 2020  ))..            
+00000dd0: 2020 2020 2020 2020 6c6f 6767 696e 672e          logging.
+00000de0: 7761 726e 696e 6728 6622 416e 6e44 6174  warning(f"AnnDat
+00000df0: 6120 6f62 6a65 6374 2063 6f6e 7461 696e  a object contain
+00000e00: 7320 634e 4d46 2067 656e 6520 7374 6174  s cNMF gene stat
+00000e10: 7320 7768 6963 6820 7769 6c6c 206f 7665  s which will ove
+00000e20: 7272 6964 6520 636f 6c75 6d6e 7320 696e  rride columns in
+00000e30: 2061 6461 7461 2e76 6172 3a20 7b6f 7665   adata.var: {ove
+00000e40: 726c 6170 7069 6e67 5f63 6f6c 7374 727d  rlapping_colstr}
+00000e50: 2229 0d0a 2020 2020 2020 2020 2020 2020  ")..            
+00000e60: 2020 2020 6164 6174 612e 7661 7220 3d20      adata.var = 
+00000e70: 7064 2e6d 6572 6765 286c 6566 743d 6164  pd.merge(left=ad
+00000e80: 6174 612e 7661 722c 2072 6967 6874 3d61  ata.var, right=a
+00000e90: 6461 7461 2e75 6e73 5b22 6f64 6722 5d5b  data.uns["odg"][
+00000ea0: 2267 656e 655f 7374 6174 7322 5d2c 2068  "gene_stats"], h
+00000eb0: 6f77 3d22 6c65 6674 222c 206c 6566 745f  ow="left", left_
+00000ec0: 696e 6465 783d 5472 7565 2c20 7269 6768  index=True, righ
+00000ed0: 745f 696e 6465 783d 5472 7565 290d 0a20  t_index=True).. 
+00000ee0: 2020 2020 2020 2020 2020 2020 2020 2064                 d
+00000ef0: 656c 2061 6461 7461 2e75 6e73 5b22 6f64  el adata.uns["od
+00000f00: 6722 5d5b 2267 656e 655f 7374 6174 7322  g"]["gene_stats"
+00000f10: 5d0d 0a20 2020 2020 2020 2020 2020 200d  ]..            .
+00000f20: 0a20 2020 2020 2020 2020 2020 2023 2063  .            # c
+00000f30: 7265 6174 6520 6e65 7720 416e 6e44 6174  reate new AnnDat
+00000f40: 6120 6f62 6a65 6374 0d0a 2020 2020 2020  a object..      
+00000f50: 2020 2020 2020 6e65 775f 6164 6174 6120        new_adata 
+00000f60: 3d20 6164 2e41 6e6e 4461 7461 2858 3d72  = ad.AnnData(X=r
+00000f70: 6177 2c20 6f62 733d 6164 6174 612e 6f62  aw, obs=adata.ob
+00000f80: 732c 2076 6172 3d61 6461 7461 2e76 6172  s, var=adata.var
+00000f90: 2c20 7661 726d 3d61 6461 7461 2e76 6172  , varm=adata.var
+00000fa0: 6d2c 206f 6273 6d3d 6164 6174 612e 6f62  m, obsm=adata.ob
+00000fb0: 736d 2c20 756e 733d 6164 6174 612e 756e  sm, uns=adata.un
+00000fc0: 7329 0d0a 2020 2020 2020 2020 2020 2020  s)..            
+00000fd0: 6966 2022 6869 7374 6f72 7922 206e 6f74  if "history" not
+00000fe0: 2069 6e20 6e65 775f 6164 6174 612e 756e   in new_adata.un
+00000ff0: 733a 0d0a 2020 2020 2020 2020 2020 2020  s:..            
+00001000: 2020 2020 6e65 775f 6164 6174 612e 756e      new_adata.un
+00001010: 735b 2268 6973 746f 7279 225d 203d 207b  s["history"] = {
+00001020: 7d0d 0a0d 0a20 2020 2020 2020 2020 2020  }....           
+00001030: 2023 2075 7064 6174 6520 6461 7461 7365   # update datase
+00001040: 7420 6f62 6a65 6374 0d0a 2020 2020 2020  t object..      
+00001050: 2020 2020 2020 7365 6c66 2e61 6461 7461        self.adata
+00001060: 203d 206e 6577 5f61 6461 7461 0d0a 2020   = new_adata..  
+00001070: 2020 2020 2020 2020 2020 7365 6c66 2e69            self.i
+00001080: 735f 6e6f 726d 616c 697a 6564 203d 2069  s_normalized = i
+00001090: 735f 6e6f 726d 616c 697a 6564 0d0a 2020  s_normalized..  
+000010a0: 2020 2020 2020 2020 2020 7365 6c66 2e63            self.c
+000010b0: 6e6d 6673 6e73 5f76 6572 7369 6f6e 203d  nmfsns_version =
+000010c0: 205f 5f76 6572 7369 6f6e 5f5f 0d0a 2020   __version__..  
+000010d0: 2020 0d0a 2020 2020 4063 6c61 7373 6d65    ..    @classme
+000010e0: 7468 6f64 0d0a 2020 2020 6465 6620 6672  thod..    def fr
+000010f0: 6f6d 5f64 6628 636c 732c 0d0a 2020 2020  om_df(cls,..    
+00001100: 2020 2020 2020 2020 2020 2020 2020 6461                da
+00001110: 7461 3a20 7064 2e44 6174 6146 7261 6d65  ta: pd.DataFrame
+00001120: 2c0d 0a20 2020 2020 2020 2020 2020 2020  ,..             
+00001130: 2020 2020 2069 735f 6e6f 726d 616c 697a       is_normaliz
+00001140: 6564 3a20 626f 6f6c 2c0d 0a20 2020 2020  ed: bool,..     
+00001150: 2020 2020 2020 2020 2020 2020 2073 7061               spa
+00001160: 7273 6966 793a 2062 6f6f 6c20 3d20 4661  rsify: bool = Fa
+00001170: 6c73 652c 0d0a 2020 2020 2020 2020 2020  lse,..          
+00001180: 2020 2020 2020 2020 6f62 733a 204f 7074          obs: Opt
+00001190: 696f 6e61 6c5b 7064 2e44 6174 6146 7261  ional[pd.DataFra
+000011a0: 6d65 5d20 3d20 4e6f 6e65 2c0d 0a20 2020  me] = None,..   
+000011b0: 2020 2020 2020 2020 2020 2020 2020 2076                 v
+000011c0: 6172 3a20 4f70 7469 6f6e 616c 5b70 642e  ar: Optional[pd.
+000011d0: 4461 7461 4672 616d 655d 203d 204e 6f6e  DataFrame] = Non
+000011e0: 652c 0d0a 2020 2020 2020 2020 2020 2020  e,..            
+000011f0: 2020 2020 2020 7061 7469 656e 745f 6964        patient_id
+00001200: 5f63 6f6c 3a20 4f70 7469 6f6e 616c 5b73  _col: Optional[s
+00001210: 7472 5d20 3d20 4e6f 6e65 0d0a 2020 2020  tr] = None..    
+00001220: 2020 2020 2020 2020 2020 2020 2020 293a                ):
+00001230: 0d0a 2020 2020 2020 2020 2222 2243 7265  ..        """Cre
+00001240: 6174 6573 2061 203a 636c 6173 733a 607e  ates a :class:`~
+00001250: 636e 6d66 736e 732e 6461 7461 7365 742e  cnmfsns.dataset.
+00001260: 4461 7461 7365 7460 206f 626a 6563 7420  Dataset` object 
+00001270: 6672 6f6d 2061 2070 616e 6461 7320 4461  from a pandas Da
+00001280: 7461 4672 616d 652e 0d0a 0d0a 2020 2020  taFrame.....    
+00001290: 2020 2020 3a70 6172 616d 2064 6174 613a      :param data:
+000012a0: 2041 6e20 6f62 7365 7276 6174 696f 6e73   An observations
+000012b0: 20c3 9720 7661 7269 6162 6c65 7320 6461   .. variables da
+000012c0: 7461 0d0a 2020 2020 2020 2020 3a74 7970  ta..        :typ
+000012d0: 6520 6461 7461 3a20 7064 2e44 6174 6146  e data: pd.DataF
+000012e0: 7261 6d65 0d0a 2020 2020 2020 2020 3a70  rame..        :p
+000012f0: 6172 616d 2069 735f 6e6f 726d 616c 697a  aram is_normaliz
+00001300: 6564 3a20 5370 6563 6966 7920 6966 2064  ed: Specify if d
+00001310: 6174 6120 6973 2061 6c72 6561 6479 206e  ata is already n
+00001320: 6f72 6d61 6c69 7a65 6420 6f72 2077 6865  ormalized or whe
+00001330: 7468 6572 206e 6f74 2e20 5261 7720 6461  ther not. Raw da
+00001340: 7461 2077 696c 6c20 6265 2054 504d 206e  ta will be TPM n
+00001350: 6f72 6d61 6c69 7a65 6420 7072 696f 7220  ormalized prior 
+00001360: 746f 206f 7665 7264 6973 7065 7273 6564  to overdispersed
+00001370: 2067 656e 6520 7365 6c65 6374 696f 6e2c   gene selection,
+00001380: 2077 6865 7265 6173 2061 6c72 6561 6479   whereas already
+00001390: 206e 6f72 6d61 6c69 7a65 6420 6461 7461   normalized data
+000013a0: 2077 696c 6c20 6e6f 742e 0d0a 2020 2020   will not...    
+000013b0: 2020 2020 3a74 7970 6520 6973 5f6e 6f72      :type is_nor
+000013c0: 6d61 6c69 7a65 643a 2062 6f6f 6c0d 0a20  malized: bool.. 
+000013d0: 2020 2020 2020 203a 7061 7261 6d20 7370         :param sp
+000013e0: 6172 7369 6679 3a20 5374 6f72 6520 6461  arsify: Store da
+000013f0: 7461 2061 7320 6120 7370 6172 7365 206d  ta as a sparse m
+00001400: 6174 7269 782e 205b 4e6f 7465 2074 6861  atrix. [Note tha
+00001410: 7420 7468 6973 2066 6561 7475 7265 2069  t this feature i
+00001420: 7320 6578 7065 7269 6d65 6e74 616c 5d2c  s experimental],
+00001430: 2064 6566 6175 6c74 7320 746f 2046 616c   defaults to Fal
+00001440: 7365 0d0a 2020 2020 2020 2020 3a74 7970  se..        :typ
+00001450: 6520 7370 6172 7369 6679 3a20 626f 6f6c  e sparsify: bool
+00001460: 2c20 6f70 7469 6f6e 616c 0d0a 2020 2020  , optional..    
+00001470: 2020 2020 3a70 6172 616d 206f 6273 3a20      :param obs: 
+00001480: 416e 206f 6273 6572 7661 7469 6f6e 7320  An observations 
+00001490: c397 206d 6574 6164 6174 6120 6d61 7472  .. metadata matr
+000014a0: 6978 2c20 6465 6661 756c 7473 2074 6f20  ix, defaults to 
+000014b0: 4e6f 6e65 0d0a 2020 2020 2020 2020 3a74  None..        :t
+000014c0: 7970 6520 6f62 733a 2060 7064 2e44 6174  ype obs: `pd.Dat
+000014d0: 6146 7261 6d65 602c 206f 7074 696f 6e61  aFrame`, optiona
+000014e0: 6c0d 0a20 2020 2020 2020 203a 7061 7261  l..        :para
+000014f0: 6d20 7661 723a 2041 2076 6172 6961 626c  m var: A variabl
+00001500: 6573 20c3 9720 6d65 7461 6461 7461 206d  es .. metadata m
+00001510: 6174 7269 782c 2064 6566 6175 6c74 7320  atrix, defaults 
+00001520: 746f 204e 6f6e 650d 0a20 2020 2020 2020  to None..       
+00001530: 203a 7479 7065 2076 6172 3a20 6070 642e   :type var: `pd.
+00001540: 4461 7461 4672 616d 6560 2c20 6f70 7469  DataFrame`, opti
+00001550: 6f6e 616c 0d0a 2020 2020 2020 2020 3a70  onal..        :p
+00001560: 6172 616d 2070 6174 6965 6e74 5f69 645f  aram patient_id_
+00001570: 636f 6c3a 204e 616d 6520 6f66 206d 6574  col: Name of met
+00001580: 6164 6174 6120 6c61 7965 7220 7769 7468  adata layer with
+00001590: 2070 6174 6965 6e74 2049 4420 696e 666f   patient ID info
+000015a0: 726d 6174 696f 6e2c 2064 6566 6175 6c74  rmation, default
+000015b0: 7320 746f 204e 6f6e 650d 0a20 2020 2020  s to None..     
+000015c0: 2020 203a 7479 7065 2070 6174 6965 6e74     :type patient
+000015d0: 5f69 645f 636f 6c3a 2073 7472 2c20 6f70  _id_col: str, op
+000015e0: 7469 6f6e 616c 0d0a 2020 2020 2020 2020  tional..        
+000015f0: 3a72 6574 7572 6e3a 204f 626a 6563 7420  :return: Object 
+00001600: 7769 7468 2065 7870 7265 7373 696f 6e20  with expression 
+00001610: 616e 6420 6d65 7461 6461 7461 0d0a 2020  and metadata..  
+00001620: 2020 2020 2020 3a72 7479 7065 3a20 3a63        :rtype: :c
+00001630: 6c61 7373 3a60 7e63 6e6d 6673 6e73 2e64  lass:`~cnmfsns.d
+00001640: 6174 6173 6574 2e44 6174 6173 6574 600d  ataset.Dataset`.
+00001650: 0a20 2020 2020 2020 2022 2222 0d0a 2020  .        """..  
+00001660: 2020 2020 2020 6966 2076 6172 2069 7320        if var is 
+00001670: 6e6f 7420 4e6f 6e65 3a0d 0a20 2020 2020  not None:..     
+00001680: 2020 2020 2020 2076 6172 203d 2076 6172         var = var
+00001690: 2e72 6569 6e64 6578 2864 6174 612e 636f  .reindex(data.co
+000016a0: 6c75 6d6e 7329 0d0a 2020 2020 2020 2020  lumns)..        
+000016b0: 6966 2073 7061 7273 6966 793a 0d0a 2020  if sparsify:..  
+000016c0: 2020 2020 2020 2020 2020 6461 7461 203d            data =
+000016d0: 2073 702e 6373 725f 6d61 7472 6978 2864   sp.csr_matrix(d
+000016e0: 6174 612e 7661 6c75 6573 290d 0a20 2020  ata.values)..   
+000016f0: 2020 2020 2064 6174 6120 3d20 6461 7461       data = data
+00001700: 2e61 7374 7970 6528 2266 6c6f 6174 3332  .astype("float32
+00001710: 2229 0d0a 2020 2020 2020 2020 756e 7320  ")..        uns 
+00001720: 3d20 7b22 6869 7374 6f72 7922 3a20 7b7d  = {"history": {}
+00001730: 2c20 226f 6467 223a 7b7d 7d0d 0a20 2020  , "odg":{}}..   
+00001740: 2020 2020 2061 6461 7461 203d 2061 642e       adata = ad.
+00001750: 416e 6e44 6174 6128 583d 6461 7461 2c20  AnnData(X=data, 
+00001760: 7661 723d 7661 722c 2075 6e73 3d75 6e73  var=var, uns=uns
+00001770: 290d 0a20 2020 2020 2020 2064 6174 6173  )..        datas
+00001780: 6574 203d 2063 6c73 2861 6461 7461 290d  et = cls(adata).
+00001790: 0a20 2020 2020 2020 2069 6620 6f62 7320  .        if obs 
+000017a0: 6973 206e 6f74 204e 6f6e 653a 0d0a 2020  is not None:..  
+000017b0: 2020 2020 2020 2020 2020 6461 7461 7365            datase
+000017c0: 742e 7570 6461 7465 5f6f 6273 286f 6273  t.update_obs(obs
+000017d0: 3d6f 6273 2920 0d0a 2020 2020 2020 2020  =obs) ..        
+000017e0: 6461 7461 7365 742e 7061 7469 656e 745f  dataset.patient_
+000017f0: 6964 5f63 6f6c 203d 2070 6174 6965 6e74  id_col = patient
+00001800: 5f69 645f 636f 6c20 2023 206d 7573 7420  _id_col  # must 
+00001810: 6265 2064 6f6e 6520 6166 7465 7220 7570  be done after up
+00001820: 6461 7469 6e67 2074 6865 206f 6273 2066  dating the obs f
+00001830: 7261 6d65 2073 696e 6365 2069 7420 7769  rame since it wi
+00001840: 6c6c 2063 6865 636b 2066 6f72 2074 6865  ll check for the
+00001850: 2063 6f6c 756d 6e73 2065 7869 7374 656e   columns existen
+00001860: 6365 2069 6e20 6f62 730d 0a20 2020 2020  ce in obs..     
+00001870: 2020 2064 6174 6173 6574 2e69 735f 6e6f     dataset.is_no
+00001880: 726d 616c 697a 6564 203d 2069 735f 6e6f  rmalized = is_no
+00001890: 726d 616c 697a 6564 0d0a 2020 2020 2020  rmalized..      
+000018a0: 2020 7265 7475 726e 2064 6174 6173 6574    return dataset
+000018b0: 0d0a 2020 2020 0d0a 2020 2020 4063 6c61  ..    ..    @cla
+000018c0: 7373 6d65 7468 6f64 0d0a 2020 2020 6465  ssmethod..    de
+000018d0: 6620 6672 6f6d 5f68 3561 6428 636c 732c  f from_h5ad(cls,
+000018e0: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
+000018f0: 2020 2020 6835 6164 5f66 696c 653a 2073      h5ad_file: s
+00001900: 7472 2c0d 0a20 2020 2020 2020 2020 2020  tr,..           
+00001910: 2020 2020 2020 2066 6f72 6365 5f6d 6967         force_mig
+00001920: 7261 7465 3d46 616c 7365 2c20 6261 636b  rate=False, back
+00001930: 6564 3d46 616c 7365 0d0a 2020 2020 2020  ed=False..      
+00001940: 2020 2020 2020 2020 2020 2020 293a 0d0a              ):..
+00001950: 2020 2020 2020 2020 2222 2243 7265 6174          """Creat
+00001960: 6573 2061 203a 636c 6173 733a 607e 636e  es a :class:`~cn
+00001970: 6d66 736e 732e 6461 7461 7365 742e 4461  mfsns.dataset.Da
+00001980: 7461 7365 7460 206f 626a 6563 7420 6672  taset` object fr
+00001990: 6f6d 2061 6e20 416e 6e44 6174 612d 636f  om an AnnData-co
+000019a0: 6d70 6174 6962 6c65 202e 6835 6164 2066  mpatible .h5ad f
+000019b0: 696c 652e 0d0a 0d0a 2020 2020 2020 2020  ile.....        
+000019c0: 3a70 6172 616d 2068 3561 645f 6669 6c65  :param h5ad_file
+000019d0: 3a20 5061 7468 2074 6f20 2e68 3561 6420  : Path to .h5ad 
+000019e0: 6669 6c65 2070 726f 6475 6365 6420 6279  file produced by
+000019f0: 2073 6361 6e70 792c 2041 6e6e 4461 7461   scanpy, AnnData
+00001a00: 2c20 6f72 2063 4e4d 462d 534e 530d 0a20  , or cNMF-SNS.. 
+00001a10: 2020 2020 2020 203a 7479 7065 2068 3561         :type h5a
+00001a20: 645f 6669 6c65 3a20 7374 720d 0a20 2020  d_file: str..   
+00001a30: 2020 2020 203a 7061 7261 6d20 666f 7263       :param forc
+00001a40: 655f 6d69 6772 6174 653a 2066 6f72 6365  e_migrate: force
+00001a50: 7320 636f 6e76 6572 7369 6f6e 206f 6620  s conversion of 
+00001a60: 416e 6e44 6174 6120 6f62 6a65 6374 7320  AnnData objects 
+00001a70: 6576 656e 2077 6865 6e20 6164 6174 612e  even when adata.
+00001a80: 5820 616e 6420 6164 6174 612e 7261 772e  X and adata.raw.
+00001a90: 5820 6172 6520 6e6f 7420 6c69 6e65 6172  X are not linear
+00001aa0: 6c79 2073 6361 6c65 6420 7265 6c61 7469  ly scaled relati
+00001ab0: 7665 2074 6f20 6561 6368 206f 7468 6572  ve to each other
+00001ac0: 2c20 6465 6661 756c 7473 2074 6f20 4661  , defaults to Fa
+00001ad0: 6c73 650d 0a20 2020 2020 2020 203a 7479  lse..        :ty
+00001ae0: 7065 2066 6f72 6365 5f6d 6967 7261 7465  pe force_migrate
+00001af0: 3a20 626f 6f6c 2c20 6f70 7469 6f6e 616c  : bool, optional
+00001b00: 0d0a 2020 2020 2020 2020 3a70 6172 616d  ..        :param
+00001b10: 2062 6163 6b65 643a 2055 7365 2062 6163   backed: Use bac
+00001b20: 6b65 6420 6d6f 6465 2074 6f20 6f70 656e  ked mode to open
+00001b30: 2068 3561 6420 6669 6c65 2e20 5468 6973   h5ad file. This
+00001b40: 2063 616e 2073 6176 6520 6d65 6d6f 7279   can save memory
+00001b50: 2077 6865 6e20 7468 6520 6461 7461 7365   when the datase
+00001b60: 7420 6973 2076 6572 7920 6c61 7267 652c  t is very large,
+00001b70: 2062 7574 2069 7320 6e6f 7420 636f 6d70   but is not comp
+00001b80: 6174 6962 6c65 2077 6974 6820 6835 6164  atible with h5ad
+00001b90: 2066 696c 6573 2070 726f 6475 6365 6420   files produced 
+00001ba0: 6f75 7473 6964 6520 6f66 2063 4e4d 462d  outside of cNMF-
+00001bb0: 534e 532c 2064 6566 6175 6c74 7320 746f  SNS, defaults to
+00001bc0: 2046 616c 7365 0d0a 2020 2020 2020 2020   False..        
+00001bd0: 3a74 7970 6520 6261 636b 6564 3a20 626f  :type backed: bo
+00001be0: 6f6c 2c20 6f70 7469 6f6e 616c 0d0a 2020  ol, optional..  
+00001bf0: 2020 2020 2020 3a72 6574 7572 6e3a 204f        :return: O
+00001c00: 626a 6563 7420 7769 7468 2065 7870 7265  bject with expre
+00001c10: 7373 696f 6e20 616e 6420 6d65 7461 6461  ssion and metada
+00001c20: 7461 0d0a 2020 2020 2020 2020 3a72 7479  ta..        :rty
+00001c30: 7065 3a20 3a63 6c61 7373 3a60 7e63 6e6d  pe: :class:`~cnm
+00001c40: 6673 6e73 2e64 6174 6173 6574 2e44 6174  fsns.dataset.Dat
+00001c50: 6173 6574 600d 0a20 2020 2020 2020 2022  aset`..        "
+00001c60: 2222 0d0a 2020 2020 2020 2020 6164 6174  ""..        adat
+00001c70: 6120 3d20 6164 2e72 6561 645f 6835 6164  a = ad.read_h5ad
+00001c80: 2868 3561 645f 6669 6c65 2c20 6261 636b  (h5ad_file, back
+00001c90: 6564 3d62 6163 6b65 6429 0d0a 2020 2020  ed=backed)..    
+00001ca0: 2020 2020 6461 7461 7365 7420 3d20 636c      dataset = cl
+00001cb0: 7328 6164 6174 613d 6164 6174 612c 2066  s(adata=adata, f
+00001cc0: 6f72 6365 5f6d 6967 7261 7465 3d66 6f72  orce_migrate=for
+00001cd0: 6365 5f6d 6967 7261 7465 290d 0a20 2020  ce_migrate)..   
+00001ce0: 2020 2020 2072 6574 7572 6e20 6461 7461       return data
+00001cf0: 7365 740d 0a20 2020 200d 0a20 2020 2040  set..    ..    @
+00001d00: 7072 6f70 6572 7479 0d0a 2020 2020 6465  property..    de
+00001d10: 6620 6973 5f6e 6f72 6d61 6c69 7a65 6428  f is_normalized(
+00001d20: 7365 6c66 293a 0d0a 2020 2020 2020 2020  self):..        
+00001d30: 2222 224f 7574 7075 7473 2074 6865 206e  """Outputs the n
+00001d40: 6f72 6d61 6c69 7a61 7469 6f6e 2073 7461  ormalization sta
+00001d50: 7475 7320 6f66 2074 6865 2064 6174 6173  tus of the datas
+00001d60: 6574 2e0d 0a0d 0a20 2020 2020 2020 203a  et.....        :
+00001d70: 7265 7475 726e 3a20 5472 7565 2069 6620  return: True if 
+00001d80: 6461 7461 7365 7420 636f 6e74 6169 6e73  dataset contains
+00001d90: 206e 6f72 6d61 6c69 7a65 6420 6461 7461   normalized data
+00001da0: 2c20 4661 6c73 6520 6966 2069 7420 6973  , False if it is
+00001db0: 2072 6177 2064 6174 612e 0d0a 2020 2020   raw data...    
+00001dc0: 2020 2020 3a72 7479 7065 3a20 626f 6f6c      :rtype: bool
+00001dd0: 0d0a 2020 2020 2020 2020 2222 220d 0a20  ..        """.. 
+00001de0: 2020 2020 2020 2072 6574 7572 6e20 7365         return se
+00001df0: 6c66 2e61 6461 7461 2e75 6e73 5b22 6973  lf.adata.uns["is
+00001e00: 5f6e 6f72 6d61 6c69 7a65 6422 5d0d 0a20  _normalized"].. 
+00001e10: 2020 200d 0a20 2020 2040 6973 5f6e 6f72     ..    @is_nor
+00001e20: 6d61 6c69 7a65 642e 7365 7474 6572 0d0a  malized.setter..
+00001e30: 2020 2020 6465 6620 6973 5f6e 6f72 6d61      def is_norma
+00001e40: 6c69 7a65 6428 7365 6c66 2c20 7661 6c75  lized(self, valu
+00001e50: 653a 2062 6f6f 6c29 3a0d 0a20 2020 2020  e: bool):..     
+00001e60: 2020 2073 656c 662e 6164 6174 612e 756e     self.adata.un
+00001e70: 735b 2269 735f 6e6f 726d 616c 697a 6564  s["is_normalized
+00001e80: 225d 203d 2076 616c 7565 0d0a 2020 2020  "] = value..    
+00001e90: 0d0a 2020 2020 4070 726f 7065 7274 790d  ..    @property.
+00001ea0: 0a20 2020 2064 6566 2070 6174 6965 6e74  .    def patient
+00001eb0: 5f69 645f 636f 6c28 7365 6c66 293a 0d0a  _id_col(self):..
+00001ec0: 2020 2020 2020 2020 2222 224f 7574 7075          """Outpu
+00001ed0: 7473 2074 6865 206e 6f72 6d61 6c69 7a61  ts the normaliza
+00001ee0: 7469 6f6e 2073 7461 7475 7320 6f66 2074  tion status of t
+00001ef0: 6865 2064 6174 6173 6574 2e0d 0a0d 0a20  he dataset..... 
+00001f00: 2020 2020 2020 203a 7265 7475 726e 3a20         :return: 
+00001f10: 5472 7565 2069 6620 6461 7461 7365 7420  True if dataset 
+00001f20: 636f 6e74 6169 6e73 206e 6f72 6d61 6c69  contains normali
+00001f30: 7a65 6420 6461 7461 2c20 4661 6c73 6520  zed data, False 
+00001f40: 6966 2069 7420 6973 2072 6177 2064 6174  if it is raw dat
+00001f50: 612e 0d0a 2020 2020 2020 2020 3a72 7479  a...        :rty
+00001f60: 7065 3a20 626f 6f6c 0d0a 2020 2020 2020  pe: bool..      
+00001f70: 2020 2222 220d 0a20 2020 2020 2020 2072    """..        r
+00001f80: 6574 7572 6e20 7365 6c66 2e61 6461 7461  eturn self.adata
+00001f90: 2e75 6e73 5b22 7061 7469 656e 745f 6964  .uns["patient_id
+00001fa0: 5f63 6f6c 225d 0d0a 2020 2020 0d0a 2020  _col"]..    ..  
+00001fb0: 2020 4070 6174 6965 6e74 5f69 645f 636f    @patient_id_co
+00001fc0: 6c2e 7365 7474 6572 0d0a 2020 2020 6465  l.setter..    de
+00001fd0: 6620 7061 7469 656e 745f 6964 5f63 6f6c  f patient_id_col
+00001fe0: 2873 656c 662c 2076 616c 7565 3a20 7374  (self, value: st
+00001ff0: 7229 3a0d 0a20 2020 2020 2020 200d 0a20  r):..        .. 
+00002000: 2020 2020 2020 2069 6620 7661 6c75 6520         if value 
+00002010: 6973 206e 6f74 204e 6f6e 6520 616e 6420  is not None and 
+00002020: 7661 6c75 6520 6e6f 7420 696e 2073 656c  value not in sel
+00002030: 662e 6164 6174 612e 6f62 732e 636f 6c75  f.adata.obs.colu
+00002040: 6d6e 733a 0d0a 2020 2020 2020 2020 2020  mns:..          
+00002050: 2020 6176 6169 6c5f 636f 6c75 6d6e 7320    avail_columns 
+00002060: 3d20 222c 2022 2e6a 6f69 6e28 7365 6c66  = ", ".join(self
+00002070: 2e61 6461 7461 2e6f 6273 2e63 6f6c 756d  .adata.obs.colum
+00002080: 6e73 290d 0a20 2020 2020 2020 2020 2020  ns)..           
+00002090: 2072 6169 7365 2056 616c 7565 4572 726f   raise ValueErro
+000020a0: 7228 6622 7b76 616c 7565 7d20 6973 206e  r(f"{value} is n
+000020b0: 6f74 2061 2076 616c 6964 2063 6f6c 756d  ot a valid colum
+000020c0: 6e20 696e 2074 6865 206d 6574 6164 6174  n in the metadat
+000020d0: 6120 6d61 7472 6978 2e20 4176 6169 6c61  a matrix. Availa
+000020e0: 626c 6520 636f 6c75 6d6e 7320 6172 653a  ble columns are:
+000020f0: 207b 6176 6169 6c5f 636f 6c75 6d6e 737d   {avail_columns}
+00002100: 2229 0d0a 2020 2020 2020 2020 7365 6c66  ")..        self
+00002110: 2e61 6461 7461 2e75 6e73 5b22 7061 7469  .adata.uns["pati
+00002120: 656e 745f 6964 5f63 6f6c 225d 203d 2076  ent_id_col"] = v
+00002130: 616c 7565 0d0a 0d0a 2020 2020 4070 726f  alue....    @pro
+00002140: 7065 7274 790d 0a20 2020 2064 6566 2063  perty..    def c
+00002150: 6e6d 6673 6e73 5f76 6572 7369 6f6e 2873  nmfsns_version(s
+00002160: 656c 6629 3a0d 0a20 2020 2020 2020 2022  elf):..        "
+00002170: 2222 634e 4d46 2d53 4e53 2076 6572 7369  ""cNMF-SNS versi
+00002180: 6f6e 2075 7365 6420 746f 2063 7265 6174  on used to creat
+00002190: 6520 7468 6520 6461 7461 7365 740d 0a0d  e the dataset...
+000021a0: 0a20 2020 2020 2020 203a 7265 7475 726e  .        :return
+000021b0: 3a20 7665 7273 696f 6e0d 0a20 2020 2020  : version..     
+000021c0: 2020 203a 7274 7970 653a 2073 7472 0d0a     :rtype: str..
+000021d0: 2020 2020 2020 2020 2222 220d 0a20 2020          """..   
+000021e0: 2020 2020 2069 6620 2263 6e6d 6673 6e73       if "cnmfsns
+000021f0: 5f76 6572 7369 6f6e 2220 696e 2073 656c  _version" in sel
+00002200: 662e 6164 6174 612e 756e 733a 0d0a 2020  f.adata.uns:..  
+00002210: 2020 2020 2020 2020 2020 7665 7273 696f            versio
+00002220: 6e20 3d20 7365 6c66 2e61 6461 7461 2e75  n = self.adata.u
+00002230: 6e73 5b22 636e 6d66 736e 735f 7665 7273  ns["cnmfsns_vers
+00002240: 696f 6e22 5d0d 0a20 2020 2020 2020 2065  ion"]..        e
+00002250: 6c73 653a 0d0a 2020 2020 2020 2020 2020  lse:..          
+00002260: 2020 7665 7273 696f 6e20 3d20 4e6f 6e65    version = None
+00002270: 0d0a 2020 2020 2020 2020 7265 7475 726e  ..        return
+00002280: 2076 6572 7369 6f6e 0d0a 2020 2020 2020   version..      
+00002290: 2020 0d0a 2020 2020 4070 726f 7065 7274    ..    @propert
+000022a0: 790d 0a20 2020 2064 6566 2068 6173 5f63  y..    def has_c
+000022b0: 6e6d 665f 7265 7375 6c74 7328 7365 6c66  nmf_results(self
+000022c0: 293a 0d0a 2020 2020 2020 2020 2222 2254  ):..        """T
+000022d0: 6573 7420 666f 7220 7765 6874 6865 7220  est for wehther 
+000022e0: 4461 7461 7365 7420 636f 6e74 6169 6e73  Dataset contains
+000022f0: 2063 4e4d 4620 7265 7375 6c74 7320 666f   cNMF results fo
+00002300: 7220 7468 6520 6461 7461 7365 740d 0a0d  r the dataset...
+00002310: 0a20 2020 2020 2020 203a 7265 7475 726e  .        :return
+00002320: 3a20 5768 6574 6865 7220 636f 6d70 6c65  : Whether comple
+00002330: 7465 2063 4e4d 4620 7265 7375 6c74 7320  te cNMF results 
+00002340: 6172 6520 636f 6e74 6169 6e65 6420 666f  are contained fo
+00002350: 7220 6174 206c 6561 7374 2031 2072 616e  r at least 1 ran
+00002360: 6b20 286b 290d 0a20 2020 2020 2020 203a  k (k)..        :
+00002370: 7274 7970 653a 2062 6f6f 6c0d 0a20 2020  rtype: bool..   
+00002380: 2020 2020 2022 2222 0d0a 2020 2020 2020       """..      
+00002390: 2020 6d61 7472 6978 5f63 6865 636b 7320    matrix_checks 
+000023a0: 3d20 5b0d 0a20 2020 2020 2020 2020 2020  = [..           
+000023b0: 2022 636e 6d66 5f75 7361 6765 2220 696e   "cnmf_usage" in
+000023c0: 2073 656c 662e 6164 6174 612e 6f62 736d   self.adata.obsm
+000023d0: 2c0d 0a20 2020 2020 2020 2020 2020 2022  ,..            "
+000023e0: 636e 6d66 5f67 6570 5f73 636f 7265 2220  cnmf_gep_score" 
+000023f0: 696e 2073 656c 662e 6164 6174 612e 7661  in self.adata.va
+00002400: 726d 2c0d 0a20 2020 2020 2020 2020 2020  rm,..           
+00002410: 2022 636e 6d66 5f67 6570 5f74 706d 2220   "cnmf_gep_tpm" 
+00002420: 696e 2073 656c 662e 6164 6174 612e 7661  in self.adata.va
+00002430: 726d 2c0d 0a20 2020 2020 2020 2020 2020  rm,..           
+00002440: 2022 636e 6d66 5f67 6570 5f72 6177 2220   "cnmf_gep_raw" 
+00002450: 696e 2073 656c 662e 6164 6174 612e 7661  in self.adata.va
+00002460: 726d 2c0d 0a20 2020 2020 2020 2020 2020  rm,..           
+00002470: 2022 6b76 616c 7322 2069 6e20 7365 6c66   "kvals" in self
+00002480: 2e61 6461 7461 2e75 6e73 0d0a 2020 2020  .adata.uns..    
+00002490: 2020 2020 5d0d 0a20 2020 2020 2020 2072      ]..        r
+000024a0: 6574 7572 6e20 616c 6c28 6d61 7472 6978  eturn all(matrix
+000024b0: 5f63 6865 636b 7329 0d0a 2020 2020 2020  _checks)..      
+000024c0: 2020 0d0a 2020 2020 4070 726f 7065 7274    ..    @propert
+000024d0: 790d 0a20 2020 2064 6566 206f 7665 7264  y..    def overd
+000024e0: 6973 7065 7273 6564 5f67 656e 6573 2873  ispersed_genes(s
+000024f0: 656c 6629 3a0d 0a20 2020 2020 2020 2022  elf):..        "
+00002500: 2222 4f76 6572 6469 7370 6572 7365 6420  ""Overdispersed 
+00002510: 6765 6e65 206c 6973 7420 7573 6564 2066  gene list used f
+00002520: 6f72 2063 4e4d 460d 0a0d 0a20 2020 2020  or cNMF....     
+00002530: 2020 203a 7265 7475 726e 3a20 6765 6e65     :return: gene
+00002540: 206c 6973 740d 0a20 2020 2020 2020 203a   list..        :
+00002550: 7274 7970 653a 206c 6973 740d 0a20 2020  rtype: list..   
+00002560: 2020 2020 2022 2222 0d0a 2020 2020 2020       """..      
+00002570: 2020 7265 7475 726e 2073 656c 662e 6164    return self.ad
+00002580: 6174 612e 756e 735b 2267 656e 655f 6c69  ata.uns["gene_li
+00002590: 7374 225d 0d0a 0d0a 0d0a 2020 2020 4063  st"]......    @c
+000025a0: 6e6d 6673 6e73 5f76 6572 7369 6f6e 2e73  nmfsns_version.s
+000025b0: 6574 7465 720d 0a20 2020 2064 6566 2063  etter..    def c
+000025c0: 6e6d 6673 6e73 5f76 6572 7369 6f6e 2873  nmfsns_version(s
+000025d0: 656c 662c 2076 616c 7565 3a20 626f 6f6c  elf, value: bool
+000025e0: 293a 0d0a 2020 2020 2020 2020 7365 6c66  ):..        self
+000025f0: 2e61 6461 7461 2e75 6e73 5b22 636e 6d66  .adata.uns["cnmf
+00002600: 736e 735f 7665 7273 696f 6e22 5d20 3d20  sns_version"] = 
+00002610: 7661 6c75 650d 0a0d 0a20 2020 2020 2020  value....       
+00002620: 200d 0a20 2020 2064 6566 2075 7064 6174   ..    def updat
+00002630: 655f 6f62 7328 7365 6c66 2c20 6f62 7329  e_obs(self, obs)
+00002640: 3a0d 0a20 2020 2020 2020 2022 2222 5570  :..        """Up
+00002650: 6461 7465 2074 6865 206f 6273 6572 7661  date the observa
+00002660: 7469 6f6e 206d 6574 6164 6174 6120 7769  tion metadata wi
+00002670: 7468 2061 206e 6577 206d 6574 6164 6174  th a new metadat
+00002680: 6120 6d61 7472 6978 0d0a 0d0a 2020 2020  a matrix....    
+00002690: 2020 2020 3a70 6172 616d 206f 6273 3a20      :param obs: 
+000026a0: 416e 206f 6273 6572 7661 7469 6f6e 7320  An observations 
+000026b0: c397 206d 6574 6164 6174 6120 6d61 7472  .. metadata matr
+000026c0: 6978 2c20 6465 6661 756c 7473 2074 6f20  ix, defaults to 
+000026d0: 4e6f 6e65 0d0a 2020 2020 2020 2020 3a74  None..        :t
+000026e0: 7970 6520 6f62 733a 2060 7064 2e44 6174  ype obs: `pd.Dat
+000026f0: 6146 7261 6d65 602c 206f 7074 696f 6e61  aFrame`, optiona
+00002700: 6c0d 0a20 2020 2020 2020 2022 2222 0d0a  l..        """..
+00002710: 2020 2020 2020 2020 2320 636f 6e76 6572          # conver
+00002720: 7420 276f 626a 6563 7427 2064 7479 7065  t 'object' dtype
+00002730: 2074 6f20 6361 7465 676f 7269 6361 6c2c   to categorical,
+00002740: 2063 6f6e 7665 7274 696e 6720 626f 6f6c   converting bool
+00002750: 2076 616c 7565 7320 746f 2073 7472 696e   values to strin
+00002760: 6773 2061 7320 7468 6573 6520 6172 6520  gs as these are 
+00002770: 6e6f 7420 7375 7070 6f72 7465 6420 6279  not supported by
+00002780: 2041 6e6e 4461 7461 206f 6e2d 6469 736b   AnnData on-disk
+00002790: 2066 6f72 6d61 740d 0a20 2020 2020 2020   format..       
+000027a0: 2066 6f72 2063 6f6c 2069 6e20 6f62 732e   for col in obs.
+000027b0: 7365 6c65 6374 5f64 7479 7065 7328 696e  select_dtypes(in
+000027c0: 636c 7564 653d 2822 626f 6f6c 222c 2022  clude=("bool", "
+000027d0: 6f62 6a65 6374 2229 292e 636f 6c75 6d6e  object")).column
+000027e0: 733a 0d0a 2020 2020 2020 2020 2020 2020  s:..            
+000027f0: 6f62 735b 636f 6c5d 203d 206f 6273 5b63  obs[col] = obs[c
+00002800: 6f6c 5d2e 6173 7479 7065 2822 7374 7222  ol].astype("str"
+00002810: 292e 6173 7479 7065 2822 6361 7465 676f  ).astype("catego
+00002820: 7279 2229 0d0a 2020 2020 2020 2020 6d69  ry")..        mi
+00002830: 7373 696e 675f 7361 6d70 6c65 735f 696e  ssing_samples_in
+00002840: 5f58 203d 206f 6273 2e69 6e64 6578 2e64  _X = obs.index.d
+00002850: 6966 6665 7265 6e63 6528 7365 6c66 2e61  ifference(self.a
+00002860: 6461 7461 2e6f 6273 2e69 6e64 6578 292e  data.obs.index).
+00002870: 6173 7479 7065 2873 7472 292e 746f 5f6c  astype(str).to_l
+00002880: 6973 7428 290d 0a20 2020 2020 2020 2069  ist()..        i
+00002890: 6620 6d69 7373 696e 675f 7361 6d70 6c65  f missing_sample
+000028a0: 735f 696e 5f58 3a0d 0a20 2020 2020 2020  s_in_X:..       
+000028b0: 2020 2020 206c 6f67 6769 6e67 2e77 6172       logging.war
+000028c0: 6e69 6e67 2822 5468 6520 666f 6c6c 6f77  ning("The follow
+000028d0: 696e 6720 7361 6d70 6c65 7320 696e 2074  ing samples in t
+000028e0: 6865 206d 6574 6164 6174 6120 7765 7265  he metadata were
+000028f0: 206e 6f74 2070 7265 7365 6e74 2069 6e20   not present in 
+00002900: 7468 6520 6461 7461 2028 6061 6461 7461  the data (`adata
+00002910: 2e58 6029 3a5c 6e20 202d 2022 202b 2022  .X`):\n  - " + "
+00002920: 5c6e 2020 2d20 222e 6a6f 696e 286d 6973  \n  - ".join(mis
+00002930: 7369 6e67 5f73 616d 706c 6573 5f69 6e5f  sing_samples_in_
+00002940: 5829 290d 0a20 2020 2020 2020 206d 6973  X))..        mis
+00002950: 7369 6e67 5f73 616d 706c 6573 5f69 6e5f  sing_samples_in_
+00002960: 6d64 203d 2073 656c 662e 6164 6174 612e  md = self.adata.
+00002970: 6f62 732e 696e 6465 782e 6469 6666 6572  obs.index.differ
+00002980: 656e 6365 286f 6273 2e69 6e64 6578 292e  ence(obs.index).
+00002990: 6173 7479 7065 2873 7472 292e 746f 5f6c  astype(str).to_l
+000029a0: 6973 7428 290d 0a20 2020 2020 2020 2069  ist()..        i
+000029b0: 6620 6d69 7373 696e 675f 7361 6d70 6c65  f missing_sample
+000029c0: 735f 696e 5f6d 643a 0d0a 2020 2020 2020  s_in_md:..      
+000029d0: 2020 2020 2020 6c6f 6767 696e 672e 7761        logging.wa
+000029e0: 726e 696e 6728 2254 6865 2066 6f6c 6c6f  rning("The follo
+000029f0: 7769 6e67 2073 616d 706c 6573 2069 6e20  wing samples in 
+00002a00: 7468 6520 6461 7461 2028 6061 6461 7461  the data (`adata
+00002a10: 2e58 6029 2077 6572 6520 6162 7365 6e74  .X`) were absent
+00002a20: 2069 6e20 7468 6520 6d65 7461 6461 7461   in the metadata
+00002a30: 3a5c 6e20 202d 2022 202b 2022 5c6e 2020  :\n  - " + "\n  
+00002a40: 2d20 222e 6a6f 696e 286d 6973 7369 6e67  - ".join(missing
+00002a50: 5f73 616d 706c 6573 5f69 6e5f 6d64 2929  _samples_in_md))
+00002a60: 0d0a 2020 2020 2020 2020 7365 6c66 2e61  ..        self.a
+00002a70: 6461 7461 2e6f 6273 203d 206f 6273 2e72  data.obs = obs.r
+00002a80: 6569 6e64 6578 2873 656c 662e 6164 6174  eindex(self.adat
+00002a90: 612e 6f62 732e 696e 6465 7829 0d0a 2020  a.obs.index)..  
+00002aa0: 2020 0d0a 2020 2020 6465 6620 6765 745f    ..    def get_
+00002ab0: 6d65 7461 6461 7461 5f74 7970 655f 7375  metadata_type_su
+00002ac0: 6d6d 6172 7928 7365 6c66 293a 0d0a 2020  mmary(self):..  
+00002ad0: 2020 2020 2020 2222 2252 6574 7572 6e20        """Return 
+00002ae0: 6120 7072 696e 7461 626c 6520 7375 6d6d  a printable summ
+00002af0: 6172 7920 6f66 206d 6574 6164 6174 6120  ary of metadata 
+00002b00: 616e 6420 7661 6c75 6520 7479 7065 7320  and value types 
+00002b10: 666f 7220 6561 6368 206d 6574 6164 6174  for each metadat
+00002b20: 6120 6c61 7965 722e 0d0a 0d0a 2020 2020  a layer.....    
+00002b30: 2020 2020 3a72 6574 7572 6e3a 2053 756d      :return: Sum
+00002b40: 6d61 7279 206f 6620 6d65 7461 6461 7461  mary of metadata
+00002b50: 0d0a 2020 2020 2020 2020 3a72 7479 7065  ..        :rtype
+00002b60: 3a20 7374 720d 0a20 2020 2020 2020 2022  : str..        "
+00002b70: 2222 0d0a 2020 2020 2020 2020 6d73 6720  ""..        msg 
+00002b80: 3d20 2222 0d0a 2020 2020 2020 2020 666f  = ""..        fo
+00002b90: 7220 636f 6c20 696e 2073 656c 662e 6164  r col in self.ad
+00002ba0: 6174 612e 6f62 732e 636f 6c75 6d6e 733a  ata.obs.columns:
+00002bb0: 0d0a 2020 2020 2020 2020 2020 2020 6d73  ..            ms
+00002bc0: 6720 2b3d 2022 2020 2020 436f 6c75 6d6e  g += "    Column
+00002bd0: 3a20 2220 2b20 636f 6c20 2b20 225c 6e22  : " + col + "\n"
+00002be0: 0d0a 2020 2020 2020 2020 2020 2020 666f  ..            fo
+00002bf0: 7220 7661 6c75 655f 7479 7065 2c20 636f  r value_type, co
+00002c00: 756e 7420 696e 2073 656c 662e 6164 6174  unt in self.adat
+00002c10: 612e 6f62 735b 636f 6c5d 2e64 726f 706e  a.obs[col].dropn
+00002c20: 6128 292e 6d61 7028 7479 7065 292e 7661  a().map(type).va
+00002c30: 6c75 655f 636f 756e 7473 2829 2e69 7465  lue_counts().ite
+00002c40: 6d73 2829 3a0d 0a20 2020 2020 2020 2020  ms():..         
+00002c50: 2020 2020 2020 206d 7367 202b 3d20 6622         msg += f"
+00002c60: 2020 2020 2020 2020 7b76 616c 7565 5f74          {value_t
+00002c70: 7970 657d 3a20 7b63 6f75 6e74 7d5c 6e22  ype}: {count}\n"
+00002c80: 0d0a 2020 2020 2020 2020 7265 7475 726e  ..        return
+00002c90: 206d 7367 0d0a 2020 2020 0d0a 2020 2020   msg..    ..    
+00002ca0: 6465 6620 7772 6974 655f 6835 6164 2873  def write_h5ad(s
+00002cb0: 656c 662c 0d0a 2020 2020 2020 2020 2020  elf,..          
+00002cc0: 2020 2020 2020 2020 2066 696c 656e 616d           filenam
+00002cd0: 653a 2073 7472 293a 0d0a 2020 2020 2020  e: str):..      
+00002ce0: 2020 2222 2257 7269 7465 2064 6174 6173    """Write datas
+00002cf0: 6574 2074 6f20 2e68 3561 6420 6669 6c65  et to .h5ad file
+00002d00: 2e0d 0a0d 0a20 2020 2020 2020 203a 7061  .....        :pa
+00002d10: 7261 6d20 6669 6c65 6e61 6d65 3a20 6669  ram filename: fi
+00002d20: 6c65 7061 7468 0d0a 2020 2020 2020 2020  lepath..        
+00002d30: 3a74 7970 6520 6669 6c65 6e61 6d65 3a20  :type filename: 
+00002d40: 7374 720d 0a20 2020 2020 2020 2022 2222  str..        """
+00002d50: 0d0a 2020 2020 2020 2020 6669 6c65 6e61  ..        filena
+00002d60: 6d65 203d 206f 732e 7061 7468 2e61 6273  me = os.path.abs
+00002d70: 7061 7468 2866 696c 656e 616d 6529 0d0a  path(filename)..
+00002d80: 2020 2020 2020 2020 6c6f 6767 696e 672e          logging.
+00002d90: 696e 666f 2866 2257 7269 7469 6e67 2074  info(f"Writing t
+00002da0: 6f20 7b66 696c 656e 616d 657d 2229 0d0a  o {filename}")..
+00002db0: 2020 2020 2020 2020 7365 6c66 2e61 6461          self.ada
+00002dc0: 7461 2e77 7269 7465 5f68 3561 6428 6669  ta.write_h5ad(fi
+00002dd0: 6c65 6e61 6d65 290d 0a20 2020 2020 2020  lename)..       
+00002de0: 206c 6f67 6769 6e67 2e69 6e66 6f28 6622   logging.info(f"
+00002df0: 446f 6e65 2229 0d0a 2020 2020 0d0a 2020  Done")..    ..  
+00002e00: 2020 6465 6620 746f 5f64 6628 7365 6c66    def to_df(self
+00002e10: 2c0d 0a20 2020 2020 2020 2020 2020 2020  ,..             
+00002e20: 206e 6f72 6d61 6c69 7a65 643a 2062 6f6f   normalized: boo
+00002e30: 6c20 3d20 4661 6c73 6529 3a0d 0a20 2020  l = False):..   
+00002e40: 2020 2020 2022 2222 4765 7420 6461 7461       """Get data
+00002e50: 206d 6174 7269 7820 6173 2061 2060 7064   matrix as a `pd
+00002e60: 2e44 6174 6146 7261 6d65 600d 0a0d 0a20  .DataFrame`.... 
+00002e70: 2020 2020 2020 203a 7061 7261 6d20 6e6f         :param no
+00002e80: 726d 616c 697a 6564 3a20 5365 7420 7472  rmalized: Set tr
+00002e90: 7565 2066 6f72 2054 504d 206e 6f72 6d61  ue for TPM norma
+00002ea0: 6c69 7a65 6420 6f75 7470 7574 2c20 6465  lized output, de
+00002eb0: 6661 756c 7473 2074 6f20 4661 6c73 650d  faults to False.
+00002ec0: 0a20 2020 2020 2020 203a 7479 7065 206e  .        :type n
+00002ed0: 6f72 6d61 6c69 7a65 643a 2062 6f6f 6c2c  ormalized: bool,
+00002ee0: 206f 7074 696f 6e61 6c0d 0a20 2020 2020   optional..     
+00002ef0: 2020 203a 7265 7475 726e 3a20 6f62 7365     :return: obse
+00002f00: 7276 6174 696f 6e73 20c3 9720 7661 7269  rvations .. vari
+00002f10: 6162 6c65 7320 6461 7461 206d 6174 7269  ables data matri
+00002f20: 780d 0a20 2020 2020 2020 203a 7274 7970  x..        :rtyp
+00002f30: 653a 2070 642e 4461 7461 4672 616d 650d  e: pd.DataFrame.
+00002f40: 0a20 2020 2020 2020 2022 2222 0d0a 2020  .        """..  
+00002f50: 2020 2020 2020 6466 203d 2073 656c 662e        df = self.
+00002f60: 6164 6174 612e 746f 5f64 6628 290d 0a20  adata.to_df().. 
+00002f70: 2020 2020 2020 2069 6620 6e6f 726d 616c         if normal
+00002f80: 697a 6564 2061 6e64 206e 6f74 2073 656c  ized and not sel
+00002f90: 662e 6973 5f6e 6f72 6d61 6c69 7a65 643a  f.is_normalized:
+00002fa0: 0d0a 2020 2020 2020 2020 2020 2020 6466  ..            df
+00002fb0: 203d 2064 662e 6469 7628 6466 2e73 756d   = df.div(df.sum
+00002fc0: 2861 7869 733d 3129 2c20 6178 6973 3d30  (axis=1), axis=0
+00002fd0: 2920 2a20 3165 3620 2023 2054 504d 206e  ) * 1e6  # TPM n
+00002fe0: 6f72 6d61 6c69 7a61 7469 6f6e 0d0a 2020  ormalization..  
+00002ff0: 2020 2020 2020 7265 7475 726e 2064 660d        return df.
+00003000: 0a20 2020 2020 2020 200d 0a20 2020 2064  .        ..    d
+00003010: 6566 2072 656d 6f76 655f 756e 6661 6374  ef remove_unfact
+00003020: 6f72 697a 6162 6c65 5f67 656e 6573 2873  orizable_genes(s
+00003030: 656c 6629 3a0d 0a20 2020 2020 2020 2022  elf):..        "
+00003040: 2222 5265 6d6f 7665 7320 6765 6e65 7320  ""Removes genes 
+00003050: 7769 7468 206d 6973 7369 6e67 2076 616c  with missing val
+00003060: 7565 7320 6f72 207a 6572 6f20 7661 7269  ues or zero vari
+00003070: 616e 6365 2066 726f 6d20 7468 6520 6461  ance from the da
+00003080: 7461 206d 6174 7269 782e 0d0a 2020 2020  ta matrix...    
+00003090: 2020 2020 2222 220d 0a20 2020 2020 2020      """..       
+000030a0: 2064 6620 3d20 7365 6c66 2e74 6f5f 6466   df = self.to_df
+000030b0: 286e 6f72 6d61 6c69 7a65 643d 4661 6c73  (normalized=Fals
+000030c0: 6529 0d0a 2020 2020 2020 2020 2320 4368  e)..        # Ch
+000030d0: 6563 6b20 666f 7220 7661 7269 6162 6c65  eck for variable
+000030e0: 7320 7769 7468 206d 6973 7369 6e67 2076  s with missing v
+000030f0: 616c 7565 730d 0a20 2020 2020 2020 2067  alues..        g
+00003100: 656e 6573 5f77 6974 685f 6d69 7373 696e  enes_with_missin
+00003110: 6776 616c 7565 7320 3d20 6466 2e69 736e  gvalues = df.isn
+00003120: 756c 6c28 292e 616e 7928 290d 0a20 2020  ull().any()..   
+00003130: 2020 2020 200d 0a20 2020 2020 2020 2069       ..        i
+00003140: 6620 6765 6e65 735f 7769 7468 5f6d 6973  f genes_with_mis
+00003150: 7369 6e67 7661 6c75 6573 2e61 6e79 2829  singvalues.any()
+00003160: 3a0d 0a20 2020 2020 2020 2020 2020 206e  :..            n
+00003170: 5f6d 6973 7369 6e67 203d 2067 656e 6573  _missing = genes
+00003180: 5f77 6974 685f 6d69 7373 696e 6776 616c  _with_missingval
+00003190: 7565 732e 7375 6d28 290d 0a20 2020 2020  ues.sum()..     
+000031a0: 2020 2020 2020 206c 6f67 6769 6e67 2e77         logging.w
+000031b0: 6172 6e69 6e67 2866 227b 6e5f 6d69 7373  arning(f"{n_miss
+000031c0: 696e 677d 206f 6620 7b73 656c 662e 6164  ing} of {self.ad
+000031d0: 6174 612e 6e5f 7661 7273 7d20 7661 7269  ata.n_vars} vari
+000031e0: 6162 6c65 7320 6172 6520 6d69 7373 696e  ables are missin
+000031f0: 6720 7661 6c75 6573 2028 6061 6461 7461  g values (`adata
+00003200: 2e58 6029 2e22 290d 0a20 2020 2020 2020  .X`).")..       
+00003210: 2020 2020 206c 6f67 6769 6e67 2e77 6172       logging.war
+00003220: 6e69 6e67 2866 2253 7562 7365 7474 696e  ning(f"Subsettin
+00003230: 6720 7661 7269 6162 6c65 7320 746f 2074  g variables to t
+00003240: 686f 7365 2077 6974 6820 6e6f 206d 6973  hose with no mis
+00003250: 7369 6e67 2076 616c 7565 732e 2229 0d0a  sing values.")..
+00003260: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00003270: 0d0a 2020 2020 2020 2020 2320 4368 6563  ..        # Chec
+00003280: 6b20 666f 7220 6765 6e65 7320 7769 7468  k for genes with
+00003290: 207a 6572 6f20 7661 7269 616e 6365 0d0a   zero variance..
+000032a0: 2020 2020 2020 2020 7a65 726f 7661 7267          zerovarg
+000032b0: 656e 6573 203d 2028 6466 2e76 6172 2829  enes = (df.var()
+000032c0: 203d 3d20 3029 0d0a 2020 2020 2020 2020   == 0)..        
+000032d0: 6966 207a 6572 6f76 6172 6765 6e65 732e  if zerovargenes.
+000032e0: 616e 7928 293a 0d0a 2020 2020 2020 2020  any():..        
+000032f0: 2020 2020 6e5f 7a65 726f 7661 7220 3d20      n_zerovar = 
+00003300: 7a65 726f 7661 7267 656e 6573 2e73 756d  zerovargenes.sum
+00003310: 2829 0d0a 2020 2020 2020 2020 2020 2020  ()..            
+00003320: 6c6f 6767 696e 672e 7761 726e 696e 6728  logging.warning(
+00003330: 6622 7b6e 5f7a 6572 6f76 6172 7d20 6f66  f"{n_zerovar} of
+00003340: 207b 7365 6c66 2e61 6461 7461 2e6e 5f76   {self.adata.n_v
+00003350: 6172 737d 2076 6172 6961 626c 6573 2068  ars} variables h
+00003360: 6176 6520 6120 7661 7269 616e 6365 206f  ave a variance o
+00003370: 6620 7a65 726f 2069 6e20 636f 756e 7473  f zero in counts
+00003380: 2064 6174 6120 2860 6164 6174 612e 7261   data (`adata.ra
+00003390: 772e 5860 292e 2229 0d0a 2020 2020 2020  w.X`).")..      
+000033a0: 2020 2020 2020 6c6f 6767 696e 672e 7761        logging.wa
+000033b0: 726e 696e 6728 6622 5375 6273 6574 7469  rning(f"Subsetti
+000033c0: 6e67 2076 6172 6961 626c 6573 2074 6f20  ng variables to 
+000033d0: 7468 6f73 6520 7769 7468 206e 6f6e 7a65  those with nonze
+000033e0: 726f 2076 6172 6961 6e63 652e 2229 0d0a  ro variance.")..
+000033f0: 2020 2020 2020 2020 0d0a 2020 2020 2020          ..      
+00003400: 2020 6765 6e65 735f 746f 5f6b 6565 7020    genes_to_keep 
+00003410: 3d20 287e 6765 6e65 735f 7769 7468 5f6d  = (~genes_with_m
+00003420: 6973 7369 6e67 7661 6c75 6573 2920 2620  issingvalues) & 
+00003430: 287e 7a65 726f 7661 7267 656e 6573 290d  (~zerovargenes).
+00003440: 0a20 2020 2020 2020 200d 0a20 2020 2020  .        ..     
+00003450: 2020 2073 656c 662e 6164 6174 6120 3d20     self.adata = 
+00003460: 7365 6c66 2e61 6461 7461 5b3a 2c67 656e  self.adata[:,gen
+00003470: 6573 5f74 6f5f 6b65 6570 5d0d 0a0d 0a20  es_to_keep].... 
+00003480: 2020 2064 6566 2063 6f6d 7075 7465 5f67     def compute_g
+00003490: 656e 655f 7374 6174 7328 7365 6c66 2c20  ene_stats(self, 
+000034a0: 6f64 675f 6465 6661 756c 745f 7370 6c69  odg_default_spli
+000034b0: 6e65 5f64 6567 7265 653a 2069 6e74 203d  ne_degree: int =
+000034c0: 2033 2c20 6f64 675f 6465 6661 756c 745f   3, odg_default_
+000034d0: 646f 663a 2069 6e74 203d 2038 293a 0d0a  dof: int = 8):..
+000034e0: 2020 2020 2020 2020 2222 220d 0a20 2020          """..   
+000034f0: 2020 2020 2043 6f6d 7075 7465 7320 6765       Computes ge
+00003500: 6e65 2073 7461 7469 7374 6963 7320 616e  ne statistics an
+00003510: 6420 6669 7473 2074 776f 206d 6f64 656c  d fits two model
+00003520: 7320 6f66 206d 6561 6e20 616e 6420 7661  s of mean and va
+00003530: 7269 616e 6365 206f 6620 6765 6e65 7320  riance of genes 
+00003540: 696e 2074 6865 2064 6174 6173 6574 2e20  in the dataset. 
+00003550: 5468 6520 6669 7273 7420 6d65 7468 6f64  The first method
+00003560: 2069 7320 7468 650d 0a20 2020 2020 2020   is the..       
+00003570: 2067 656e 6572 616c 697a 6564 2061 6464   generalized add
+00003580: 6974 6976 6520 6d6f 6465 6c20 7769 7468  itive model with
+00003590: 2073 6d6f 6f74 6820 636f 6d70 6f6e 656e   smooth componen
+000035a0: 7473 2028 422d 7370 6c69 6e65 7329 2074  ts (B-splines) t
+000035b0: 6f20 6d6f 6465 6c20 7468 6520 7265 6c61  o model the rela
+000035c0: 7469 6f6e 7368 6970 206f 6620 6d65 616e  tionship of mean
+000035d0: 2061 6e64 2076 6172 6961 6e63 650d 0a20   and variance.. 
+000035e0: 2020 2020 2020 2062 6574 7765 656e 2067         between g
+000035f0: 656e 6573 2069 6e20 7468 6520 6461 7461  enes in the data
+00003600: 7365 742e 2049 7420 7072 6f64 7563 6573  set. It produces
+00003610: 2061 6e20 6f64 7363 6f72 6520 6d65 7472   an odscore metr
+00003620: 6963 2066 6f72 206f 7665 7264 6973 7065  ic for overdispe
+00003630: 7273 696f 6e2e 2054 6865 2073 6563 6f6e  rsion. The secon
+00003640: 6420 6973 2074 6865 2063 6f75 6e74 2d73  d is the count-s
+00003650: 7461 7469 7374 6963 730d 0a20 2020 2020  tatistics..     
+00003660: 2020 206d 6574 686f 6420 666f 756e 6420     method found 
+00003670: 696e 2074 6865 2063 4e4d 4620 7061 636b  in the cNMF pack
+00003680: 6167 652c 2077 6869 6368 2070 726f 6475  age, which produ
+00003690: 6365 7320 6120 6d6f 6469 6669 6564 2076  ces a modified v
+000036a0: 2d73 636f 7265 206d 6574 7269 632e 2041  -score metric. A
+000036b0: 6c6c 2067 656e 6520 7374 6174 6973 7469  ll gene statisti
+000036c0: 6373 2061 7265 2073 746f 7265 6420 7769  cs are stored wi
+000036d0: 7468 696e 2074 6865 0d0a 2020 2020 2020  thin the..      
+000036e0: 2020 6461 7461 7365 7420 6f62 6a65 6374    dataset object
+000036f0: 2061 6e64 2061 7265 2061 6363 6573 7369   and are accessi
+00003700: 626c 6520 7573 696e 6720 6064 6174 6173  ble using `datas
+00003710: 6574 2e61 6e6e 6461 7461 2e76 6172 602e  et.anndata.var`.
+00003720: 0d0a 0d0a 2020 2020 2020 2020 3a70 6172  ....        :par
+00003730: 616d 206f 6467 5f64 6566 6175 6c74 5f73  am odg_default_s
+00003740: 706c 696e 655f 6465 6772 6565 3a20 422d  pline_degree: B-
+00003750: 5370 6c69 6e65 2064 6567 7265 6520 666f  Spline degree fo
+00003760: 7220 474c 4d2d 4741 4d20 6d6f 6465 6c6c  r GLM-GAM modell
+00003770: 696e 6720 6f66 206d 6561 6e2d 7661 7269  ing of mean-vari
+00003780: 616e 6365 2072 656c 6174 696f 6e73 6869  ance relationshi
+00003790: 702c 2064 6566 6175 6c74 7320 746f 2033  p, defaults to 3
+000037a0: 0d0a 2020 2020 2020 2020 3a74 7970 6520  ..        :type 
+000037b0: 6f64 675f 6465 6661 756c 745f 7370 6c69  odg_default_spli
+000037c0: 6e65 5f64 6567 7265 653a 2069 6e74 2c20  ne_degree: int, 
+000037d0: 6f70 7469 6f6e 616c 0d0a 2020 2020 2020  optional..      
+000037e0: 2020 3a70 6172 616d 206f 6467 5f64 6566    :param odg_def
+000037f0: 6175 6c74 5f64 6f66 3a20 4465 6772 6565  ault_dof: Degree
+00003800: 7320 6f66 2066 7265 6564 6f6d 2066 6f72  s of freedom for
+00003810: 2047 4c4d 2d47 414d 206d 6f64 656c 6c69   GLM-GAM modelli
+00003820: 6e67 206f 6620 6d65 616e 2d76 6172 616e  ng of mean-varan
+00003830: 6365 2c20 6465 6661 756c 7473 2074 6f20  ce, defaults to 
+00003840: 380d 0a20 2020 2020 2020 203a 7479 7065  8..        :type
+00003850: 206f 6467 5f64 6566 6175 6c74 5f64 6f66   odg_default_dof
+00003860: 3a20 696e 742c 206f 7074 696f 6e61 6c0d  : int, optional.
+00003870: 0a20 2020 2020 2020 2022 2222 0d0a 2020  .        """..  
+00003880: 2020 2020 2020 6461 7461 5f72 6177 203d        data_raw =
+00003890: 2073 656c 662e 746f 5f64 6628 290d 0a20   self.to_df().. 
+000038a0: 2020 2020 2020 2064 6174 615f 6e6f 726d         data_norm
+000038b0: 616c 697a 6564 203d 2073 656c 662e 746f  alized = self.to
+000038c0: 5f64 6628 6e6f 726d 616c 697a 6564 3d54  _df(normalized=T
+000038d0: 7275 6529 0d0a 2020 2020 2020 2020 0d0a  rue)..        ..
+000038e0: 2020 2020 2020 2020 2320 6372 6561 7465          # create
+000038f0: 2064 6174 6166 7261 6d65 206f 6620 7065   dataframe of pe
+00003900: 722d 6765 6e65 2073 7461 7469 7374 6963  r-gene statistic
+00003910: 730d 0a20 2020 2020 2020 2073 656c 662e  s..        self.
+00003920: 6164 6174 612e 7661 725b 226d 6561 6e22  adata.var["mean"
+00003930: 5d20 3d20 6461 7461 5f6e 6f72 6d61 6c69  ] = data_normali
+00003940: 7a65 642e 6d65 616e 2829 0d0a 2020 2020  zed.mean()..    
+00003950: 2020 2020 7365 6c66 2e61 6461 7461 2e76      self.adata.v
+00003960: 6172 5b22 7261 6e6b 5f6d 6561 6e22 5d20  ar["rank_mean"] 
+00003970: 3d20 7365 6c66 2e61 6461 7461 2e76 6172  = self.adata.var
+00003980: 5b22 6d65 616e 225d 2e72 616e 6b28 290d  ["mean"].rank().
+00003990: 0a20 2020 2020 2020 2073 656c 662e 6164  .        self.ad
+000039a0: 6174 612e 7661 725b 2276 6172 6961 6e63  ata.var["varianc
+000039b0: 6522 5d20 3d20 6461 7461 5f6e 6f72 6d61  e"] = data_norma
+000039c0: 6c69 7a65 642e 7661 7228 290d 0a20 2020  lized.var()..   
+000039d0: 2020 2020 2073 656c 662e 6164 6174 612e       self.adata.
+000039e0: 7661 725b 2273 6422 5d20 3d20 6461 7461  var["sd"] = data
+000039f0: 5f6e 6f72 6d61 6c69 7a65 642e 7374 6428  _normalized.std(
+00003a00: 290d 0a20 2020 2020 2020 2073 656c 662e  )..        self.
+00003a10: 6164 6174 612e 7661 725b 226d 6973 7369  adata.var["missi
+00003a20: 6e67 6e65 7373 225d 203d 2064 6174 615f  ngness"] = data_
+00003a30: 6e6f 726d 616c 697a 6564 2e69 736e 756c  normalized.isnul
+00003a40: 6c28 292e 7375 6d28 2920 2f20 6461 7461  l().sum() / data
+00003a50: 5f6e 6f72 6d61 6c69 7a65 642e 7368 6170  _normalized.shap
+00003a60: 655b 305d 0d0a 2020 2020 2020 2020 7365  e[0]..        se
+00003a70: 6c66 2e61 6461 7461 2e76 6172 5b5b 226c  lf.adata.var[["l
+00003a80: 6f67 5f6d 6561 6e22 2c20 226c 6f67 5f76  og_mean", "log_v
+00003a90: 6172 6961 6e63 6522 5d5d 203d 206e 702e  ariance"]] = np.
+00003aa0: 6c6f 6731 3028 7365 6c66 2e61 6461 7461  log10(self.adata
+00003ab0: 2e76 6172 5b5b 226d 6561 6e22 2c20 2276  .var[["mean", "v
+00003ac0: 6172 6961 6e63 6522 5d5d 290d 0a20 2020  ariance"]])..   
+00003ad0: 2020 2020 2073 656c 662e 6164 6174 612e       self.adata.
+00003ae0: 7661 725b 226d 6561 6e5f 636f 756e 7473  var["mean_counts
+00003af0: 225d 203d 2064 6174 615f 7261 772e 6d65  "] = data_raw.me
+00003b00: 616e 2829 0d0a 2020 2020 2020 2020 7365  an()..        se
+00003b10: 6c66 2e61 6461 7461 2e76 6172 5b22 6f64  lf.adata.var["od
+00003b20: 7363 6f72 655f 6578 636c 7564 6564 225d  score_excluded"]
+00003b30: 203d 2028 2873 656c 662e 6164 6174 612e   = ((self.adata.
+00003b40: 7661 725b 226d 6973 7369 6e67 6e65 7373  var["missingness
+00003b50: 225d 203e 2030 2920 7c0d 0a20 2020 2020  "] > 0) |..     
+00003b60: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00003b70: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00003b80: 2020 2020 2020 2020 2073 656c 662e 6164           self.ad
+00003b90: 6174 612e 7661 725b 226c 6f67 5f6d 6561  ata.var["log_mea
+00003ba0: 6e22 5d2e 6973 6e75 6c6c 2829 207c 0d0a  n"].isnull() |..
+00003bb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00003bc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00003bd0: 2020 2020 2020 2020 2020 2020 2020 2873                (s
+00003be0: 656c 662e 6164 6174 612e 7661 725b 226d  elf.adata.var["m
+00003bf0: 6561 6e22 5d20 3d3d 2030 2920 7c0d 0a20  ean"] == 0) |.. 
+00003c00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00003c10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00003c20: 2020 2020 2020 2020 2020 2020 2073 656c               sel
+00003c30: 662e 6164 6174 612e 7661 725b 226c 6f67  f.adata.var["log
+00003c40: 5f76 6172 6961 6e63 6522 5d2e 6973 6e75  _variance"].isnu
+00003c50: 6c6c 2829 290d 0a0d 0a20 2020 2020 2020  ll())....       
+00003c60: 2023 206d 6f64 656c 206d 6561 6e2d 7661   # model mean-va
+00003c70: 7269 616e 6365 2072 656c 6174 696f 6e73  riance relations
+00003c80: 6869 7020 7573 696e 6720 6765 6e65 7261  hip using genera
+00003c90: 6c69 7a65 6420 6164 6469 7469 7665 206d  lized additive m
+00003ca0: 6f64 656c 2077 6974 6820 736d 6f6f 7468  odel with smooth
+00003cb0: 2063 6f6d 706f 6e65 6e74 730d 0a20 2020   components..   
+00003cc0: 2020 2020 2064 665f 6d6f 6465 6c20 3d20       df_model = 
+00003cd0: 7365 6c66 2e61 6461 7461 2e76 6172 5b7e  self.adata.var[~
+00003ce0: 7365 6c66 2e61 6461 7461 2e76 6172 5b22  self.adata.var["
+00003cf0: 6f64 7363 6f72 655f 6578 636c 7564 6564  odscore_excluded
+00003d00: 225d 5d0d 0a20 2020 2020 2020 2062 7320  "]]..        bs 
+00003d10: 3d20 4253 706c 696e 6573 2864 665f 6d6f  = BSplines(df_mo
+00003d20: 6465 6c5b 226d 6561 6e22 5d2c 2064 663d  del["mean"], df=
+00003d30: 6f64 675f 6465 6661 756c 745f 646f 662c  odg_default_dof,
+00003d40: 2064 6567 7265 653d 6f64 675f 6465 6661   degree=odg_defa
+00003d50: 756c 745f 7370 6c69 6e65 5f64 6567 7265  ult_spline_degre
+00003d60: 6529 0d0a 2020 2020 2020 2020 6761 6d20  e)..        gam 
+00003d70: 3d20 474c 4d47 616d 2e66 726f 6d5f 666f  = GLMGam.from_fo
+00003d80: 726d 756c 6128 226c 6f67 5f76 6172 6961  rmula("log_varia
+00003d90: 6e63 6520 7e20 6c6f 675f 6d65 616e 222c  nce ~ log_mean",
+00003da0: 2064 6174 613d 6466 5f6d 6f64 656c 2c20   data=df_model, 
+00003db0: 736d 6f6f 7468 6572 3d62 7329 2e66 6974  smoother=bs).fit
+00003dc0: 2829 0d0a 2020 2020 2020 2020 7365 6c66  ()..        self
+00003dd0: 2e61 6461 7461 2e76 6172 5b22 7265 7369  .adata.var["resi
+00003de0: 645f 6c6f 675f 7661 7269 616e 6365 225d  d_log_variance"]
+00003df0: 203d 2067 616d 2e72 6573 6964 5f72 6573   = gam.resid_res
+00003e00: 706f 6e73 650d 0a20 2020 2020 2020 2073  ponse..        s
+00003e10: 656c 662e 6164 6174 612e 7661 725b 226f  elf.adata.var["o
+00003e20: 6473 636f 7265 225d 203d 206e 702e 7371  dscore"] = np.sq
+00003e30: 7274 2831 3020 2a2a 2073 656c 662e 6164  rt(10 ** self.ad
+00003e40: 6174 612e 7661 725b 2272 6573 6964 5f6c  ata.var["resid_l
+00003e50: 6f67 5f76 6172 6961 6e63 6522 5d29 0d0a  og_variance"])..
+00003e60: 2020 2020 2020 2020 7365 6c66 2e61 6461          self.ada
+00003e70: 7461 2e76 6172 5b22 6761 6d5f 6669 7474  ta.var["gam_fitt
+00003e80: 6564 7661 6c75 6573 225d 203d 2067 616d  edvalues"] = gam
+00003e90: 2e66 6974 7465 6476 616c 7565 730d 0a0d  .fittedvalues...
+00003ea0: 0a0d 0a20 2020 2020 2020 2023 206d 6f64  ...        # mod
+00003eb0: 656c 206d 6561 6e2d 7661 7269 616e 6365  el mean-variance
+00003ec0: 2072 656c 6174 696f 6e73 6869 7020 7573   relationship us
+00003ed0: 696e 6720 634e 4d46 2773 206d 6574 686f  ing cNMF's metho
+00003ee0: 6420 6261 7365 6420 6f6e 2076 2d73 636f  d based on v-sco
+00003ef0: 7265 2061 6e64 206d 696e 696d 756d 2065  re and minimum e
+00003f00: 7870 7265 7373 696f 6e20 7468 7265 7368  xpression thresh
+00003f10: 6f6c 640d 0a20 2020 2020 2020 2076 7363  old..        vsc
+00003f20: 6f72 655f 7374 6174 7320 3d20 7064 2e44  ore_stats = pd.D
+00003f30: 6174 6146 7261 6d65 2863 6e6d 662e 6765  ataFrame(cnmf.ge
+00003f40: 745f 6869 6768 7661 725f 6765 6e65 7328  t_highvar_genes(
+00003f50: 696e 7075 745f 636f 756e 7473 3d64 6174  input_counts=dat
+00003f60: 615f 6e6f 726d 616c 697a 6564 2e76 616c  a_normalized.val
+00003f70: 7565 732c 206d 696e 696d 616c 5f6d 6561  ues, minimal_mea
+00003f80: 6e3d 3029 5b30 5d29 0d0a 2020 2020 2020  n=0)[0])..      
+00003f90: 2020 7673 636f 7265 5f73 7461 7473 2e69    vscore_stats.i
+00003fa0: 6e64 6578 203d 2064 6174 615f 6e6f 726d  ndex = data_norm
+00003fb0: 616c 697a 6564 2e63 6f6c 756d 6e73 0d0a  alized.columns..
+00003fc0: 2020 2020 2020 2020 7365 6c66 2e61 6461          self.ada
+00003fd0: 7461 2e76 6172 5b22 7673 636f 7265 225d  ta.var["vscore"]
+00003fe0: 203d 2076 7363 6f72 655f 7374 6174 735b   = vscore_stats[
+00003ff0: 2266 616e 6f5f 7261 7469 6f22 5d0d 0a20  "fano_ratio"].. 
+00004000: 2020 2020 2020 2073 656c 662e 6164 6174         self.adat
+00004010: 612e 756e 735b 226f 6467 225d 5b22 6f64  a.uns["odg"]["od
+00004020: 675f 6465 6661 756c 745f 7370 6c69 6e65  g_default_spline
+00004030: 5f64 6567 7265 6522 5d20 3d20 6f64 675f  _degree"] = odg_
+00004040: 6465 6661 756c 745f 7370 6c69 6e65 5f64  default_spline_d
+00004050: 6567 7265 650d 0a20 2020 2020 2020 2073  egree..        s
+00004060: 656c 662e 6164 6174 612e 756e 735b 226f  elf.adata.uns["o
+00004070: 6467 225d 5b22 6f64 675f 6465 6661 756c  dg"]["odg_defaul
+00004080: 745f 646f 6622 5d20 3d20 6f64 675f 6465  t_dof"] = odg_de
+00004090: 6661 756c 745f 646f 660d 0a20 2020 2020  fault_dof..     
+000040a0: 2020 2073 656c 662e 6170 7065 6e64 5f74     self.append_t
+000040b0: 6f5f 6869 7374 6f72 7928 2247 656e 652d  o_history("Gene-
+000040c0: 6c65 7665 6c20 7374 6174 6973 7469 6373  level statistics
+000040d0: 2061 6e64 206f 7665 7264 6973 7065 7273   and overdispers
+000040e0: 696f 6e20 6d6f 6465 6c6c 696e 6720 636f  ion modelling co
+000040f0: 6d70 6c65 7465 642e 2229 0d0a 2020 2020  mpleted.")..    
+00004100: 2020 2020 0d0a 2020 2020 6465 6620 7365      ..    def se
+00004110: 6c65 6374 5f6f 7665 7264 6973 7065 7273  lect_overdispers
+00004120: 6564 5f67 656e 6573 5f66 726f 6d5f 6765  ed_genes_from_ge
+00004130: 6e65 6c69 7374 2873 656c 662c 2067 656e  nelist(self, gen
+00004140: 6573 3a20 436f 6c6c 6563 7469 6f6e 2c20  es: Collection, 
+00004150: 6d69 6e5f 6d65 616e 3d30 293a 0d0a 2020  min_mean=0):..  
+00004160: 2020 2020 2020 2222 2253 656c 6563 7420        """Select 
+00004170: 6f76 6572 6469 7370 6572 7365 6420 6765  overdispersed ge
+00004180: 6e65 732f 6665 6174 7572 6573 2075 7369  nes/features usi
+00004190: 6e67 2061 2063 7573 746f 6d20 6c69 7374  ng a custom list
+000041a0: 2e20 4765 6e65 732f 6665 6174 7572 6573  . Genes/features
+000041b0: 206e 6f74 2070 7265 7365 6e74 2069 6e20   not present in 
+000041c0: 7468 6520 6461 7461 7365 7420 6172 6520  the dataset are 
+000041d0: 6175 746f 6d61 7469 6361 6c6c 7920 6669  automatically fi
+000041e0: 6c74 6572 6564 206f 7574 2e0d 0a0d 0a20  ltered out..... 
+000041f0: 2020 2020 2020 203a 7061 7261 6d20 6765         :param ge
+00004200: 6e65 733a 2067 656e 6520 6c69 7374 0d0a  nes: gene list..
+00004210: 2020 2020 2020 2020 3a74 7970 6520 6765          :type ge
+00004220: 6e65 733a 2043 6f6c 6c65 6374 696f 6e0d  nes: Collection.
+00004230: 0a20 2020 2020 2020 203a 7061 7261 6d20  .        :param 
+00004240: 6d69 6e5f 6d65 616e 3a20 6d69 6e69 6d75  min_mean: minimu
+00004250: 6d20 6765 6e65 2065 7870 7265 7373 696f  m gene expressio
+00004260: 6e20 666f 7220 6765 6e65 7320 746f 2062  n for genes to b
+00004270: 6520 636f 756e 7465 6420 6173 206f 7665  e counted as ove
+00004280: 7264 6973 7065 7273 6564 2c20 6465 6661  rdispersed, defa
+00004290: 756c 7473 2074 6f20 300d 0a20 2020 2020  ults to 0..     
+000042a0: 2020 203a 7479 7065 206d 696e 5f6d 6561     :type min_mea
+000042b0: 6e3a 2069 6e74 2c20 6f70 7469 6f6e 616c  n: int, optional
+000042c0: 0d0a 2020 2020 2020 2020 2222 220d 0a20  ..        """.. 
+000042d0: 2020 2020 2020 2073 656c 662e 6164 6174         self.adat
+000042e0: 612e 7661 725b 2273 656c 6563 7465 6422  a.var["selected"
+000042f0: 5d20 3d20 7365 6c66 2e61 6461 7461 2e76  ] = self.adata.v
+00004300: 6172 2e69 6e64 6578 2e69 7369 6e28 6765  ar.index.isin(ge
+00004310: 6e65 7329 2026 2073 656c 662e 6164 6174  nes) & self.adat
+00004320: 612e 7661 725b 226d 6561 6e5f 636f 756e  a.var["mean_coun
+00004330: 7473 225d 203e 3d20 6d69 6e5f 6d65 616e  ts"] >= min_mean
+00004340: 0d0a 2020 2020 2020 2020 7365 6c66 2e61  ..        self.a
+00004350: 6461 7461 2e75 6e73 5b22 6f64 6722 5d5b  data.uns["odg"][
+00004360: 226f 7665 7264 6973 7065 7273 696f 6e5f  "overdispersion_
+00004370: 6d65 7472 6963 225d 203d 2022 220d 0a20  metric"] = "".. 
+00004380: 2020 2020 2020 2073 656c 662e 6164 6174         self.adat
+00004390: 612e 756e 735b 226f 6467 225d 5b22 6d69  a.uns["odg"]["mi
+000043a0: 6e5f 6d65 616e 225d 203d 206d 696e 5f6d  n_mean"] = min_m
+000043b0: 6561 6e0d 0a20 2020 2020 2020 2073 656c  ean..        sel
+000043c0: 662e 6164 6174 612e 756e 735b 226f 6467  f.adata.uns["odg
+000043d0: 225d 5b22 6d69 6e5f 7363 6f72 6522 5d20  "]["min_score"] 
+000043e0: 3d20 2222 0d0a 2020 2020 2020 2020 7365  = ""..        se
+000043f0: 6c66 2e61 6461 7461 2e75 6e73 5b22 6f64  lf.adata.uns["od
+00004400: 6722 5d5b 2274 6f70 5f6e 225d 203d 2022  g"]["top_n"] = "
+00004410: 220d 0a20 2020 2020 2020 2073 656c 662e  "..        self.
+00004420: 6164 6174 612e 756e 735b 226f 6467 225d  adata.uns["odg"]
+00004430: 5b22 7175 616e 7469 6c65 225d 203d 2022  ["quantile"] = "
+00004440: 220d 0a20 2020 2020 2020 2073 656c 662e  "..        self.
+00004450: 6170 7065 6e64 5f74 6f5f 6869 7374 6f72  append_to_histor
+00004460: 7928 224f 7665 7264 6973 7065 7273 6564  y("Overdispersed
+00004470: 2067 656e 6573 2073 656c 6563 7465 6420   genes selected 
+00004480: 6672 6f6d 2063 7573 746f 6d20 6765 6e65  from custom gene
+00004490: 206c 6973 7422 290d 0a20 2020 2020 2020   list")..       
+000044a0: 200d 0a20 2020 2064 6566 2073 656c 6563   ..    def selec
+000044b0: 745f 6f76 6572 6469 7370 6572 7365 645f  t_overdispersed_
+000044c0: 6765 6e65 7328 7365 6c66 2c0d 0a20 2020  genes(self,..   
+000044d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000044e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000044f0: 6f76 6572 6469 7370 6572 7369 6f6e 5f6d  overdispersion_m
+00004500: 6574 7269 633a 2073 7472 203d 2022 6f64  etric: str = "od
+00004510: 7363 6f72 6522 2c0d 0a20 2020 2020 2020  score",..       
+00004520: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004530: 2020 2020 2020 2020 2020 2020 6d69 6e5f              min_
+00004540: 6d65 616e 3a20 666c 6f61 7420 3d20 302c  mean: float = 0,
+00004550: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
+00004560: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004570: 2020 2020 206d 696e 5f73 636f 7265 3a20       min_score: 
+00004580: 666c 6f61 7420 3d20 312e 302c 0d0a 2020  float = 1.0,..  
+00004590: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000045a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000045b0: 2074 6f70 5f6e 3a20 696e 7420 3d20 4e6f   top_n: int = No
+000045c0: 6e65 2c0d 0a20 2020 2020 2020 2020 2020  ne,..           
+000045d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000045e0: 2020 2020 2020 2020 7175 616e 7469 6c65          quantile
+000045f0: 3a20 666c 6f61 7420 3d20 4e6f 6e65 293a  : float = None):
+00004600: 0d0a 2020 2020 2020 2020 2222 2253 656c  ..        """Sel
+00004610: 6563 7420 6f76 6572 6469 7370 6572 7365  ect overdisperse
+00004620: 6420 6765 6e65 732f 6665 6174 7572 6573  d genes/features
+00004630: 2075 7369 6e67 2061 6e20 6f76 6572 6469   using an overdi
+00004640: 7370 6572 7369 6f6e 206d 6574 7269 632e  spersion metric.
+00004650: 204f 7074 696f 6e61 6c6c 7920 7365 7420   Optionally set 
+00004660: 6120 6d69 6e69 6d75 6d20 6765 6e65 2065  a minimum gene e
+00004670: 7870 7265 7373 696f 6e20 6c65 7665 6c2e  xpression level.
+00004680: 0d0a 2020 2020 2020 2020 5365 7420 6120  ..        Set a 
+00004690: 7468 7265 7368 6f6c 6420 7573 696e 6720  threshold using 
+000046a0: 7468 6520 746f 7020 4e20 2827 746f 705f  the top N ('top_
+000046b0: 6e27 292c 206d 696e 696d 756d 2073 636f  n'), minimum sco
+000046c0: 7265 2028 276d 696e 5f73 636f 7265 2729  re ('min_score')
+000046d0: 2c20 6f72 2070 726f 706f 7274 696f 6e20  , or proportion 
+000046e0: 6f66 2066 6561 7475 7265 7320 2827 7175  of features ('qu
+000046f0: 616e 7469 6c65 2729 206d 6574 686f 6473  antile') methods
+00004700: 2e0d 0a20 2020 2020 2020 204f 7665 7264  ...        Overd
+00004710: 6973 7065 7273 6564 2067 656e 6520 6c69  ispersed gene li
+00004720: 7374 2069 7320 7361 7665 6420 696e 2074  st is saved in t
+00004730: 6865 2044 6174 6173 6574 206f 626a 6563  he Dataset objec
+00004740: 742e 0d0a 0d0a 2020 2020 2020 2020 3a70  t.....        :p
+00004750: 6172 616d 206f 7665 7264 6973 7065 7273  aram overdispers
+00004760: 696f 6e5f 6d65 7472 6963 3a20 226f 6473  ion_metric: "ods
+00004770: 636f 7265 2220 6f72 2022 7673 636f 7265  core" or "vscore
+00004780: 222c 2064 6566 6175 6c74 7320 746f 2022  ", defaults to "
+00004790: 6f64 7363 6f72 6522 0d0a 2020 2020 2020  odscore"..      
+000047a0: 2020 3a74 7970 6520 6f76 6572 6469 7370    :type overdisp
+000047b0: 6572 7369 6f6e 5f6d 6574 7269 633a 2073  ersion_metric: s
+000047c0: 7472 2c20 6f70 7469 6f6e 616c 0d0a 2020  tr, optional..  
+000047d0: 2020 2020 2020 3a70 6172 616d 206d 696e        :param min
+000047e0: 5f6d 6561 6e3a 206d 696e 696d 756d 2067  _mean: minimum g
+000047f0: 656e 6520 6578 7072 6573 7369 6f6e 2066  ene expression f
+00004800: 6f72 2067 656e 6573 2074 6f20 6265 2063  or genes to be c
+00004810: 6f75 6e74 6564 2061 7320 6f76 6572 6469  ounted as overdi
+00004820: 7370 6572 7365 642c 2064 6566 6175 6c74  spersed, default
+00004830: 7320 746f 2030 0d0a 2020 2020 2020 2020  s to 0..        
+00004840: 3a74 7970 6520 6d69 6e5f 6d65 616e 3a20  :type min_mean: 
+00004850: 696e 742c 206f 7074 696f 6e61 6c0d 0a20  int, optional.. 
+00004860: 2020 2020 2020 203a 7061 7261 6d20 6d69         :param mi
+00004870: 6e5f 7363 6f72 653a 206d 696e 696d 756d  n_score: minimum
+00004880: 2073 636f 7265 2066 6f72 206f 7665 7264   score for overd
+00004890: 6973 7065 7273 696f 6e2c 2064 6566 6175  ispersion, defau
+000048a0: 6c74 7320 746f 2031 2e30 0d0a 2020 2020  lts to 1.0..    
+000048b0: 2020 2020 3a74 7970 6520 6d69 6e5f 7363      :type min_sc
+000048c0: 6f72 653a 2066 6c6f 6174 2c20 6f70 7469  ore: float, opti
+000048d0: 6f6e 616c 0d0a 2020 2020 2020 2020 3a70  onal..        :p
+000048e0: 6172 616d 2074 6f70 5f6e 3a20 4368 6f6f  aram top_n: Choo
+000048f0: 7365 2074 6865 2074 6f70 204e 206d 6f73  se the top N mos
+00004900: 7420 6f76 6572 6469 7370 6572 7365 6420  t overdispersed 
+00004910: 6765 6e65 732c 2064 6566 6175 6c74 7320  genes, defaults 
+00004920: 746f 204e 6f6e 650d 0a20 2020 2020 2020  to None..       
+00004930: 203a 7479 7065 2074 6f70 5f6e 3a20 696e   :type top_n: in
+00004940: 742c 206f 7074 696f 6e61 6c0d 0a20 2020  t, optional..   
+00004950: 2020 2020 203a 7061 7261 6d20 7175 616e       :param quan
+00004960: 7469 6c65 3a20 4368 6f6f 7365 2061 2071  tile: Choose a q
+00004970: 7561 6e74 696c 6520 6f66 206f 7665 7264  uantile of overd
+00004980: 6973 7065 7273 696f 6e2e 2046 6f72 2065  ispersion. For e
+00004990: 7861 6d70 6c65 2c20 7468 6520 746f 7020  xample, the top 
+000049a0: 3130 2520 6f66 206f 7665 7264 6973 7065  10% of overdispe
+000049b0: 7273 6564 2067 656e 6573 2077 6f75 6c64  rsed genes would
+000049c0: 2062 6520 302e 3130 2e20 4465 6661 756c   be 0.10. Defaul
+000049d0: 7473 2074 6f20 4e6f 6e65 0d0a 2020 2020  ts to None..    
+000049e0: 2020 2020 3a74 7970 6520 7175 616e 7469      :type quanti
+000049f0: 6c65 3a20 666c 6f61 742c 206f 7074 696f  le: float, optio
+00004a00: 6e61 6c0d 0a20 2020 2020 2020 203a 7261  nal..        :ra
+00004a10: 6973 6573 2056 616c 7565 4572 726f 723a  ises ValueError:
+00004a20: 2045 7272 6f72 2069 6620 696e 7661 6c69   Error if invali
+00004a30: 6420 6f76 6572 6469 7370 6572 7369 6f6e  d overdispersion
+00004a40: 206d 6574 7269 6320 6973 2063 686f 7365   metric is chose
+00004a50: 6e2e 0d0a 2020 2020 2020 2020 2222 220d  n...        """.
+00004a60: 0a20 2020 2020 2020 200d 0a20 2020 2020  .        ..     
+00004a70: 2020 2069 6620 6f76 6572 6469 7370 6572     if overdisper
+00004a80: 7369 6f6e 5f6d 6574 7269 6320 6e6f 7420  sion_metric not 
+00004a90: 696e 2073 656c 662e 6164 6174 612e 7661  in self.adata.va
+00004aa0: 722e 636f 6c75 6d6e 733a 0d0a 2020 2020  r.columns:..    
+00004ab0: 2020 2020 2020 2020 6966 206f 7665 7264          if overd
+00004ac0: 6973 7065 7273 696f 6e5f 6d65 7472 6963  ispersion_metric
+00004ad0: 2069 6e20 2822 6f64 7363 6f72 6522 2c20   in ("odscore", 
+00004ae0: 2276 7363 6f72 6522 293a 0d0a 2020 2020  "vscore"):..    
+00004af0: 2020 2020 2020 2020 2020 2020 7261 6973              rais
+00004b00: 6520 5661 6c75 6545 7272 6f72 280d 0a20  e ValueError(.. 
+00004b10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004b20: 2020 2066 227b 6f76 6572 6469 7370 6572     f"{overdisper
+00004b30: 7369 6f6e 5f6d 6574 7269 637d 2068 6173  sion_metric} has
+00004b40: 206e 6f74 2062 6565 6e20 6361 6c63 756c   not been calcul
+00004b50: 6174 6564 2066 6f72 2074 6869 7320 6461  ated for this da
+00004b60: 7461 7365 742e 2022 0d0a 2020 2020 2020  taset. "..      
+00004b70: 2020 2020 2020 2020 2020 2020 2020 2245                "E
+00004b80: 6e73 7572 6520 7468 6174 2079 6f75 2063  nsure that you c
+00004b90: 616c 6c20 7468 6520 6044 6174 6173 6574  all the `Dataset
+00004ba0: 2e63 6f6d 7075 7465 5f67 656e 655f 7374  .compute_gene_st
+00004bb0: 6174 7328 2960 2066 6972 7374 2e22 0d0a  ats()` first."..
+00004bc0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004bd0: 2020 2020 290d 0a20 2020 2020 2020 2020      )..         
+00004be0: 2020 2065 6c73 653a 0d0a 2020 2020 2020     else:..      
+00004bf0: 2020 2020 2020 2020 2020 7261 6973 6520            raise 
+00004c00: 5661 6c75 6545 7272 6f72 280d 0a20 2020  ValueError(..   
+00004c10: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004c20: 2066 227b 6f76 6572 6469 7370 6572 7369   f"{overdispersi
+00004c30: 6f6e 5f6d 6574 7269 637d 2069 7320 6e6f  on_metric} is no
+00004c40: 7420 6120 7661 6c69 6420 6f76 6572 6469  t a valid overdi
+00004c50: 7370 6572 7369 6f6e 206d 6574 7269 632e  spersion metric.
+00004c60: 220d 0a20 2020 2020 2020 2020 2020 2020  "..             
+00004c70: 2020 2020 2020 2029 0d0a 2020 2020 2020         )..      
+00004c80: 2020 0d0a 2020 2020 2020 2020 2320 7761    ..        # wa
+00004c90: 726e 2069 6620 6d75 6c74 6970 6c65 206d  rn if multiple m
+00004ca0: 6574 686f 6473 2061 7265 2073 656c 6563  ethods are selec
+00004cb0: 7465 640d 0a20 2020 2020 2020 2073 656c  ted..        sel
+00004cc0: 6563 7465 645f 6d65 7468 6f64 7320 3d20  ected_methods = 
+00004cd0: 5b5d 0d0a 2020 2020 2020 2020 6966 206d  []..        if m
+00004ce0: 696e 5f73 636f 7265 2069 7320 6e6f 7420  in_score is not 
+00004cf0: 4e6f 6e65 3a0d 0a20 2020 2020 2020 2020  None:..         
+00004d00: 2020 2073 656c 6563 7465 645f 6d65 7468     selected_meth
+00004d10: 6f64 732e 6170 7065 6e64 2822 6d69 6e5f  ods.append("min_
+00004d20: 7363 6f72 6522 290d 0a20 2020 2020 2020  score")..       
+00004d30: 2069 6620 746f 705f 6e20 6973 206e 6f74   if top_n is not
+00004d40: 204e 6f6e 653a 0d0a 2020 2020 2020 2020   None:..        
+00004d50: 2020 2020 7365 6c65 6374 6564 5f6d 6574      selected_met
+00004d60: 686f 6473 2e61 7070 656e 6428 2274 6f70  hods.append("top
+00004d70: 5f6e 2229 0d0a 2020 2020 2020 2020 6966  _n")..        if
+00004d80: 2071 7561 6e74 696c 6520 6973 206e 6f74   quantile is not
+00004d90: 204e 6f6e 653a 0d0a 2020 2020 2020 2020   None:..        
+00004da0: 2020 2020 7365 6c65 6374 6564 5f6d 6574      selected_met
+00004db0: 686f 6473 2e61 7070 656e 6428 2271 7561  hods.append("qua
+00004dc0: 6e74 696c 6522 290d 0a20 2020 2020 2020  ntile")..       
+00004dd0: 2069 6620 6c65 6e28 7365 6c65 6374 6564   if len(selected
+00004de0: 5f6d 6574 686f 6473 2920 3e20 313a 0d0a  _methods) > 1:..
+00004df0: 2020 2020 2020 2020 2020 2020 6d65 7468              meth
+00004e00: 6f64 7761 726e 7374 7220 3d20 222c 2022  odwarnstr = ", "
+00004e10: 2e6a 6f69 6e28 7365 6c65 6374 6564 5f6d  .join(selected_m
+00004e20: 6574 686f 6473 290d 0a20 2020 2020 2020  ethods)..       
+00004e30: 2020 2020 206c 6f67 6769 6e67 2e77 6172       logging.war
+00004e40: 6e69 6e67 2866 224d 756c 7469 706c 6520  ning(f"Multiple 
+00004e50: 636f 6e66 6c69 6374 696e 6720 6f76 6572  conflicting over
+00004e60: 6469 7370 6572 7365 6420 6765 6e65 2073  dispersed gene s
+00004e70: 656c 6563 7469 6f6e 2063 7269 7465 7269  election criteri
+00004e80: 6120 6861 7665 2062 6565 6e20 7365 6c65  a have been sele
+00004e90: 6374 6564 3a20 7b6d 6574 686f 6477 6172  cted: {methodwar
+00004ea0: 6e73 7472 7d2e 2022 0d0a 2020 2020 2020  nstr}. "..      
+00004eb0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00004ec0: 2020 2020 2020 224f 6e6c 7920 7468 6520        "Only the 
+00004ed0: 696e 7465 7273 6563 7469 6f6e 206f 6620  intersection of 
+00004ee0: 7468 6573 6520 6d65 7468 6f64 7320 7769  these methods wi
+00004ef0: 6c6c 2062 6520 7365 6c65 6374 6564 2e22  ll be selected."
+00004f00: 290d 0a20 2020 2020 2020 2020 2020 2020  )..             
+00004f10: 2020 200d 0a20 2020 2020 2020 2023 206d     ..        # m
+00004f20: 696e 5f6d 6561 6e20 6669 6c74 6572 0d0a  in_mean filter..
+00004f30: 2020 2020 2020 2020 7365 6c65 6374 6564          selected
+00004f40: 5f67 656e 6573 203d 2073 656c 662e 6164  _genes = self.ad
+00004f50: 6174 612e 7661 725b 226d 6561 6e5f 636f  ata.var["mean_co
+00004f60: 756e 7473 225d 203e 3d20 6d69 6e5f 6d65  unts"] >= min_me
+00004f70: 616e 0d0a 2020 2020 2020 2020 2320 6d69  an..        # mi
+00004f80: 6e5f 7363 6f72 6520 6669 6c74 6572 0d0a  n_score filter..
+00004f90: 2020 2020 2020 2020 6966 206d 696e 5f73          if min_s
+00004fa0: 636f 7265 2069 7320 6e6f 7420 4e6f 6e65  core is not None
+00004fb0: 3a0d 0a20 2020 2020 2020 2020 2020 2073  :..            s
+00004fc0: 656c 6563 7465 645f 6765 6e65 7320 3d20  elected_genes = 
+00004fd0: 7365 6c65 6374 6564 5f67 656e 6573 2026  selected_genes &
+00004fe0: 2028 7365 6c66 2e61 6461 7461 2e76 6172   (self.adata.var
+00004ff0: 5b6f 7665 7264 6973 7065 7273 696f 6e5f  [overdispersion_
+00005000: 6d65 7472 6963 5d20 3e3d 206d 696e 5f73  metric] >= min_s
+00005010: 636f 7265 290d 0a20 2020 2020 2020 2023  core)..        #
+00005020: 2074 6f70 5f6e 2066 696c 7465 720d 0a20   top_n filter.. 
+00005030: 2020 2020 2020 2069 6620 746f 705f 6e20         if top_n 
+00005040: 6973 206e 6f74 204e 6f6e 653a 0d0a 2020  is not None:..  
+00005050: 2020 2020 2020 2020 2020 6765 6e65 7320            genes 
+00005060: 3d20 7365 6c66 2e61 6461 7461 2e76 6172  = self.adata.var
+00005070: 5b6f 7665 7264 6973 7065 7273 696f 6e5f  [overdispersion_
+00005080: 6d65 7472 6963 5d2e 736f 7274 5f76 616c  metric].sort_val
+00005090: 7565 7328 6173 6365 6e64 696e 673d 4661  ues(ascending=Fa
+000050a0: 6c73 6529 2e68 6561 6428 696e 7428 746f  lse).head(int(to
+000050b0: 705f 6e29 292e 696e 6465 780d 0a20 2020  p_n)).index..   
+000050c0: 2020 2020 2020 2020 2073 656c 6563 7465           selecte
+000050d0: 645f 6765 6e65 7320 3d20 7365 6c65 6374  d_genes = select
+000050e0: 6564 5f67 656e 6573 2026 2073 656c 662e  ed_genes & self.
+000050f0: 6164 6174 612e 7661 722e 696e 6465 782e  adata.var.index.
+00005100: 6973 696e 2867 656e 6573 290d 0a20 2020  isin(genes)..   
+00005110: 2020 2020 2023 2071 7561 6e74 696c 6520       # quantile 
+00005120: 6669 6c74 6572 0d0a 2020 2020 2020 2020  filter..        
+00005130: 6966 2071 7561 6e74 696c 6520 6973 206e  if quantile is n
+00005140: 6f74 204e 6f6e 653a 0d0a 2020 2020 2020  ot None:..      
+00005150: 2020 2020 2020 6e5f 746f 7461 6c5f 6765        n_total_ge
+00005160: 6e65 7320 3d20 7365 6c66 2e61 6461 7461  nes = self.adata
+00005170: 2e76 6172 5b6f 7665 7264 6973 7065 7273  .var[overdispers
+00005180: 696f 6e5f 6d65 7472 6963 5d2e 6e6f 746e  ion_metric].notn
+00005190: 756c 6c28 292e 7375 6d28 290d 0a20 2020  ull().sum()..   
+000051a0: 2020 2020 2020 2020 2067 656e 6573 203d           genes =
+000051b0: 2073 656c 662e 6164 6174 612e 7661 725b   self.adata.var[
+000051c0: 6f76 6572 6469 7370 6572 7369 6f6e 5f6d  overdispersion_m
+000051d0: 6574 7269 635d 2e73 6f72 745f 7661 6c75  etric].sort_valu
+000051e0: 6573 2861 7363 656e 6469 6e67 3d46 616c  es(ascending=Fal
+000051f0: 7365 292e 6865 6164 2869 6e74 2871 7561  se).head(int(qua
+00005200: 6e74 696c 6520 2a20 6e5f 746f 7461 6c5f  ntile * n_total_
+00005210: 6765 6e65 7329 292e 696e 6465 780d 0a20  genes)).index.. 
+00005220: 2020 2020 2020 2020 2020 2073 656c 6563             selec
+00005230: 7465 645f 6765 6e65 7320 3d20 7365 6c65  ted_genes = sele
+00005240: 6374 6564 5f67 656e 6573 2026 2073 656c  cted_genes & sel
+00005250: 662e 6164 6174 612e 7661 722e 696e 6465  f.adata.var.inde
+00005260: 782e 6973 696e 2867 656e 6573 290d 0a0d  x.isin(genes)...
+00005270: 0a20 2020 2020 2020 206e 5f73 656c 6563  .        n_selec
+00005280: 7465 645f 6765 6e65 7320 3d20 7365 6c65  ted_genes = sele
+00005290: 6374 6564 5f67 656e 6573 2e73 756d 2829  cted_genes.sum()
+000052a0: 0d0a 2020 2020 2020 2020 6c6f 6767 696e  ..        loggin
+000052b0: 672e 696e 666f 2866 227b 6e5f 7365 6c65  g.info(f"{n_sele
+000052c0: 6374 6564 5f67 656e 6573 7d20 6765 6e65  cted_genes} gene
+000052d0: 7320 7365 6c65 6374 6564 2066 6f72 2066  s selected for f
+000052e0: 6163 746f 7269 7a61 7469 6f6e 2229 0d0a  actorization")..
+000052f0: 2020 2020 2020 2020 0d0a 2020 2020 2020          ..      
+00005300: 2020 2320 6d61 6b65 2063 6861 6e67 6573    # make changes
+00005310: 2074 6f20 4461 7461 7365 7420 6f62 6a65   to Dataset obje
+00005320: 6374 0d0a 2020 2020 2020 2020 7365 6c66  ct..        self
+00005330: 2e61 6461 7461 2e76 6172 5b22 7365 6c65  .adata.var["sele
+00005340: 6374 6564 225d 203d 2073 656c 6563 7465  cted"] = selecte
+00005350: 645f 6765 6e65 730d 0a20 2020 2020 2020  d_genes..       
+00005360: 2073 656c 662e 6164 6174 612e 756e 735b   self.adata.uns[
+00005370: 226f 6467 225d 5b22 6f76 6572 6469 7370  "odg"]["overdisp
+00005380: 6572 7369 6f6e 5f6d 6574 7269 6322 5d20  ersion_metric"] 
+00005390: 3d20 6f76 6572 6469 7370 6572 7369 6f6e  = overdispersion
+000053a0: 5f6d 6574 7269 630d 0a20 2020 2020 2020  _metric..       
+000053b0: 2073 656c 662e 6164 6174 612e 756e 735b   self.adata.uns[
+000053c0: 226f 6467 225d 5b22 6d69 6e5f 6d65 616e  "odg"]["min_mean
+000053d0: 225d 203d 206d 696e 5f6d 6561 6e0d 0a20  "] = min_mean.. 
+000053e0: 2020 2020 2020 2073 656c 662e 6164 6174         self.adat
+000053f0: 612e 756e 735b 226f 6467 225d 5b22 6d69  a.uns["odg"]["mi
+00005400: 6e5f 7363 6f72 6522 5d20 3d20 6d69 6e5f  n_score"] = min_
+00005410: 7363 6f72 6520 6966 206d 696e 5f73 636f  score if min_sco
+00005420: 7265 2069 7320 6e6f 7420 4e6f 6e65 2065  re is not None e
+00005430: 6c73 6520 2222 0d0a 2020 2020 2020 2020  lse ""..        
+00005440: 7365 6c66 2e61 6461 7461 2e75 6e73 5b22  self.adata.uns["
+00005450: 6f64 6722 5d5b 2274 6f70 5f6e 225d 203d  odg"]["top_n"] =
+00005460: 2074 6f70 5f6e 2069 6620 746f 705f 6e20   top_n if top_n 
+00005470: 6973 206e 6f74 204e 6f6e 6520 656c 7365  is not None else
+00005480: 2022 220d 0a20 2020 2020 2020 2073 656c   ""..        sel
+00005490: 662e 6164 6174 612e 756e 735b 226f 6467  f.adata.uns["odg
+000054a0: 225d 5b22 7175 616e 7469 6c65 225d 203d  "]["quantile"] =
+000054b0: 2071 7561 6e74 696c 6520 6966 2071 7561   quantile if qua
+000054c0: 6e74 696c 6520 6973 206e 6f74 204e 6f6e  ntile is not Non
+000054d0: 6520 656c 7365 2022 220d 0a20 2020 2020  e else ""..     
+000054e0: 2020 2073 656c 662e 6170 7065 6e64 5f74     self.append_t
+000054f0: 6f5f 6869 7374 6f72 7928 224f 7665 7264  o_history("Overd
+00005500: 6973 7065 7273 6564 2067 656e 6573 2073  ispersed genes s
+00005510: 656c 6563 7465 6422 290d 0a20 2020 200d  elected")..    .
+00005520: 0a20 2020 2064 6566 2069 6e69 7469 616c  .    def initial
+00005530: 697a 655f 636e 6d66 2873 656c 662c 2063  ize_cnmf(self, c
+00005540: 6e6d 665f 6f75 7470 7574 5f64 6972 3a20  nmf_output_dir: 
+00005550: 7374 722c 0d0a 2020 2020 2020 2020 2020  str,..          
+00005560: 2020 2020 2020 2020 2020 2020 2020 636e                cn
+00005570: 6d66 5f6e 616d 653a 2073 7472 2c0d 0a20  mf_name: str,.. 
+00005580: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00005590: 2020 2020 2020 206b 7661 6c73 3a20 436f         kvals: Co
+000055a0: 6c6c 6563 7469 6f6e 203d 2072 616e 6765  llection = range
+000055b0: 2832 2c20 3631 292c 0d0a 2020 2020 2020  (2, 61),..      
+000055c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000055d0: 2020 6e5f 6974 6572 3a20 696e 7420 3d20    n_iter: int = 
+000055e0: 3230 302c 0d0a 2020 2020 2020 2020 2020  200,..          
+000055f0: 2020 2020 2020 2020 2020 2020 2020 6265                be
+00005600: 7461 5f6c 6f73 733a 2073 7472 203d 2022  ta_loss: str = "
+00005610: 6b75 6c6c 6261 636b 2d6c 6569 626c 6572  kullback-leibler
+00005620: 222c 0d0a 2020 2020 2020 2020 2020 2020  ",..            
+00005630: 2020 2020 2020 2020 2020 2020 7365 6564              seed
+00005640: 3a20 4f70 7469 6f6e 616c 5b69 6e74 5d20  : Optional[int] 
+00005650: 3d20 4e6f 6e65 2920 2d3e 2063 6e6d 662e  = None) -> cnmf.
+00005660: 634e 4d46 3a0d 0a20 2020 2020 2020 2022  cNMF:..        "
+00005670: 2222 496e 6974 6961 6c69 7a65 2061 2063  ""Initialize a c
+00005680: 4e4d 4620 7275 6e20 666f 7220 7375 6273  NMF run for subs
+00005690: 6571 7565 6e74 2066 6163 746f 7269 7a61  equent factoriza
+000056a0: 7469 6f6e 2e0d 0a0d 0a20 2020 2020 2020  tion.....       
+000056b0: 203a 7061 7261 6d20 636e 6d66 5f6f 7574   :param cnmf_out
+000056c0: 7075 745f 6469 723a 204f 7574 7075 7420  put_dir: Output 
+000056d0: 6469 7265 6374 6f72 7920 666f 7220 634e  directory for cN
+000056e0: 4d46 2072 6573 756c 7473 0d0a 2020 2020  MF results..    
+000056f0: 2020 2020 3a74 7970 6520 636e 6d66 5f6f      :type cnmf_o
+00005700: 7574 7075 745f 6469 723a 2073 7472 0d0a  utput_dir: str..
+00005710: 2020 2020 2020 2020 3a70 6172 616d 2063          :param c
+00005720: 6e6d 665f 6e61 6d65 3a20 4e61 6d65 206f  nmf_name: Name o
+00005730: 6620 7468 6520 634e 4d46 2072 6573 756c  f the cNMF resul
+00005740: 7473 2e20 4669 6c65 7320 7769 6c6c 2062  ts. Files will b
+00005750: 6520 6f75 7470 7574 2074 6f20 5b63 6e6d  e output to [cnm
+00005760: 665f 6f75 7470 7574 5f64 6972 5d2f 5b63  f_output_dir]/[c
+00005770: 6e6d 665f 6e61 6d65 5d2f 0d0a 2020 2020  nmf_name]/..    
+00005780: 2020 2020 3a74 7970 6520 636e 6d66 5f6e      :type cnmf_n
+00005790: 616d 653a 2073 7472 0d0a 2020 2020 2020  ame: str..      
+000057a0: 2020 3a70 6172 616d 206b 7661 6c73 3a20    :param kvals: 
+000057b0: 5261 6e6b 7320 666f 7220 634e 4d46 2066  Ranks for cNMF f
+000057c0: 6163 746f 7269 7a61 7469 6f6e 2c20 6465  actorization, de
+000057d0: 6661 756c 7473 2074 6f20 7261 6e67 6528  faults to range(
+000057e0: 322c 2036 3129 0d0a 2020 2020 2020 2020  2, 61)..        
+000057f0: 3a74 7970 6520 6b76 616c 733a 2043 6f6c  :type kvals: Col
+00005800: 6c65 6374 696f 6e2c 206f 7074 696f 6e61  lection, optiona
+00005810: 6c0d 0a20 2020 2020 2020 203a 7061 7261  l..        :para
+00005820: 6d20 6e5f 6974 6572 3a20 4e75 6d62 6572  m n_iter: Number
+00005830: 206f 6620 6974 6572 6174 696f 6e73 2066   of iterations f
+00005840: 726f 6d20 7768 6963 6820 746f 2062 7569  rom which to bui
+00005850: 6c64 2061 2063 6f6e 7365 6e73 7573 2073  ld a consensus s
+00005860: 6f6c 7574 696f 6e2c 2064 6566 6175 6c74  olution, default
+00005870: 7320 746f 2032 3030 0d0a 2020 2020 2020  s to 200..      
+00005880: 2020 3a74 7970 6520 6e5f 6974 6572 3a20    :type n_iter: 
+00005890: 696e 742c 206f 7074 696f 6e61 6c0d 0a20  int, optional.. 
+000058a0: 2020 2020 2020 203a 7061 7261 6d20 6265         :param be
+000058b0: 7461 5f6c 6f73 733a 2062 6574 612d 6c6f  ta_loss: beta-lo
+000058c0: 7373 2066 756e 6374 696f 6e2c 2065 6974  ss function, eit
+000058d0: 6865 7220 226b 756c 6c62 6163 6b2d 6c65  her "kullback-le
+000058e0: 6962 6c65 7222 206f 7220 2266 726f 6265  ibler" or "frobe
+000058f0: 6e69 7573 222e 2044 6566 6175 6c74 7320  nius". Defaults 
+00005900: 746f 2022 6b75 6c6c 6261 636b 2d6c 6569  to "kullback-lei
+00005910: 626c 6572 220d 0a20 2020 2020 2020 203a  bler"..        :
+00005920: 7479 7065 2062 6574 615f 6c6f 7373 3a20  type beta_loss: 
+00005930: 7374 722c 206f 7074 696f 6e61 6c0d 0a20  str, optional.. 
+00005940: 2020 2020 2020 203a 7061 7261 6d20 7365         :param se
+00005950: 6564 3a20 5261 6e64 6f6d 2073 6565 6420  ed: Random seed 
+00005960: 666f 7220 7265 7072 6f64 7563 6962 696c  for reproducibil
+00005970: 6974 792c 2064 6566 6175 6c74 7320 746f  ity, defaults to
+00005980: 204e 6f6e 650d 0a20 2020 2020 2020 203a   None..        :
+00005990: 7479 7065 2073 6565 643a 204f 7074 696f  type seed: Optio
+000059a0: 6e61 6c5b 696e 745d 2c20 6f70 7469 6f6e  nal[int], option
+000059b0: 616c 0d0a 2020 2020 2020 2020 3a72 6574  al..        :ret
+000059c0: 7572 6e3a 2063 4e4d 4620 6f62 6a65 6374  urn: cNMF object
+000059d0: 0d0a 2020 2020 2020 2020 3a72 7479 7065  ..        :rtype
+000059e0: 3a20 3a63 6c61 7373 3a60 636e 6d66 736e  : :class:`cnmfsn
+000059f0: 732e 636e 6d66 2e63 4e4d 4660 0d0a 2020  s.cnmf.cNMF`..  
+00005a00: 2020 2020 2020 2222 220d 0a20 2020 2020        """..     
+00005a10: 2020 2063 6e6d 665f 6f62 6a20 3d20 636e     cnmf_obj = cn
+00005a20: 6d66 2e63 4e4d 4628 6f75 7470 7574 5f64  mf.cNMF(output_d
+00005a30: 6972 3d63 6e6d 665f 6f75 7470 7574 5f64  ir=cnmf_output_d
+00005a40: 6972 2c20 6e61 6d65 3d63 6e6d 665f 6e61  ir, name=cnmf_na
+00005a50: 6d65 290d 0a20 2020 2020 2020 200d 0a20  me)..        .. 
+00005a60: 2020 2020 2020 2023 2077 7269 7465 2054         # write T
+00005a70: 504d 2028 6e6f 726d 616c 697a 6564 2920  PM (normalized) 
+00005a80: 6461 7461 0d0a 2020 2020 2020 2020 7470  data..        tp
+00005a90: 6d20 3d20 6164 2e41 6e6e 4461 7461 2873  m = ad.AnnData(s
+00005aa0: 656c 662e 746f 5f64 6628 6e6f 726d 616c  elf.to_df(normal
+00005ab0: 697a 6564 3d54 7275 6529 290d 0a20 2020  ized=True))..   
+00005ac0: 2020 2020 2074 706d 2e77 7269 7465 5f68       tpm.write_h
+00005ad0: 3561 6428 636e 6d66 5f6f 626a 2e70 6174  5ad(cnmf_obj.pat
+00005ae0: 6873 5b22 7470 6d22 5d29 0d0a 0d0a 2020  hs["tpm"])....  
+00005af0: 2020 2020 2020 6765 6e65 5f74 706d 5f6d        gene_tpm_m
+00005b00: 6561 6e20 3d20 6e70 2e61 7272 6179 2874  ean = np.array(t
+00005b10: 706d 2e58 2e6d 6561 6e28 6178 6973 3d30  pm.X.mean(axis=0
+00005b20: 2929 2e72 6573 6861 7065 282d 3129 0d0a  )).reshape(-1)..
+00005b30: 2020 2020 2020 2020 6765 6e65 5f74 706d          gene_tpm
+00005b40: 5f73 7464 6465 7620 3d20 6e70 2e61 7272  _stddev = np.arr
+00005b50: 6179 2874 706d 2e58 2e73 7464 2861 7869  ay(tpm.X.std(axi
+00005b60: 733d 302c 2064 646f 663d 3029 292e 7265  s=0, ddof=0)).re
+00005b70: 7368 6170 6528 2d31 290d 0a20 2020 2020  shape(-1)..     
+00005b80: 2020 2069 6e70 7574 5f74 706d 5f73 7461     input_tpm_sta
+00005b90: 7473 203d 2070 642e 4461 7461 4672 616d  ts = pd.DataFram
+00005ba0: 6528 5b67 656e 655f 7470 6d5f 6d65 616e  e([gene_tpm_mean
+00005bb0: 2c20 6765 6e65 5f74 706d 5f73 7464 6465  , gene_tpm_stdde
+00005bc0: 765d 2c20 696e 6465 7820 3d20 5b27 5f5f  v], index = ['__
+00005bd0: 6d65 616e 272c 2027 5f5f 7374 6427 5d29  mean', '__std'])
+00005be0: 2e54 0d0a 2020 2020 2020 2020 7574 696c  .T..        util
+00005bf0: 732e 7361 7665 5f64 665f 746f 5f6e 707a  s.save_df_to_npz
+00005c00: 2869 6e70 7574 5f74 706d 5f73 7461 7473  (input_tpm_stats
+00005c10: 2c20 636e 6d66 5f6f 626a 2e70 6174 6873  , cnmf_obj.paths
+00005c20: 5b27 7470 6d5f 7374 6174 7327 5d29 0d0a  ['tpm_stats'])..
+00005c30: 2020 2020 2020 2020 6f76 6572 6469 7370          overdisp
+00005c40: 6572 7365 645f 6765 6e65 7320 3d20 7365  ersed_genes = se
+00005c50: 6c66 2e61 6461 7461 2e76 6172 5b22 7365  lf.adata.var["se
+00005c60: 6c65 6374 6564 225d 5b73 656c 662e 6164  lected"][self.ad
+00005c70: 6174 612e 7661 725b 2273 656c 6563 7465  ata.var["selecte
+00005c80: 6422 5d5d 2e69 6e64 6578 0d0a 2020 2020  d"]].index..    
+00005c90: 2020 2020 6e6f 726d 5f63 6f75 6e74 7320      norm_counts 
+00005ca0: 3d20 636e 6d66 5f6f 626a 2e67 6574 5f6e  = cnmf_obj.get_n
+00005cb0: 6f72 6d5f 636f 756e 7473 2873 656c 662e  orm_counts(self.
+00005cc0: 6164 6174 612c 2074 706d 2c20 6869 6768  adata, tpm, high
+00005cd0: 5f76 6172 6961 6e63 655f 6765 6e65 735f  _variance_genes_
+00005ce0: 6669 6c74 6572 3d6f 7665 7264 6973 7065  filter=overdispe
+00005cf0: 7273 6564 5f67 656e 6573 290d 0a20 2020  rsed_genes)..   
+00005d00: 2020 2020 2069 6620 6e6f 726d 5f63 6f75       if norm_cou
+00005d10: 6e74 732e 582e 6474 7970 6520 213d 206e  nts.X.dtype != n
+00005d20: 702e 666c 6f61 7436 343a 0d0a 2020 2020  p.float64:..    
+00005d30: 2020 2020 2020 2020 6e6f 726d 5f63 6f75          norm_cou
+00005d40: 6e74 732e 5820 3d20 6e6f 726d 5f63 6f75  nts.X = norm_cou
+00005d50: 6e74 732e 582e 6173 7479 7065 286e 702e  nts.X.astype(np.
+00005d60: 666c 6f61 7436 3429 0d0a 2020 2020 2020  float64)..      
+00005d70: 2020 636e 6d66 5f6f 626a 2e73 6176 655f    cnmf_obj.save_
+00005d80: 6e6f 726d 5f63 6f75 6e74 7328 6e6f 726d  norm_counts(norm
+00005d90: 5f63 6f75 6e74 7329 0d0a 0d0a 2020 2020  _counts)....    
+00005da0: 2020 2020 2320 7361 7665 2070 6172 616d      # save param
+00005db0: 6574 6572 7320 666f 7220 6661 6374 6f72  eters for factor
+00005dc0: 697a 6174 696f 6e20 7374 6570 0d0a 2020  ization step..  
+00005dd0: 2020 2020 2020 636e 6d66 5f6f 626a 2e73        cnmf_obj.s
+00005de0: 6176 655f 6e6d 665f 6974 6572 5f70 6172  ave_nmf_iter_par
+00005df0: 616d 7328 2a63 6e6d 665f 6f62 6a2e 6765  ams(*cnmf_obj.ge
+00005e00: 745f 6e6d 665f 6974 6572 5f70 6172 616d  t_nmf_iter_param
+00005e10: 7328 6b73 3d6b 7661 6c73 2c20 6e5f 6974  s(ks=kvals, n_it
+00005e20: 6572 3d6e 5f69 7465 722c 2072 616e 646f  er=n_iter, rando
+00005e30: 6d5f 7374 6174 655f 7365 6564 3d73 6565  m_state_seed=see
+00005e40: 642c 2062 6574 615f 6c6f 7373 3d62 6574  d, beta_loss=bet
+00005e50: 615f 6c6f 7373 2929 0d0a 0d0a 2020 2020  a_loss))....    
+00005e60: 2020 2020 2320 7361 7665 2070 6172 616d      # save param
+00005e70: 6574 6572 7320 696e 2041 6e6e 4461 7461  eters in AnnData
+00005e80: 206f 626a 6563 740d 0a20 2020 2020 2020   object..       
+00005e90: 2073 656c 662e 6164 6174 612e 756e 735b   self.adata.uns[
+00005ea0: 2263 6e6d 6622 5d20 3d20 636e 6d66 5f6f  "cnmf"] = cnmf_o
+00005eb0: 626a 2e67 6574 5f6e 6d66 5f69 7465 725f  bj.get_nmf_iter_
+00005ec0: 7061 7261 6d73 286b 733d 6b76 616c 732c  params(ks=kvals,
+00005ed0: 206e 5f69 7465 723d 6e5f 6974 6572 2c20   n_iter=n_iter, 
+00005ee0: 7261 6e64 6f6d 5f73 7461 7465 5f73 6565  random_state_see
+00005ef0: 643d 7365 6564 2c20 6265 7461 5f6c 6f73  d=seed, beta_los
+00005f00: 733d 6265 7461 5f6c 6f73 7329 5b31 5d20  s=beta_loss)[1] 
+00005f10: 2023 2064 6963 7420 6f66 2063 6e6d 6620   # dict of cnmf 
+00005f20: 7061 7261 6d65 7465 7273 0d0a 2020 2020  parameters..    
+00005f30: 2020 2020 0d0a 2020 2020 2020 2020 7365      ..        se
+00005f40: 6c66 2e61 7070 656e 645f 746f 5f68 6973  lf.append_to_his
+00005f50: 746f 7279 2866 2263 4e4d 4620 7061 7261  tory(f"cNMF para
+00005f60: 6d65 7465 7273 2061 6464 6564 2e20 634e  meters added. cN
+00005f70: 4d46 2069 6e70 7574 7320 696e 6974 6961  MF inputs initia
+00005f80: 6c69 7a65 6420 696e 207b 636e 6d66 5f6f  lized in {cnmf_o
+00005f90: 7574 7075 745f 6469 727d 2f7b 636e 6d66  utput_dir}/{cnmf
+00005fa0: 5f6e 616d 657d 2229 0d0a 2020 2020 2020  _name}")..      
+00005fb0: 2020 7265 7475 726e 2063 6e6d 665f 6f62    return cnmf_ob
+00005fc0: 6a0d 0a20 2020 200d 0a20 2020 2064 6566  j..    ..    def
+00005fd0: 2061 6464 5f63 6e6d 665f 7265 7375 6c74   add_cnmf_result
+00005fe0: 7328 7365 6c66 2c20 636e 6d66 5f6f 7574  s(self, cnmf_out
+00005ff0: 7075 745f 6469 722c 2063 6e6d 665f 6e61  put_dir, cnmf_na
+00006000: 6d65 2c20 6c6f 6361 6c5f 6465 6e73 6974  me, local_densit
+00006010: 795f 7468 7265 7368 6f6c 643a 2066 6c6f  y_threshold: flo
+00006020: 6174 203d 204e 6f6e 652c 206c 6f63 616c  at = None, local
+00006030: 5f6e 6569 6768 626f 7268 6f6f 645f 7369  _neighborhood_si
+00006040: 7a65 3a20 666c 6f61 7420 3d20 4e6f 6e65  ze: float = None
+00006050: 293a 0d0a 2020 2020 2020 2020 2222 220d  ):..        """.
+00006060: 0a20 2020 2020 2020 2041 6674 6572 2066  .        After f
+00006070: 6163 746f 7269 7a61 7469 6f6e 2c20 6164  actorization, ad
+00006080: 6420 636f 6d70 6c65 7465 6420 634e 4d46  d completed cNMF
+00006090: 2072 6573 756c 7473 2069 6e20 5b63 6e6d   results in [cnm
+000060a0: 665f 6f75 7470 7574 5f64 6972 5d2f 5b63  f_output_dir]/[c
+000060b0: 6e6d 665f 6e61 6d65 5d20 746f 2074 6865  nmf_name] to the
+000060c0: 2064 6174 6173 6574 206f 626a 6563 742e   dataset object.
+000060d0: 0d0a 0d0a 2020 2020 2020 2020 3a70 6172  ....        :par
+000060e0: 616d 2063 6e6d 665f 6f75 7470 7574 5f64  am cnmf_output_d
+000060f0: 6972 3a20 4f75 7470 7574 2064 6972 6563  ir: Output direc
+00006100: 746f 7279 2066 6f72 2063 4e4d 4620 7265  tory for cNMF re
+00006110: 7375 6c74 730d 0a20 2020 2020 2020 203a  sults..        :
+00006120: 7479 7065 2063 6e6d 665f 6f75 7470 7574  type cnmf_output
+00006130: 5f64 6972 3a20 7374 720d 0a20 2020 2020  _dir: str..     
+00006140: 2020 203a 7061 7261 6d20 636e 6d66 5f6e     :param cnmf_n
+00006150: 616d 653a 204e 616d 6520 6f66 2074 6865  ame: Name of the
+00006160: 2063 4e4d 4620 7265 7375 6c74 732e 2046   cNMF results. F
+00006170: 696c 6573 2077 696c 6c20 6265 206f 7574  iles will be out
+00006180: 7075 7420 746f 205b 636e 6d66 5f6f 7574  put to [cnmf_out
+00006190: 7075 745f 6469 725d 2f5b 636e 6d66 5f6e  put_dir]/[cnmf_n
+000061a0: 616d 655d 2f0d 0a20 2020 2020 2020 203a  ame]/..        :
+000061b0: 7479 7065 2063 6e6d 665f 6e61 6d65 3a20  type cnmf_name: 
+000061c0: 7374 720d 0a20 2020 2020 2020 203a 7061  str..        :pa
+000061d0: 7261 6d20 6c6f 6361 6c5f 6465 6e73 6974  ram local_densit
+000061e0: 795f 7468 7265 7368 6f6c 643a 2054 6872  y_threshold: Thr
+000061f0: 6573 686f 6c64 2066 6f72 2074 6865 206c  eshold for the l
+00006200: 6f63 616c 2064 656e 7369 7479 2066 696c  ocal density fil
+00006210: 7465 7269 6e67 2070 7269 6f72 2074 6f20  tering prior to 
+00006220: 4745 5020 636f 6e73 656e 7375 732e 2041  GEP consensus. A
+00006230: 6363 6570 7461 626c 6520 7468 7265 7368  cceptable thresh
+00006240: 6f6c 6473 2061 7265 203e 2030 2061 6e64  olds are > 0 and
+00006250: 203c 3d20 3220 2832 2e30 2069 7320 6e6f   <= 2 (2.0 is no
+00006260: 2066 696c 7465 7269 6e67 292e 2044 6566   filtering). Def
+00006270: 6175 6c74 7320 746f 204e 6f6e 652e 0d0a  aults to None...
+00006280: 2020 2020 2020 2020 3a74 7970 6520 6c6f          :type lo
+00006290: 6361 6c5f 6465 6e73 6974 795f 7468 7265  cal_density_thre
+000062a0: 7368 6f6c 643a 2066 6c6f 6174 2c20 6f70  shold: float, op
+000062b0: 7469 6f6e 616c 0d0a 2020 2020 2020 2020  tional..        
+000062c0: 3a70 6172 616d 206c 6f63 616c 5f6e 6569  :param local_nei
+000062d0: 6768 626f 7268 6f6f 645f 7369 7a65 3a20  ghborhood_size: 
+000062e0: 4672 6163 7469 6f6e 206f 6620 7468 6520  Fraction of the 
+000062f0: 6e75 6d62 6572 206f 6620 7265 706c 6963  number of replic
+00006300: 6174 6573 2074 6f20 7573 6520 6173 206e  ates to use as n
+00006310: 6561 7265 7374 206e 6569 6768 626f 7273  earest neighbors
+00006320: 2066 6f72 206c 6f63 616c 2064 656e 7369   for local densi
+00006330: 7479 2066 696c 7465 7269 6e67 2e20 4465  ty filtering. De
+00006340: 6661 756c 7473 2074 6f20 4e6f 6e65 0d0a  faults to None..
+00006350: 2020 2020 2020 2020 3a74 7970 6520 6c6f          :type lo
+00006360: 6361 6c5f 6e65 6967 6862 6f72 686f 6f64  cal_neighborhood
+00006370: 5f73 697a 653a 2066 6c6f 6174 2c20 6f70  _size: float, op
+00006380: 7469 6f6e 616c 0d0a 2020 2020 2020 2020  tional..        
+00006390: 2222 220d 0a20 2020 2020 2020 2073 656c  """..        sel
+000063a0: 662e 6164 6174 612e 756e 735b 2263 6e6d  f.adata.uns["cnm
+000063b0: 665f 6e61 6d65 225d 203d 2063 6e6d 665f  f_name"] = cnmf_
+000063c0: 6e61 6d65 0d0a 0d0a 2020 2020 2020 2020  name....        
+000063d0: 2320 696e 6665 7220 6672 6f6d 2066 696c  # infer from fil
+000063e0: 656e 616d 6573 2077 6869 6368 206c 6f63  enames which loc
+000063f0: 616c 2064 656e 7369 7479 2074 6872 6573  al density thres
+00006400: 686f 6c64 2077 6173 2075 7365 640d 0a20  hold was used.. 
+00006410: 2020 2020 2020 2073 656e 7365 645f 6c64         sensed_ld
+00006420: 7473 203d 2073 6574 2829 0d0a 2020 2020  ts = set()..    
+00006430: 2020 2020 666f 7220 666e 2069 6e20 676c      for fn in gl
+00006440: 6f62 286f 732e 7061 7468 2e6a 6f69 6e28  ob(os.path.join(
+00006450: 636e 6d66 5f6f 7574 7075 745f 6469 722c  cnmf_output_dir,
+00006460: 2063 6e6d 665f 6e61 6d65 2c20 6622 7b63   cnmf_name, f"{c
+00006470: 6e6d 665f 6e61 6d65 7d2a 2e2a 7370 6563  nmf_name}*.*spec
+00006480: 7472 612a 2e6b 5f2a 2229 293a 0d0a 2020  tra*.k_*")):..  
+00006490: 2020 2020 2020 2020 2020 6c64 745f 7374            ldt_st
+000064a0: 7220 3d20 6f73 2e70 6174 682e 6261 7365  r = os.path.base
+000064b0: 6e61 6d65 2866 6e29 2e73 706c 6974 2822  name(fn).split("
+000064c0: 2e22 295b 2d33 5d0d 0a20 2020 2020 2020  .")[-3]..       
+000064d0: 2020 2020 2074 7279 3a0d 0a20 2020 2020       try:..     
+000064e0: 2020 2020 2020 2020 2020 206c 6474 203d             ldt =
+000064f0: 2066 6c6f 6174 286c 6474 5f73 7472 2e72   float(ldt_str.r
+00006500: 6570 6c61 6365 2822 6474 5f22 2c20 2222  eplace("dt_", ""
+00006510: 292e 7265 706c 6163 6528 225f 222c 2022  ).replace("_", "
+00006520: 2e22 2929 0d0a 2020 2020 2020 2020 2020  ."))..          
+00006530: 2020 6578 6365 7074 2056 616c 7565 4572    except ValueEr
+00006540: 726f 723a 0d0a 2020 2020 2020 2020 2020  ror:..          
+00006550: 2020 2020 2020 7061 7373 0d0a 2020 2020        pass..    
+00006560: 2020 2020 2020 2020 656c 7365 3a0d 0a20          else:.. 
+00006570: 2020 2020 2020 2020 2020 2020 2020 2073                 s
+00006580: 656e 7365 645f 6c64 7473 2e61 6464 2828  ensed_ldts.add((
+00006590: 6c64 745f 7374 722c 206c 6474 2929 0d0a  ldt_str, ldt))..
+000065a0: 2020 2020 2020 2020 6966 206c 6f63 616c          if local
+000065b0: 5f64 656e 7369 7479 5f74 6872 6573 686f  _density_thresho
+000065c0: 6c64 2069 7320 4e6f 6e65 2061 6e64 206c  ld is None and l
+000065d0: 656e 2873 656e 7365 645f 6c64 7473 2920  en(sensed_ldts) 
+000065e0: 3d3d 2031 3a0d 0a20 2020 2020 2020 2020  == 1:..         
+000065f0: 2020 206c 6474 5f73 7472 2c20 6c64 7420     ldt_str, ldt 
+00006600: 3d20 7365 6e73 6564 5f6c 6474 732e 706f  = sensed_ldts.po
+00006610: 7028 290d 0a20 2020 2020 2020 2065 6c69  p()..        eli
+00006620: 6620 6c6f 6361 6c5f 6465 6e73 6974 795f  f local_density_
+00006630: 7468 7265 7368 6f6c 6420 696e 2028 6c64  threshold in (ld
+00006640: 745b 315d 2066 6f72 206c 6474 2069 6e20  t[1] for ldt in 
+00006650: 7365 6e73 6564 5f6c 6474 7329 3a0d 0a20  sensed_ldts):.. 
+00006660: 2020 2020 2020 2020 2020 206c 6474 5f73             ldt_s
+00006670: 7472 2c20 6c64 7420 3d20 5b28 6c64 745f  tr, ldt = [(ldt_
+00006680: 7374 722c 206c 6474 2920 666f 7220 6c64  str, ldt) for ld
+00006690: 745f 7374 722c 206c 6474 2069 6e20 7365  t_str, ldt in se
+000066a0: 6e73 6564 5f6c 6474 7320 6966 206c 6474  nsed_ldts if ldt
+000066b0: 203d 3d20 6c6f 6361 6c5f 6465 6e73 6974   == local_densit
+000066c0: 795f 7468 7265 7368 6f6c 645d 2e70 6f70  y_threshold].pop
+000066d0: 2829 0d0a 2020 2020 2020 2020 656c 7365  ()..        else
+000066e0: 3a0d 0a20 2020 2020 2020 2020 2020 206c  :..            l
+000066f0: 6f67 6769 6e67 2e65 7272 6f72 2866 226c  ogging.error(f"l
+00006700: 6f63 616c 5f64 656e 7369 7479 5f74 6872  ocal_density_thr
+00006710: 6573 686f 6c64 206f 6620 7b6c 6f63 616c  eshold of {local
+00006720: 5f64 656e 7369 7479 5f74 6872 6573 686f  _density_thresho
+00006730: 6c64 7d20 646f 6573 206e 6f74 206d 6174  ld} does not mat
+00006740: 6368 2077 6861 7420 6973 2069 6e20 7468  ch what is in th
+00006750: 6520 634e 4d46 2072 6573 756c 7420 6469  e cNMF result di
+00006760: 7265 6374 6f72 793a 207b 7365 6e73 6564  rectory: {sensed
+00006770: 5f6c 6474 737d 2229 0d0a 2020 2020 2020  _ldts}")..      
+00006780: 2020 2020 2020 7379 732e 6578 6974 2831        sys.exit(1
+00006790: 290d 0a20 2020 2020 2020 2073 656c 662e  )..        self.
+000067a0: 6164 6174 612e 756e 735b 226c 6474 225d  adata.uns["ldt"]
+000067b0: 203d 206c 6474 0d0a 2020 2020 2020 2020   = ldt..        
+000067c0: 7365 6c66 2e61 6461 7461 2e75 6e73 5b22  self.adata.uns["
+000067d0: 6c6e 7322 5d20 3d20 6c6f 6361 6c5f 6e65  lns"] = local_ne
+000067e0: 6967 6862 6f72 686f 6f64 5f73 697a 650d  ighborhood_size.
+000067f0: 0a20 2020 2020 2020 2020 2020 200d 0a20  .            .. 
+00006800: 2020 2020 2020 2023 2049 6d70 6f72 7420         # Import 
+00006810: 4745 5073 0d0a 2020 2020 2020 2020 7265  GEPs..        re
+00006820: 7375 6c74 5f74 7970 6573 203d 207b 0d0a  sult_types = {..
+00006830: 2020 2020 2020 2020 2020 2020 2267 656e              "gen
+00006840: 655f 7370 6563 7472 615f 7363 6f72 6522  e_spectra_score"
+00006850: 3a20 2263 6e6d 665f 6765 705f 7363 6f72  : "cnmf_gep_scor
+00006860: 6522 2c0d 0a20 2020 2020 2020 2020 2020  e",..           
+00006870: 2022 6765 6e65 5f73 7065 6374 7261 5f74   "gene_spectra_t
+00006880: 706d 223a 2022 636e 6d66 5f67 6570 5f74  pm": "cnmf_gep_t
+00006890: 706d 222c 0d0a 2020 2020 2020 2020 2020  pm",..          
+000068a0: 2020 2273 7065 6374 7261 223a 2022 636e    "spectra": "cn
+000068b0: 6d66 5f67 6570 5f72 6177 220d 0a20 2020  mf_gep_raw"..   
+000068c0: 2020 2020 2020 2020 207d 0d0a 2020 2020           }..    
+000068d0: 2020 2020 666f 7220 6d61 7463 6873 7472      for matchstr
+000068e0: 2c20 7265 7375 6c74 5f74 7970 6520 696e  , result_type in
+000068f0: 2072 6573 756c 745f 7479 7065 732e 6974   result_types.it
+00006900: 656d 7328 293a 0d0a 2020 2020 2020 2020  ems():..        
+00006910: 2020 2020 6c6f 6767 696e 672e 696e 666f      logging.info
+00006920: 2866 2249 6d70 6f72 7469 6e67 2047 4550  (f"Importing GEP
+00006930: 733a 207b 6d61 7463 6873 7472 7d22 2920  s: {matchstr}") 
+00006940: 200d 0a20 2020 2020 2020 2020 2020 206d   ..            m
+00006950: 6574 615f 7720 3d20 5b5d 0d0a 2020 2020  eta_w = []..    
+00006960: 2020 2020 2020 2020 666f 7220 666e 2069          for fn i
+00006970: 6e20 676c 6f62 286f 732e 7061 7468 2e6a  n glob(os.path.j
+00006980: 6f69 6e28 636e 6d66 5f6f 7574 7075 745f  oin(cnmf_output_
+00006990: 6469 722c 2063 6e6d 665f 6e61 6d65 2c20  dir, cnmf_name, 
+000069a0: 6622 7b63 6e6d 665f 6e61 6d65 7d2a 2e7b  f"{cnmf_name}*.{
+000069b0: 6d61 7463 6873 7472 7d2e 6b5f 2a2e 7b6c  matchstr}.k_*.{l
+000069c0: 6474 5f73 7472 7d2e 2a74 7874 2229 293a  dt_str}.*txt")):
+000069d0: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
+000069e0: 2020 6b20 3d20 696e 7428 6f73 2e70 6174    k = int(os.pat
+000069f0: 682e 6261 7365 6e61 6d65 2866 6e29 2e72  h.basename(fn).r
+00006a00: 656d 6f76 6570 7265 6669 7828 6622 7b63  emoveprefix(f"{c
+00006a10: 6e6d 665f 6e61 6d65 7d2e 7b6d 6174 6368  nmf_name}.{match
+00006a20: 7374 727d 2e22 292e 7370 6c69 7428 222e  str}.").split(".
+00006a30: 2229 5b30 5d2e 7265 706c 6163 6528 226b  ")[0].replace("k
+00006a40: 5f22 2c20 2222 2929 0d0a 2020 2020 2020  _", ""))..      
+00006a50: 2020 2020 2020 2020 2020 7720 3d20 7064            w = pd
+00006a60: 2e72 6561 645f 7461 626c 6528 666e 2c20  .read_table(fn, 
+00006a70: 696e 6465 785f 636f 6c3d 3029 0d0a 2020  index_col=0)..  
+00006a80: 2020 2020 2020 2020 2020 2020 2020 772e                w.
+00006a90: 696e 6465 7820 3d20 7374 7228 6b29 202b  index = str(k) +
+00006aa0: 2022 2e22 202b 2077 2e69 6e64 6578 2e61   "." + w.index.a
+00006ab0: 7374 7970 6528 7374 7229 0d0a 2020 2020  stype(str)..    
+00006ac0: 2020 2020 2020 2020 2020 2020 6d65 7461              meta
+00006ad0: 5f77 2e61 7070 656e 6428 7729 0d0a 2020  _w.append(w)..  
+00006ae0: 2020 2020 2020 2020 2020 6d65 7461 5f77            meta_w
+00006af0: 203d 2070 642e 636f 6e63 6174 286d 6574   = pd.concat(met
+00006b00: 615f 772c 2061 7869 733d 3029 2e54 2e72  a_w, axis=0).T.r
+00006b10: 6569 6e64 6578 2873 656c 662e 6164 6174  eindex(self.adat
+00006b20: 612e 7661 722e 696e 6465 7829 2e72 656e  a.var.index).ren
+00006b30: 616d 655f 6178 6973 285b 226b 2e67 6570  ame_axis(["k.gep
+00006b40: 225d 2c20 6178 6973 3d31 290d 0a20 2020  "], axis=1)..   
+00006b50: 2020 2020 2020 2020 2073 656c 662e 6164           self.ad
+00006b60: 6174 612e 7661 726d 5b72 6573 756c 745f  ata.varm[result_
+00006b70: 7479 7065 5d20 3d20 6d65 7461 5f77 0d0a  type] = meta_w..
+00006b80: 0d0a 2020 2020 2020 2020 2320 496d 706f  ..        # Impo
+00006b90: 7274 2055 7361 6765 730d 0a20 2020 2020  rt Usages..     
+00006ba0: 2020 206c 6f67 6769 6e67 2e69 6e66 6f28     logging.info(
+00006bb0: 6622 496d 706f 7274 696e 6720 5573 6167  f"Importing Usag
+00006bc0: 6573 2229 2020 0d0a 2020 2020 2020 2020  es")  ..        
+00006bd0: 7573 6167 6520 3d20 5b5d 0d0a 2020 2020  usage = []..    
+00006be0: 2020 2020 666f 7220 666e 2069 6e20 676c      for fn in gl
+00006bf0: 6f62 286f 732e 7061 7468 2e6a 6f69 6e28  ob(os.path.join(
+00006c00: 636e 6d66 5f6f 7574 7075 745f 6469 722c  cnmf_output_dir,
+00006c10: 2063 6e6d 665f 6e61 6d65 2c20 6622 7b63   cnmf_name, f"{c
+00006c20: 6e6d 665f 6e61 6d65 7d2a 2e75 7361 6765  nmf_name}*.usage
+00006c30: 732e 6b5f 2a2e 7b6c 6474 5f73 7472 7d2e  s.k_*.{ldt_str}.
+00006c40: 2a74 7874 2229 293a 0d0a 2020 2020 2020  *txt")):..      
+00006c50: 2020 2020 2020 6b20 3d20 696e 7428 6f73        k = int(os
+00006c60: 2e70 6174 682e 6261 7365 6e61 6d65 2866  .path.basename(f
+00006c70: 6e29 2e72 656d 6f76 6570 7265 6669 7828  n).removeprefix(
+00006c80: 6622 7b63 6e6d 665f 6e61 6d65 7d2e 7573  f"{cnmf_name}.us
+00006c90: 6167 6573 2e22 292e 7370 6c69 7428 222e  ages.").split(".
+00006ca0: 2229 5b30 5d2e 7265 706c 6163 6528 226b  ")[0].replace("k
+00006cb0: 5f22 2c20 2222 2929 0d0a 2020 2020 2020  _", ""))..      
+00006cc0: 2020 2020 2020 6820 3d20 7064 2e72 6561        h = pd.rea
+00006cd0: 645f 7461 626c 6528 666e 2c20 696e 6465  d_table(fn, inde
+00006ce0: 785f 636f 6c3d 3029 0d0a 2020 2020 2020  x_col=0)..      
+00006cf0: 2020 2020 2020 682e 636f 6c75 6d6e 7320        h.columns 
+00006d00: 3d20 7374 7228 6b29 202b 2022 2e22 202b  = str(k) + "." +
+00006d10: 2068 2e63 6f6c 756d 6e73 2e61 7374 7970   h.columns.astyp
+00006d20: 6528 7374 7229 0d0a 2020 2020 2020 2020  e(str)..        
+00006d30: 2020 2020 7573 6167 652e 6170 7065 6e64      usage.append
+00006d40: 2868 290d 0a20 2020 2020 2020 2073 656c  (h)..        sel
+00006d50: 662e 6164 6174 612e 6f62 736d 5b22 636e  f.adata.obsm["cn
+00006d60: 6d66 5f75 7361 6765 225d 203d 2070 642e  mf_usage"] = pd.
+00006d70: 636f 6e63 6174 2875 7361 6765 2c20 6178  concat(usage, ax
+00006d80: 6973 3d31 292e 736f 7274 5f69 6e64 6578  is=1).sort_index
+00006d90: 2861 7869 733d 3129 2e72 656e 616d 655f  (axis=1).rename_
+00006da0: 6178 6973 285b 226b 2e67 6570 225d 2c20  axis(["k.gep"], 
+00006db0: 6178 6973 3d31 290d 0a20 2020 2020 2020  axis=1)..       
+00006dc0: 200d 0a20 2020 2020 2020 2023 2049 6d70   ..        # Imp
+00006dd0: 6f72 7420 6765 6e65 206c 6973 7420 7573  ort gene list us
+00006de0: 6564 2066 6f72 2066 6163 746f 7269 7a61  ed for factoriza
+00006df0: 7469 6f6e 0d0a 2020 2020 2020 2020 7769  tion..        wi
+00006e00: 7468 206f 7065 6e28 6f73 2e70 6174 682e  th open(os.path.
+00006e10: 6a6f 696e 2863 6e6d 665f 6f75 7470 7574  join(cnmf_output
+00006e20: 5f64 6972 2c20 636e 6d66 5f6e 616d 652c  _dir, cnmf_name,
+00006e30: 2066 227b 636e 6d66 5f6e 616d 657d 2e6f   f"{cnmf_name}.o
+00006e40: 7665 7264 6973 7065 7273 6564 5f67 656e  verdispersed_gen
+00006e50: 6573 2e74 7874 2229 2920 6173 2066 3a0d  es.txt")) as f:.
+00006e60: 0a20 2020 2020 2020 2020 2020 2073 656c  .            sel
+00006e70: 662e 6164 6174 612e 756e 735b 2267 656e  f.adata.uns["gen
+00006e80: 655f 6c69 7374 225d 203d 205b 6c69 6e65  e_list"] = [line
+00006e90: 2e73 7472 6970 2829 2066 6f72 206c 696e  .strip() for lin
+00006ea0: 6520 696e 2066 2e72 6561 646c 696e 6573  e in f.readlines
+00006eb0: 2829 5d0d 0a0d 0a20 2020 2020 2020 2023  ()]....        #
+00006ec0: 2049 6d70 6f72 7420 4b2d 7365 6c65 6374   Import K-select
+00006ed0: 696f 6e20 7374 6174 730d 0a20 2020 2020  ion stats..     
+00006ee0: 2020 206b 7661 6c73 203d 2070 642e 4461     kvals = pd.Da
+00006ef0: 7461 4672 616d 6528 2a2a 6e70 2e6c 6f61  taFrame(**np.loa
+00006f00: 6428 6f73 2e70 6174 682e 6a6f 696e 2863  d(os.path.join(c
+00006f10: 6e6d 665f 6f75 7470 7574 5f64 6972 2c20  nmf_output_dir, 
+00006f20: 636e 6d66 5f6e 616d 652c 2066 227b 636e  cnmf_name, f"{cn
+00006f30: 6d66 5f6e 616d 657d 2e6b 5f73 656c 6563  mf_name}.k_selec
+00006f40: 7469 6f6e 5f73 7461 7473 2e64 662e 6e70  tion_stats.df.np
+00006f50: 7a22 292c 2061 6c6c 6f77 5f70 6963 6b6c  z"), allow_pickl
+00006f60: 653d 5472 7565 2929 2e73 6574 5f69 6e64  e=True)).set_ind
+00006f70: 6578 2822 6b22 295b 5b22 7374 6162 696c  ex("k")[["stabil
+00006f80: 6974 7922 2c20 2270 7265 6469 6374 696f  ity", "predictio
+00006f90: 6e5f 6572 726f 7222 5d5d 0d0a 2020 2020  n_error"]]..    
+00006fa0: 2020 2020 6b76 616c 732e 696e 6465 7820      kvals.index 
+00006fb0: 3d20 6b76 616c 732e 696e 6465 782e 6173  = kvals.index.as
+00006fc0: 7479 7065 2869 6e74 290d 0a20 2020 2020  type(int)..     
+00006fd0: 2020 2073 656c 662e 6164 6174 612e 756e     self.adata.un
+00006fe0: 735b 226b 7661 6c73 225d 203d 206b 7661  s["kvals"] = kva
+00006ff0: 6c73 0d0a 2020 2020 2020 2020 7365 6c66  ls..        self
+00007000: 2e61 7070 656e 645f 746f 5f68 6973 746f  .append_to_histo
+00007010: 7279 2822 634e 4d46 2072 6573 756c 7473  ry("cNMF results
+00007020: 2061 6464 6564 2066 726f 6d20 6f75 7470   added from outp
+00007030: 7574 2064 6972 6563 746f 7279 207b 636e  ut directory {cn
+00007040: 6d66 5f6f 7574 7075 745f 6469 727d 2f7b  mf_output_dir}/{
+00007050: 636e 6d66 5f6e 616d 657d 2229 0d0a 0d0a  cnmf_name}")....
+00007060: 2020 2020 6465 6620 7265 6d6f 7665 5f63      def remove_c
+00007070: 6e6d 665f 7265 7375 6c74 7328 7365 6c66  nmf_results(self
+00007080: 293a 0d0a 2020 2020 2020 2020 666f 7220  ):..        for 
+00007090: 7265 7375 6c74 5f74 7970 6520 696e 2028  result_type in (
+000070a0: 2263 6e6d 665f 6765 705f 7363 6f72 6522  "cnmf_gep_score"
+000070b0: 2c20 2263 6e6d 665f 6765 705f 7470 6d22  , "cnmf_gep_tpm"
+000070c0: 2c20 2263 6e6d 665f 6765 705f 7261 7722  , "cnmf_gep_raw"
+000070d0: 293a 0d0a 2020 2020 2020 2020 2020 2020  ):..            
+000070e0: 7365 6c66 2e61 6461 7461 2e76 6172 6d2e  self.adata.varm.
+000070f0: 706f 7028 7265 7375 6c74 5f74 7970 6529  pop(result_type)
+00007100: 0d0a 2020 2020 2020 2020 7365 6c66 2e61  ..        self.a
+00007110: 6461 7461 2e6f 6273 6d2e 706f 7028 2263  data.obsm.pop("c
+00007120: 6e6d 665f 7573 6167 6522 290d 0a20 2020  nmf_usage")..   
+00007130: 2020 2020 2073 656c 662e 6164 6174 612e       self.adata.
+00007140: 756e 732e 706f 7028 2267 656e 655f 6c69  uns.pop("gene_li
+00007150: 7374 2229 0d0a 2020 2020 2020 2020 7365  st")..        se
+00007160: 6c66 2e61 6461 7461 2e75 6e73 2e70 6f70  lf.adata.uns.pop
+00007170: 2822 6b76 616c 7322 290d 0a20 2020 2020  ("kvals")..     
+00007180: 2020 2073 656c 662e 6170 7065 6e64 5f74     self.append_t
+00007190: 6f5f 6869 7374 6f72 7928 2263 4e4d 4620  o_history("cNMF 
+000071a0: 7265 7375 6c74 7320 7265 6d6f 7665 642e  results removed.
+000071b0: 2229 0d0a 0d0a 0d0a 2020 2020 6465 6620  ")......    def 
+000071c0: 6765 745f 7573 6167 6573 2873 656c 662c  get_usages(self,
+000071d0: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
+000071e0: 2020 2020 206b 3a20 556e 696f 6e5b 696e       k: Union[in
+000071f0: 742c 2049 7465 7261 626c 655d 203d 204e  t, Iterable] = N
+00007200: 6f6e 652c 0d0a 2020 2020 2020 2020 2020  one,..          
+00007210: 2020 2020 2020 2020 2064 6973 6372 6574           discret
+00007220: 697a 653a 2062 6f6f 6c20 3d20 4661 6c73  ize: bool = Fals
+00007230: 652c 0d0a 2020 2020 2020 2020 2020 2020  e,..            
+00007240: 2020 2020 2020 206e 6f72 6d61 6c69 7a65         normalize
+00007250: 3a20 626f 6f6c 203d 2046 616c 7365 0d0a  : bool = False..
+00007260: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007270: 2020 2029 202d 3e20 7064 2e44 6174 6146     ) -> pd.DataF
+00007280: 7261 6d65 3a0d 0a20 2020 2020 2020 2022  rame:..        "
+00007290: 2222 0d0a 2020 2020 2020 2020 4361 6c63  ""..        Calc
+000072a0: 756c 6174 6520 7573 6167 6520 6f66 2065  ulate usage of e
+000072b0: 6163 6820 4745 5020 696e 2065 6163 6820  ach GEP in each 
+000072c0: 7361 6d70 6c65 2f6f 6273 6572 7661 7469  sample/observati
+000072d0: 6f6e 2e0d 0a0d 0a20 2020 2020 2020 203a  on.....        :
+000072e0: 7061 7261 6d20 6b3a 2049 6620 616e 2069  param k: If an i
+000072f0: 6e74 6567 6572 206f 7220 6c69 7374 206f  nteger or list o
+00007300: 6620 696e 7465 6765 7273 2c20 7265 7475  f integers, retu
+00007310: 726e 7320 7573 6167 6573 206f 6e6c 7920  rns usages only 
+00007320: 666f 7220 7370 6563 6966 6965 6420 7261  for specified ra
+00007330: 6e6b 732e 204f 7468 6572 7769 7365 2c20  nks. Otherwise, 
+00007340: 7265 7475 726e 7320 7573 6167 6520 6f66  returns usage of
+00007350: 2061 6c6c 2047 4550 7320 6163 726f 7373   all GEPs across
+00007360: 2072 616e 6b73 2e20 4465 6661 756c 7473   ranks. Defaults
+00007370: 2074 6f20 4e6f 6e65 0d0a 2020 2020 2020   to None..      
+00007380: 2020 3a74 7970 6520 6b3a 2069 6e74 2c20    :type k: int, 
+00007390: 6f70 7469 6f6e 616c 0d0a 2020 2020 2020  optional..      
+000073a0: 2020 3a70 6172 616d 2064 6973 6372 6574    :param discret
+000073b0: 697a 653a 2044 6973 6372 6574 697a 6573  ize: Discretizes
+000073c0: 2074 6865 2075 7361 6765 206d 6174 7269   the usage matri
+000073d0: 7820 7375 6368 2074 6861 7420 666f 7220  x such that for 
+000073e0: 6561 6368 2076 616c 7565 206f 6620 6b2c  each value of k,
+000073f0: 2065 6163 6820 7361 6d70 6c65 2068 6173   each sample has
+00007400: 2075 7361 6765 206f 6620 6f6e 6c79 2031   usage of only 1
+00007410: 2047 4550 2028 7468 6520 6f6e 6520 7769   GEP (the one wi
+00007420: 7468 2074 6865 206d 6178 696d 756d 2075  th the maximum u
+00007430: 7361 6765 292e 2044 6566 6175 6c74 7320  sage). Defaults 
+00007440: 746f 2046 616c 7365 0d0a 2020 2020 2020  to False..      
+00007450: 2020 3a74 7970 6520 6469 7363 7265 7469    :type discreti
+00007460: 7a65 3a20 626f 6f6c 2c20 6f70 7469 6f6e  ze: bool, option
+00007470: 616c 0d0a 2020 2020 2020 2020 3a70 6172  al..        :par
+00007480: 616d 206e 6f72 6d61 6c69 7a65 3a20 4e6f  am normalize: No
+00007490: 726d 616c 697a 6520 7468 6520 4745 5020  rmalize the GEP 
+000074a0: 7573 6167 6520 6d61 7472 6978 2073 7563  usage matrix suc
+000074b0: 6820 7468 6174 2066 6f72 2065 6163 6820  h that for each 
+000074c0: 7661 6c75 6520 6f66 206b 2c20 7573 6167  value of k, usag
+000074d0: 6520 6f66 2061 6c6c 2047 4550 7320 7375  e of all GEPs su
+000074e0: 6d73 2074 6f20 312e 2044 6566 6175 6c74  ms to 1. Default
+000074f0: 7320 746f 2046 616c 7365 0d0a 2020 2020  s to False..    
+00007500: 2020 2020 3a74 7970 6520 6e6f 726d 616c      :type normal
+00007510: 697a 653a 2062 6f6f 6c2c 206f 7074 696f  ize: bool, optio
+00007520: 6e61 6c0d 0a20 2020 2020 2020 203a 7265  nal..        :re
+00007530: 7475 726e 3a20 6f62 7365 7276 6174 696f  turn: observatio
+00007540: 6e20 c397 2047 4550 206d 6174 7269 780d  n .. GEP matrix.
+00007550: 0a20 2020 2020 2020 203a 7274 7970 653a  .        :rtype:
+00007560: 2070 642e 4461 7461 4672 616d 650d 0a20   pd.DataFrame.. 
+00007570: 2020 2020 2020 2022 2222 0d0a 2020 2020         """..    
+00007580: 2020 2020 6466 203d 2073 656c 662e 6164      df = self.ad
+00007590: 6174 612e 6f62 736d 5b22 636e 6d66 5f75  ata.obsm["cnmf_u
+000075a0: 7361 6765 225d 2e63 6f70 7928 290d 0a20  sage"].copy().. 
+000075b0: 2020 2020 2020 2064 662e 636f 6c75 6d6e         df.column
+000075c0: 7320 3d20 7064 2e4d 756c 7469 496e 6465  s = pd.MultiInde
+000075d0: 782e 6672 6f6d 5f74 7570 6c65 7328 6466  x.from_tuples(df
+000075e0: 2e63 6f6c 756d 6e73 2e73 7472 2e73 706c  .columns.str.spl
+000075f0: 6974 2822 2e22 292e 746f 5f6c 6973 7428  it(".").to_list(
+00007600: 2929 0d0a 2020 2020 2020 2020 6466 2e63  ))..        df.c
+00007610: 6f6c 756d 6e73 203d 2064 662e 636f 6c75  olumns = df.colu
+00007620: 6d6e 732e 7365 745f 6c65 7665 6c73 285b  mns.set_levels([
+00007630: 6c2e 6173 7479 7065 2822 696e 7422 2920  l.astype("int") 
+00007640: 666f 7220 6c20 696e 2064 662e 636f 6c75  for l in df.colu
+00007650: 6d6e 732e 6c65 7665 6c73 5d29 0d0a 2020  mns.levels])..  
+00007660: 2020 2020 2020 6966 206e 6f72 6d61 6c69        if normali
+00007670: 7a65 3a0d 0a20 2020 2020 2020 2020 2020  ze:..           
+00007680: 206e 6f72 6d61 6c69 7a65 6420 3d20 5b5d   normalized = []
+00007690: 0d0a 2020 2020 2020 2020 2020 2020 666f  ..            fo
+000076a0: 7220 5f2c 2073 7562 6466 2069 6e20 6466  r _, subdf in df
+000076b0: 2e67 726f 7570 6279 2861 7869 733d 312c  .groupby(axis=1,
+000076c0: 206c 6576 656c 3d30 293a 0d0a 2020 2020   level=0):..    
+000076d0: 2020 2020 2020 2020 2020 2020 6e6f 726d              norm
+000076e0: 616c 697a 6564 2e61 7070 656e 6428 7375  alized.append(su
+000076f0: 6264 662e 6469 7628 7375 6264 662e 7375  bdf.div(subdf.su
+00007700: 6d28 6178 6973 3d31 292c 2061 7869 733d  m(axis=1), axis=
+00007710: 3029 290d 0a20 2020 2020 2020 2020 2020  0))..           
+00007720: 2064 6620 3d20 7064 2e63 6f6e 6361 7428   df = pd.concat(
+00007730: 6e6f 726d 616c 697a 6564 2c20 6178 6973  normalized, axis
+00007740: 3d31 290d 0a20 2020 2020 2020 2069 6620  =1)..        if 
+00007750: 6469 7363 7265 7469 7a65 3a0d 0a20 2020  discretize:..   
+00007760: 2020 2020 2020 2020 2064 6973 6372 6574           discret
+00007770: 697a 6564 203d 205b 5d0d 0a20 2020 2020  ized = []..     
+00007780: 2020 2020 2020 2066 6f72 205f 2c20 7375         for _, su
+00007790: 6264 6620 696e 2064 662e 6772 6f75 7062  bdf in df.groupb
+000077a0: 7928 6178 6973 3d31 2c20 6c65 7665 6c3d  y(axis=1, level=
+000077b0: 3029 3a0d 0a20 2020 2020 2020 2020 2020  0):..           
+000077c0: 2020 2020 2064 6973 6372 6574 697a 6564       discretized
+000077d0: 2e61 7070 656e 6428 7375 6264 662e 6571  .append(subdf.eq
+000077e0: 2873 7562 6466 2e6d 6178 2861 7869 733d  (subdf.max(axis=
+000077f0: 3129 2c20 6178 6973 3d30 292e 6173 7479  1), axis=0).asty
+00007800: 7065 2869 6e74 2929 0d0a 2020 2020 2020  pe(int))..      
+00007810: 2020 2020 2020 6466 203d 2070 642e 636f        df = pd.co
+00007820: 6e63 6174 2864 6973 6372 6574 697a 6564  ncat(discretized
+00007830: 2c20 6178 6973 3d31 2920 2020 2020 2020  , axis=1)       
+00007840: 200d 0a20 2020 2020 2020 2069 6620 6b20   ..        if k 
+00007850: 6973 206e 6f74 204e 6f6e 653a 0d0a 2020  is not None:..  
+00007860: 2020 2020 2020 2020 2020 6466 203d 2064            df = d
+00007870: 662e 6c6f 635b 3a2c 206b 5d0d 0a20 2020  f.loc[:, k]..   
+00007880: 2020 2020 2064 6620 3d20 6466 2e73 6f72       df = df.sor
+00007890: 745f 696e 6465 7828 6178 6973 3d30 292e  t_index(axis=0).
+000078a0: 736f 7274 5f69 6e64 6578 2861 7869 733d  sort_index(axis=
+000078b0: 3129 2020 200d 0a20 2020 2020 2020 2072  1)   ..        r
+000078c0: 6574 7572 6e20 6466 0d0a 2020 2020 0d0a  eturn df..    ..
+000078d0: 2020 2020 6465 6620 6765 745f 6765 7073      def get_geps
+000078e0: 2873 656c 662c 0d0a 2020 2020 2020 2020  (self,..        
+000078f0: 2020 2020 2020 2020 206b 3a20 556e 696f           k: Unio
+00007900: 6e5b 696e 742c 2049 7465 7261 626c 655d  n[int, Iterable]
+00007910: 203d 204e 6f6e 652c 0d0a 2020 2020 2020   = None,..      
+00007920: 2020 2020 2020 2020 2020 2074 7970 653d             type=
+00007930: 2263 6e6d 665f 6765 705f 7363 6f72 6522  "cnmf_gep_score"
+00007940: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
+00007950: 2020 2029 202d 3e20 7064 2e44 6174 6146     ) -> pd.DataF
+00007960: 7261 6d65 3a0d 0a20 2020 2020 2020 2022  rame:..        "
+00007970: 2222 0d0a 2020 2020 2020 2020 4765 7420  ""..        Get 
+00007980: 4745 5073 2e0d 0a0d 0a20 2020 2020 2020  GEPs.....       
+00007990: 203a 7061 7261 6d20 6b3a 2049 6620 616e   :param k: If an
+000079a0: 2069 6e74 6567 6572 206f 7220 6c69 7374   integer or list
+000079b0: 206f 6620 696e 7465 6765 7273 2c20 7265   of integers, re
+000079c0: 7475 726e 7320 4745 5073 206f 6e6c 7920  turns GEPs only 
+000079d0: 666f 7220 7370 6563 6966 6965 6420 7261  for specified ra
+000079e0: 6e6b 732e 204f 7468 6572 7769 7365 2c20  nks. Otherwise, 
+000079f0: 7265 7475 726e 7320 4745 5073 2066 726f  returns GEPs fro
+00007a00: 6d20 616c 6c20 7261 6e6b 732e 2044 6566  m all ranks. Def
+00007a10: 6175 6c74 7320 746f 204e 6f6e 650d 0a20  aults to None.. 
+00007a20: 2020 2020 2020 203a 7479 7065 206b 3a20         :type k: 
+00007a30: 556e 696f 6e5b 696e 742c 2049 7465 7261  Union[int, Itera
+00007a40: 626c 655d 2c20 6f70 7469 6f6e 616c 0d0a  ble], optional..
+00007a50: 2020 2020 2020 2020 3a70 6172 616d 2074          :param t
+00007a60: 7970 653a 2022 636e 6d66 5f67 6570 5f73  ype: "cnmf_gep_s
+00007a70: 636f 7265 2220 6f72 2022 636e 6d66 5f67  core" or "cnmf_g
+00007a80: 6570 5f74 706d 222c 2064 6566 6175 6c74  ep_tpm", default
+00007a90: 7320 746f 2022 636e 6d66 5f67 6570 5f73  s to "cnmf_gep_s
+00007aa0: 636f 7265 220d 0a20 2020 2020 2020 203a  core"..        :
+00007ab0: 7479 7065 2074 7970 653a 2073 7472 2c20  type type: str, 
+00007ac0: 6f70 7469 6f6e 616c 0d0a 2020 2020 2020  optional..      
+00007ad0: 2020 3a72 6574 7572 6e3a 2066 6561 7475    :return: featu
+00007ae0: 7265 7320 c397 2047 4550 206d 6174 7269  res .. GEP matri
+00007af0: 780d 0a20 2020 2020 2020 203a 7274 7970  x..        :rtyp
+00007b00: 653a 2070 642e 4461 7461 4672 616d 650d  e: pd.DataFrame.
+00007b10: 0a20 2020 2020 2020 2022 2222 0d0a 2020  .        """..  
+00007b20: 2020 2020 2020 6466 203d 2073 656c 662e        df = self.
+00007b30: 6164 6174 612e 7661 726d 5b74 7970 655d  adata.varm[type]
+00007b40: 2e63 6f70 7928 290d 0a20 2020 2020 2020  .copy()..       
+00007b50: 2064 662e 636f 6c75 6d6e 7320 3d20 7064   df.columns = pd
+00007b60: 2e4d 756c 7469 496e 6465 782e 6672 6f6d  .MultiIndex.from
+00007b70: 5f74 7570 6c65 7328 6466 2e63 6f6c 756d  _tuples(df.colum
+00007b80: 6e73 2e73 7472 2e73 706c 6974 2822 2e22  ns.str.split("."
+00007b90: 292e 746f 5f6c 6973 7428 2929 0d0a 2020  ).to_list())..  
+00007ba0: 2020 2020 2020 6466 2e63 6f6c 756d 6e73        df.columns
+00007bb0: 203d 2064 662e 636f 6c75 6d6e 732e 7365   = df.columns.se
+00007bc0: 745f 6c65 7665 6c73 285b 6c2e 6173 7479  t_levels([l.asty
+00007bd0: 7065 2822 696e 7422 2920 666f 7220 6c20  pe("int") for l 
+00007be0: 696e 2064 662e 636f 6c75 6d6e 732e 6c65  in df.columns.le
+00007bf0: 7665 6c73 5d29 0d0a 2020 2020 2020 2020  vels])..        
+00007c00: 6966 2069 7369 6e73 7461 6e63 6528 6b2c  if isinstance(k,
+00007c10: 2028 696e 742c 2049 7465 7261 626c 6529   (int, Iterable)
+00007c20: 293a 0d0a 2020 2020 2020 2020 2020 2020  ):..            
+00007c30: 6466 203d 2064 662e 6c6f 635b 3a2c 206b  df = df.loc[:, k
+00007c40: 5d0d 0a20 2020 2020 2020 2064 6620 3d20  ]..        df = 
+00007c50: 6466 2e73 6f72 745f 696e 6465 7828 6178  df.sort_index(ax
+00007c60: 6973 3d31 290d 0a20 2020 2020 2020 2072  is=1)..        r
+00007c70: 6574 7572 6e20 6466 0d0a 2020 2020 0d0a  eturn df..    ..
+00007c80: 2020 2020 6465 6620 6765 745f 6d65 7461      def get_meta
+00007c90: 6461 7461 5f64 6628 7365 6c66 2c0d 0a20  data_df(self,.. 
+00007ca0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007cb0: 2020 2020 2020 2069 6e63 6c75 6465 5f63         include_c
+00007cc0: 6174 6567 6f72 6963 616c 3a20 626f 6f6c  ategorical: bool
+00007cd0: 203d 2054 7275 652c 0d0a 2020 2020 2020   = True,..      
+00007ce0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00007cf0: 2020 696e 636c 7564 655f 6e75 6d65 7269    include_numeri
+00007d00: 6361 6c3a 2062 6f6f 6c20 3d20 5472 7565  cal: bool = True
+00007d10: 0d0a 2020 2020 2020 2020 2020 2020 2020  ..              
+00007d20: 2020 2020 2020 2020 2020 2920 2d3e 2070            ) -> p
+00007d30: 642e 4461 7461 4672 616d 653a 0d0a 2020  d.DataFrame:..  
+00007d40: 2020 2020 2020 2222 2247 6574 2073 616d        """Get sam
+00007d50: 706c 652f 6f62 7365 7276 6174 696f 6e20  ple/observation 
+00007d60: 6d65 7461 6461 7461 2e0d 0a0d 0a20 2020  metadata.....   
+00007d70: 2020 2020 203a 7061 7261 6d20 696e 636c       :param incl
+00007d80: 7564 655f 6361 7465 676f 7269 6361 6c3a  ude_categorical:
+00007d90: 2049 6e63 6c75 6465 2063 6174 6567 6f72   Include categor
+00007da0: 6963 616c 206d 6574 6164 6174 6120 6c61  ical metadata la
+00007db0: 7965 7273 2c20 6465 6661 756c 7473 2074  yers, defaults t
+00007dc0: 6f20 5472 7565 0d0a 2020 2020 2020 2020  o True..        
+00007dd0: 3a74 7970 6520 696e 636c 7564 655f 6361  :type include_ca
+00007de0: 7465 676f 7269 6361 6c3a 2062 6f6f 6c2c  tegorical: bool,
+00007df0: 206f 7074 696f 6e61 6c0d 0a20 2020 2020   optional..     
+00007e00: 2020 203a 7061 7261 6d20 696e 636c 7564     :param includ
+00007e10: 655f 6e75 6d65 7269 6361 6c3a 2049 6e63  e_numerical: Inc
+00007e20: 6c75 6465 206e 756d 6572 6963 616c 206d  lude numerical m
+00007e30: 6574 6164 6174 6120 6c61 7965 7273 2c20  etadata layers, 
+00007e40: 6465 6661 756c 7473 2074 6f20 5472 7565  defaults to True
+00007e50: 0d0a 2020 2020 2020 2020 3a74 7970 6520  ..        :type 
+00007e60: 696e 636c 7564 655f 6e75 6d65 7269 6361  include_numerica
+00007e70: 6c3a 2062 6f6f 6c2c 206f 7074 696f 6e61  l: bool, optiona
+00007e80: 6c0d 0a20 2020 2020 2020 203a 7261 6973  l..        :rais
+00007e90: 6573 2056 616c 7565 4572 726f 723a 2045  es ValueError: E
+00007ea0: 7272 6f72 2069 6620 6d65 7461 6461 7461  rror if metadata
+00007eb0: 2074 7970 6573 2061 7265 206e 6f74 2072   types are not r
+00007ec0: 6563 6f67 6e69 7a65 640d 0a20 2020 2020  ecognized..     
+00007ed0: 2020 203a 7265 7475 726e 3a20 6f62 7365     :return: obse
+00007ee0: 7276 6174 696f 6e73 20c3 9720 6d65 7461  rvations .. meta
+00007ef0: 6461 7461 206d 6174 7269 780d 0a20 2020  data matrix..   
+00007f00: 2020 2020 203a 7274 7970 653a 2070 642e       :rtype: pd.
+00007f10: 4461 7461 4672 616d 650d 0a20 2020 2020  DataFrame..     
+00007f20: 2020 2022 2222 0d0a 2020 2020 2020 2020     """..        
+00007f30: 6474 7970 6573 203d 205b 5d0d 0a20 2020  dtypes = []..   
+00007f40: 2020 2020 2069 6620 696e 636c 7564 655f       if include_
+00007f50: 6361 7465 676f 7269 6361 6c3a 0d0a 2020  categorical:..  
+00007f60: 2020 2020 2020 2020 2020 6474 7970 6573            dtypes
+00007f70: 2e61 7070 656e 6428 2263 6174 6567 6f72  .append("categor
+00007f80: 7922 290d 0a20 2020 2020 2020 2069 6620  y")..        if 
+00007f90: 696e 636c 7564 655f 6e75 6d65 7269 6361  include_numerica
+00007fa0: 6c3a 0d0a 2020 2020 2020 2020 2020 2020  l:..            
+00007fb0: 6474 7970 6573 202b 3d20 5b22 666c 6f61  dtypes += ["floa
+00007fc0: 7422 2c20 2269 6e74 225d 0d0a 2020 2020  t", "int"]..    
+00007fd0: 2020 2020 756e 6578 706c 6169 6e65 645f      unexplained_
+00007fe0: 636f 6c73 203d 2073 656c 662e 6164 6174  cols = self.adat
+00007ff0: 612e 6f62 732e 7365 6c65 6374 5f64 7479  a.obs.select_dty
+00008000: 7065 7328 6578 636c 7564 653d 2822 6361  pes(exclude=("ca
+00008010: 7465 676f 7279 222c 2022 666c 6f61 7422  tegory", "float"
+00008020: 2c20 2269 6e74 2229 292e 636f 6c75 6d6e  , "int")).column
+00008030: 730d 0a20 2020 2020 2020 2069 6620 6c65  s..        if le
+00008040: 6e28 756e 6578 706c 6169 6e65 645f 636f  n(unexplained_co
+00008050: 6c73 2920 3e20 303a 0d0a 2020 2020 2020  ls) > 0:..      
+00008060: 2020 2020 2020 756e 6578 706c 6169 6e65        unexplaine
+00008070: 645f 636f 6c5f 7374 7220 3d20 222c 2022  d_col_str = ", "
+00008080: 2e6a 6f69 6e28 756e 6578 706c 6169 6e65  .join(unexplaine
+00008090: 645f 636f 6c73 290d 0a20 2020 2020 2020  d_cols)..       
+000080a0: 2020 2020 2072 6169 7365 2056 616c 7565       raise Value
+000080b0: 4572 726f 7228 6622 7b75 6e65 7870 6c61  Error(f"{unexpla
+000080c0: 696e 6564 5f63 6f6c 5f73 7472 7d20 6d65  ined_col_str} me
+000080d0: 7461 6461 7461 2063 6f6c 756d 6e73 2068  tadata columns h
+000080e0: 6176 6520 756e 7265 636f 676e 697a 6564  ave unrecognized
+000080f0: 2064 7479 7065 732e 2229 0d0a 2020 2020   dtypes.")..    
+00008100: 2020 2020 6466 203d 2073 656c 662e 6164      df = self.ad
+00008110: 6174 612e 6f62 732e 7365 6c65 6374 5f64  ata.obs.select_d
+00008120: 7479 7065 7328 696e 636c 7564 653d 6474  types(include=dt
+00008130: 7970 6573 290d 0a20 2020 2020 2020 2064  ypes)..        d
+00008140: 6620 3d20 6466 2e64 726f 706e 6128 6178  f = df.dropna(ax
+00008150: 6973 3d31 2c20 686f 773d 2261 6c6c 2229  is=1, how="all")
+00008160: 0d0a 2020 2020 2020 2020 7265 7475 726e  ..        return
+00008170: 2064 660d 0a20 2020 200d 0a20 2020 2064   df..    ..    d
+00008180: 6566 2067 6574 5f63 6174 6567 6f72 795f  ef get_category_
+00008190: 6f76 6572 7265 7072 6573 656e 7461 7469  overrepresentati
+000081a0: 6f6e 2873 656c 662c 0d0a 2020 2020 2020  on(self,..      
+000081b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
 000081c0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000081d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000081e0: 2020 2074 7275 6e63 6174 655f 6e65 6761     truncate_nega
-000081f0: 7469 7665 3a20 626f 6f6c 203d 2054 7275  tive: bool = Tru
-00008200: 650d 0a20 2020 2020 2020 2020 2020 2020  e..             
-00008210: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008220: 2020 2020 2020 2020 2020 2029 202d 3e20             ) -> 
-00008230: 7064 2e44 6174 6146 7261 6d65 3a0d 0a20  pd.DataFrame:.. 
-00008240: 2020 2020 2020 2022 2222 4361 6c63 756c         """Calcul
-00008250: 6174 6520 5065 6172 736f 6e20 7265 7369  ate Pearson resi
-00008260: 6475 616c 206f 6620 6368 692d 7371 7561  dual of chi-squa
-00008270: 7265 6420 7465 7374 2c20 6173 736f 6369  red test, associ
-00008280: 6174 696e 6720 4745 5073 2066 6f72 2065  ating GEPs for e
-00008290: 6163 6820 7261 6e6b 2028 6b29 2074 6f20  ach rank (k) to 
-000082a0: 6361 7465 676f 7269 6573 206f 6620 7361  categories of sa
-000082b0: 6d70 6c65 732f 6f62 7365 7276 6174 696f  mples/observatio
-000082c0: 6e73 2e20 4279 2064 6566 6175 6c74 2c20  ns. By default, 
-000082d0: 7472 756e 6361 7465 7320 6e65 6761 7469  truncates negati
-000082e0: 7665 2076 616c 7565 732e 0d0a 0d0a 2020  ve values.....  
-000082f0: 2020 2020 2020 3a70 6172 616d 206c 6179        :param lay
-00008300: 6572 3a20 6e61 6d65 206f 6620 6361 7465  er: name of cate
-00008310: 676f 7269 6361 6c20 6461 7461 206c 6179  gorical data lay
-00008320: 6572 0d0a 2020 2020 2020 2020 3a74 7970  er..        :typ
-00008330: 6520 6c61 7965 723a 2073 7472 0d0a 2020  e layer: str..  
-00008340: 2020 2020 2020 3a70 6172 616d 2074 7275        :param tru
-00008350: 6e63 6174 655f 6e65 6761 7469 7665 3a20  ncate_negative: 
-00008360: 5472 756e 6361 7465 206e 6567 6174 6976  Truncate negativ
-00008370: 6520 7265 7369 6475 616c 7320 746f 2030  e residuals to 0
-00008380: 2c20 6465 6661 756c 7473 2074 6f20 5472  , defaults to Tr
-00008390: 7565 0d0a 2020 2020 2020 2020 3a74 7970  ue..        :typ
-000083a0: 6520 7472 756e 6361 7465 5f6e 6567 6174  e truncate_negat
-000083b0: 6976 653a 2062 6f6f 6c2c 206f 7074 696f  ive: bool, optio
-000083c0: 6e61 6c0d 0a20 2020 2020 2020 203a 7265  nal..        :re
-000083d0: 7475 726e 3a20 6361 7465 676f 7279 20c3  turn: category .
-000083e0: 9720 4745 5020 6d61 7472 6978 206f 6620  . GEP matrix of 
-000083f0: 6f76 6572 7265 7072 6573 656e 7461 7469  overrepresentati
-00008400: 6f6e 2076 616c 7565 730d 0a20 2020 2020  on values..     
-00008410: 2020 203a 7274 7970 653a 2070 642e 4461     :rtype: pd.Da
-00008420: 7461 4672 616d 650d 0a20 2020 2020 2020  taFrame..       
-00008430: 2022 2222 0d0a 2020 2020 2020 2020 7573   """..        us
-00008440: 6167 6520 3d20 7365 6c66 2e67 6574 5f75  age = self.get_u
-00008450: 7361 6765 7328 6e6f 726d 616c 697a 653d  sages(normalize=
-00008460: 5472 7565 292e 636f 7079 2829 0d0a 2020  True).copy()..  
-00008470: 2020 2020 2020 7361 6d70 6c65 5f74 6f5f        sample_to_
-00008480: 636c 6173 7320 3d20 7365 6c66 2e67 6574  class = self.get
-00008490: 5f6d 6574 6164 6174 615f 6466 2829 5b6c  _metadata_df()[l
-000084a0: 6179 6572 5d0d 0a20 2020 2020 2020 2075  ayer]..        u
-000084b0: 7361 6765 2e69 6e64 6578 203d 2075 7361  sage.index = usa
-000084c0: 6765 2e69 6e64 6578 2e6d 6170 2873 616d  ge.index.map(sam
-000084d0: 706c 655f 746f 5f63 6c61 7373 290d 0a20  ple_to_class).. 
-000084e0: 2020 2020 2020 206f 6273 6572 7665 6420         observed 
-000084f0: 3d20 7573 6167 652e 6772 6f75 7062 7928  = usage.groupby(
-00008500: 6178 6973 3d30 2c20 6c65 7665 6c3d 3029  axis=0, level=0)
-00008510: 2e73 756d 2829 0d0a 0d0a 2020 2020 2020  .sum()....      
-00008520: 2020 6e5f 6361 7465 676f 7269 6573 203d    n_categories =
-00008530: 206f 6273 6572 7665 642e 7368 6170 655b   observed.shape[
-00008540: 305d 0d0a 2020 2020 2020 2020 6966 206e  0]..        if n
-00008550: 5f63 6174 6567 6f72 6965 7320 3c20 323a  _categories < 2:
-00008560: 0d0a 2020 2020 2020 2020 2020 2020 6c6f  ..            lo
-00008570: 6767 696e 672e 7761 726e 696e 6728 6622  gging.warning(f"
-00008580: 4f76 6572 7265 7072 6573 656e 7461 7469  Overrepresentati
-00008590: 6f6e 2063 6f75 6c64 206e 6f74 2062 6520  on could not be 
-000085a0: 6361 6c63 756c 6174 6564 2066 6f72 206c  calculated for l
-000085b0: 6179 6572 2027 7b6c 6179 6572 7d27 2c20  ayer '{layer}', 
-000085c0: 6173 206f 6e6c 7920 7b6e 5f63 6174 6567  as only {n_categ
-000085d0: 6f72 6965 737d 2063 6174 6567 6f72 6965  ories} categorie
-000085e0: 7320 7765 7265 2066 6f75 6e64 2069 6e20  s were found in 
-000085f0: 7468 6520 6461 7461 2e20 220d 0a20 2020  the data. "..   
-00008600: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008610: 2020 2020 2020 2020 2066 224e 6f74 6520           f"Note 
-00008620: 7468 6174 2065 6d70 7479 2076 616c 7565  that empty value
-00008630: 7320 696e 2074 6865 206d 6574 6164 6174  s in the metadat
-00008640: 6120 6172 6520 6e6f 7420 636f 6e73 6964  a are not consid
-00008650: 6572 6564 2061 2063 6174 6567 6f72 792e  ered a category.
-00008660: 2022 0d0a 2020 2020 2020 2020 2020 2020   "..            
-00008670: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008680: 6622 4f76 6572 7265 7072 6573 656e 7461  f"Overrepresenta
-00008690: 7469 6f6e 2063 616e 6e6f 7420 6265 2063  tion cannot be c
-000086a0: 616c 6375 6c61 7465 6420 7769 7468 2066  alculated with f
-000086b0: 6577 6572 2074 6861 6e20 3220 6361 7465  ewer than 2 cate
-000086c0: 676f 7269 6573 2066 6f72 2065 6163 6820  gories for each 
-000086d0: 6c61 7965 722e 2022 290d 0a20 2020 2020  layer. ")..     
-000086e0: 2020 2020 2020 2072 6574 7572 6e20 7064         return pd
-000086f0: 2e44 6174 6146 7261 6d65 286e 702e 4e61  .DataFrame(np.Na
-00008700: 4e2c 2069 6e64 6578 203d 206f 6273 6572  N, index = obser
-00008710: 7665 642e 696e 6465 782c 2063 6f6c 756d  ved.index, colum
-00008720: 6e73 3d6f 6273 6572 7665 642e 636f 6c75  ns=observed.colu
-00008730: 6d6e 7329 0d0a 2020 2020 2020 2020 6578  mns)..        ex
-00008740: 7065 6374 6564 203d 205b 5d0d 0a20 2020  pected = []..   
-00008750: 2020 2020 2066 6f72 206b 2c20 6f62 735f       for k, obs_
-00008760: 6b20 696e 206f 6273 6572 7665 642e 6772  k in observed.gr
-00008770: 6f75 7062 7928 6178 6973 3d31 2c20 6c65  oupby(axis=1, le
-00008780: 7665 6c3d 3129 3a0d 0a20 2020 2020 2020  vel=1):..       
-00008790: 2020 2020 2065 7870 5f6b 203d 2070 642e       exp_k = pd.
-000087a0: 4461 7461 4672 616d 6528 6f62 735f 6b2e  DataFrame(obs_k.
-000087b0: 7375 6d28 6178 6973 3d31 2929 2040 2070  sum(axis=1)) @ p
-000087c0: 642e 4461 7461 4672 616d 6528 6f62 735f  d.DataFrame(obs_
-000087d0: 6b2e 7375 6d28 6178 6973 3d30 2929 2e54  k.sum(axis=0)).T
-000087e0: 202f 206f 6273 5f6b 2e73 756d 2829 2e73   / obs_k.sum().s
-000087f0: 756d 2829 0d0a 2020 2020 2020 2020 2020  um()..          
-00008800: 2020 6578 7065 6374 6564 2e61 7070 656e    expected.appen
-00008810: 6428 6578 705f 6b29 0d0a 2020 2020 2020  d(exp_k)..      
-00008820: 2020 6578 7065 6374 6564 203d 2070 642e    expected = pd.
-00008830: 636f 6e63 6174 2865 7870 6563 7465 642c  concat(expected,
-00008840: 2061 7869 733d 3129 0d0a 2020 2020 2020   axis=1)..      
-00008850: 2020 6368 6973 715f 7265 7369 6420 3d20    chisq_resid = 
-00008860: 286f 6273 6572 7665 6420 2d20 6578 7065  (observed - expe
-00008870: 6374 6564 2920 2f20 6e70 2e73 7172 7428  cted) / np.sqrt(
-00008880: 6578 7065 6374 6564 2920 2023 2070 6561  expected)  # pea
-00008890: 7273 6f6e 2072 6573 6964 7561 6c20 6f66  rson residual of
-000088a0: 2063 6869 2d73 7175 6172 6564 2074 6573   chi-squared tes
-000088b0: 7420 6f66 2063 6f6e 7469 6e67 656e 6379  t of contingency
-000088c0: 2074 6162 6c65 0d0a 2020 2020 2020 2020   table..        
-000088d0: 6966 2074 7275 6e63 6174 655f 6e65 6761  if truncate_nega
-000088e0: 7469 7665 3a0d 0a20 2020 2020 2020 2020  tive:..         
-000088f0: 2020 2063 6869 7371 5f72 6573 6964 203d     chisq_resid =
-00008900: 2063 6869 7371 5f72 6573 6964 2e63 6c69   chisq_resid.cli
-00008910: 7028 6c6f 7765 723d 3029 0d0a 2020 2020  p(lower=0)..    
-00008920: 2020 2020 7265 7475 726e 2063 6869 7371      return chisq
-00008930: 5f72 6573 6964 0d0a 2020 2020 0d0a 2020  _resid..    ..  
-00008940: 2020 2020 2020 0d0a 2020 2020 6465 6620        ..    def 
-00008950: 6765 745f 6d65 7461 6461 7461 5f63 6f72  get_metadata_cor
-00008960: 7265 6c61 7469 6f6e 2873 656c 662c 200d  relation(self, .
-00008970: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
-00008980: 2020 2020 2020 2020 2020 2020 2020 2020                  
-00008990: 2020 6c61 7965 723a 2073 7472 2c0d 0a20    layer: str,.. 
+000081d0: 2020 6c61 7965 723a 2073 7472 2c0d 0a20    layer: str,.. 
+000081e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000081f0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008200: 2020 2020 2020 2074 7275 6e63 6174 655f         truncate_
+00008210: 6e65 6761 7469 7665 3a20 626f 6f6c 203d  negative: bool =
+00008220: 2054 7275 650d 0a20 2020 2020 2020 2020   True..         
+00008230: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008240: 2020 2020 2020 2020 2020 2020 2020 2029                 )
+00008250: 202d 3e20 7064 2e44 6174 6146 7261 6d65   -> pd.DataFrame
+00008260: 3a0d 0a20 2020 2020 2020 2022 2222 4361  :..        """Ca
+00008270: 6c63 756c 6174 6520 5065 6172 736f 6e20  lculate Pearson 
+00008280: 7265 7369 6475 616c 206f 6620 6368 692d  residual of chi-
+00008290: 7371 7561 7265 6420 7465 7374 2c20 6173  squared test, as
+000082a0: 736f 6369 6174 696e 6720 4745 5073 2066  sociating GEPs f
+000082b0: 6f72 2065 6163 6820 7261 6e6b 2028 6b29  or each rank (k)
+000082c0: 2074 6f20 6361 7465 676f 7269 6573 206f   to categories o
+000082d0: 6620 7361 6d70 6c65 732f 6f62 7365 7276  f samples/observ
+000082e0: 6174 696f 6e73 2e20 4279 2064 6566 6175  ations. By defau
+000082f0: 6c74 2c20 7472 756e 6361 7465 7320 6e65  lt, truncates ne
+00008300: 6761 7469 7665 2076 616c 7565 732e 0d0a  gative values...
+00008310: 0d0a 2020 2020 2020 2020 3a70 6172 616d  ..        :param
+00008320: 206c 6179 6572 3a20 6e61 6d65 206f 6620   layer: name of 
+00008330: 6361 7465 676f 7269 6361 6c20 6461 7461  categorical data
+00008340: 206c 6179 6572 0d0a 2020 2020 2020 2020   layer..        
+00008350: 3a74 7970 6520 6c61 7965 723a 2073 7472  :type layer: str
+00008360: 0d0a 2020 2020 2020 2020 3a70 6172 616d  ..        :param
+00008370: 2074 7275 6e63 6174 655f 6e65 6761 7469   truncate_negati
+00008380: 7665 3a20 5472 756e 6361 7465 206e 6567  ve: Truncate neg
+00008390: 6174 6976 6520 7265 7369 6475 616c 7320  ative residuals 
+000083a0: 746f 2030 2c20 6465 6661 756c 7473 2074  to 0, defaults t
+000083b0: 6f20 5472 7565 0d0a 2020 2020 2020 2020  o True..        
+000083c0: 3a74 7970 6520 7472 756e 6361 7465 5f6e  :type truncate_n
+000083d0: 6567 6174 6976 653a 2062 6f6f 6c2c 206f  egative: bool, o
+000083e0: 7074 696f 6e61 6c0d 0a20 2020 2020 2020  ptional..       
+000083f0: 203a 7265 7475 726e 3a20 6361 7465 676f   :return: catego
+00008400: 7279 20c3 9720 4745 5020 6d61 7472 6978  ry .. GEP matrix
+00008410: 206f 6620 6f76 6572 7265 7072 6573 656e   of overrepresen
+00008420: 7461 7469 6f6e 2076 616c 7565 730d 0a20  tation values.. 
+00008430: 2020 2020 2020 203a 7274 7970 653a 2070         :rtype: p
+00008440: 642e 4461 7461 4672 616d 650d 0a20 2020  d.DataFrame..   
+00008450: 2020 2020 2022 2222 0d0a 2020 2020 2020       """..      
+00008460: 2020 7573 6167 6520 3d20 7365 6c66 2e67    usage = self.g
+00008470: 6574 5f75 7361 6765 7328 6e6f 726d 616c  et_usages(normal
+00008480: 697a 653d 5472 7565 292e 636f 7079 2829  ize=True).copy()
+00008490: 0d0a 2020 2020 2020 2020 7361 6d70 6c65  ..        sample
+000084a0: 5f74 6f5f 636c 6173 7320 3d20 7365 6c66  _to_class = self
+000084b0: 2e67 6574 5f6d 6574 6164 6174 615f 6466  .get_metadata_df
+000084c0: 2829 5b6c 6179 6572 5d0d 0a20 2020 2020  ()[layer]..     
+000084d0: 2020 2075 7361 6765 2e69 6e64 6578 203d     usage.index =
+000084e0: 2075 7361 6765 2e69 6e64 6578 2e6d 6170   usage.index.map
+000084f0: 2873 616d 706c 655f 746f 5f63 6c61 7373  (sample_to_class
+00008500: 290d 0a20 2020 2020 2020 206f 6273 6572  )..        obser
+00008510: 7665 6420 3d20 7573 6167 652e 6772 6f75  ved = usage.grou
+00008520: 7062 7928 6178 6973 3d30 2c20 6c65 7665  pby(axis=0, leve
+00008530: 6c3d 3029 2e73 756d 2829 0d0a 0d0a 2020  l=0).sum()....  
+00008540: 2020 2020 2020 6e5f 6361 7465 676f 7269        n_categori
+00008550: 6573 203d 206f 6273 6572 7665 642e 7368  es = observed.sh
+00008560: 6170 655b 305d 0d0a 2020 2020 2020 2020  ape[0]..        
+00008570: 6966 206e 5f63 6174 6567 6f72 6965 7320  if n_categories 
+00008580: 3c20 323a 0d0a 2020 2020 2020 2020 2020  < 2:..          
+00008590: 2020 6c6f 6767 696e 672e 7761 726e 696e    logging.warnin
+000085a0: 6728 6622 4f76 6572 7265 7072 6573 656e  g(f"Overrepresen
+000085b0: 7461 7469 6f6e 2063 6f75 6c64 206e 6f74  tation could not
+000085c0: 2062 6520 6361 6c63 756c 6174 6564 2066   be calculated f
+000085d0: 6f72 206c 6179 6572 2027 7b6c 6179 6572  or layer '{layer
+000085e0: 7d27 2c20 6173 206f 6e6c 7920 7b6e 5f63  }', as only {n_c
+000085f0: 6174 6567 6f72 6965 737d 2063 6174 6567  ategories} categ
+00008600: 6f72 6965 7320 7765 7265 2066 6f75 6e64  ories were found
+00008610: 2069 6e20 7468 6520 6461 7461 2e20 220d   in the data. ".
+00008620: 0a20 2020 2020 2020 2020 2020 2020 2020  .               
+00008630: 2020 2020 2020 2020 2020 2020 2066 224e               f"N
+00008640: 6f74 6520 7468 6174 2065 6d70 7479 2076  ote that empty v
+00008650: 616c 7565 7320 696e 2074 6865 206d 6574  alues in the met
+00008660: 6164 6174 6120 6172 6520 6e6f 7420 636f  adata are not co
+00008670: 6e73 6964 6572 6564 2061 2063 6174 6567  nsidered a categ
+00008680: 6f72 792e 2022 0d0a 2020 2020 2020 2020  ory. "..        
+00008690: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000086a0: 2020 2020 6622 4f76 6572 7265 7072 6573      f"Overrepres
+000086b0: 656e 7461 7469 6f6e 2063 616e 6e6f 7420  entation cannot 
+000086c0: 6265 2063 616c 6375 6c61 7465 6420 7769  be calculated wi
+000086d0: 7468 2066 6577 6572 2074 6861 6e20 3220  th fewer than 2 
+000086e0: 6361 7465 676f 7269 6573 2066 6f72 2065  categories for e
+000086f0: 6163 6820 6c61 7965 722e 2022 290d 0a20  ach layer. ").. 
+00008700: 2020 2020 2020 2020 2020 2072 6574 7572             retur
+00008710: 6e20 7064 2e44 6174 6146 7261 6d65 286e  n pd.DataFrame(n
+00008720: 702e 4e61 4e2c 2069 6e64 6578 203d 206f  p.NaN, index = o
+00008730: 6273 6572 7665 642e 696e 6465 782c 2063  bserved.index, c
+00008740: 6f6c 756d 6e73 3d6f 6273 6572 7665 642e  olumns=observed.
+00008750: 636f 6c75 6d6e 7329 0d0a 2020 2020 2020  columns)..      
+00008760: 2020 6578 7065 6374 6564 203d 205b 5d0d    expected = [].
+00008770: 0a20 2020 2020 2020 2066 6f72 206b 2c20  .        for k, 
+00008780: 6f62 735f 6b20 696e 206f 6273 6572 7665  obs_k in observe
+00008790: 642e 6772 6f75 7062 7928 6178 6973 3d31  d.groupby(axis=1
+000087a0: 2c20 6c65 7665 6c3d 3129 3a0d 0a20 2020  , level=1):..   
+000087b0: 2020 2020 2020 2020 2065 7870 5f6b 203d           exp_k =
+000087c0: 2070 642e 4461 7461 4672 616d 6528 6f62   pd.DataFrame(ob
+000087d0: 735f 6b2e 7375 6d28 6178 6973 3d31 2929  s_k.sum(axis=1))
+000087e0: 2040 2070 642e 4461 7461 4672 616d 6528   @ pd.DataFrame(
+000087f0: 6f62 735f 6b2e 7375 6d28 6178 6973 3d30  obs_k.sum(axis=0
+00008800: 2929 2e54 202f 206f 6273 5f6b 2e73 756d  )).T / obs_k.sum
+00008810: 2829 2e73 756d 2829 0d0a 2020 2020 2020  ().sum()..      
+00008820: 2020 2020 2020 6578 7065 6374 6564 2e61        expected.a
+00008830: 7070 656e 6428 6578 705f 6b29 0d0a 2020  ppend(exp_k)..  
+00008840: 2020 2020 2020 6578 7065 6374 6564 203d        expected =
+00008850: 2070 642e 636f 6e63 6174 2865 7870 6563   pd.concat(expec
+00008860: 7465 642c 2061 7869 733d 3129 0d0a 2020  ted, axis=1)..  
+00008870: 2020 2020 2020 6368 6973 715f 7265 7369        chisq_resi
+00008880: 6420 3d20 286f 6273 6572 7665 6420 2d20  d = (observed - 
+00008890: 6578 7065 6374 6564 2920 2f20 6e70 2e73  expected) / np.s
+000088a0: 7172 7428 6578 7065 6374 6564 2920 2023  qrt(expected)  #
+000088b0: 2070 6561 7273 6f6e 2072 6573 6964 7561   pearson residua
+000088c0: 6c20 6f66 2063 6869 2d73 7175 6172 6564  l of chi-squared
+000088d0: 2074 6573 7420 6f66 2063 6f6e 7469 6e67   test of conting
+000088e0: 656e 6379 2074 6162 6c65 0d0a 2020 2020  ency table..    
+000088f0: 2020 2020 6966 2074 7275 6e63 6174 655f      if truncate_
+00008900: 6e65 6761 7469 7665 3a0d 0a20 2020 2020  negative:..     
+00008910: 2020 2020 2020 2063 6869 7371 5f72 6573         chisq_res
+00008920: 6964 203d 2063 6869 7371 5f72 6573 6964  id = chisq_resid
+00008930: 2e63 6c69 7028 6c6f 7765 723d 3029 0d0a  .clip(lower=0)..
+00008940: 2020 2020 2020 2020 7265 7475 726e 2063          return c
+00008950: 6869 7371 5f72 6573 6964 0d0a 2020 2020  hisq_resid..    
+00008960: 0d0a 2020 2020 2020 2020 0d0a 2020 2020  ..        ..    
+00008970: 6465 6620 6765 745f 6d65 7461 6461 7461  def get_metadata
+00008980: 5f63 6f72 7265 6c61 7469 6f6e 2873 656c  _correlation(sel
+00008990: 662c 200d 0a20 2020 2020 2020 2020 2020  f, ..           
 000089a0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000089b0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000089c0: 6d65 7468 6f64 3a20 7374 7220 3d20 2270  method: str = "p
-000089d0: 6561 7273 6f6e 220d 0a20 2020 2020 2020  earson"..       
-000089e0: 2020 2020 2020 2020 2020 2020 2020 2020                  
-000089f0: 2020 2020 2020 2020 2020 2920 2d3e 2070            ) -> p
-00008a00: 642e 5365 7269 6573 3a0d 0a20 2020 2020  d.Series:..     
-00008a10: 2020 2022 2222 4361 6c63 756c 6174 6520     """Calculate 
-00008a20: 5065 6172 736f 6e20 636f 7272 656c 6174  Pearson correlat
-00008a30: 696f 6e20 6f66 2047 4550 2075 7361 6765  ion of GEP usage
-00008a40: 2074 6f20 6e75 6d65 7269 6361 6c20 6d65   to numerical me
-00008a50: 7461 6461 7461 2061 6372 6f73 7320 7361  tadata across sa
-00008a60: 6d70 6c65 732f 6f62 7365 7276 6174 696f  mples/observatio
-00008a70: 6e73 2e0d 0a0d 0a20 2020 2020 2020 203a  ns.....        :
-00008a80: 7061 7261 6d20 6c61 7965 723a 206e 616d  param layer: nam
-00008a90: 6520 6f66 206e 756d 6572 6963 616c 2064  e of numerical d
-00008aa0: 6174 6120 6c61 7965 720d 0a20 2020 2020  ata layer..     
-00008ab0: 2020 203a 7479 7065 206c 6179 6572 3a20     :type layer: 
-00008ac0: 7374 720d 0a20 2020 2020 2020 203a 7061  str..        :pa
-00008ad0: 7261 6d20 6d65 7468 6f64 3a20 436f 7272  ram method: Corr
-00008ae0: 656c 6174 696f 6e20 6d65 7468 6f64 3a20  elation method: 
-00008af0: 2270 6561 7273 6f6e 222c 2022 7370 6561  "pearson", "spea
-00008b00: 726d 616e 222c 206f 7220 226b 656e 6461  rman", or "kenda
-00008b10: 6c6c 222e 2044 6566 6175 6c74 7320 746f  ll". Defaults to
-00008b20: 2022 7065 6172 736f 6e22 0d0a 2020 2020   "pearson"..    
-00008b30: 2020 2020 3a74 7970 6520 6d65 7468 6f64      :type method
-00008b40: 3a20 7374 722c 206f 7074 696f 6e61 6c0d  : str, optional.
-00008b50: 0a20 2020 2020 2020 203a 7265 7475 726e  .        :return
-00008b60: 3a20 636f 7272 656c 6174 696f 6e20 6f66  : correlation of
-00008b70: 2047 4550 2074 6f20 6d65 7461 6461 7461   GEP to metadata
-00008b80: 0d0a 2020 2020 2020 2020 3a72 7479 7065  ..        :rtype
-00008b90: 3a20 7064 2e53 6572 6965 730d 0a20 2020  : pd.Series..   
-00008ba0: 2020 2020 2022 2222 0d0a 2020 2020 2020       """..      
-00008bb0: 2020 7573 6167 6520 3d20 7365 6c66 2e67    usage = self.g
-00008bc0: 6574 5f75 7361 6765 7328 292e 636f 7079  et_usages().copy
-00008bd0: 2829 0d0a 2020 2020 2020 2020 6d65 7461  ()..        meta
-00008be0: 6461 7461 203d 2073 656c 662e 6765 745f  data = self.get_
-00008bf0: 6d65 7461 6461 7461 5f64 6628 295b 6c61  metadata_df()[la
-00008c00: 7965 725d 0d0a 2020 2020 2020 2020 6d64  yer]..        md
-00008c10: 5f63 6f72 7220 3d20 7573 6167 652e 636f  _corr = usage.co
-00008c20: 7272 7769 7468 286d 6574 6164 6174 612c  rrwith(metadata,
-00008c30: 206d 6574 686f 643d 6d65 7468 6f64 290d   method=method).
-00008c40: 0a20 2020 2020 2020 2072 6574 7572 6e20  .        return 
-00008c50: 6d64 5f63 6f72 720d 0a20 2020 2020 2020  md_corr..       
-00008c60: 200d 0a20 2020 2064 6566 2061 7070 656e   ..    def appen
-00008c70: 645f 746f 5f68 6973 746f 7279 2873 656c  d_to_history(sel
-00008c80: 662c 2065 6e74 7279 293a 0d0a 2020 2020  f, entry):..    
-00008c90: 2020 2020 2222 2241 6464 2065 6e74 7279      """Add entry
-00008ca0: 2074 6f20 4461 7461 7365 7420 6869 7374   to Dataset hist
-00008cb0: 6f72 792e 0d0a 0d0a 2020 2020 2020 2020  ory.....        
-00008cc0: 3a70 6172 616d 2065 6e74 7279 3a20 4465  :param entry: De
-00008cd0: 7363 7269 7074 696f 6e20 6f66 2065 7665  scription of eve
-00008ce0: 6e74 2074 6f20 7265 636f 7264 2069 6e20  nt to record in 
-00008cf0: 7468 6520 6869 7374 6f72 792e 0d0a 2020  the history...  
-00008d00: 2020 2020 2020 3a74 7970 6520 656e 7472        :type entr
-00008d10: 793a 2073 7472 0d0a 2020 2020 2020 2020  y: str..        
-00008d20: 2222 220d 0a20 2020 2020 2020 2073 656c  """..        sel
-00008d30: 662e 6164 6174 612e 756e 735b 2268 6973  f.adata.uns["his
-00008d40: 746f 7279 225d 5b64 6174 6574 696d 652e  tory"][datetime.
-00008d50: 7574 636e 6f77 2829 2e69 736f 666f 726d  utcnow().isoform
-00008d60: 6174 2829 5d20 3d20 656e 7472 790d 0a20  at()] = entry.. 
-00008d70: 2020 2020 2020 200d 0a20 2020 2064 6566         ..    def
-00008d80: 2067 6574 5f68 6973 746f 7279 2873 656c   get_history(sel
-00008d90: 6629 3a0d 0a20 2020 2020 2020 2022 2222  f):..        """
-00008da0: 5265 7475 726e 7320 7469 6d65 7374 616d  Returns timestam
-00008db0: 7065 6420 6869 7374 6f72 7920 6f66 2044  ped history of D
-00008dc0: 6174 6173 6574 206f 626a 6563 742e 0d0a  ataset object...
-00008dd0: 0d0a 2020 2020 2020 2020 3a72 6574 7572  ..        :retur
-00008de0: 6e3a 2068 6973 746f 7279 0d0a 2020 2020  n: history..    
-00008df0: 2020 2020 3a72 7479 7065 3a20 6469 6374      :rtype: dict
-00008e00: 0d0a 2020 2020 2020 2020 2222 220d 0a20  ..        """.. 
-00008e10: 2020 2020 2020 2072 6574 7572 6e20 7365         return se
-00008e20: 6c66 2e61 6461 7461 2e75 6e73 5b22 6869  lf.adata.uns["hi
-00008e30: 7374 6f72 7922 5d                        story"]
+000089b0: 2020 2020 2020 6c61 7965 723a 2073 7472        layer: str
+000089c0: 2c0d 0a20 2020 2020 2020 2020 2020 2020  ,..             
+000089d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
+000089e0: 2020 2020 6d65 7468 6f64 3a20 7374 7220      method: str 
+000089f0: 3d20 2270 6561 7273 6f6e 220d 0a20 2020  = "pearson"..   
+00008a00: 2020 2020 2020 2020 2020 2020 2020 2020                  
+00008a10: 2020 2020 2020 2020 2020 2020 2020 2920                ) 
+00008a20: 2d3e 2070 642e 5365 7269 6573 3a0d 0a20  -> pd.Series:.. 
+00008a30: 2020 2020 2020 2022 2222 4361 6c63 756c         """Calcul
+00008a40: 6174 6520 5065 6172 736f 6e20 636f 7272  ate Pearson corr
+00008a50: 656c 6174 696f 6e20 6f66 2047 4550 2075  elation of GEP u
+00008a60: 7361 6765 2074 6f20 6e75 6d65 7269 6361  sage to numerica
+00008a70: 6c20 6d65 7461 6461 7461 2061 6372 6f73  l metadata acros
+00008a80: 7320 7361 6d70 6c65 732f 6f62 7365 7276  s samples/observ
+00008a90: 6174 696f 6e73 2e0d 0a0d 0a20 2020 2020  ations.....     
+00008aa0: 2020 203a 7061 7261 6d20 6c61 7965 723a     :param layer:
+00008ab0: 206e 616d 6520 6f66 206e 756d 6572 6963   name of numeric
+00008ac0: 616c 2064 6174 6120 6c61 7965 720d 0a20  al data layer.. 
+00008ad0: 2020 2020 2020 203a 7479 7065 206c 6179         :type lay
+00008ae0: 6572 3a20 7374 720d 0a20 2020 2020 2020  er: str..       
+00008af0: 203a 7061 7261 6d20 6d65 7468 6f64 3a20   :param method: 
+00008b00: 436f 7272 656c 6174 696f 6e20 6d65 7468  Correlation meth
+00008b10: 6f64 3a20 2270 6561 7273 6f6e 222c 2022  od: "pearson", "
+00008b20: 7370 6561 726d 616e 222c 206f 7220 226b  spearman", or "k
+00008b30: 656e 6461 6c6c 222e 2044 6566 6175 6c74  endall". Default
+00008b40: 7320 746f 2022 7065 6172 736f 6e22 0d0a  s to "pearson"..
+00008b50: 2020 2020 2020 2020 3a74 7970 6520 6d65          :type me
+00008b60: 7468 6f64 3a20 7374 722c 206f 7074 696f  thod: str, optio
+00008b70: 6e61 6c0d 0a20 2020 2020 2020 203a 7265  nal..        :re
+00008b80: 7475 726e 3a20 636f 7272 656c 6174 696f  turn: correlatio
+00008b90: 6e20 6f66 2047 4550 2074 6f20 6d65 7461  n of GEP to meta
+00008ba0: 6461 7461 0d0a 2020 2020 2020 2020 3a72  data..        :r
+00008bb0: 7479 7065 3a20 7064 2e53 6572 6965 730d  type: pd.Series.
+00008bc0: 0a20 2020 2020 2020 2022 2222 0d0a 2020  .        """..  
+00008bd0: 2020 2020 2020 7573 6167 6520 3d20 7365        usage = se
+00008be0: 6c66 2e67 6574 5f75 7361 6765 7328 292e  lf.get_usages().
+00008bf0: 636f 7079 2829 0d0a 2020 2020 2020 2020  copy()..        
+00008c00: 6d65 7461 6461 7461 203d 2073 656c 662e  metadata = self.
+00008c10: 6765 745f 6d65 7461 6461 7461 5f64 6628  get_metadata_df(
+00008c20: 295b 6c61 7965 725d 0d0a 2020 2020 2020  )[layer]..      
+00008c30: 2020 6d64 5f63 6f72 7220 3d20 7573 6167    md_corr = usag
+00008c40: 652e 636f 7272 7769 7468 286d 6574 6164  e.corrwith(metad
+00008c50: 6174 612c 206d 6574 686f 643d 6d65 7468  ata, method=meth
+00008c60: 6f64 290d 0a20 2020 2020 2020 2072 6574  od)..        ret
+00008c70: 7572 6e20 6d64 5f63 6f72 720d 0a20 2020  urn md_corr..   
+00008c80: 2020 2020 200d 0a20 2020 2064 6566 2061       ..    def a
+00008c90: 7070 656e 645f 746f 5f68 6973 746f 7279  ppend_to_history
+00008ca0: 2873 656c 662c 2065 6e74 7279 293a 0d0a  (self, entry):..
+00008cb0: 2020 2020 2020 2020 2222 2241 6464 2065          """Add e
+00008cc0: 6e74 7279 2074 6f20 4461 7461 7365 7420  ntry to Dataset 
+00008cd0: 6869 7374 6f72 792e 0d0a 0d0a 2020 2020  history.....    
+00008ce0: 2020 2020 3a70 6172 616d 2065 6e74 7279      :param entry
+00008cf0: 3a20 4465 7363 7269 7074 696f 6e20 6f66  : Description of
+00008d00: 2065 7665 6e74 2074 6f20 7265 636f 7264   event to record
+00008d10: 2069 6e20 7468 6520 6869 7374 6f72 792e   in the history.
+00008d20: 0d0a 2020 2020 2020 2020 3a74 7970 6520  ..        :type 
+00008d30: 656e 7472 793a 2073 7472 0d0a 2020 2020  entry: str..    
+00008d40: 2020 2020 2222 220d 0a20 2020 2020 2020      """..       
+00008d50: 2073 656c 662e 6164 6174 612e 756e 735b   self.adata.uns[
+00008d60: 2268 6973 746f 7279 225d 5b64 6174 6574  "history"][datet
+00008d70: 696d 652e 7574 636e 6f77 2829 2e69 736f  ime.utcnow().iso
+00008d80: 666f 726d 6174 2829 5d20 3d20 656e 7472  format()] = entr
+00008d90: 790d 0a20 2020 2020 2020 200d 0a20 2020  y..        ..   
+00008da0: 2064 6566 2067 6574 5f68 6973 746f 7279   def get_history
+00008db0: 2873 656c 6629 3a0d 0a20 2020 2020 2020  (self):..       
+00008dc0: 2022 2222 5265 7475 726e 7320 7469 6d65   """Returns time
+00008dd0: 7374 616d 7065 6420 6869 7374 6f72 7920  stamped history 
+00008de0: 6f66 2044 6174 6173 6574 206f 626a 6563  of Dataset objec
+00008df0: 742e 0d0a 0d0a 2020 2020 2020 2020 3a72  t.....        :r
+00008e00: 6574 7572 6e3a 2068 6973 746f 7279 0d0a  eturn: history..
+00008e10: 2020 2020 2020 2020 3a72 7479 7065 3a20          :rtype: 
+00008e20: 6469 6374 0d0a 2020 2020 2020 2020 2222  dict..        ""
+00008e30: 220d 0a20 2020 2020 2020 2072 6574 7572  "..        retur
+00008e40: 6e20 7365 6c66 2e61 6461 7461 2e75 6e73  n self.adata.uns
+00008e50: 5b22 6869 7374 6f72 7922 5d              ["history"]
```

### Comparing `cnmfsns-1.6.5/src/cnmfsns/integration.py` & `cnmfsns-1.7.0/src/cnmfsns/integration.py`

 * *Files 6% similar despite different names*

```diff
@@ -321,36 +321,70 @@
         nodetable = pd.DataFrame(nodetable)
         nodetable.columns.rename(["Node filter", "Edge Filter"], inplace=True)
         return nodetable
     
     def get_metadata_df(self,
                         include_categorical: bool = True,
                         include_numerical: bool = True,
-                        prepend_dataset_column: bool = False
+                        prepend_dataset_column: bool = False,
+                        subset_datasets: Optional[Union[str, Iterable[str]]] = None
                         ) -> pd.DataFrame:
         """Get sample/observation metadata for all datasets.
 
         :param include_categorical: Include categorical metadata layers, defaults to True
         :type include_categorical: bool, optional
         :param include_numerical: Include numerical metadata layers, defaults to True
         :type include_numerical: bool, optional
         :param prepend_dataset_column: Prepend dataframe with dataset name column, defaults to False
         :type prepend_dataset_column: bool, optional
+        :param subset_datasets: dataset name or iterable of dataset names to subset the results, defaults to None
+        :type subset_datasets: str or Iterable[str], optional
         :return: observations × metadata matrix
         :rtype: pd.DataFrame
         """
+
+        if subset_datasets is None:
+            subset_datasets = self.datasets.keys()
+        elif isinstance(subset_datasets, str):
+            subset_datasets = [subset_datasets]
+        elif isinstance(subset_datasets, Iterable):
+            pass
+        else:
+            raise ValueError
+
+
         df = {}
-        for dataset_name, dataset in self.datasets.items():
-            df[dataset_name] = dataset.get_metadata_df(include_categorical=include_categorical,
+        for dataset_name in subset_datasets:
+            df[dataset_name] = self.datasets[dataset_name].get_metadata_df(include_categorical=include_categorical,
                                                        include_numerical=include_numerical)
         df = pd.concat(df, axis=0)
         if prepend_dataset_column:
             df.insert(0, "Dataset", df.index.get_level_values(0))
         return df
     
+    def get_features_overlap_table(self):
+
+        all_features = set()
+        for ds in self.datasets.values():
+            all_features |= set(ds.adata.var_names)
+        df = pd.DataFrame(False, index=sorted(list(all_features)), columns=self.datasets.keys())
+        for dsname, ds in self.datasets.items():
+            df.loc[ds.adata.var_names, dsname] = True
+        return df
+
+    def get_overdispersed_features_overlap_table(self):
+
+        all_features = set()
+        for ds in self.datasets.values():
+            all_features |= set(ds.adata.var_names)
+        df = pd.DataFrame(False, index=sorted(list(all_features)), columns=self.datasets.keys())
+        for dsname, ds in self.datasets.items():
+            df.loc[ds.overdispersed_genes, dsname] = True
+        return df
+    
     def get_category_overrepresentation(self,
                                         layer: str,
                                         subset_datasets: Optional[Union[str, Iterable[str]]] = None,
                                         truncate_negative: bool = True) -> pd.DataFrame:
         """
         Calculate Pearson residual of chi-squared test, associating GEPs for each rank (k) to categories of samples/observations. By default, truncates negative values.
```

### Comparing `cnmfsns-1.6.5/src/cnmfsns/plots.py` & `cnmfsns-1.7.0/src/cnmfsns/plots.py`

 * *Files 4% similar despite different names*

```diff
@@ -109,56 +109,14 @@
         else:
             sns.histplot(df, x=y_unit, bins=100, linewidth=0, ax=ax_plot, color="blue")
     
     if ax is None:
         return fig
 
 
-    # if all_plots:
-    #     # Figure: cnmf model mean and variance
-    #     fig, axes = plt.subplots(1, 3, figsize=[12, 4], layout="tight")
-    #     ax = axes[0]
-    #     if show_selected:
-    #         sns.histplot(df, x="log_mean", y="log_variance", hue="selected", bins=[100,100], ax=ax, alpha=0.5, palette={True: "red", False: "blue"})
-    #     else:
-    #         sns.histplot(df, x="log_mean", y="log_variance", bins=[100,100], ax=ax, color="blue")
-    #     ax.set_title("Mean-Variance")
-    #     ax.set_xlabel("log10(mean)")
-    #     ax.set_ylabel("log10(variance)")
-    #     ax = axes[1]
-    #     if show_selected:
-    #         sns.histplot(df, x="log_mean", y="vscore", hue="selected", bins=[100,100], ax=ax, palette={True: "red", False: "blue"})
-    #     else:
-    #         sns.histplot(df, x="log_mean", y="vscore", bins=[100,100], ax=ax, color="blue")
-    #     ax.set_title("Mean-Overdispersion (cnmf)")
-    #     ax.set_xlabel("log10(mean)")
-    #     ax.set_ylabel("v-score")
-    #     ax=axes[2]  # thresholds
-    #     ax.set_title("v-score Distribution")
-    #     if df["vscore"].notnull().any():e
-    #         if show_selected:
-    #             sns.histplot(df, x="vscore", hue="selected", bins=100, linewidth=0, ax=ax, palette={True: "red", False: "blue"})
-    #             ax.legend(handles=[
-    #                 Patch(color="blue", alpha=0.5, label="False"),
-    #                 Patch(color="red", alpha=0.5, label="True")
-    #             ], title="selected")
-    #         else:
-    #             sns.histplot(df, x="vscore", bins=100, linewidth=0, ax=ax, color="blue")
-    #     ax.set_xlabel("v-score")
-    #     figs["cnmf"] = fig
-
-    #     fig, ax = plt.subplots(figsize=[12,12], layout="tight")
-    #     sns.histplot(data=df.fillna(0), x="odscore", y="vscore", bins=[100, 100], ax=ax)
-    #     ax.set_xlabel("od-score")
-    #     ax.set_ylabel("v-score")
-    #     figs["score_comparison"] = fig
-        
-    # return figs
-
-
 def plot_stability_error(dataset: Dataset, figsize=(6, 4)):
     '''
     Borrowed from Alexandrov Et Al. 2013 Deciphering Mutational Signatures
     publication in Cell Reports
     '''
     
     stats = dataset.adata.uns["kvals"]
@@ -181,15 +139,15 @@
     ax1.grid('on')
     return fig
 
 
 
 def annotated_heatmap(
         data, title, metadata=None, metadata_colors=None, missing_data_color="#BBBBBB", heatmap_cmap='YlOrRd', 
-        row_cluster=True, col_cluster=True, plot_col_dendrogram=True, show_sample_labels=True, ylabel=""):
+        row_cluster=True, col_cluster=True, plot_col_dendrogram=True, show_sample_labels=True, ylabel=None, cbar_label=None):
     n_columns = data.shape[1]
     if metadata is None:
         n_metadata_columns = 0
     else:
         n_metadata_columns = metadata.shape[1]
 
     fig = plt.figure(figsize=[20, 2 + n_columns/3 + n_metadata_columns/4])
@@ -271,15 +229,15 @@
             ax.spines['right'].set_visible(False)
             ax.spines['bottom'].set_visible(False)
             ax.spines['left'].set_visible(False)
     
     # Colormap for heatmap
     ax = fig.add_subplot(gs0[1])
     ax.set_axis_off()
-    plt.colorbar(im_heatmap, ax=ax, location="top")
+    plt.colorbar(im_heatmap, ax=ax, location="top", label=cbar_label)
     return fig
 
 def plot_usage_heatmap(dataset: Dataset, k: Optional[int], colors, subset_metadata = None, subset_samples = None, title = None, cluster_geps = False, cluster_samples = True, show_sample_labels = True):
     assert dataset.has_cnmf_results
     df = dataset.get_usages(k=k)
     if subset_samples is not None:
         df = df.loc[subset_samples]
@@ -293,15 +251,15 @@
                             metadata_colors=metadata_colors, 
                             missing_data_color=colors.missing_data_color, 
                             title=title,
                             row_cluster=cluster_geps,
                             col_cluster=cluster_samples,
                             show_sample_labels=show_sample_labels,
                             plot_col_dendrogram=True,
-                            ylabel="GEP")
+                            ylabel="Program")
     return fig
 
 
 
 #############################
 # Integration-related plots #
 #############################
@@ -313,15 +271,15 @@
                         cmap=cmap, center=0, vmin=-1, vmax=1,
                         row_colors=ds_color_track, col_colors=ds_color_track)
     cg.ax_heatmap.set_ylabel("")
     cg.ax_heatmap.set_xlabel("")
     if hide_gep_labels:
         cg.ax_heatmap.set_xticks([])
         cg.ax_heatmap.set_yticks([])
-    cg.ax_col_dendrogram.set_title("GEP correlations")
+    cg.ax_col_dendrogram.set_title("Program correlations")
     return cg.figure
 
 
 def plot_rank_reduction(integration: Integration, figsize=None):
     n_ranks = integration.k_table.shape[0]
     n_datasets = len(integration.datasets)
     if figsize is None:
@@ -337,116 +295,199 @@
         ax.set_title(dataset_name)
         ax.set_xticks(df["max_k"])
         sns.lineplot(data=df, x="max_k", y="max_k_median_corr", hue="max_k_filter_pass", ax=ax, marker="o")
         ax.get_legend().remove()
         plt.tight_layout()
     return fig
 
-def plot_pairwise_corr(integration: Integration, subplot_size = [3, 3.5], overlaid=False, bins=50):
+
+def plot_pairwise_corr(integration: Integration, subplot_size: Collection = [3, 3.5], overlaid=False, bins=50, inside_padding = 0.02) -> Figure:
+    """Plot histograms of the pairwise correlation distribution for each dataset pair.
+
+    :param integration: Integration object
+    :type integration: Integration
+    :param subplot_size: width and height of each subplot, defaults to [3, 3.5]
+    :type subplot_size: Collection, optional
+    :param overlaid: Overlay negative and positive parts to compare the distribution, defaults to False
+    :type overlaid: bool, optional
+    :param bins: Number of bins for histogram, defaults to 50
+    :type bins: int, optional
+    :param inside_padding: Amount to shrink data within the axes, defaults to 0.02
+    :type inside_padding: float, optional
+    :return: figure
+    :rtype: Figure
+    """
     tril = integration.get_corr_matrix_lowertriangle()
     sps_width, sps_height = subplot_size
     n_datasets = integration.n_datasets
     fig, axes = plt.subplots(n_datasets, n_datasets,
                              figsize=[sps_width * n_datasets, sps_height * n_datasets], sharex=True, sharey=True, squeeze=False, layout="tight")
-    fig.suptitle(f"Correlation distribution between datasets")
-    fig.supxlabel("Correlation coefficient")
+    fig.supxlabel("Correlation")
     for row, dataset_row in enumerate(tril.index.levels[0]):
         for col, dataset_col in enumerate(tril.columns.levels[0]):
             ax = axes[row,col]
-            if row < col:
+            if row == 0 and col == integration.n_datasets - 1:
+                handles = [
+                    Rectangle((0, 0), width = 0.8, height = 0.8, facecolor='red', label="Included"),
+                    Rectangle((0, 0), width = 0.8, height = 0.8, facecolor='grey', label="Excluded")
+                ]
+                ax.legend(handles=handles, loc='center', frameon=False)
+                ax.set_axis_off()
+            elif row < col:
                 ax.set_axis_off()
             else:
                 corr = pd.Series(tril.loc[dataset_row, dataset_col].values.flatten()).dropna()
+                max_corr = corr.max()
                 if integration.pairwise_thresholds is not None:
                     min_corr = integration.pairwise_thresholds.loc[(dataset_row, dataset_col)]
                     hist_kwargs = {
                         "hue":[("Included" if c else "Excluded") for c in (corr > min_corr)],
                         "palette": {"Included": "red", "Excluded": "gray"},
                         "hue_order": ["Excluded", "Included"]
                     }
                     included_fraction = (corr > min_corr).sum() / corr.shape[0]
-                    ax.text(x=0.01, y=1.01, s=f"quantile={included_fraction:.3f}\nmin_corr={min_corr:.3f}", size=8, ha='left', va='bottom', transform=ax.transAxes, color="black")
+                    ax.text(x=0.01, y=1.08, s=f"AUC = {included_fraction:.3f}", size=8, ha='left', va='bottom', transform=ax.transAxes, color="black")
+                    ax.text(x=0.01, y=1.01, s=f"range = {min_corr:.3f}-{max_corr:.3f}", size=8, ha='left', va='bottom', transform=ax.transAxes, color="red")
                 else:
                     hist_kwargs = {"color": "gray"}
-                sns.histplot(x=corr, ax=ax,legend=(row == 0)&(col == 0), bins=bins, linewidth=0, **hist_kwargs)
+                sns.histplot(x=corr, ax=ax,legend=False, bins=bins, linewidth=0, multiple="stack", **hist_kwargs)
 
                 ax.set_ylabel(dataset_row)
                 ax.set_xlabel(dataset_col)
-                ax.set_xlim(-1,1)
+                ax.set_xlim(-1 - inside_padding, 1 + inside_padding)
+
+    ymax = max([ax.get_ylim()[1] for row in axes for ax in row])
+    for row in axes:
+        for ax in row:
+            ax.set_ylim((
+                - ymax * inside_padding,
+                ymax + ymax * inside_padding
+            ))
+
     return fig
 
-def plot_pairwise_corr_overlaid(integration: Integration, subplot_size = [3, 3.5], bins=50):
+def plot_pairwise_corr_overlaid(integration: Integration, subplot_size = [3, 3.5], bins=50, inside_padding = 0.02):
     tril = integration.get_corr_matrix_lowertriangle(max_k_filter=True)
     n_datasets = integration.n_datasets
     sps_width, sps_height = subplot_size
     fig, axes = plt.subplots(n_datasets, n_datasets,
                              figsize=[sps_width * n_datasets, sps_height * n_datasets],
                              sharex=True, sharey=True, squeeze=False, layout="tight")
-    fig.suptitle(f"Correlation distribution between datasets")
-    fig.supxlabel("Correlation coefficient")
+    fig.supxlabel("Correlation")
     for row, dataset_row in enumerate(tril.index.levels[0]):
         for col, dataset_col in enumerate(tril.columns.levels[0]):
             ax = axes[row,col]
-            if row < col:
+            if row == 0 and col == integration.n_datasets - 1:
+                handles = [
+                    Rectangle((0, 0), width = 0.8, height = 0.8, facecolor='red', alpha=0.4, label="Positive"),
+                    Rectangle((0, 0), width = 0.8, height = 0.8, facecolor='blue', alpha=0.4, label="Negative")
+                ]
+                ax.legend(handles=handles, loc='center', frameon=False)
+                ax.set_axis_off()
+            elif row < col:
                 ax.set_axis_off()
             else:
                 corr = pd.DataFrame({"corr": tril.loc[dataset_row, dataset_col].values.flatten()}).dropna()
+                max_corr = corr["corr"].max()
                 corr["sign"] = (corr["corr"] >= 0).map({True: "Positive", False: "Negative"})
                 corr["abscorr"] = corr["corr"].abs()
-                sns.histplot(data=corr, x="abscorr", hue="sign", palette= {"Positive": "red", "Negative": "blue"}, bins=bins, alpha=0.5, linewidth=0,
-                             hue_order= ["Negative", "Positive"], ax=ax,legend=(row == 0)&(col == 0))
+                sns.histplot(data=corr, x="abscorr", hue="sign", palette= {"Positive": "red", "Negative": "blue"}, bins=bins, alpha=0.4, linewidth=0,
+                             hue_order= ["Negative", "Positive"], ax=ax,legend=False)
+
 
                 # show min_corr as text in top left of plot and vertical line
                 min_corr = integration.pairwise_thresholds.loc[(dataset_row, dataset_col)]
                 included_fraction = (corr["corr"] > min_corr).sum() / corr.shape[0]  # could also show the quantile of the min_corr threshold
-                ax.text(x=0.01, y=1.01, s=f"quantile={included_fraction:.3f}\nmin_corr={min_corr:.3f}", size=8, ha='left', va='bottom', transform=ax.transAxes, color="black")
+                ax.text(x=0.01, y=1.01, s=f"AUC = {included_fraction:.3f}\nrange = {min_corr:.3f}-{max_corr:.3f}",
+                        size=8, ha='left', va='bottom', transform=ax.transAxes, color="black")
                 ax.axvline(min_corr, color="black")
                 ax.set_ylabel(dataset_row)
                 ax.set_xlabel(dataset_col)
-                ax.set_xlim(0,1)
+                ax.set_xlim(0 - inside_padding, 1 + inside_padding)
+
+    ymax = max([ax.get_ylim()[1] for row in axes for ax in row])
+    for row in axes:
+        for ax in row:
+            ax.set_ylim((
+                - ymax * inside_padding,
+                ymax + ymax * inside_padding
+            ))
+
     return fig
 
-def plot_overdispersed_features_upset(integration: Integration, figsize=[6, 4]):
+def plot_overdispersed_features_upset(integration: Integration, figsize: Collection = (6, 4)) -> Figure:
+    """Plot overlaps of overdispersed features between datasets
+
+    :param integration: integration object
+    :type integration: Integration
+    :param figsize: width and height of figure, defaults to [6, 4]
+    :type figsize: Collection, optional
+    :return: figure
+    :rtype: Figure
+    """
     overdispersed_feature_lists = {dataset_name: dataset.overdispersed_genes for dataset_name, dataset in integration.datasets.items()}
     fig = Figure(figsize=figsize)
     upsetplot.UpSet(upsetplot.from_contents(overdispersed_feature_lists)).plot(fig=fig)
     fig.suptitle("Overdispersed features")
     return fig
 
 def plot_features_upset(integration: Integration, figsize=[6, 4]):
+    """Plot overlaps of features between datasets
+
+    :param integration: integration object
+    :type integration: Integration
+    :param figsize: width and height of figure, defaults to [6, 4]
+    :type figsize: Collection, optional
+    :return: figure
+    :rtype: Figure
+    """
     feature_lists = {dataset_name: list(dataset.adata.var.index) for dataset_name, dataset in integration.datasets.items()}
     fig = Figure(figsize=figsize)
     upsetplot.UpSet(upsetplot.from_contents(feature_lists)).plot(fig=fig)
     fig.suptitle("Features")
     return fig
 
 #####################
 # SNS-related plots #
 #####################
 
 def plot_community_usage_heatmap(snsmap: SNS,
                                  colors: Colors,
                                  subset_metadata = None,
-                                 subset_datasets = None,
+                                 subset_datasets: Optional[Union[str, Collection[str]]] = None,
                                  subset_samples = None,
                                  title = None,
                                  cluster_geps = False,
                                  cluster_samples = True,
                                  show_sample_labels = True,
                                  prepend_dataset_colors = True):
+    
+
+    
+    if subset_datasets is None:
+        subset_datasets = snsmap.integration.datasets.keys()
+    elif isinstance(subset_datasets, str):
+        subset_datasets = [subset_datasets]
+    elif isinstance(subset_datasets, Iterable):
+        pass
+    else:
+        raise ValueError
+    
     df = snsmap.get_community_usage()
+    
     if subset_samples is not None:
         df = df.loc[subset_samples]
-    if subset_datasets is not None:
-        df = df.loc[subset_datasets]
-        
     
+    df = df.loc[subset_datasets]
+            
     metadata = snsmap.integration.get_metadata_df()
+    metadata = metadata[metadata.index.get_level_values(0).isin(subset_datasets)]
     if subset_metadata is not None:
         metadata = metadata.loc[:, subset_metadata]
+    metadata = metadata.dropna(axis=1, how="all")
 
     metadata_colors = {col: colors.get_metadata_colors(col) for col in metadata.columns}
     if prepend_dataset_colors:
         metadata.insert(0, "dataset", metadata.index.get_level_values(0))
         metadata_colors["dataset"] = colors.dataset_colors
     
     
@@ -454,15 +495,16 @@
                             metadata_colors=metadata_colors, 
                             missing_data_color=colors.missing_data_color, 
                             title=title,
                             row_cluster=cluster_geps,
                             col_cluster=cluster_samples,
                             show_sample_labels=show_sample_labels,
                             plot_col_dendrogram=True,
-                            ylabel="Community")
+                            ylabel="Community",
+                            cbar_label="Normalized Program Usage")
     return fig
 
 
 def plot_community_usage_per_sample(snsmap: SNS,
                                     colors: Colors,
                                     dataset_name: str,
                                     layer: str
@@ -481,28 +523,33 @@
     for i, (celltype, subdf) in enumerate(df.groupby(axis=0, level=0)):
         ax = axes[i // n_col, i % n_col]
         sns.violinplot(data=subdf, linewidth=0.1, ax=ax, cut=0, color=colors.community_colors)
         ax.set_title(celltype)
         is_last_row = i // n_col == n_row - 1
         if not is_last_row:
             ax.set_xlabel("")
+    
+    return fig
 
-
-def plot_community_by_dataset_rank(snsmap: SNS, colors: Colors, figsize: Collection = None):
+def plot_community_by_dataset_rank(snsmap: SNS, colors: Colors, figsize: Collection = None, highlight_central_gep: bool = True):
     """
     Plot communities by dataset and rank representation
     """
 
     marker_style = {
         1: ("s", 30),  # 1 factor: square markers, size 30
         2: (2, 30)     # 2 factors: marker #2 (up tick), size 30
         }
+
+    if highlight_central_gep:
+        central_ranks = pd.DataFrame(snsmap.get_central_geps()).reset_index().set_index(["Community", "dataset"])["k"]
+
     n_datasets = snsmap.integration.n_datasets
     if figsize is None:
-        figsize = [1 + n_datasets * 5, 1 + len(snsmap.communities)/4]
+        figsize = [2 + n_datasets * 5.5, 1 + len(snsmap.communities)/4]
     fig, axes = plt.subplots(1, n_datasets+1, figsize=figsize, sharex=True, sharey=True, layout="tight")
     for dataset, ax in zip(snsmap.integration.datasets, axes):
         for y, community in enumerate(snsmap.ordered_community_names):
             members = snsmap.communities[community]
             counts = pd.Series([m.rpartition("|")[0] for m in members]).value_counts()
             
             # plot line if any factors are present
@@ -522,29 +569,36 @@
                 scatter_y = []
                 for x, rank in enumerate(snsmap.integration.selected_k[dataset]):
                     factor_prefix = f"{dataset}|{rank}"
                     if factor_prefix in counts.index and counts[factor_prefix] == count:
                         scatter_x.append(x)
                         scatter_y.append(y)
                 ax.scatter(scatter_x, scatter_y, color=colors.dataset_colors[dataset], marker=style[0], s=style[1])
+            
+            if highlight_central_gep and (community, dataset) in central_ranks:
+                x = snsmap.integration.selected_k[dataset].index(central_ranks[(community, dataset)])
+                ax.add_patch(Rectangle((x - 0.4, y-0.4), width = 0.8, height = 0.8, edgecolor="k", facecolor='none'))
+
         ax.set_yticks(list(range(len(snsmap.ordered_community_names))))
         ax.set_yticklabels(snsmap.ordered_community_names)
         ax.set_xticks(list(range(len(snsmap.integration.selected_k[dataset]))))
         ax.set_xticklabels(snsmap.integration.selected_k[dataset])
         ax.set_title(dataset)
 
     fig.supxlabel("Rank (k)")
     fig.supylabel("Community")
 
 
     # Add legend
     cbdrlegend = []
-    cbdrlegend.append(Line2D([0],[0], marker='s', color='black', label="1 GEP", markerfacecolor="black", markersize=8))
-    cbdrlegend.append(Line2D([0],[0], marker=2, color='black', label="2 GEPs", markerfacecolor="black", markersize=8))
-    cbdrlegend.append(Line2D([0],[0], marker=None, color='black', label="3+ GEPs", markerfacecolor="black", markersize=8))
+    cbdrlegend.append(Line2D([0],[0], marker='s', color='black', label="1 Program", markerfacecolor="black", markersize=8))
+    cbdrlegend.append(Line2D([0],[0], marker=2, color='black', label="2 Programs", markerfacecolor="black", markersize=8))
+    cbdrlegend.append(Line2D([0],[0], marker=None, color='black', label="3+ Programs", markerfacecolor="black", markersize=8))
+    if highlight_central_gep:
+        cbdrlegend.append(Rectangle((0, 0), width = 0.8, height = 0.8, edgecolor="k", facecolor='none', label="Central Program"))
     axes[-1].legend(handles=cbdrlegend, loc='center', frameon=False)
     axes[-1].set_axis_off()
     return fig
 
 
 ###########################################
 #   Helper functions for circle bar plots #
@@ -570,57 +624,59 @@
             if draw_labels:
                 a = (previous - np.pi / len(enrichments))
                 x_offset = (size * 1.1) * np.cos(a)
                 y_offset = (size * 1.1) * np.sin(a)
                 ax.text(x+x_offset, y+y_offset, label, rotation=np.rad2deg(a), ha="left", va="center", rotation_mode='anchor', fontsize=label_font_size)
         previous = this
 
-def draw_circle_bar_scale(position, size, ax, scale_factor, label_font_size, linewidth=0.5, rings=[0.5,1]):
+def draw_circle_bar_scale(position, size, ax, scale_factor, label_font_size = 6, linewidth=0.8, rings=[0.2, 0.5, 1]):
     x, y = position
     for ring in rings:
         ax.add_patch(plt.Circle(position, np.sqrt(ring) * size, color="black", fill=False, linewidth=linewidth))
     ax.add_patch(Rectangle(position, size * 1.01, size * 1.01, color="#FFFFFF"))
+    ax.scatter(position[0], position[1], s=linewidth*2, color="black", marker=".")
     for ring in rings:
         value = ring/scale_factor
         ax.text(
             x + size * 0.05,
             y + np.sqrt(ring) * size,
-            f"{value:.3f}",
+            f"{value:.2f}",
             fontsize=label_font_size,
             verticalalignment="center")
 
 
-#######################
-#   GEP Network Plots #
-#######################
+###########################
+#  Program Network Plots  #
+###########################
 
 
 def plot_overrepresentation_gep_network(snsmap: SNS,
                                         colors: Colors,
                                         layer: str,
                                         subset_datasets: Optional[Union[str, Collection[str]]] = None,
                                         ax: Optional[Axes] = None,
-                                        pie_size: float = 0.1,
+                                        pie_size: float = 0.3,
                                         figsize: Collection = (9, 6),
                                         edge_weights: Optional[str] = None,
                                         edge_color: str = "#AAAAAA88",
                                         metric: str = "pearson_residual",
+                                        legend_pie_size: float = 0.1,
                                         show_legend: bool = True) -> Optional[Figure]:
     
     if show_legend:
         assert ax is None
     
     if ax is None:
         fig, (ax_plot, ax_legend) = plt.subplots(1, 2, figsize=figsize, sharey=True, gridspec_kw={"width_ratios": [2, 1]}, layout="tight")
     else:
         ax_plot = ax
         ax_legend = ax
     ax_plot.set_aspect("equal")
     ax_plot.set_axis_off()
-    ax_plot.set_title("GEP Network")
+    ax_plot.set_title("Programs")
     ax_legend.set_xlim([-0.5, 0.5])
     ax_legend.set_aspect("equal")
     ax_legend.set_axis_off()
     
     if edge_weights is None or not snsmap.gep_graph.edges:
         width = 0.2
     else:
@@ -652,22 +708,22 @@
     if show_legend:
         # Add legends
         draw_circle_bar_plot(
             position=(0, 0.5),
             enrichments=pd.Series(max_or, index=gep_or.index.sort_values().unique()),
             colors=overrepresentation.index.map(colors.get_metadata_colors(layer)),
             scale_factor=scale_factor,
-            size=pie_size,
+            size=legend_pie_size,
             draw_labels=True,
             label_font_size=6, ax=ax_legend)
         draw_circle_bar_scale(
-            position=(0, -0.5),
+            position=(0, -0.75),
             scale_factor=scale_factor,
             size=pie_size,
-            label_font_size=4, ax=ax_legend)
+            label_font_size=5, ax=ax_legend)
         ax_legend.set_title(f"{layer}")
         ax_legend.text(0, -0.25, "overrepresentation", ha="center", va="center", )
     
     # assert ax_plot.get_xlim() == ax_plot.get_ylim()
     # assert ax_plot.get_ylim() == ax_legend.get_ylim()
     
     if ax is None:
@@ -693,15 +749,15 @@
     if ax is None:
         fig, (ax_plot, ax_legend) = plt.subplots(1, 2, figsize=figsize, sharey=True, gridspec_kw={"width_ratios": [6, 1]}, layout="tight")
     else:
         ax_plot = ax
         ax_legend = ax
     ax_plot.set_aspect("equal")
     ax_plot.set_axis_off()
-    ax_plot.set_title(f"GEPs correlated with\n{layer}")
+    ax_plot.set_title(f"Programs correlated with\n{layer}")
     ax_legend.set_aspect("equal")
     ax_legend.set_axis_off()
     
     if edge_weights is None or not snsmap.gep_graph.edges:
         width = 0.2
     else:
         width = np.array(list(nx.get_edge_attributes(snsmap.gep_graph, edge_weights).values()))
@@ -717,17 +773,15 @@
         if gep in md_corr.index:
             node_colors.append(color_map.to_rgba(md_corr[gep]))
         else:
             node_colors.append(mpl.colors.to_rgba(colors.missing_data_color))
     nx.draw(snsmap.gep_graph, pos=snsmap.layout, node_color=node_colors, node_size=node_size, linewidths=0, width=width, edge_color=edge_color, with_labels=False, ax=ax_plot, font_size=20)
     
     if show_legend:
-        # Add legend
-        ax_legend.set_title(f"{method.capitalize()} correlation", fontsize=6)
-        plt.colorbar(color_map, ax=ax_legend, location="left", anchor=(0, 1), panchor=(1, 1), shrink=0.5, fraction=0.15)
+        plt.colorbar(color_map, ax=ax_legend, location="left", anchor=(0, 1), panchor=(1, 1), shrink=0.5, fraction=0.15, label=f"{method.capitalize()} correlation")
     
     # assert ax_plot.get_xlim() == ax_plot.get_ylim()
     # assert ax_plot.get_ylim() == ax_legend.get_ylim()
     
     if ax is None:
         return fig
 
@@ -737,14 +791,15 @@
     if ax is None:
         fig, (ax_plot, ax_legend) = plt.subplots(1, 2, figsize=figsize, sharey=True, gridspec_kw={"width_ratios": [2, 1]}, layout="tight")
     else:
         ax_plot = ax
         ax_legend = ax
     ax_plot.set_aspect("equal")
     ax_plot.set_axis_off()
+    ax_plot.set_title("Programs")
     ax_legend.set_xlim([-0.5, 0.5])
     ax_legend.set_axis_off()
     colors.plot_dataset_colors_legend(ax=ax_legend)
     
     node_colors = []
     for node in snsmap.gep_graph:
         node_colors.append(colors.dataset_colors[node.split("|")[0]])
@@ -772,15 +827,14 @@
             node_color=node_colors,
             labels=node_labels,
             node_size=node_sizes,
             linewidths=0,
             width=0.2,
             edge_color=edge_color,
             font_size=4, ax=ax_plot)
-    ax_plot.set_title("GEP Network")
     return fig
 
 
 def plot_gep_network_communities(snsmap: SNS,
                                  colors: Colors,
                                  figsize = (9,6),
                                  edge_color = "#AAAAAA88",
@@ -791,14 +845,15 @@
     if ax is None:
         fig, (ax_plot, ax_legend) = plt.subplots(1, 2, figsize=figsize, sharey=True, gridspec_kw={"width_ratios": [2, 1]}, layout="tight")
     else:
         ax_plot = ax
         ax_legend = ax
     ax_plot.set_aspect("equal")
     ax_plot.set_axis_off()
+    ax_plot.set_title("Programs")
     ax_legend.set_xlim([-0.5, 0.5])
     ax_legend.set_axis_off()
     colors.plot_community_colors_legend(ax=ax_legend)
     
     node_colors = []
     for node in snsmap.gep_graph:
         if node in snsmap.gep_communities:
@@ -830,15 +885,14 @@
             node_color=node_colors,
             labels=labels,
             node_size=node_sizes,
             linewidths=0,
             width=0.2,
             edge_color=edge_color,
             font_size=4, ax=ax_plot)
-    ax_plot.set_title("GEP Network")
     if ax is None:
         return fig
 
 
 def plot_gep_network_nsamples(snsmap: SNS,
                              colors: Colors,
                              figsize: Collection = (9, 6),
@@ -851,14 +905,15 @@
     if ax is None:
         fig, (ax_plot, ax_legend) = plt.subplots(1, 2, figsize=figsize, sharey=True, gridspec_kw={"width_ratios": [2, 1]}, layout="tight")
     else:
         ax_plot = ax
         ax_legend = ax
     ax_plot.set_aspect("equal")
     ax_plot.set_axis_off()
+    ax_plot.set_title("Programs")
     ax_legend.set_xlim([-0.5, 0.5])
     ax_legend.set_axis_off()
     colors.plot_dataset_colors_legend(ax=ax_legend)
 
     usages = snsmap.integration.get_usages(discretize = discretize,
                                            normalize = True)
 
@@ -894,21 +949,24 @@
         fig, (ax_plot, ax_legend) = plt.subplots(1, 2, figsize=figsize, sharey=True, gridspec_kw={"width_ratios": [2, 1]}, layout="tight")
         ax_legend.set_axis_off()
     else:
         ax_plot = ax
         ax_legend = ax
     ax_plot.set_aspect("equal")
     ax_plot.set_axis_off()
+    ax_plot.set_title("Programs")
     ax_legend.set_xlim([-0.5, 0.5])
     colors.plot_dataset_colors_legend(ax=ax_legend)
 
     usages = snsmap.integration.get_usages(discretize = True).fillna(0).astype(bool)
-    if snsmap.integration.sample_to_patient is None:
+    s2p = snsmap.integration.sample_to_patient
+    if s2p is None:
         raise ValueError("No samples have valid patient IDs. Make sure to set the patient_id_col property for each Dataset")
-    usages.index = usages.index.map(snsmap.integration.sample_to_patient)
+    usages = usages[usages.index.isin(s2p.index)]
+    usages.index = usages.index.map(s2p)
     usages = usages.groupby(axis=0, level=[0,1]).any()
         
     labels = usages[snsmap.geps_in_graph].sum().apply(lambda x: str(int(x))).to_dict()
     labels = {f"{k[0]}|{k[1]}|{k[2]}": v for k,v in labels.items()}
     
     scale_factor = node_size / usages[snsmap.geps_in_graph].sum().max()
     sizes = (usages[snsmap.geps_in_graph].sum() * scale_factor).to_dict()
@@ -920,15 +978,15 @@
     nx.draw(snsmap.gep_graph, snsmap.layout,
     with_labels=True, labels=labels, node_color=colors, node_size=node_sizes, linewidths=0, width=0.2, edge_color=edge_color, font_size=font_size, ax=ax_plot)
     if ax is None:
         return fig
 
 
 #################################
-#   GEP Community Network Plots #
+#   Community Network Plots #
 #################################
 
 
 def plot_summary_community_network(snsmap: SNS,
                                    colors: Colors,
                                    figsize = (4, 4),
                                    edge_color = "#AAAAAA88",
@@ -937,15 +995,15 @@
                                    ax: Axes = None):
     if ax is None:
         fig, ax_plot = plt.subplots(figsize=figsize, layout="tight")
     else:
         ax_plot = ax
     ax_plot.set_aspect("equal")
     ax_plot.set_axis_off()
-    ax_plot.set_title("Community Network")
+    ax_plot.set_title("Communities")
     
     G = snsmap.comm_graph
     if G.edges:
         width = np.array(list(nx.get_edge_attributes(G, "n_edges").values()))
         width = 20 * width / np.max(width)
     else:
         width = None
@@ -981,15 +1039,15 @@
     if ax is None:
         fig, (ax_plot, ax_legend) = plt.subplots(1, 2, figsize=figsize, sharey=True, gridspec_kw={"width_ratios": [4, 1]}, layout="tight")
     else:
         ax_plot = ax
         ax_legend = ax
     ax_plot.set_aspect("equal")
     ax_plot.set_axis_off()
-    ax_plot.set_title(f"GEP Communities correlated with\n{layer}")
+    ax_plot.set_title(f"Communities\n{layer}")
     ax_legend.set_aspect("equal")
     ax_legend.set_axis_off()
     
     if snsmap.comm_graph.edges:
         width = np.array(list(nx.get_edge_attributes(snsmap.comm_graph, "n_edges").values()))
         width = 20 * width / np.max(width)
     else:
@@ -1005,27 +1063,26 @@
         if community in md_corr.index:
             node_colors.append(color_map.to_rgba(md_corr[community]))
         else:
             node_colors.append(mpl.colors.to_rgba(colors.missing_data_color))
     nx.draw(snsmap.comm_graph, pos=snsmap.comm_layout, node_color=node_colors, node_size=sizes, linewidths=0, width=width, edge_color=edge_color, with_labels=True, ax=ax_plot, font_size=20)
     
     if show_legend:
-        ax_legend.set_title(f"{method.capitalize()} correlation", fontsize=6)
-        plt.colorbar(color_map, ax=ax_legend, location="left", anchor=(0, 1), panchor=(1, 1), shrink=0.5, fraction=0.15)
+        plt.colorbar(color_map, ax=ax_legend, location="left", anchor=(0, 1), panchor=(1, 1), shrink=0.5, fraction=0.15, label=f"{method.capitalize()} correlation")
         
     if ax is None:
         return fig
 
 
 def plot_overrepresentation_community_network(snsmap: SNS,
                                         colors: Colors,
                                         layer: str,
                                         subset_datasets: Optional[Union[str, Collection[str]]] = None,
                                         ax: Optional[Axes] = None,
-                                        pie_size: float = 0.1,
+                                        pie_size: float = 0.3,
                                         figsize: Collection = (6, 4),
                                         edge_color: str = "#AAAAAA88",
                                         metric: str = "pearson_residual",
                                         legend_pie_size: float = 0.1,
                                         show_legend: bool = True) -> Optional[Figure]:
     
     if show_legend:
@@ -1038,15 +1095,15 @@
             ax_legend.set_aspect("equal")
             ax_legend.set_axis_off()
     else:
         ax_plot = ax
         ax_legend = ax
     ax_plot.set_aspect("equal")
     ax_plot.set_axis_off()
-    ax_plot.set_title("Community Network")
+    ax_plot.set_title("Communities")
     
     if snsmap.comm_graph.edges:
         width = np.array(list(nx.get_edge_attributes(snsmap.comm_graph, "n_edges").values()))
         width = 20 * width / np.max(width)
     else:
         width = None
 
@@ -1075,30 +1132,30 @@
             enrichments=pd.Series(max_or, index=comm_or.index.sort_values().unique()),
             colors=overrepresentation.index.map(colors.get_metadata_colors(layer)),
             scale_factor=scale_factor,
             size=legend_pie_size,
             draw_labels=True,
             label_font_size=6, ax=ax_legend)
         draw_circle_bar_scale(
-            position=(0, -0.5),
+            position=(0, -0.75),
             scale_factor=scale_factor,
             size=pie_size,
-            label_font_size=4, ax=ax_legend)
+            label_font_size=5, ax=ax_legend)
         ax_legend.set_title(f"{layer}")
         ax_legend.text(0, 0 , "overrepresentation", ha="center", va="center", )
     
     # assert ax_plot.get_xlim() == ax_plot.get_ylim()
     # assert ax_plot.get_ylim() == ax_legend.get_ylim()
     
     if ax is None:
         return fig
 
 
 #################
-# GEP bar plots #
+# Program-level bar plots #
 #################
 
 
 # overrepresentation bar plots
 def plot_overrepresentation_gep_bar(snsmap: SNS,
                                      colors: Colors,
                                      dataset_name: str,
@@ -1123,15 +1180,15 @@
     
     fig, axes = plt.subplots(
         len(metadata_layers) + 1, snsmap.n_communities,
         figsize = figsize,
         sharey='row', squeeze=False,
         gridspec_kw={"width_ratios": community_gep_counts},
         layout="tight")
-    fig.supxlabel("GEP")
+    fig.supxlabel("Program")
     fig.supylabel("Overrepresentation")
     fig.suptitle("Community")
     
     # first subplot row has k-values
     axes[0, 0].set_ylabel("rank (k)")
     for col, community in enumerate(snsmap.ordered_community_names):
         ax = axes[0, col]
@@ -1199,15 +1256,15 @@
     
     fig, axes = plt.subplots(
         len(metadata_layers) + 1, snsmap.n_communities,
         figsize = figsize,
         sharey='row', squeeze=False,
         gridspec_kw={"width_ratios": community_gep_counts},
         layout="tight")
-    fig.supxlabel("GEP")
+    fig.supxlabel("Program")
     fig.supylabel(method.capitalize() + " Correlation")
     fig.suptitle("Community")
     
     # first subplot row has k-values
     axes[0, 0].set_ylabel("rank (k)")
     for col, community in enumerate(snsmap.ordered_community_names):
         ax = axes[0, col]
@@ -1244,15 +1301,15 @@
             if row == 0:
                 ax.set_title(community, size=14)
 
     return fig
     
 
 ###########################
-# GEP community bar plots #
+# Community bar plots #
 ###########################
 
 
 def plot_overrepresentation_community_bar(snsmap: SNS,
                                       colors: Colors,
                                       layer: str,
                                       subset_datasets: Optional[Union[str, Collection[str]]] = None,
@@ -1271,15 +1328,15 @@
         ax_plot = ax
         ax_legend = ax
         
     # Overrepresentation plot
     df.T.plot.bar(stacked=True, ax=ax_plot, width = 0.9, color=colors.get_metadata_colors(layer), legend=False)
     ax_plot.set_xticklabels(ax_plot.get_xticklabels(), rotation=0)
     ax_plot.set_ylabel("Median overrepresentation")
-    ax_plot.set_xlabel("GEP Community")
+    ax_plot.set_xlabel("Community")
     
     # Legend
     colors.plot_metadata_colors_legend(layer=layer, ax=ax_legend)
 
     if ax is None:
         return fig
 
@@ -1302,15 +1359,15 @@
     else:
         ax_plot = ax
         
     # Overrepresentation plot
     md_corr.plot.bar(ax=ax_plot, width = 0.9, color="#888888")
     ax_plot.set_xticklabels(ax_plot.get_xticklabels(), rotation=0)
     ax_plot.set_ylabel(f"Median {method.capitalize()} Correlation")
-    ax_plot.set_xlabel("GEP Community")
+    ax_plot.set_xlabel("Community")
     
     if ax is None:
         return fig
 
 
 #################################
 # Community-usage Entropy plots #
@@ -1319,15 +1376,15 @@
 
 def plot_sample_entropy(snsmap: SNS,
                      colors: Colors,
                      layer: str = "Dataset",
                      subset_datasets: Optional[Union[str, Collection[str]]] = None,
                      figsize = None
                      ):    
-    metadata = snsmap.integration.get_metadata_df(prepend_dataset_column=True)[layer]
+    metadata = snsmap.integration.get_metadata_df(prepend_dataset_column=True, subset_datasets=subset_datasets)[layer]
     diversity = snsmap.get_sample_entropy()
     df = pd.concat([diversity.rename("entropy"), metadata], axis=1)
     
     if figsize == None:
         figsize= [df[metadata.name].nunique() * 0.5 + 1,3]
     fig, ax = plt.subplots(figsize=figsize)
     if metadata.name == "Dataset":
@@ -1344,10 +1401,10 @@
         y="entropy",
         data=df,
         showfliers=False,
         showbox=False,
         showcaps=False,
         ax=ax)
     ax.set_ybound(lower=0)
-    ax.set_title("Sample Entropy")
+    ax.set_title("Shannon Entropy")
     ax.set_xticklabels(ax.get_xticklabels(), rotation=90)
     return fig
```

### Comparing `cnmfsns-1.6.5/src/cnmfsns/sns.py` & `cnmfsns-1.7.0/src/cnmfsns/sns.py`

 * *Files identical despite different names*

### Comparing `cnmfsns-1.6.5/src/cnmfsns/utils.py` & `cnmfsns-1.7.0/src/cnmfsns/utils.py`

 * *Files identical despite different names*

### Comparing `cnmfsns-1.6.5/src/cnmfsns.egg-info/PKG-INFO` & `cnmfsns-1.7.0/src/cnmfsns.egg-info/PKG-INFO`

 * *Files 18% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 Metadata-Version: 2.1
 Name: cnmfsns
-Version: 1.6.5
+Version: 1.7.0
 Summary: cNMF Solution Network Space
 Home-page: https://github.com/MorrissyLab/cNMF-SNS
 Author: Ted Verhey
 Author-email: tbverhey@ucalgary.ca
 Project-URL: Bug Tracker, https://github.com/MorrissyLab/cNMF-SNS/issues
 Classifier: Programming Language :: Python :: 3
 Classifier: License :: OSI Approved :: MIT License
@@ -16,15 +16,15 @@
 
 ![cNMF-SNS logo](logo.png)
 
 -----------------
 
 # cNMF-SNS: powerful factorization-based multi-omics integration toolkit
 
-![version badge](https://img.shields.io/badge/version-1.6.5-blue)
+![version badge](https://img.shields.io/badge/version-1.7.0-blue)
 [![PyPI Latest Release](https://img.shields.io/pypi/v/cnmfsns.svg)](https://pypi.org/project/cnmfsns/)
 [![Conda Latest Release](https://img.shields.io/conda/vn/conda-forge/cnmfsns)](https://anaconda.org/conda-forge/cnmfsns/)
 [![Documentation status](https://readthedocs.org/projects/cnmf-sns/badge/?version=latest&style=flat)]()
 [![Downloads](https://static.pepy.tech/badge/cnmfsns)](https://pepy.tech/project/cnmfsns)
 [![License](https://img.shields.io/pypi/l/cnmfsns.svg)](https://github.com/MorrissyLab/cNMF-SNS/blob/main/LICENSE)
 
 Authors: [Ted Verhey](https://github.com/verheytb), [Heewon Seo](https://github.com/lootpiz), [Sorana Morrissy](https://github.com/ancasorana)
@@ -38,31 +38,33 @@
 Communities with GEPs from multiple datasets can be annotated with dataset-specific
 annotations to facilitate interpretation.
 
 ## ⚡Main Features
 
 Here are just a few of the things that cNMF-SNS does well:
 
-- Identifies interpretable, additive non-negative gene expression programs
+- Identifies interpretable, non-negative programs at multiple resolutions
 - Mosaic integration does not require subsetting features/genes to
   a shared or overdispersed subset
 - Ideal for incremental integration (adding datasets one at a time) since
-  deconvolution is performed independently on each dataset generating invariant GEPs
-- Integration does not require comparable data sparsity or sequencing depth (single-cell, microarray, bulk)
+  deconvolution is performed independently on each dataset
+- Integration performs well even when the datasets have mismatched features
+  (eg. Microarray, RNA-Seq, Proteomics) or sparsity (eg single-cell vs bulk RNA-Seq and ATAC-Seq)
 - Two interfaces: command-line interface for rapid data exploration and python
   interface for extensibility and flexibility
 
 ## 🔧 Install
 
 ### ☁️ Public Release
 
 Install the package with conda (in an isolated conda environment)
 ```bash
-conda env create cnmfsns -c conda-forge cnmfsns
+conda create -n cnmfsns -c conda-forge cnmfsns
 conda activate cnmfsns
+
 ```
 
 ### ✨ Latest version from GitHub
 
 Before installing cNMF-SNS using pip, it is recommended to first set up a separate conda environment and have conda manage as many dependencies as possible.
 
 ```bash
```

